{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b35cc53d",
   "metadata": {},
   "source": [
    "# üê¢ Generate RDF files for Bgee SPARQL queries\n",
    "\n",
    "Queries from https://github.com/RIKEN-DKO/Generation_SPARQL/blob/main/augment_queries/augment_queries.ipynb\n",
    "\n",
    "To be added to https://github.com/sib-swiss/sparql-examples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1b9f0d7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Bgee queries (with variants) from tutorial online:\n",
    "# https://www.bgee.org/support/tutorial-query-bgee-knowledge-graph-sparql#querying-with-controlled-vocabularies-and-identifiers\n",
    "\n",
    "qs = [None] * 16\n",
    "\n",
    "qs[0] = []\n",
    "\n",
    "qs[1] = [\"\"\"\n",
    "PREFIX up: <http://purl.uniprot.org/core/>\n",
    "\n",
    "SELECT ?species WHERE {\n",
    "\t?species a up:Taxon .\n",
    "}\n",
    "\"\"\"]\n",
    "\n",
    "qs[2] = [\"\"\"\n",
    "PREFIX up: <http://purl.uniprot.org/core/>\n",
    "\n",
    "SELECT ?species ?sci_name ?common_name WHERE {\n",
    "\t?species a up:Taxon ;\n",
    "\t\tup:scientificName ?sci_name ;\n",
    "        up:rank up:Species .\n",
    "\tOPTIONAL { ?species up:commonName ?common_name . }\n",
    "}\n",
    "\"\"\"]\n",
    "\n",
    "qs[3] = [\"\"\"\n",
    "PREFIX orth: <http://purl.org/net/orth#>\n",
    "PREFIX genex: <http://purl.org/genex#>\n",
    "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
    "\n",
    "SELECT DISTINCT ?anat ?anatName WHERE {\n",
    "\t?seq a orth:Gene ;\n",
    "\t\tgenex:isExpressedIn ?anat ;\n",
    "\t\trdfs:label \"APOC1\" .\n",
    "\t?anat a genex:AnatomicalEntity ;\n",
    "\t\trdfs:label ?anatName .\n",
    "}\n",
    "\"\"\"]\n",
    "\n",
    "qs[4] = [\"\"\"\n",
    "PREFIX orth: <http://purl.org/net/orth#>\n",
    "PREFIX genex: <http://purl.org/genex#>\n",
    "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
    "PREFIX up: <http://purl.uniprot.org/core/>\n",
    "\n",
    "SELECT DISTINCT ?anat ?anatName WHERE {\n",
    "\t?seq a orth:Gene ;\n",
    "\t\tgenex:isExpressedIn ?anat ;\n",
    "\t\trdfs:label \"APOC1\" .\n",
    "\t?anat a genex:AnatomicalEntity ;\n",
    "\t\trdfs:label ?anatName .\n",
    "\t?seq orth:organism ?organism .\n",
    "\t?organism obo:RO_0002162  ?species .\n",
    "\t?species a up:Taxon ;\n",
    "\t\tup:scientificName \"Homo sapiens\" .\n",
    "}\n",
    "\"\"\"]\n",
    "\n",
    "qs[5] = [\"\"\"\n",
    "PREFIX orth: <http://purl.org/net/orth#>\n",
    "PREFIX genex: <http://purl.org/genex#>\n",
    "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
    "PREFIX efo: <http://www.ebi.ac.uk/efo/>\n",
    "\n",
    "SELECT DISTINCT ?anat ?anatName {\n",
    "\t?seq a orth:Gene ;\n",
    "\t\tgenex:isExpressedIn ?condition ;\n",
    "\t\trdfs:label \"APOC1\" .\n",
    "    ?condition a genex:ExpressionCondition ;\n",
    "\t\tgenex:hasAnatomicalEntity ?anat ;\n",
    "\t\tgenex:hasAnatomicalEntity obo:GO_0005575 ;\n",
    "\t\tgenex:hasDevelopmentalStage ?stage ;\n",
    "\t\tgenex:hasSex \"any\" ;\n",
    "\t\tgenex:hasStrain ?strain .\n",
    "    ?anat a genex:AnatomicalEntity ;\n",
    "\t\trdfs:label ?anatName .\n",
    "    ?stage a efo:EFO_0000399 ;\n",
    "\t\trdfs:label \"life cycle\" .\n",
    "\t?strain rdfs:label \"wild-type\" .\n",
    "}\n",
    "\"\"\"]\n",
    "\n",
    "qs[6] = [\"\"\"\n",
    "PREFIX orth: <http://purl.org/net/orth#>\n",
    "PREFIX genex: <http://purl.org/genex#>\n",
    "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
    "PREFIX up: <http://purl.uniprot.org/core/>\n",
    "PREFIX efo: <http://www.ebi.ac.uk/efo/>\n",
    "\n",
    "SELECT DISTINCT ?anat ?anatName ?stage WHERE {\n",
    "\t?seq a orth:Gene .\n",
    "\t?seq genex:isExpressedIn ?condition.\n",
    "    ?condition a genex:ExpressionCondition.\n",
    "\t?seq rdfs:label \"APOC1\" .\n",
    "\t?condition genex:hasAnatomicalEntity ?anat .\n",
    "\t?anat rdfs:label ?anatName .\n",
    "\t\t?condition genex:hasAnatomicalEntity obo:GO_0005575 .\n",
    "\t\t?condition genex:hasDevelopmentalStage ?stage .\n",
    "        ?stage a efo:EFO_0000399 .\n",
    "\t?stage rdfs:label \"post-juvenile\" .\n",
    "\t?condition genex:hasSex \"any\".\n",
    "\t?condition genex:hasStrain ?strain .\n",
    "    ?strain a efo:EFO_0005135.\n",
    "\t\t?strain rdfs:label \"wild-type\" .\n",
    "\t\t?seq orth:organism ?organism .\n",
    "\t\t?organism obo:RO_0002162  ?species .\n",
    "\t\t?species a up:Taxon .\n",
    "\t\t?species up:commonName \"human\" .\n",
    "}\n",
    "\"\"\"]\n",
    "\n",
    "qs[7] = [\"\"\"\n",
    "PREFIX orth: <http://purl.org/net/orth#>\n",
    "PREFIX genex: <http://purl.org/genex#>\n",
    "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
    "PREFIX up: <http://purl.uniprot.org/core/>\n",
    "PREFIX efo: <http://www.ebi.ac.uk/efo/>\n",
    "\n",
    "SELECT DISTINCT ?anat ?anatName ?stage WHERE {\n",
    "\t?seq a orth:Gene .\n",
    "\t?seq genex:isExpressedIn ?condition.\n",
    "\t?seq rdfs:label \"APOC1\" .\n",
    "    ?condition a genex:ExpressionCondition.\n",
    "\t?condition genex:hasAnatomicalEntity ?anat .\n",
    "    ?anat a genex:AnatomicalEntity.\n",
    "\t?anat rdfs:label ?anatName .\n",
    "\t\t?condition genex:hasAnatomicalEntity obo:GO_0005575 .\n",
    "\t\t?condition genex:hasDevelopmentalStage ?stage .\n",
    "        ?stage a efo:EFO_0000399 .\n",
    "\t?stage rdfs:label \"post-juvenile\" .\n",
    "\t\t?seq orth:organism ?organism .\n",
    "\t\t?organism obo:RO_0002162  ?species .\n",
    "\t\t?species a up:Taxon .\n",
    "\t\t?species up:commonName \"human\" .\n",
    "}\n",
    "\"\"\"]\n",
    "\n",
    "qs[8] = [None] * 3\n",
    "\n",
    "qs[8][0] = \"\"\"\n",
    "PREFIX orth: <http://purl.org/net/orth#>\n",
    "PREFIX genex: <http://purl.org/genex#>\n",
    "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
    "PREFIX up: <http://purl.uniprot.org/core/>\n",
    "\n",
    "SELECT DISTINCT ?anat ?anatName ?score ?stage WHERE {\n",
    "\t?seq a orth:Gene .\n",
    "\t?expression a genex:Expression .\n",
    "\t?expression genex:hasExpressionCondition ?condition .\n",
    "\t?expression genex:hasExpressionLevel ?score .\n",
    "\t?expression genex:hasSequenceUnit ?seq .\n",
    "\t?seq rdfs:label \"APOC1\" .\n",
    "\t?condition a genex:ExpressionCondition.\n",
    "\t?condition genex:hasAnatomicalEntity ?anat .\n",
    "\t?condition genex:hasAnatomicalEntity obo:GO_0005575 .\n",
    "\t?anat a genex:AnatomicalEntity.\n",
    "\t?anat rdfs:label ?anatName .\n",
    "\t?condition genex:hasDevelopmentalStage ?stage .\n",
    "\t?stage rdfs:label \"post-juvenile\" .\n",
    "\t?condition genex:hasSex \"any\".\n",
    "\t?condition genex:hasStrain ?strain .\n",
    "\t?strain rdfs:label \"wild-type\" .\n",
    "\t?seq orth:organism ?organism .\n",
    "\t?organism obo:RO_0002162  ?species .\n",
    "\t?species a up:Taxon .\n",
    "\t?species up:commonName \"human\" .\n",
    "FILTER (?anat !=  obo:GO_0005575)\n",
    "} ORDER BY DESC(?score)\n",
    "\"\"\"\n",
    "\n",
    "qs[8][1] = \"\"\"\n",
    "PREFIX orth: <http://purl.org/net/orth#>\n",
    "PREFIX genex: <http://purl.org/genex#>\n",
    "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
    "PREFIX up: <http://purl.uniprot.org/core/>\n",
    "PREFIX up-taxon:<http://purl.uniprot.org/taxonomy/>\n",
    "PREFIX ensembl: <http://rdf.ebi.ac.uk/resource/ensembl/>\n",
    "PREFIX lscr: <http://purl.org/lscr#>\n",
    "PREFIX efo: <http://www.ebi.ac.uk/efo/>\n",
    "\n",
    "SELECT DISTINCT ?anat ?anatName ?score WHERE {\n",
    "\t?seq a orth:Gene .\n",
    "\t?expression a genex:Expression ;\n",
    "\t\tgenex:hasExpressionCondition ?condition ;\n",
    "\t\tgenex:hasExpressionLevel ?score ;\n",
    "\t\tgenex:hasSequenceUnit ?seq .\n",
    "\t?seq lscr:xrefEnsemblGene ensembl:ENSG00000130208 .\n",
    "\t?condition a genex:ExpressionCondition ;\n",
    "\t\tgenex:hasAnatomicalEntity ?anat .\n",
    "\t?anat a genex:AnatomicalEntity ;\n",
    "\t\trdfs:label ?anatName .\n",
    "\t?condition genex:hasDevelopmentalStage obo:UBERON_0000113 ;\n",
    "\t\tgenex:hasSex \"any\" ;\n",
    "\t\tgenex:hasStrain ?strain .\n",
    "    ?strain a efo:EFO_0005135.\n",
    "\t?strain rdfs:label \"wild-type\" .\n",
    "\t?seq orth:organism ?organism .\n",
    "\t?organism obo:RO_0002162  up-taxon:9606 .\n",
    "    FILTER (?anat !=  obo:GO_0005575)\n",
    "} ORDER BY DESC(?score)\n",
    "\"\"\"\n",
    "\n",
    "qs[8][2] = \"\"\"\n",
    "PREFIX orth: <http://purl.org/net/orth#>\n",
    "PREFIX genex: <http://purl.org/genex#>\n",
    "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
    "PREFIX up: <http://purl.uniprot.org/core/>\n",
    "PREFIX up-taxon:<http://purl.uniprot.org/taxonomy/>\n",
    "PREFIX ensembl: <http://rdf.ebi.ac.uk/resource/ensembl/>\n",
    "PREFIX lscr: <http://purl.org/lscr#>\n",
    "\n",
    "SELECT DISTINCT ?anat ?anatName ?score  {\n",
    "\t?seq a orth:Gene .\n",
    "\t?expression a genex:Expression .\n",
    "\t?expression genex:hasExpressionCondition ?condition .\n",
    "\t?expression genex:hasExpressionLevel ?score .\n",
    "\t?expression genex:hasSequenceUnit ?seq .\n",
    "\t?seq lscr:xrefEnsemblGene ensembl:ENSG00000130208 .\n",
    "    ?condition a genex:ExpressionCondition.\n",
    "\t?condition genex:hasAnatomicalEntity ?anat .\n",
    "    ?anat a genex:AnatomicalEntity.\n",
    "\t?anat rdfs:label ?anatName .\n",
    "\t?condition genex:hasDevelopmentalStage obo:UBERON_0000113 .\n",
    "\t?condition genex:hasSex \"any\".\n",
    "\t?condition genex:hasStrain ?strain .\n",
    "    ?strain a efo:EFO_0005135.\n",
    "\t?strain rdfs:label \"wild-type\" .\n",
    "FILTER (?anat !=  obo:GO_0005575)\n",
    "} ORDER BY DESC(?score)\n",
    "\"\"\"\n",
    "\n",
    "qs[9] = [\"\"\"\n",
    "PREFIX orth: <http://purl.org/net/orth#>\n",
    "PREFIX genex: <http://purl.org/genex#>\n",
    "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
    "PREFIX up: <http://purl.uniprot.org/core/>\n",
    "PREFIX efo: <http://www.ebi.ac.uk/efo/>\n",
    "\n",
    "SELECT DISTINCT ?anat ?cellType ?anatName ?cellTypeName ?score ?stage WHERE {\n",
    "\t?seq a orth:Gene .\n",
    "\t?expression a genex:Expression .\n",
    "\t?expression genex:hasExpressionCondition ?condition .\n",
    "\t?expression genex:hasExpressionLevel ?score .\n",
    "\t?expression genex:hasSequenceUnit ?seq .\n",
    "\t?seq rdfs:label \"APOC1\" .\n",
    "    ?condition a genex:ExpressionCondition.\n",
    "\t?condition genex:hasAnatomicalEntity ?anat .\n",
    "    ?anat a genex:AnatomicalEntity.\n",
    "\t?anat rdfs:label ?anatName .\n",
    "\t?condition genex:hasAnatomicalEntity ?cellType .\n",
    "\t?cellType rdfs:label ?cellTypeName .\n",
    "\t?condition genex:hasDevelopmentalStage ?stage .\n",
    "\t?stage rdfs:label \"post-juvenile\" .\n",
    "\t?condition genex:hasSex \"any\".\n",
    "\t?condition genex:hasStrain ?strain .\n",
    "    ?strain a efo:EFO_0005135.\n",
    "\t?strain rdfs:label \"wild-type\" .\n",
    "\t?seq orth:organism ?organism .\n",
    "\t?organism obo:RO_0002162  ?species .\n",
    "\t?species a up:Taxon .\n",
    "\t?species up:commonName \"human\" .\n",
    "    FILTER (?anat != obo:GO_0005575)\n",
    "    FILTER (?anat != ?cellType)\n",
    "} ORDER BY DESC(?score)\n",
    "\"\"\"]\n",
    "\n",
    "qs[10] = [\"\"\"\n",
    "PREFIX orth: <http://purl.org/net/orth#>\n",
    "PREFIX genex: <http://purl.org/genex#>\n",
    "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
    "PREFIX up: <http://purl.uniprot.org/core/>\n",
    "PREFIX efo: <http://www.ebi.ac.uk/efo/>\n",
    "\n",
    "SELECT DISTINCT ?anat ?cellType ?anatName ?cellTypeName ?score ?stage WHERE {\n",
    "\t?seq a orth:Gene .\n",
    "\t?expression a genex:Expression .\n",
    "\t?expression genex:hasExpressionCondition ?condition .\n",
    "\t?expression genex:hasExpressionLevel ?score .\n",
    "\t?expression genex:hasSequenceUnit ?seq .\n",
    "\t?seq rdfs:label \"APOC1\" .\n",
    "    ?condition a genex:ExpressionCondition.\n",
    "\t?condition genex:hasAnatomicalEntity ?anat .\n",
    "    ?anat a genex:AnatomicalEntity.\n",
    "\t?anat rdfs:label ?anatName .\n",
    "\t?condition genex:hasAnatomicalEntity ?cellType .\n",
    "\t?cellType rdfs:label ?cellTypeName .\n",
    "\t?condition genex:hasDevelopmentalStage ?stage .\n",
    "    ?stage rdf:type efo:EFO_0000399 .\n",
    "\t?stage rdfs:label \"post-juvenile\" .\n",
    "\t?seq orth:organism ?organism .\n",
    "\t?organism obo:RO_0002162  ?species .\n",
    "\t?species a up:Taxon .\n",
    "\t?species up:commonName \"human\" .\n",
    "    FILTER (?anat !=  obo:GO_0005575)\n",
    "    FILTER (?anat != ?cellType)\n",
    "} ORDER BY DESC(?score)\n",
    "\"\"\"]\n",
    "\n",
    "qs[11] = [None] * 2\n",
    "\n",
    "qs[11][1] = \"\"\"\n",
    "PREFIX dcterms: <http://purl.org/dc/terms/>\n",
    "PREFIX efo: <http://www.ebi.ac.uk/efo/EFO_0000399>\n",
    "\n",
    "SELECT DISTINCT ?stage ?stageName ?stageDescription WHERE {\n",
    "\t?stage rdf:type efo:EFO_0000399 .\n",
    "\t?stage rdfs:label ?stageName .\n",
    "\t?stage dcterms:description ?stageDescription .\n",
    "}\n",
    "\"\"\"\n",
    "\n",
    "qs[11][0] = \"\"\"\n",
    "PREFIX dcterms: <http://purl.org/dc/terms/>\n",
    "PREFIX efo: <http://www.ebi.ac.uk/efo/EFO_0000399>\n",
    "\n",
    "SELECT DISTINCT ?stage ?stageName ?stageDescription WHERE {\n",
    "        ?stage rdf:type efo:EFO_0000399 .\n",
    "\t\t?stage rdfs:label ?stageName .\n",
    "\t\t?stage dcterms:description ?stageDescription .\n",
    "\t\tFILTER (CONTAINS(?stageName,\"post-juvenile\"))\n",
    "}\n",
    "\"\"\"\n",
    "\n",
    "qs[12] = [None] * 3\n",
    "\n",
    "qs[12][0] = \"\"\"\n",
    "PREFIX orth: <http://purl.org/net/orth#>\n",
    "PREFIX genex: <http://purl.org/genex#>\n",
    "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
    "PREFIX up: <http://purl.uniprot.org/core/>\n",
    "PREFIX up-taxon:<http://purl.uniprot.org/taxonomy/>\n",
    "PREFIX lscr: <http://purl.org/lscr#>\n",
    "PREFIX efo: <http://www.ebi.ac.uk/efo/>\n",
    "\n",
    "SELECT DISTINCT ?anat ?anatName ?stageIRI ?score WHERE {\n",
    "\t?seq a orth:Gene .\n",
    "\t?expression a genex:Expression .\n",
    "\t?expression genex:hasExpressionCondition ?condition .\n",
    "\t?expression genex:hasExpressionLevel ?score .\n",
    "\t?expression genex:hasSequenceUnit ?seq .\n",
    "\t?seq lscr:xrefNCBIGene <https://www.ncbi.nlm.nih.gov/gene/118230125> .\n",
    "\t?condition genex:hasAnatomicalEntity ?anat .\n",
    "    ?anat a genex:AnatomicalEntity.\n",
    "\t?anat rdfs:label ?anatName .\n",
    "    ?condition a genex:ExpressionCondition.\n",
    "\t?condition genex:hasDevelopmentalStage ?stageIRI .\n",
    "    ?stageIRI rdf:type efo:EFO_0000399 .\n",
    "\t?condition genex:hasSex \"any\".\n",
    "\t?condition genex:hasStrain ?strain .\n",
    "    ?strain a efo:EFO_0005135.\n",
    "\t?strain rdfs:label \"wild-type\" .\n",
    "    FILTER (?anat != obo:GO_0005575)\n",
    "} ORDER BY DESC(?score)\n",
    "\"\"\"\n",
    "\n",
    "qs[12][1] = \"\"\"\n",
    "PREFIX orth: <http://purl.org/net/orth#>\n",
    "PREFIX genex: <http://purl.org/genex#>\n",
    "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
    "PREFIX up: <http://purl.uniprot.org/core/>\n",
    "PREFIX up-taxon:<http://purl.uniprot.org/taxonomy/>\n",
    "PREFIX lscr: <http://purl.org/lscr#>\n",
    "PREFIX dcterms: <http://purl.org/dc/terms/>\n",
    "\n",
    "SELECT DISTINCT ?anat ?anatName ?stageIRI ?score WHERE {\n",
    "\t?seq a orth:Gene .\n",
    "\t?expression a genex:Expression .\n",
    "\t?expression genex:hasExpressionCondition ?condition .\n",
    "\t?expression genex:hasExpressionLevel ?score .\n",
    "\t?expression genex:hasSequenceUnit ?seq .\n",
    "\t?seq dcterms:identifier \"118230125\" .\n",
    "    ?condition a genex:ExpressionCondition.\n",
    "\t?condition genex:hasAnatomicalEntity ?anat .\n",
    "\t?anat rdfs:label ?anatName .\n",
    "\t?condition genex:hasDevelopmentalStage ?stageIRI .\n",
    "\t?condition genex:hasSex \"any\".\n",
    "\t?condition genex:hasStrain ?strain .\n",
    "    ?strain a efo:EFO_0005135.\n",
    "\t?strain rdfs:label \"wild-type\" .\n",
    "FILTER (?anat !=  obo:GO_0005575)\n",
    "} ORDER BY DESC(?score)\n",
    "\"\"\"\n",
    "\n",
    "qs[13] = [\"\"\"\n",
    "PREFIX orth: <http://purl.org/net/orth#>\n",
    "PREFIX genex: <http://purl.org/genex#>\n",
    "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
    "PREFIX up-protein:<http://purl.uniprot.org/uniprot/>\n",
    "PREFIX lscr: <http://purl.org/lscr#>\n",
    "\n",
    "SELECT DISTINCT ?anat ?anatName WHERE {\n",
    "\t?seq a orth:Gene .\n",
    "\t?seq genex:isExpressedIn ?anat .\n",
    "\t?seq lscr:xrefUniprot up-protein:P02654 .\n",
    "\t?anat a genex:AnatomicalEntity .\n",
    "\t?anat rdfs:label ?anatName .\n",
    "}\n",
    "\"\"\"]\n",
    "\n",
    "qs[14] = [\"\"\"\n",
    "PREFIX orth: <http://purl.org/net/orth#>\n",
    "PREFIX lscr: <http://purl.org/lscr#>\n",
    "PREFIX dcterms: <http://purl.org/dc/terms/>\n",
    "\n",
    "SELECT DISTINCT ?symbol ?description ?id\n",
    "?links ?organism ?uniprot ?ensembl ?ncbi WHERE {\n",
    "    ?seq a orth:Gene .\n",
    "    ?seq rdfs:label ?symbol .\n",
    "    ?seq rdfs:seeAlso ?links .\n",
    "    ?seq dcterms:description ?description .\n",
    "    ?seq dcterms:identifier ?id .\n",
    "    ?seq orth:organism ?organism .\n",
    "    OPTIONAL{?seq lscr:xrefUniprot ?uniprot .}\n",
    "    OPTIONAL{?seq lscr:xrefEnsemblGene ?ensembl .}\n",
    "    OPTIONAL{?seq lscr:xrefNCBIGene ?ncbi .}\n",
    "    FILTER (?id = \"ENSG00000130208\")\n",
    "}\n",
    "\"\"\"]\n",
    "\n",
    "qs[15] = [\"\"\"\n",
    "PREFIX orth: <http://purl.org/net/orth#>\n",
    "PREFIX genex: <http://purl.org/genex#>\n",
    "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
    "PREFIX up: <http://purl.uniprot.org/core/>\n",
    "\n",
    "SELECT DISTINCT ?anat ?anatName WHERE {\n",
    "\t?seq a orth:Gene .\n",
    "\t?seq genex:isAbsentIn ?anat.\n",
    "\t?seq rdfs:label \"APOC1\" .\n",
    "\t?anat a genex:AnatomicalEntity .\n",
    "\t?anat rdfs:label ?anatName .\n",
    "\t?seq orth:organism ?organism .\n",
    "\t?organism obo:RO_0002162  ?species .\n",
    "\t?species a up:Taxon .\n",
    "\t?species up:scientificName \"Homo sapiens\" .\n",
    "}\n",
    "\"\"\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "eed8c0b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# also store NL questions corresponding to the queries\n",
    "\n",
    "nlqs = [None] *16\n",
    "\n",
    "nlqs[0] = \"\"\n",
    "nlqs[1] = \"What are the species present in Bgee?\"\n",
    "nlqs[2] = \"What are the species present in Bgee and their scientific and common names?\"\n",
    "nlqs[3] = \"What are the anatomical entities where the APOC1 gene is expressed?\"\n",
    "nlqs[4] = \"What are the anatomical entities where the APOC1 Homo sapiens gene is expressed?\"\n",
    "nlqs[5] = \"What are the anatomical entities where the APOC1 gene is expressed independently of the developmental stage, sex, strain and cell type?\"\n",
    "nlqs[6] = \"What are the anatomical entities where the human gene APOC1 is expressed in the post-juvenile stage?\"\n",
    "nlqs[7] = \"What are the anatomical entities where the human gene APOC1 is expressed in the post-juvenile stage?\"\n",
    "nlqs[8] = \"What are the anatomical entities where the human gene APOC1 is expressed in the post-juvenile stage along with its expression score independently of the strain, sex, and cell type?\"\n",
    "nlqs[9] = \"What are the anatomical entities including cell types, if any, where the human gene APOC1 is expressed at the post-juvenile stage along with its expression score independently of the strain and sex?\"\n",
    "nlqs[10] = \"What are the anatomical entities including cell types, if any, where the human gene APOC1 is expressed at the post-juvenile stage along with its expression score independently of the strain and sex?\"\n",
    "nlqs[11] = \"What are the developmental stages present in Bgee?\"\n",
    "nlqs[12] = \"What are the anatomical entities where the eel gene apoc1 is expressed along with its expression score independently of the strain, sex, and cell type?\"\n",
    "nlqs[13] = \"What are the anatomical entities where the P02654 gene is expressed? Note that P02654 is a UniProtKB identifier of the APOC1 human gene.\"\n",
    "nlqs[14] = \"What is all the metadata related to the ENSG00000130208 gene, where ENSG00000130208 is the identifier of the APOC1 human gene. \"\n",
    "nlqs[15] = \"What are the anatomical entities where the APOC1 Homo sapiens gene is not expressed, that is where is APOC1 absent ?\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5aa0300c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "‚úÖ Query 1 got 52 results\n",
      "‚úÖ Query 2 got 52 results\n",
      "‚úÖ Query 3 got 584 results\n",
      "‚úÖ Query 4 got 526 results\n",
      "‚úÖ Query 5 got 585 results\n",
      "‚úÖ Query 6 got 457 results\n",
      "‚úÖ Query 7 got 457 results\n",
      "‚úÖ Query 8 got 456 results\n",
      "‚úÖ Query 9 got 456 results\n",
      "‚úÖ Query 10 got 456 results\n",
      "‚úÖ Query 11 got 0 results\n",
      "‚úÖ Query 12 got 99 results\n",
      "‚úÖ Query 13 got 526 results\n",
      "‚úÖ Query 14 got 16 results\n",
      "‚úÖ Query 15 got 6 results\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "from rdflib import RDF, RDFS, BNode, Graph, Literal, Namespace\n",
    "from SPARQLWrapper import JSON, SPARQLWrapper\n",
    "\n",
    "queries_dir = \"../../sparql-examples/bgee\"\n",
    "# queries_dir = \"../data/bgee_queries\"\n",
    "os.makedirs(queries_dir, exist_ok=True)\n",
    "\n",
    "EX = Namespace(\"https://www.bgee.org/sparql/.well-known/sparql-examples/\")\n",
    "UP = Namespace(\"http://purl.uniprot.org/core/\")\n",
    "SH = Namespace(\"http://www.w3.org/ns/shacl#\")\n",
    "\n",
    "sparql_endpoint = SPARQLWrapper(\"https://www.bgee.org/sparql/\")\n",
    "\n",
    "for i in range(1, len(nlqs)):\n",
    "    g = Graph()\n",
    "    g.bind(\"ex\", EX)\n",
    "    g.bind(\"up\", UP)\n",
    "    g.bind(\"sh\", SH)\n",
    "    g.bind(\"rdf\", RDF)\n",
    "    g.bind(\"rdfs\", RDFS)\n",
    "\n",
    "    query_resource = EX[str(i)]\n",
    "    g.add((query_resource, RDF.type, SH.SPARQLSelectExecutable))\n",
    "    g.add((query_resource, RDF.type, SH.SPARQLExecutable))\n",
    "    g.add((query_resource, SH.prefixes, BNode(\"sparql_examples_prefixes\")))\n",
    "\n",
    "    g.add((query_resource, RDFS.comment, Literal(nlqs[i])))\n",
    "    # TODO: what do we do for 8, 11 and 12 that have multiple queries for 1 description?\n",
    "    query = qs[i][0]\n",
    "    g.add((query_resource, SH.select, Literal(query)))\n",
    "\n",
    "    # Check query gets results\n",
    "    try:\n",
    "        sparql_endpoint.setQuery(query)\n",
    "        sparql_endpoint.setReturnFormat(JSON)\n",
    "        results = sparql_endpoint.query().convert()\n",
    "        n_res = len(results[\"results\"][\"bindings\"])\n",
    "        print(f\"‚úÖ Query {i} got {n_res} results\")\n",
    "    except Exception as e:\n",
    "        print(f\"‚ùå Query {i} failed: {e}\")\n",
    "\n",
    "    # NOTE: Dirty hack to get the BNode formatted as we want\n",
    "    ttl_str = g.serialize(format=\"turtle\").replace(\"sh:prefixes [ ] ;\", \"sh:prefixes _:sparql_examples_prefixes ;\")\n",
    "    with open(f\"{queries_dir}/{i}.ttl\", \"w\") as f:\n",
    "        f.write(ttl_str)\n",
    "\n",
    "    # g.serialize(f\"../data/bgee_queries/{i}.ttl\", format=\"turtle\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
