{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Parsing a SPARQL query to extract basic graph patterns\n",
    "\n",
    "Function to parse a SPARQL query and extract the BGP it contains to load them in a RDFLib graph, with support for federated queries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Subject ?pathVar6 in endpoint https://sparql.omabrowser.org/sparql does not support the predicate http://purl.obolibrary.org/obo/RO_0002163 according to the VOID description. Correct predicate might be one of the following: http://purl.obolibrary.org/obo/RO_0002162 (we inferred this variable might be of the type http://purl.org/net/orth#Organism)\n"
     ]
    }
   ],
   "source": [
    "import json\n",
    "from typing import Tuple\n",
    "from unittest import result\n",
    "from SPARQLWrapper import SPARQLWrapper\n",
    "\n",
    "from rdflib import Namespace, URIRef, Variable, ConjunctiveGraph\n",
    "from rdflib.plugins.sparql.parser import parseQuery\n",
    "from rdflib.plugins.sparql.algebra import translateQuery\n",
    "from rdflib.plugins.sparql.sparql import Query\n",
    "from rdflib.paths import Path, SequencePath, MulPath\n",
    "from expasy_chat.utils import validate_sparql_with_void\n",
    "\n",
    "\n",
    "# sparql_query = \"\"\"\n",
    "# PREFIX up: <http://purl.uniprot.org/core/>\n",
    "# PREFIX rh: <http://rdf.rhea-db.org/>\n",
    "# SELECT ?uniprotCount ?rhea ?accession ?equation\n",
    "# WHERE {\n",
    "#   SERVICE <https://sparql.uniprot.org/sparql> {\n",
    "#       SELECT ?rhea (count(?uniprot) as ?uniprotCount) {\n",
    "#           ?uniprot up:annotation ?rhea .\n",
    "#           ?uniprot up:toast ?toooast .\n",
    "#       }\n",
    "#       GROUP BY ?rhea\n",
    "#   }\n",
    "#   ?rhea rh:accession ?accession .\n",
    "#   ?rhea rh:equation ?equation .\n",
    "# }\n",
    "# \"\"\"\n",
    "\n",
    "# # Hard query\n",
    "sparql_query = \"\"\"PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
    "PREFIX up:<http://purl.uniprot.org/core/>\n",
    "PREFIX taxon:<http://purl.uniprot.org/taxonomy/>\n",
    "PREFIX rdfs:<http://www.w3.org/2000/01/rdf-schema#>\n",
    "PREFIX orth:<http://purl.org/net/orth#>\n",
    "PREFIX dcterms:<http://purl.org/dc/terms/>\n",
    "PREFIX obo:<http://purl.obolibrary.org/obo/>\n",
    "PREFIX lscr:<http://purl.org/lscr#>\n",
    "PREFIX genex:<http://purl.org/genex#>\n",
    "PREFIX sio: <http://semanticscience.org/resource/>\n",
    "SELECT DISTINCT ?diseaseLabel ?humanProtein ?hgncSymbol ?orthologRatProtein ?orthologRatGene\n",
    "WHERE {\n",
    "    SERVICE <https://sparql.uniprot.org/sparql> {\n",
    "        SELECT DISTINCT * WHERE {\n",
    "            ?humanProtein a up:Protein ;\n",
    "                up:organism/up:scientificName 'Homo sapiens' ;\n",
    "                up:annotation ?annotation ;\n",
    "                rdfs:seeAlso ?hgnc .\n",
    "            ?hgnc up:database <http://purl.uniprot.org/database/HGNC> ;\n",
    "                rdfs:label ?hgncSymbol . # comment\n",
    "            ?annotation a up:Disease_Annotation ;\n",
    "                up:disease ?disease .\n",
    "            ?disease a up:Disease ;\n",
    "                rdfs:label ?diseaseLabel . # skos:prefLabel\n",
    "            FILTER CONTAINS(?diseaseLabel, \"cancer\")\n",
    "        }\n",
    "    }\n",
    "    SERVICE <https://sparql.omabrowser.org/sparql/> {\n",
    "        SELECT ?humanProtein ?orthologRatProtein ?orthologRatGene WHERE {\n",
    "            ?humanProteinOma a orth:Protein ;\n",
    "                lscr:xrefUniprot ?humanProtein .\n",
    "            ?orthologRatProtein a orth:Protein ;\n",
    "                sio:SIO_010078 ?orthologRatGene ; # 79\n",
    "                orth:organism/obo:RO_0002162/up:scientificNam 'Rattus norvegicus' .\n",
    "            ?cluster a orth:OrthologsCluster .\n",
    "            ?cluster orth:hasHomologousMember ?node1 .\n",
    "            ?cluster orth:hasHomologousMember ?node2 .\n",
    "            ?node1 orth:hasHomologousMember* ?humanProteinOma .\n",
    "            ?node2 orth:hasHomologousMember* ?orthologRatProtein .\n",
    "            FILTER(?node1 != ?node2)\n",
    "        }\n",
    "    }\n",
    "    SERVICE <https://www.bgee.org/sparql/> {\n",
    "        ?orthologRatGene genex:isExpressedIn ?anatEntity ;\n",
    "            orth:organism ?ratOrganism .\n",
    "        ?anatEntity rdfs:label 'brain' .\n",
    "        ?ratOrganism obo:RO_0002162 taxon:10116 .\n",
    "    }\n",
    "}\n",
    "\"\"\"\n",
    "\n",
    "try:\n",
    "    validate_sparql_with_void(sparql_query, \"https://sparql.uniprot.org/sparql\")\n",
    "except Exception as e:\n",
    "    print(e)\n",
    "\n",
    "\n",
    "# sparql_query = \"\"\"PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
    "# PREFIX up: <http://purl.uniprot.org/core/>\n",
    "# PREFIX genex: <http://purl.org/genex#>\n",
    "# PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
    "# PREFIX orth: <http://purl.org/net/orth#>\n",
    "# PREFIX dcterms: <http://purl.org/dc/terms/>\n",
    "# PREFIX sio: <http://semanticscience.org/resource/>\n",
    "\n",
    "# SELECT DISTINCT ?flyEnsemblGene ?orthologTaxon ?orthologEnsemblGene ?orthologOmaLink WHERE {\n",
    "# \t{\n",
    "#         SELECT DISTINCT ?gene ?flyEnsemblGene {\n",
    "#         ?gene a orth:Gene ;\n",
    "#             genex:isExpressedIn/rdfs:label 'eye' ;\n",
    "#             orth:organism/obo:RO_0002162 ?taxon ;\n",
    "#             dcterms:identifier ?flyEnsemblGene .\n",
    "#         ?taxon up:commonName 'fruit fly' .\n",
    "#         } LIMIT 100\n",
    "#     }\n",
    "#     SERVICE <https://sparql.omabrowser.org/sparql> {\n",
    "#         ?protein2 a orth:Protein .\n",
    "#         ?protein1 a orth:Protein .\n",
    "#         ?clusterPrimates a orth:OrthologsCluster .\n",
    "#         ?cluster a orth:OrthologsCluster ;\n",
    "#             orth:hasHomologousMember ?node1 ;\n",
    "#             orth:hasHomologousMember ?node2 .\n",
    "#         ?node1 orth:hasHomologousMember* ?protein1 .\n",
    "#         ?node2 orth:hasHomologousMember* ?clusterPrimates .\n",
    "#         ?clusterPrimates orth:hasHomologousMember* ?protein2 .\n",
    "#         ?protein1 sio:SIO_010079 ?gene . # is encoded by\n",
    "#         ?protein2 rdfs:seeAlso ?orthologOmaLink ;\n",
    "#             orth:organism/obo:RO_0002162 ?orthologTaxonUri ;\n",
    "#             sio:SIO_010079 ?orthologGene . # is encoded by\n",
    "#         ?clusterPrimates orth:hasTaxonomicRange ?taxRange .\n",
    "#         ?taxRange orth:taxRange 'Primates' .\n",
    "#         FILTER ( ?node1 != ?node2 )\n",
    "#     }\n",
    "#     ?orthologTaxonUri up:commonName ?orthologTaxon .\n",
    "#     ?orthologGene dcterms:identifier ?orthologEnsemblGene .\n",
    "# }\"\"\"\n",
    "\n",
    "# try:\n",
    "#     validate_sparql_with_void(sparql_query, \"https://www.bgee.org/sparql\")\n",
    "# except Exception as e:\n",
    "#     print(e)\n",
    "\n",
    "\n",
    "# # Failing query:\n",
    "# sparql_query = \"\"\"PREFIX up: <http://purl.uniprot.org/core/>\n",
    "# PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
    "# PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\n",
    "# SELECT ?protein ?sequence\n",
    "# WHERE {\n",
    "#     ?protein a up:Protein .\n",
    "#     ?protein up:reviewed true .\n",
    "#     ?protein up:organism taxon:10090 .\n",
    "#     ?protein up:matureProtein ?matureProtein .\n",
    "#     ?matureProtein rdf:value ?sequence .\n",
    "#     FILTER (STRSTARTS(?sequence, \"G\")) .\n",
    "# }\"\"\"\n",
    "\n",
    "\n",
    "# print(\"\\n\".join(error_msgs))\n",
    "# print(len(error_msgs))\n",
    "# qres = g.query(\"\"\"SELECT ?s ?p ?o WHERE {\n",
    "#     ?s ?p ?o\n",
    "# } LIMIT 10\"\"\")\n",
    "# for row in qres:\n",
    "#     print(f\"{row.s} knows {row.o}\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "expasy-chat",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
