{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generating embeddings for SIB SPARQL queries\n",
    "\n",
    "Using `fastembed-gpu` library, FlagEmbedding large 1.5 embedding model, and Qdrant vectorstore\n",
    "\n",
    "Example code for loading the large Babel concepts synonyms dataset to Qdrant: https://github.com/vemonet/concept-resolver/blob/main/src/babel_load.py\n",
    "\n",
    "Questions:\n",
    "* Which info can I get from the UniProt endpoint?\n",
    "* Give me an example to access cross references from the UniProt SPARQL endpoint to all the databases available in the endpoint"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 60 queries for https://sparql.uniprot.org/sparql/\n",
      "Found 19 queries for https://www.bgee.org/sparql/\n",
      "[{'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': 'UniProt: Was any UniProt entry integrated on the 9th of January 2013', 'query': \"PREFIX up: <http://purl.uniprot.org/core/>\\nASK\\nWHERE\\n{\\n\\t?protein a up:Protein .\\n\\t?protein up:created '2013-01-09'^^xsd:date\\n}\"}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': \"UniProt: Construct new triples of the type 'HumanProtein' from all human UniProt entries\", 'query': 'PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX taxon: <http://purl.uniprot.org/taxonomy/>\\nCONSTRUCT\\n{\\n\\t?protein a up:HumanProtein .\\n}\\nWHERE\\n{\\n\\t?protein a up:Protein .\\n\\t?protein up:organism taxon:9606\\n}'}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': 'UniProt: Select all triples that relate to the EMBL CDS entry AA089367.1: ', 'query': 'DESCRIBE <http://purl.uniprot.org/embl-cds/AAO89367.1>\\n'}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': 'UniProt: Select all triples that relate to the taxon that describes Homo sapiens in the named graph for taxonomy', 'query': 'PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\\nDESCRIBE taxon:9606 \\nFROM <http://sparql.uniprot.org/taxonomy>\\n'}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': 'UniProt: Select all taxa from the UniProt taxonomy', 'query': 'PREFIX up: <http://purl.uniprot.org/core/>\\nSELECT ?taxon\\nFROM <http://sparql.uniprot.org/taxonomy>\\nWHERE\\n{\\n    ?taxon a up:Taxon .\\n}'}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': 'UniProt: Select all human UniProt entries with a sequence variant that leads to a tyrosine to phenylalanine substitution', 'query': \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX taxon: <http://purl.uniprot.org/taxonomy/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\\nPREFIX faldo: <http://biohackathon.org/resource/faldo#>\\nSELECT ?protein ?annotation ?begin ?text\\nWHERE\\n{\\n        ?protein a up:Protein ;\\n            up:organism taxon:9606 ; \\n            up:annotation ?annotation .\\n        ?annotation a up:Natural_Variant_Annotation ;\\n            rdfs:comment ?text ;\\n            up:substitution ?substitution ;\\n            up:range/faldo:begin\\n                [ faldo:position ?begin ;\\n                  faldo:reference ?sequence ] .\\n        ?sequence rdf:value ?value .\\n        BIND (substr(?value, ?begin, 1) as ?original) .\\n        FILTER(?original = 'Y' && ?substitution = 'F') .\\n} \"}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': \"UniProt: Select all UniProt entries with annotated transmembrane regions and the regions' begin and end coordinates on the canonical sequence\", 'query': 'PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX faldo: <http://biohackathon.org/resource/faldo#>\\nSELECT ?protein ?begin ?end\\nWHERE \\n{\\n\\t?protein a up:Protein .\\n\\t?protein up:annotation ?annotation .\\n\\t?annotation a up:Transmembrane_Annotation .\\n\\t?annotation up:range ?range .\\n\\t?range faldo:begin/faldo:position ?begin .\\n\\t?range faldo:end/faldo:position ?end\\n}'}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': 'UniProt: Select all UniProt entries that were integrated on the 30th of November 2010', 'query': \"PREFIX up: <http://purl.uniprot.org/core/>\\nSELECT ?protein\\nWHERE\\n{\\n\\t?protein a up:Protein . \\n\\t?protein up:created '2010-11-30'^^xsd:date\\n} \"}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': 'UniProt: Select the average number of cross-references to the PDB database of UniProt entries that have at least one cross-reference to the PDB database', 'query': 'PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nSELECT (AVG(?linksToPdbPerEntry) AS ?avgLinksToPdbPerEntry)\\nWHERE\\n{\\n\\tSELECT ?protein (COUNT(DISTINCT ?db) AS ?linksToPdbPerEntry)\\n\\tWHERE\\n\\t{\\n\\t\\t?protein a up:Protein .\\n\\t\\t?protein rdfs:seeAlso ?db .\\n\\t\\t?db up:database <http://purl.uniprot.org/database/PDB> .\\n\\t}\\n\\tGROUP BY ?protein ORDER BY DESC(?linksToPdbPerEntry)\\n}'}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': 'UniProt: Select the number of UniProt entries for each of the EC (Enzyme Commission) top level categories', 'query': 'PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX ec: <http://purl.uniprot.org/enzyme/>\\nSELECT ?ecClass (COUNT(?protein) as ?size)\\nWHERE\\n{\\n    VALUES (?ecClass) {(ec:1.-.-.-) (ec:2.-.-.-) (ec:3.-.-.-) (ec:4.-.-.-) (ec:5.-.-.-) (ec:6.-.-.-) (ec:7.-.-.-)} .\\n    ?protein ( up:enzyme | up:domain/up:enzyme | up:component/up:enzyme ) ?enzyme .\\n    # Enzyme subclasses are materialized, do not use rdfs:subClassOf+\\n    ?enzyme rdfs:subClassOf ?ecClass .\\n}\\nGROUP BY ?ecClass ORDER BY ?ecClass\\n'}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': 'UniProt: Find all Natural Variant Annotations if associated via an evidence tag to an article with a pubmed identifier', 'query': 'PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\\nSELECT \\n    ?accession\\n    ?annotation_acc \\n    ?pubmed\\nWHERE\\n{\\n        ?protein a up:Protein ;\\n            up:annotation ?annotation .\\n        ?annotation a up:Natural_Variant_Annotation .\\n        ?linkToEvidence rdf:object ?annotation ;\\n                        up:attribution ?attribution .\\n        ?attribution up:source ?source .\\n        ?source a up:Journal_Citation .\\n  BIND(SUBSTR(STR(?protein),33) AS ?accession)\\n  BIND(IF(CONTAINS(STR(?annotation), \"#SIP\"), SUBSTR(STR(?annotation),33), SUBSTR(STR(?annotation),36))AS?annotation_acc)\\n  BIND(SUBSTR(STR(?source),35) AS ?pubmed)\\n}\\n'}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': 'UniProt: Select all bacterial taxa and their scientific name from the UniProt taxonomy', 'query': 'PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX taxon: <http://purl.uniprot.org/taxonomy/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nSELECT ?taxon ?name\\nWHERE\\n{\\n    ?taxon a up:Taxon .\\n    ?taxon up:scientificName ?name .\\n    # Taxon subclasses are materialized, do not use rdfs:subClassOf+\\n    ?taxon rdfs:subClassOf taxon:2 .\\n}'}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': 'UniProt: Find how often an article in pubmed was used in an evidence tag in a human protein (ordered by most used to least)', 'query': 'PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX taxon: <http://purl.uniprot.org/taxonomy/>\\nPREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\\nSELECT \\n    ?source \\n    (COUNT(?attribution) AS ?attribitions)\\nWHERE\\n{\\n        ?protein a up:Protein ;\\n            up:organism taxon:9606 ;\\n            up:annotation ?annotation .\\n        ?linkToEvidence rdf:object ?annotation ;\\n                        up:attribution ?attribution .\\n        ?attribution up:source ?source .\\n        ?source a up:Journal_Citation .\\n} GROUP BY ?source ORDER BY DESC(COUNT(?attribution))\\n'}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': 'UniProt: Find where disease related proteins are known to be located in the cell', 'query': 'PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX skos: <http://www.w3.org/2004/02/skos/core#>\\n\\nSELECT \\n    ?protein \\n    ?disease \\n    ?location_inside_cell \\n    ?cellcmpt\\nWHERE\\n{\\n    ?protein up:annotation ?diseaseAnnotation , ?subcellAnnotation .\\n    ?diseaseAnnotation up:disease/skos:prefLabel ?disease .\\n    ?subcellAnnotation up:locatedIn/up:cellularComponent ?cellcmpt .\\n    ?cellcmpt skos:prefLabel ?location_inside_cell .\\n}\\n'}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': 'UniProt: For two accessions find the GO term labels and group them into GO process,function and component', 'query': 'PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\n\\nSELECT \\n    (CONCAT(SUBSTR(STR(?protein), 33)) AS ?uniprot)\\n    (GROUP_CONCAT(?celtype; separator=\";\") AS ?celtypes)\\n    (GROUP_CONCAT(?biotype; separator=\";\") AS ?biotypes)\\n    (GROUP_CONCAT(?moltype; separator=\";\") AS ?moltypes)\\nWHERE\\n{\\n    VALUES (?ac) {(\"Q6GZX4\") (\"Q96375\")}\\n    BIND (IRI(CONCAT(\"http://purl.uniprot.org/uniprot/\",?ac)) AS ?protein)\\n    ?protein a up:Protein .\\n    ?protein up:classifiedWith ?goTerm .\\n    #Determine if the type is biological_process\\n    OPTIONAL {\\n        ?goTerm rdfs:subClassOf <http://purl.obolibrary.org/obo/GO_0008150>.\\n        ?goTerm rdfs:label ?biotype .\\n    }\\n    #Determine if the type is cellular_component\\n    OPTIONAL {\\n        ?goTerm rdfs:subClassOf <http://purl.obolibrary.org/obo/GO_0005575>.\\n        ?goTerm rdfs:label ?celtype .\\n    }\\n    #Determine if the type is molecular_function\\n    OPTIONAL {\\n        ?goTerm rdfs:subClassOf <http://purl.obolibrary.org/obo/GO_0003674> .\\n        ?goTerm rdfs:label ?moltype .\\n    }\\n    #Filter out the uniprot keywords\\n    FILTER(bound(?biotype) || bound(?celtype) || bound(?moltype))\\n} GROUP BY ?protein\\n'}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': 'UniProt: Number of reviewed entries (Swiss-Prot) that are related to kinase activity', 'query': 'PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX taxon: <http://purl.uniprot.org/taxonomy/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\\nPREFIX GO: <http://purl.obolibrary.org/obo/GO_>\\nSELECT\\n    (COUNT(DISTINCT(?protein)) AS ?pc)\\nWHERE\\n{   \\n    ?protein rdf:type up:Protein ;\\n        up:reviewed true  ;\\n        up:organism taxon:9606 ;\\n        up:classifiedWith|(up:classifiedWith/rdfs:subClassOf) GO:0016301 .\\n}'}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': 'UniProt: Find the release number of the uniprot data that is currently being queried', 'query': 'SELECT\\n    ?version    \\nFROM <http://sparql.uniprot.org/.well-known/void>\\nWHERE\\n{   \\n    [] <http://purl.org/pav/version> ?version\\n}'}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': \"UniProt: Find any uniprot entry which has a name 'HLA class I histocompatibility antigen, B-73 alpha chain'\", 'query': 'PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nSELECT ?protein ?anyKindOfName\\nWHERE\\n{\\n\\t\\t?protein a up:Protein .\\n\\t\\t?protein (up:recommendedName|up:alternativeName) ?structuredName .\\n\\t\\t?structuredName ?anyKindOfName  \"HLA class I histocompatibility antigen, B-73 alpha chain\" .\\n\\t\\t?anyKindOfName rdfs:subPropertyOf up:structuredNameType .\\n}'}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': \"UniProt: Find any uniprot entry, or an uniprot entries domain or component which has a name 'HLA class I histocompatibility antigen, B-73 alpha chain'\", 'query': 'PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nSELECT ?protein ?anyKindOfName\\nWHERE\\n{\\n\\t\\t?protein a up:Protein .\\n\\t\\t?protein (up:recommendedName|up:alternativeName)|((up:domain|up:component)/(up:recommendedName|up:alternativeName)) ?structuredName .\\n\\t\\t?structuredName ?anyKindOfName  \"HLA class I histocompatibility antigen, B-73 alpha chain\" .\\n\\t\\t?anyKindOfName rdfs:subPropertyOf up:structuredNameType .\\n}'}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': \"UniProt: Find all names associated with uniprot entry P05067, and if the name is associated with the entry it's domains or its components\", 'query': 'PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nSELECT\\n  ?protein\\n  ?anyKindOfName \\n  ?names \\n  ?partType\\nWHERE\\n{\\n  BIND(<http://purl.uniprot.org/uniprot/P05067> AS ?protein)\\n  ?protein a up:Protein .\\n  {\\n    ?protein (up:recommendedName|up:alternativeName) ?structuredName .\\n  }\\n    UNION\\n  {\\n    VALUES(?partType){(up:domain) (up:component)}\\n    ?protein ?partType ?part .\\n    ?part (up:recommendedName|up:alternativeName) ?structuredName .\\n  }\\n  ?structuredName ?anyKindOfName  ?names .\\n  ?anyKindOfName rdfs:subPropertyOf up:structuredNameType .\\n}'}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': 'UniProt: Get the list of uniprot entries for the chromosome of proteome UP000000625', 'query': 'PREFIX up: <http://purl.uniprot.org/core/>\\nSELECT \\n  ?protein\\n  ?proteome \\nWHERE\\n{\\n  ?protein a up:Protein ;\\n           up:reviewed true ;\\n           up:proteome ?proteome .\\n  VALUES (?proteome) {(<http://purl.uniprot.org/proteomes/UP000000625#Chromosome>)}\\n}'}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': 'UniProt: Use ALLIE a service for Abbreviation / Long Form in Japanese and English to search in UniProt using Japanese.', 'query': 'PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nSELECT\\n    ?protein ?englishLabelStr\\nWHERE {\\n    SERVICE<http://data.allie.dbcls.jp/sparql>{\\n            ?x rdfs:label \"アミロイド前駆体タンパク質\"@ja ;\\n                rdfs:label ?englishLabel .\\n            FILTER(lang(?englishLabel) = \"en\")\\n    }\\n    BIND (STR(?englishLabel) AS ?englishLabelStr)\\n    ?protein a up:Protein .\\n    {\\n        ?protein (up:recommendedName|up:alternativeName) ?structuredName .\\n    }\\n    UNION\\n    {\\n          VALUES(?partType){(up:domain) (up:component)}\\n                  ?protein ?partType ?part .\\n          ?part (up:recommendedName|up:alternativeName) ?structuredName .\\n    }\\n    ?structuredName ?anyKindOfName  ?englishLabelStr .\\n    ?anyKindOfName rdfs:subPropertyOf up:structuredNameType .\\n}\\n'}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': 'UniProt: Select all UniProt entries, and their organism and amino acid sequences (including isoforms), for E. coli K12 and all its strains', 'query': 'PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX taxon: <http://purl.uniprot.org/taxonomy/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\\nSELECT ?protein ?organism ?isoform ?sequence\\nWHERE\\n{\\n    ?protein a up:Protein .\\n    ?protein up:organism ?organism .\\n    # Taxon subclasses are materialized, do not use rdfs:subClassOf+\\n    ?organism rdfs:subClassOf taxon:83333 .\\n    ?protein up:sequence ?isoform .\\n    ?isoform rdf:value ?sequence .\\n}'}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': 'UniProt: Find UniProt entries with merged loci in Bordetella avium', 'query': \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX taxon: <http://purl.uniprot.org/taxonomy/>\\nSELECT\\n  ?protein \\n  (GROUP_CONCAT(?locusName; separator=',') AS ?locusNames)\\nWHERE \\n{ \\n  ?protein a up:Protein ;\\n    up:organism taxon:360910 ;\\n    up:encodedBy ?gene .\\n  ?gene up:locusName ?locusName .\\n} \\nGROUP BY ?protein \\nHAVING (COUNT(?locusName) > 1)\\n\"}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': 'UniProt: Find UniParc records whose sequence point to the most database entries', 'query': 'PREFIX up: <http://purl.uniprot.org/core/>\\nSELECT ?sequence ?entries\\nWHERE\\n{\\n    SELECT \\n        ?sequence \\n        (COUNT(?entry) AS ?entries)\\n    WHERE\\n    {\\n        GRAPH <http://sparql.uniprot.org/uniparc> {\\n            ?sequence up:sequenceFor ?entry .\\n        }\\n    } GROUP BY ?sequence\\n} ORDER BY DESC(?entries)   \\n'}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': 'UniProt: Find UniProtKB entries with more than 1 Topological domain annotation', 'query': 'PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\\nSELECT \\n    ?protein \\n    (GROUP_CONCAT(?comment; separator=\", \") AS ?comments)\\nWHERE\\n{\\n    ?protein a up:Protein ;\\n            up:annotation ?annotation . \\n    ?annotation rdf:type up:Topological_Domain_Annotation ;\\n            rdfs:comment ?comment .\\n} \\nGROUP BY ?protein \\nHAVING (COUNT(?annotation) > 1)\\n'}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': 'UniProt: Find longest comment text associated with a UniProtKB Natural Variant Annotation', 'query': 'PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nSELECT \\n    ?annotation ?comment\\nWHERE {\\n    ?annotation a up:Natural_Variant_Annotation ;\\n        rdfs:comment ?comment . \\n} \\nORDER BY DESC(STRLEN(?comment))\\n'}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': 'UniProt: Find the co-occurence count of Topological Domain comment text in UniProtKB entries', 'query': \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\\nSELECT \\n    ?comment1 \\n    ?comment2 \\n    (COUNT(?comment1) AS ?count1)\\nWHERE\\n{\\n    ?protein a up:Protein ;\\n               up:annotation ?annotation1 , \\n                             ?annotation2 . \\n    ?annotation1 rdf:type up:Topological_Domain_Annotation ;\\n        rdfs:comment ?rawComment1 .\\n    ?annotation2 rdf:type up:Topological_Domain_Annotation ;\\n        rdfs:comment ?rawComment2 . \\n    BIND(IF(contains(?rawComment1, ';'), \\n            STRBEFORE(?rawComment1,';'), \\n            ?rawComment1) AS ?comment1)\\n    BIND(IF(contains(?rawComment2, ';'), \\n            STRBEFORE(?rawComment2,';'), \\n            ?rawComment2) AS ?comment2)\\n    FILTER(?annotation1 != ?annotation2)\\n} \\nGROUP BY ?comment1 ?comment2 \\nORDER BY DESC(COUNT(?comment1))\\n\"}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': 'UniProt: Find the similar proteins for UniProtKB entry P05067 sorted by UniRef cluser identity', 'query': 'PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX uniprotkb: <http://purl.uniprot.org/uniprot/>\\nSELECT \\n    ?similar ?identity\\nFROM <http://sparql.uniprot.org/uniref>\\nFROM <http://sparql.uniprot.org/uniprot>\\nWHERE\\n{\\n    BIND (uniprotkb:P05607 AS ?protein)\\n    ?cluster up:member ?member ;\\n             up:member/up:sequenceFor ?protein;\\n             up:identity ?identity .\\n    ?member up:sequenceFor ?similar .\\n    FILTER(!sameTerm(?similar, ?protein))\\n} \\nORDER BY DESC(?identity)\\n'}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': 'UniProt: Find the orthologous proteins for UniProtKB entry P05067 using the OrthoDB database', 'query': 'PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX uniprotkb: <http://purl.uniprot.org/uniprot/>\\nPREFIX skos: <http://www.w3.org/2004/02/skos/core#>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\\nPREFIX orthodb: <http://purl.orthodb.org/>\\nSELECT \\n  ?protein \\n  ?orthoGroup\\n  ?scientificName \\n  ?functionComment \\n  ?prefferedGeneName \\n  ((STRLEN(?value) - ?medianLength) as ?deviationFromMedianLength)\\nWHERE\\n{\\n  uniprotkb:P05067 a up:Protein ;\\n        up:organism/up:scientificName ?scientificName ;\\n        rdfs:seeAlso ?orthoGroup ;\\n        up:encodedBy/skos:prefLabel ?prefferedGeneName ;\\n             up:sequence/rdf:value ?value .\\n  OPTIONAL {\\n     ?protein up:annotation ?functionAnnotation .\\n     ?functionAnnotation a up:Function_Annotation ;\\n                         rdfs:comment ?functionComment .\\n  }\\n  SERVICE<http://sparql.orthodb.org/sparql>{\\n    ?orthoGroup orthodb:ogMedianProteinLength ?medianLength .\\n    ?orthoGroup orthodb:hasMember ?xref .\\n    ?xref orthodb:xref/orthodb:xrefResource uniprotkb:P05067 .\\n  }\\n}\\n'}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': 'UniProt: Find the human protein which contains an Epitope VSTQ, where T is a phosporylated threonine', 'query': 'PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX taxon: <http://purl.uniprot.org/taxonomy/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\\nPREFIX faldo: <http://biohackathon.org/resource/faldo#>\\nSELECT \\n  ?protein \\n  ?comment\\n  ?begin\\n  ?end \\nWHERE\\n{\\n  ?protein a up:Protein ;\\n    up:organism taxon:9606 ; \\n    up:sequence ?sequence ;\\n    up:annotation ?annotation .\\n  ?annotation a up:Modified_Residue_Annotation ;\\n    rdfs:comment ?comment ;\\n    up:range ?range .\\n  ?range \\n    faldo:begin [ faldo:position ?begin ; faldo:reference ?sequence ] ;\\n    faldo:end [ faldo:position ?end ; faldo:reference ?sequence ] .\\n  ?sequence rdf:value ?aaSequence .\\n  FILTER (SUBSTR(?aaSequence, ?begin -2 , 4) = \"VSTQ\")     \\n  FILTER (CONTAINS(?comment, \"Phosphothreonine\"))\\n}\\n'}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': 'UniProt: For the human entry P05067 (Amyloid-beta precursor protein) find the gene start ends in WikiData', 'query': 'PREFIX wdt: <http://www.wikidata.org/prop/direct/>\\nPREFIX uniprotkb: <http://purl.uniprot.org/uniprot/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX ps: <http://www.wikidata.org/prop/statement/>\\nPREFIX pq: <http://www.wikidata.org/prop/qualifier/>\\nPREFIX p: <http://www.wikidata.org/prop/>\\n\\nSELECT \\n\\t?protein \\n\\t?begin\\n\\t?end\\n\\t?chromosome\\n\\t?assembly\\nWHERE {\\n    {\\n        BIND(uniprotkb:P05067 AS ?proteinIRI)\\n        BIND (SUBSTR(STR(?proteinIRI), STRLEN(STR(uniprotkb:))+1) AS ?protein)\\n    }\\n    SERVICE <https://query.wikidata.org/sparql> {\\n        ?wp wdt:P352 ?protein ;\\n            wdt:P702 ?wg . \\n        ?wg p:P644   ?wgss .\\n        ?wgss ps:P644        ?begin ;\\n          pq:P1057/wdt:P1813 ?chromosome ;\\n          pq:P659/rdfs:label ?assembly .\\n        ?wg p:P645 ?wgse .\\n        ?wgse ps:P645        ?end ;\\n          pq:P1057/wdt:P1813 ?chromosome ;\\n          pq:P659/rdfs:label ?assembly .\\n        FILTER(lang(?assembly) = \"en\")\\n  } \\n}\\n'}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': 'UniProt: Retrieve entries and Catalytic activities in the reviewed (Swiss-Prot) section that have experimental evidences,', 'query': 'PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\\nSELECT  \\n  ?protein\\n  ?rhea \\nWHERE {\\n  # ECO 269 is experimental evidence\\n  BIND (<http://purl.obolibrary.org/obo/ECO_0000269> as ?evidence)\\n  GRAPH <http://sparql.uniprot.org/uniprot> {\\n    ?protein up:reviewed true ;\\n      up:annotation ?a ;\\n      up:attribution ?attribution  .\\n\\n    ?a a up:Catalytic_Activity_Annotation ;\\n      up:catalyticActivity ?ca .\\n    ?ca up:catalyzedReaction ?rhea .\\n  \\n    [] rdf:subject ?a ;\\n      rdf:predicate up:catalyticActivity ;\\n      rdf:object ?ca ;\\n      up:attribution ?attribution .\\n\\n    ?attribution up:evidence ?evidence .\\n  }\\n}\\n'}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': \"UniProt: Select the UniProt entry with the mnemonic 'A4_HUMAN'\", 'query': \"PREFIX up: <http://purl.uniprot.org/core/>\\nSELECT ?protein\\nWHERE\\n{\\n    ?protein a up:Protein .\\n    ?protein up:mnemonic 'A4_HUMAN'\\n}\"}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': 'UniProt: Retrieve human enzymes that metabolize sphingolipids and are annotated in ChEMBL', 'query': 'PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX taxon: <http://purl.uniprot.org/taxonomy/>\\nPREFIX rh: <http://rdf.rhea-db.org/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX CHEBI: <http://purl.obolibrary.org/obo/CHEBI_>\\n\\nSELECT DISTINCT \\n  ?protein \\n  ?chemblEntry\\nWHERE {\\n  SERVICE <https://sparql.rhea-db.org/sparql> {\\n    ?rhea rdfs:subClassOf rh:Reaction ;\\n      rh:side/rh:contains/rh:compound/rh:chebi/rdfs:subClassOf+ CHEBI:26739 .\\n  }\\n  ?ca up:catalyzedReaction ?rhea .\\n  ?protein up:annotation/up:catalyticActivity ?ca ;\\n    up:organism taxon:9606 ;\\n    rdfs:seeAlso ?chemblEntry .\\n  ?chemblEntry up:database <http://purl.uniprot.org/database/ChEMBL> .\\n} \\n'}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': 'UniProt: Retrieve entries with sequences that are composed of fragments', 'query': 'PREFIX up: <http://purl.uniprot.org/core/>\\n\\nSELECT DISTINCT \\n  ?protein\\nWHERE {\\n  ?protein a up:Protein ;\\n    up:sequence ?sequence .\\n  MINUS { ?sequence up:fragment [] }\\n} '}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': 'UniProt: Connect patents cited in UniProtKB with those in the patent database at EPO via publication number.', 'query': \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX skos: <http://www.w3.org/2004/02/skos/core#>\\nPREFIX patent: <http://data.epo.org/linked-data/def/patent/>\\n\\nSELECT ?citation ?patent ?application ?applicationNo\\nWHERE\\n{\\n  ?citation a up:Patent_Citation ;\\n    skos:exactMatch ?patent .\\n  FILTER(CONTAINS(STR(?patent), 'EP'))\\n  BIND(SUBSTR(STR(?patent), 35) AS ?applicationNo)\\n  SERVICE<https://data.epo.org/linked-data/query>{\\n    ?application patent:publicationNumber ?applicationNo\\n  }\\n}\"}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': 'UniProt: Connect patents cited in UniProtKB with those in the patent database at EPO via publication number, whose grant date is more than twenty years in the past.', 'query': 'PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX skos: <http://www.w3.org/2004/02/skos/core#>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\\nPREFIX patent: <http://data.epo.org/linked-data/def/patent/>\\n\\nPREFIX rdf:<http://www.w3.org/1999/02/22-rdf-syntax-ns#> \\nPREFIX rdfs:<http://www.w3.org/2000/01/rdf-schema#> \\nPREFIX skos:<http://www.w3.org/2004/02/skos/core#> \\nPREFIX up:<http://purl.uniprot.org/core/> \\nprefix patent: <http://data.epo.org/linked-data/def/patent/>\\nSELECT ?grantDate ?patent ?application ?applicationNo\\nWHERE\\n{\\n    ?citation a up:Patent_Citation ;\\n  skos:exactMatch ?patent .\\n  BIND(SUBSTR(STR(?patent), 35) AS ?applicationNo)\\n  BIND(SUBSTR(STR(?patent), 33, 2) AS ?countryCode)\\n  SERVICE<https://data.epo.org/linked-data/query>{\\n    ?publication patent:publicationNumber ?applicationNo ;\\n                 patent:application ?application . \\n    ?application patent:grantDate ?grantDate .\\n  }\\n  BIND((year(now()) - 20) AS ?thisYearMinusTwenty)\\n  BIND(year(?grantDate) AS ?grantYear)\\n  FILTER(?grantYear < ?thisYearMinusTwenty)\\n} ORDER BY ?grantYear'}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': 'UniProt: Find the Rhea and Interpro combinations in UniProtKB entries.', 'query': 'PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nSELECT \\n    ?interpro\\n\\t?rhea\\nFROM <http://sparql.uniprot.org/uniprot>\\nWHERE \\n{\\n  ?protein up:reviewed true .\\n  ?protein up:annotation ?annotation .\\n  ?annotation up:catalyticActivity ?rhea .\\n  ?protein rdfs:seeAlso ?interpro .\\n  ?interpro up:database <http://purl.uniprot.org/database/InterPro> .\\n} ORDER BY ?rhea'}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': 'UniProt: Retrieve drugs that target human enzymes involved in sterol metabolism (federated query with WikiData and Rhea).', 'query': \"PREFIX wdt: <http://www.wikidata.org/prop/direct/>\\nPREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX uniprotkb: <http://purl.uniprot.org/uniprot/>\\nPREFIX taxon: <http://purl.uniprot.org/taxonomy/>\\nPREFIX rh: <http://rdf.rhea-db.org/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX owl: <http://www.w3.org/2002/07/owl#>\\nPREFIX chebihash: <http://purl.obolibrary.org/obo/chebi#>\\nPREFIX CHEBI: <http://purl.obolibrary.org/obo/CHEBI_>\\nSELECT \\n    DISTINCT\\n        ?protein\\n        ?proteinFullName\\n        ?wikiChemical\\n\\t\\t?wikiChemicalLabel\\n        ?medicalConditionTreatedLabel\\nWHERE\\n  {\\n  # ChEBI: retrieve members of the ChEBI class ChEBI:15889 (sterol)\\n  # Rhea: retrieve the reactions involving these ChEBI as participants\\n  SERVICE <https://sparql.rhea-db.org/sparql> {\\n    ?reaction rdfs:subClassOf  rh:Reaction ;\\n\\t      rh:status        rh:Approved ;\\n\\t      rh:side          ?reactionSide .\\n    ?reactionSide\\n\\t      rh:contains      ?participant .\\n    ?participant rh:compound  ?compound\\n    { \\n      ?compound  rh:chebi  ?chebi .\\n      ?chebi (rdfs:subClassOf)+ CHEBI:15889\\n    } UNION {\\n      ?compound  rh:chebi           ?chebi .\\n      ?chebi2   rdfs:subClassOf     ?chebiRestriction .\\n      ?chebiRestriction\\n\\t\\ta           owl:Restriction ;\\n\\t\\towl:onProperty      chebihash:has_major_microspecies_at_pH_7_3 ;\\n\\t\\towl:someValuesFrom  ?chebi .\\n      ?chebi2 (rdfs:subClassOf)+ CHEBI:15889\\n    }\\n  }\\n  # UniProt: retrieve the human (taxid:9606) enzymes catalyzing these Rhea reactions \\n  ?ca       up:catalyzedReaction  ?reaction .\\n  ?a        up:catalyticActivity  ?ca .\\n  ?proteinIRI  up:annotation         ?a ;\\n\\t    up:organism           taxon:9606 ;\\n\\t    up:recommendedName    ?proteinRecName .\\n  ?proteinRecName\\n\\t    up:fullName           ?proteinFullName .\\n  # Find drugs in wikidata that interact with the UniProt Proteins\\n  BIND (SUBSTR(STR(?proteinIRI), STRLEN(STR(uniprotkb:))+1) AS ?protein)\\n  SERVICE<https://query.wikidata.org/sparql> {\\n    ?wp wdt:P352  ?protein . \\n    ?wikiChemical wdt:P129  ?wp . # Physically interacts with\\n    ?wikiChemical rdfs:label ?wikiChemicalLabel .\\n    ?wikiChemical wdt:P2175 ?wmc . # Medical conndition treated\\n    ?wmc rdfs:label ?medicalConditionTreatedLabel .\\n    FILTER(lang(?medicalConditionTreatedLabel) = 'en')\\n    FILTER(lang(?wikiChemicalLabel) = 'en')\\n  }\\n}   \"}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': \"UniProt: Retrieve images of 'Anas' (Ducks) from the European Environmental Agency databases (federated query).\", 'query': \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX taxon: <http://purl.uniprot.org/taxonomy/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX eunisSpecies: <http://eunis.eea.europa.eu/rdf/species-schema.rdf#>\\nSELECT \\n    ?taxon\\n    ?ncbiTaxid\\n    ?eunisTaxon\\n    ?eunisName \\n    ?image\\nWHERE\\n{\\n    GRAPH <http://sparql.uniprot.org/taxonomy>\\n    {\\n        ?taxon a up:Taxon .\\n        # Taxon subclasses are materialized, do not use rdfs:subClassOf+\\n        ?taxon rdfs:subClassOf taxon:8835 .\\n        BIND(strafter(str(?taxon), 'onomy/') AS ?ncbiTaxid)\\n    }\\n    SERVICE <https://semantic.eea.europa.eu/sparql>\\n    {\\n        ?eunisTaxon a eunisSpecies:SpeciesSynonym ;\\n           eunisSpecies:binomialName ?eunisName ;\\n           eunisSpecies:sameSpeciesNCBI ?ncbiTaxid ;\\n           <http://xmlns.com/foaf/0.1/depiction> ?image .\\n    }\\n}\\n\"}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': 'UniProt: Retrieve the UniProt proteins, their catalyzed Rhea reactions, their encoding genes (Ensembl) and the anatomic entities where the genes are expressed (UBERON anatomic entites from Bgee expression data resource).', 'query': 'PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX taxon: <http://purl.uniprot.org/taxonomy/>\\nPREFIX rh: <http://rdf.rhea-db.org/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX lscr: <http://purl.org/lscr#>\\nPREFIX genex: <http://purl.org/genex#>\\nPREFIX CHEBI: <http://purl.obolibrary.org/obo/CHEBI_>\\nSELECT\\n        DISTINCT \\n            ?protein\\n            ?ensemblGene\\n            ?reaction\\n            ?anatomicEntityLabel\\n            ?anatomicEntity\\nWHERE {\\n  # federated query to Rhea enadpoint\\n  {\\n    SELECT DISTINCT ?reaction WHERE {\\n      SERVICE <https://sparql.rhea-db.org/sparql> {\\n        ?reaction rdfs:subClassOf rh:Reaction .\\n        ?reaction rh:equation ?reactionEquation .\\n        ?reaction rh:side ?reactionSide .\\n        ?reactionSide rh:contains ?participant .\\n        ?participant rh:compound ?compound .\\n        # compound constraint (CHEBI:16113 == cholesterol)\\n        ?compound rh:chebi CHEBI:16113 .\\n      }\\n    }\\n  }\\n  # taxonomy constraint (taxon:9606 == Homo sapiens)\\n  ?protein up:organism taxon:9606 .\\n  ?protein up:annotation ?a .\\n  ?a a up:Catalytic_Activity_Annotation .\\n  ?a up:catalyticActivity ?ca .\\n  ?ca up:catalyzedReaction ?reaction .\\n  ?protein rdfs:seeAlso / up:transcribedFrom ?ensemblGene .\\n\\n  # federated query to Bgee (expression data)\\n  SERVICE <http://biosoda.expasy.org/rdf4j-server/repositories/bgeelight> {\\n    ?gene genex:isExpressedIn ?anatomicEntity .\\n    ?gene lscr:xrefEnsemblGene ?ensemblGene .\\n    ?anatomicEntity rdfs:label ?anatomicEntityLabel .\\n  }\\n}'}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': 'UniProt: Select a mapping of UniProt to PDB entries using the UniProt cross-references to the PDB database', 'query': 'PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nSELECT ?protein ?db\\nWHERE\\n{\\n    ?protein a up:Protein .\\n    ?protein rdfs:seeAlso ?db .\\n    ?db up:database <http://purl.uniprot.org/database/PDB>\\n}'}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': 'UniProt: Retrieve drugs that target human enzymes involved in sterol metabolism (federated query with Rhea and ChEMBL via IDSM/Elixir czech republic).', 'query': 'PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX taxon: <http://purl.uniprot.org/taxonomy/>\\nPREFIX skos: <http://www.w3.org/2004/02/skos/core#>\\nPREFIX rh: <http://rdf.rhea-db.org/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX owl: <http://www.w3.org/2002/07/owl#>\\nPREFIX chebihash: <http://purl.obolibrary.org/obo/chebi#>\\nPREFIX cco: <http://rdf.ebi.ac.uk/terms/chembl#>\\nPREFIX CHEBI: <http://purl.obolibrary.org/obo/CHEBI_>\\nSELECT\\n        DISTINCT  \\n            ?protein \\n            ?proteinFullName \\n            ?activityType \\n            ?standardActivityValue \\n            ?standardActivityUnit \\n            ?chemblMolecule \\n            ?chemlbMoleculePrefLabel\\nWHERE\\n  {\\n  # ChEBI: retrieve members of the ChEBI class ChEBI:15889 (sterol)\\n  # Rhea: retrieve the reactions involving these ChEBI as participants\\n  SERVICE <https://sparql.rhea-db.org/sparql> {\\n    ?reaction rdfs:subClassOf  rh:Reaction ;\\n\\t      rh:status        rh:Approved ;\\n\\t      rh:side          ?reactionSide .\\n    ?reactionSide\\n\\t      rh:contains      ?participant .\\n    ?participant rh:compound  ?compound\\n    { \\n      ?compound  rh:chebi  ?chebi .\\n      ?chebi (rdfs:subClassOf)+ CHEBI:15889\\n    } UNION {\\n      ?compound  rh:chebi           ?chebi .\\n      ?chebi2   rdfs:subClassOf     ?chebiRestriction .\\n      ?chebiRestriction\\n\\t\\ta           owl:Restriction ;\\n\\t\\towl:onProperty      chebihash:has_major_microspecies_at_pH_7_3 ;\\n\\t\\towl:someValuesFrom  ?chebi .\\n      ?chebi2 (rdfs:subClassOf)+ CHEBI:15889\\n    }\\n  }\\n  # UniProt: retrieve the human (taxid:9606) enzymes catalyzing these Rhea reactions \\n  ?ca       up:catalyzedReaction  ?reaction .\\n  ?a        up:catalyticActivity  ?ca .\\n  ?protein  up:annotation         ?a ;\\n\\t    up:organism           taxon:9606 ;\\n\\t    up:recommendedName    ?proteinRecName .\\n  ?proteinRecName\\n\\t    up:fullName           ?proteinFullName .\\n  # Find drugs in wikidata that interact with the UniProt Proteins\\n  # ChEMBL: retrieve the corresponding targets and with drugs in clinical phase 4\\n  # Via https://idsm.elixir-czech.cz/sparql/\\n  SERVICE <https://idsm.elixir-czech.cz/sparql/endpoint/idsm> { \\n    ?activity a cco:Activity ;\\n      cco:hasMolecule ?chemblMolecule ;\\n      cco:hasAssay ?assay ;\\n      cco:standardType ?activityType ;\\n      cco:standardValue ?standardActivityValue ;\\n      cco:standardUnits ?standardActivityUnit .\\n    ?chemblMolecule cco:highestDevelopmentPhase ?highestDevelopmentPhase ;\\n      rdfs:label ?chemblMoleculeLabel ;\\n      skos:prefLabel ?chemlbMoleculePrefLabel .\\n    FILTER (?highestDevelopmentPhase > 3)\\n    ?assay cco:hasTarget ?target .\\n    ?target cco:hasTargetComponent/cco:targetCmptXref ?protein .\\n    ?protein a cco:UniprotRef .\\n  }\\n}'}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': 'UniProt: Find mouse homologs in OMABrowser of human enzymes that catalyze reactions involving Sterols (CHEBI:15889). Federating with Rhea-DB and OMABrowser.', 'query': 'PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX taxon: <http://purl.uniprot.org/taxonomy/>\\nPREFIX rh: <http://rdf.rhea-db.org/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX lscr: <http://purl.org/lscr#>\\nPREFIX CHEBI: <http://purl.obolibrary.org/obo/CHEBI_>\\nSELECT\\n  DISTINCT\\n    ?chebi\\n    ?reaction\\n    ?humanProtein\\n    ?mouseProtein\\n    ?cluster \\nWHERE {\\n  SERVICE <https://sparql.rhea-db.org/sparql> {\\n    ?reaction rdfs:subClassOf rh:Reaction .\\n    ?reaction rh:side/rh:contains/rh:compound ?compound .\\n    ?compound rh:chebi ?chebi .\\n    ?chebi rdfs:subClassOf* CHEBI:15889\\n  }\\n\\n  ?humanProtein up:organism taxon:9606 .\\n  ?humanProtein up:annotation ?a .\\n  ?a a up:Catalytic_Activity_Annotation .\\n  ?a up:catalyticActivity ?ca .\\n  ?ca up:catalyzedReaction ?reaction .\\n\\n  SERVICE  <https://sparql.omabrowser.org/sparql> {\\n    ?cluster a orth:ParalogsCluster .\\n    ?cluster orth:hasHomologousMember ?node1 , ?node2 .\\n    ?node1 orth:hasHomologousMember* ?orthoProtein1 .\\n    ?node2 orth:hasHomologousMember* ?orthoProtein2 .\\n    ?orthoProtein1 lscr:xrefUniprot ?mouseProtein .\\n    ?orthoProtein2 lscr:xrefUniprot ?humanProtein .\\n    # inTaxon mouse\\n    ?orthoProtein1 orth:organism/<http://purl.obolibrary.org/obo/RO_0002162> taxon:10090 . \\n  }\\n}'}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': 'UniProt: Proteins with annotated binding sites for ligands similar to heme', 'query': 'PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX sachem: <http://bioinfo.uochb.cas.cz/rdf/v1.0/sachem#>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nSELECT \\n  ?protein\\n  ?mnemonic\\n  ?proteinName\\n  ?ligandSimilarityScore\\n  ?ligand\\nWHERE {\\n  SERVICE <https://idsm.elixir-czech.cz/sparql/endpoint/chebi> {\\n    ?ssc sachem:compound ?ligand; \\n      sachem:score ?ligandSimilarityScore ;\\n      sachem:similaritySearch ?sss .\\n        # Smiles of Heme\\n    ?sss    sachem:query \"CC1=C(CCC([O-])=O)C2=[N+]3C1=Cc1c(C)c(C=C)c4C=C5C(C)=C(C=C)C6=[N+]5[Fe-]3(n14)n1c(=C6)c(C)c(CCC([O-])=O)c1=C2\";\\n      sachem:cutoff \"8e-1\"^^xsd:double ;\\n      sachem:aromaticityMode sachem:aromaticityDetect ;\\n      sachem:similarityRadius 1 ;\\n      sachem:tautomerMode sachem:ignoreTautomers .\\n  }\\n  ?protein up:mnemonic ?mnemonic ;\\n    up:recommendedName/up:fullName ?proteinName ;\\n    up:annotation ?annotation .\\n  ?annotation a up:Binding_Site_Annotation ;\\n      up:ligand/rdfs:subClassOf ?ligand .\\n}\\nORDER BY DESC(?ligandSimilarityScore)'}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': 'UniProt: Number of proteins with annotated binding sites for metals or metal sulfur clusters (and experimental evidence for the binding)', 'query': 'PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nPREFIX CHEBI: <http://purl.obolibrary.org/obo/CHEBI_>\\nSELECT\\n  ?ligand\\n  ?ligandName \\n  (COUNT(DISTINCT ?protein) as ?entries)\\nWHERE {\\n   ?protein up:annotation ?annotation .\\n   \\n   VALUES ?evs { obo:ECO_0000269 obo:ECO_0007744 } .\\n   VALUES ?chebids { CHEBI:25213 CHEBI:25214 } .\\n   ?st rdf:subject ?protein ; \\n       rdf:predicate up:annotation ; \\n       rdf:object ?annotation ;\\n       up:attribution/up:evidence ?evs .\\n\\n   ?annotation up:ligand/rdfs:subClassOf ?ligand .\\n   ?ligand rdfs:subClassOf+ ?chebids ;\\n     rdfs:label ?ligandName .\\n}\\nGROUP BY ?ligand ?ligandName\\nORDER BY DESC(?entries)'}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': 'UniProt: Select Enzymes that have Ligands that have a known allosteric effect', 'query': 'PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nSELECT\\n  ?protein\\n  ?ligandName \\n  ?ligandNote\\n  ?chebi\\nWHERE {\\n   ?protein up:annotation ?annotation .\\n   ?annotation a up:Binding_Site_Annotation . \\n   ?annotation up:ligand ?ligand .\\n   ?ligand rdfs:comment ?ligandNote ;\\n     rdfs:subClassOf ?chebi ;\\n     rdfs:label ?ligandName .\\n   FILTER(REGEX(?ligandNote, \"allosteric\", \"i\"))\\n}\\n'}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': 'UniProt: Map PDB identifiers plus chains to UniProt', 'query': \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nSELECT\\n  ?pdbId ?chain ?pdbChain ?uniprot\\nWHERE\\n{\\n  # A space separated list of pairs of PDB identifiers and the chain code.\\n  VALUES(?pdbId ?pdbChain) { ('6VXC' 'A') ('1BG3' 'B') }\\n\\n  # Make an IRI out of the pdbId\\n  BIND(iri(concat('http://rdf.wwpdb.org/pdb/', ?pdbId)) AS ?pdb)\\n\\n  # Map to UniProt entries\\n  ?uniprot rdfs:seeAlso ?pdb .\\n  ?pdb up:database <http://purl.uniprot.org/database/PDB> ;\\n       up:chainSequenceMapping ?chainSm .\\n  ?chainSm up:chain ?chainsPlusRange .\\n\\n  # Extract the list of chains from the text representation.\\n  BIND(STRBEFORE(?chainsPlusRange, '=') AS ?chain)\\n\\n  # Filter those that match.\\n  FILTER(CONTAINS(?chain, ?pdbChain))\\n}\"}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': 'UniProt: Map UniProt to HGNC identifiers and Symbols', 'query': \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX uniprotkb: <http://purl.uniprot.org/uniprot/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nSELECT\\n  ?uniprot\\n  ?hgnc\\n  ?hgncSymbol\\nWHERE\\n{\\n  # A space separated list of UniProt primary accessions.\\n  VALUES (?acc) {('P05067') ('P00750')}\\n  BIND(iri(concat(str(uniprotkb:), ?acc)) AS ?uniprot)\\n  ?uniprot rdfs:seeAlso ?hgnc .\\n  ?hgnc up:database <http://purl.uniprot.org/database/HGNC> ;\\n       rdfs:comment ?hgncSymbol .\\n}\"}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': 'UniProt: Count all isoforms for a given proteome', 'query': 'PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX taxon: <http://purl.uniprot.org/taxonomy/>\\nPREFIX proteome:<http://purl.uniprot.org/proteomes/>\\nSELECT\\n  (COUNT(DISTINCT ?sequence) AS ?allIsoforms)\\nWHERE\\n{\\n  ?protein up:reviewed true .\\n  ?protein up:organism taxon:9606 .\\n  ?protein up:sequence ?sequence .\\n  ?protein up:proteome/^skos:narrower proteome:UP000005640 .\\n}'}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': \"UniProt: Select all cross-references to external databases of the category '3D structure databases' of UniProt entries that are classified with the keyword 'Acetoin biosynthesis (KW-0005)'\", 'query': \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX keywords: <http://purl.uniprot.org/keywords/>\\nSELECT DISTINCT ?link\\nWHERE\\n{\\n    ?protein a up:Protein . \\n    ?protein up:classifiedWith keywords:5 .\\n    ?protein rdfs:seeAlso ?link .\\n    ?link up:database ?db .\\n    ?db up:category '3D structure databases'\\n}\"}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': \"UniProt: Select reviewed UniProt entries (Swiss-Prot), and their recommended protein name, that have a preferred gene name that contains the text 'DNA'\", 'query': \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX skos: <http://www.w3.org/2004/02/skos/core#>\\nSELECT ?protein ?name\\nWHERE\\n{\\n        ?protein a up:Protein .\\n        ?protein up:reviewed true .\\n        ?protein up:recommendedName ?recommended .\\n        ?recommended up:fullName ?name .\\n        ?protein up:encodedBy ?gene .\\n        ?gene skos:prefLabel ?text .\\n        FILTER CONTAINS(?text, 'DNA')\\n}\\n\"}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': 'UniProt: Select the preferred gene name and disease annotation of all human UniProt entries that are known to be involved in a disease', 'query': 'PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX taxon: <http://purl.uniprot.org/taxonomy/>\\nPREFIX skos: <http://www.w3.org/2004/02/skos/core#>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nSELECT ?name ?text\\nWHERE\\n{\\n        ?protein a up:Protein . \\n        ?protein up:organism taxon:9606 .\\n        ?protein up:encodedBy ?gene . \\n        ?gene skos:prefLabel ?name .\\n        ?protein up:annotation ?annotation .\\n        ?annotation a up:Disease_Annotation .\\n        ?annotation rdfs:comment ?text\\n}'}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': \"UniProt: Select all human UniProt entries with a sequence variant that leads to a 'loss of function'\", 'query': \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX taxon: <http://purl.uniprot.org/taxonomy/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nSELECT ?protein ?text\\nWHERE\\n{ \\n        ?protein a up:Protein .\\n        ?protein up:organism taxon:9606 . \\n        ?protein up:annotation ?annotation .\\n        ?annotation a up:Natural_Variant_Annotation . \\n        ?annotation rdfs:comment ?text .\\n        FILTER (CONTAINS(?text, 'loss of function'))\\n}\"}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': 'UniProt: Find UniProt entries with a transmembrane region, with an Alanine in the 15 Aminoacid region preceding the transmembrane', 'query': \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\\nPREFIX faldo: <http://biohackathon.org/resource/faldo#>\\n\\nSELECT ?protein ?from ?interestingRegion\\nWHERE\\n{\\n  ?protein up:annotation ?annotation .\\n  ?annotation a up:Transmembrane_Annotation .\\n  # Get the coordinates of the Transmembrane\\n  ?annotation up:range ?range .\\n  ?range faldo:begin ?beginI .\\n  ?beginI faldo:position ?begin .\\n  ?beginI faldo:reference ?sequence .\\n  # The aas will have the specific IUPAC aminoacids\\n  ?sequence rdf:value ?aas .\\n  # We calculate the start by substracting 10\\n  BIND(?begin - 10 AS ?tenBeforeBegin)\\n  # Can't start before the sequence starts or we might miss some results\\n  BIND(IF(?tenBeforeBegin < 1, 0, ?tenBeforeBegin) AS ?from)\\n  # Substring the IUPAC aminoacids\\n  BIND(SUBSTR(?aas, ?from, 15) AS ?interestingRegion)\\n  # The interestingRegion needds to contain an Alanine\\n  FILTER(CONTAINS(?interestingRegion, 'A'))\\n}\\n\"}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': 'UniProt: Retrieve glycosylation sites and glycans on human enzymes (federated with glyconnect)', 'query': 'PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX taxon: <http://purl.uniprot.org/taxonomy/>\\nPREFIX glycan: <http://purl.jp/bio/12/glyco/glycan#>\\nPREFIX foaf: <http://xmlns.com/foaf/0.1/>\\nPREFIX faldo: <http://biohackathon.org/resource/faldo#>\\nSELECT\\n        DISTINCT\\n            ?protein ?glycosite ?glycostructure ?glycoimage\\nWHERE{\\n  ?protein up:annotation ?annotation .\\n  ?protein up:organism taxon:9606 .\\n  ?annotation a up:Catalytic_Activity_Annotation .\\n  ?protein up:sequence ?isoform .\\n   \\n  SERVICE <http://sparql.glyconnect.org/sparql> {\\n    ?glycosite faldo:reference ?isoform .\\n    ?glycosite faldo:position ?position .\\n    ?specificglycosite faldo:location ?glycosite .\\n    ?glycoprotein glycan:glycosylated_at ?specificglycosite .\\n    ?glycostructure glycan:glycosylates_at ?specificglycosite .\\n    ?glycostructure foaf:depiction ?glycoimage .\\n  }\\n}'}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': 'UniProt: Where are the human genes encoding enzymes metabolizing cholesterol expressed? (double federated query)', 'query': 'PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX taxon: <http://purl.uniprot.org/taxonomy/>\\nPREFIX rh: <http://rdf.rhea-db.org/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX lscr: <http://purl.org/lscr#>\\nPREFIX genex: <http://purl.org/genex#>\\nPREFIX CHEBI: <http://purl.obolibrary.org/obo/CHEBI_>\\nSELECT\\n        DISTINCT \\n            ?protein\\n            ?ensemblGene\\n            ?reaction\\n            ?anatomicEntityLabel\\n            ?anatomicEntity\\nWHERE {\\n  # federated query to Rhea enadpoint\\n  {\\n    SELECT DISTINCT ?reaction WHERE {\\n      SERVICE <https://sparql.rhea-db.org/sparql> {\\n        ?reaction rdfs:subClassOf rh:Reaction .\\n        ?reaction rh:equation ?reactionEquation .\\n        ?reaction rh:side ?reactionSide .\\n        ?reactionSide rh:contains ?participant .\\n        ?participant rh:compound ?compound .\\n        # compound constraint (CHEBI:16113 == cholesterol)\\n        ?compound rh:chebi CHEBI:16113 .\\n      }\\n    }\\n  }\\n  # taxonomy constraint (taxon:9606 == Homo sapiens)\\n  ?protein up:organism taxon:9606 .\\n  ?protein up:annotation ?a .\\n  ?a a up:Catalytic_Activity_Annotation .\\n  ?a up:catalyticActivity ?ca .\\n  ?ca up:catalyzedReaction ?reaction .\\n  ?protein rdfs:seeAlso / up:transcribedFrom ?ensemblGene .\\n\\n  # federated query to Bgee (expression data)\\n  SERVICE <http://biosoda.expasy.org/rdf4j-server/repositories/bgeelight> {\\n    ?gene genex:isExpressedIn ?anatomicEntity .\\n    ?gene lscr:xrefEnsemblGene ?ensemblGene .\\n    ?anatomicEntity rdfs:label ?anatomicEntityLabel .\\n  }\\n}'}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': 'UniProt: Where are the human genes encoding enzymes metabolizing N-acyl\\nsphingosines expressed in the human body (federated query, with Rhea and Bgee)', 'query': 'PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX taxon: <http://purl.uniprot.org/taxonomy/>\\nPREFIX rh: <http://rdf.rhea-db.org/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX lscr: <http://purl.org/lscr#>\\nPREFIX genex: <http://purl.org/genex#>\\nPREFIX CHEBI: <http://purl.obolibrary.org/obo/CHEBI_>\\nSELECT\\n  DISTINCT\\n    ?chebi \\n    ?reaction \\n    ?protein ?ensemblGene\\n    ?anatomicEntityLabel\\n    ?anatomicEntity\\nWHERE {\\n  SERVICE <https://sparql.rhea-db.org/sparql> {\\n    ?reaction rdfs:subClassOf rh:Reaction .\\n    ?reaction rh:equation ?reactionEquation .\\n    ?reaction rh:side ?reactionSide .\\n    ?reactionSide rh:contains ?participant .\\n    ?participant rh:compound ?compound .\\n    ?compound rh:chebi ?chebi .\\n    ?chebi rdfs:subClassOf* CHEBI:52639\\n}\\n  ?protein up:organism taxon:9606 .\\n  ?protein up:annotation ?a .\\n  ?a a up:Catalytic_Activity_Annotation .\\n  ?a up:catalyticActivity ?ca .\\n  ?ca up:catalyzedReaction ?reaction .\\n  ?protein rdfs:seeAlso / up:transcribedFrom ?ensemblGene .\\n\\n  SERVICE  <https://bgee.org/sparql/> {\\n    ?gene genex:isExpressedIn ?anatomicEntity .\\n    ?gene lscr:xrefEnsemblGene ?ensemblGene .\\n    ?anatomicEntity rdfs:label ?anatomicEntityLabel .\\n  }\\n}'}, {'endpoint': 'https://sparql.uniprot.org/sparql/', 'comment': 'UniProt: Find all proteins linked to arachidonate (CHEBI:32395)', 'query': 'PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX skos: <http://www.w3.org/2004/02/skos/core#>\\nPREFIX rh: <http://rdf.rhea-db.org/>\\nPREFIX CHEBI: <http://purl.obolibrary.org/obo/CHEBI_>\\n\\nSELECT \\n    DISTINCT\\n        ?uniprot\\n        ?uniprotID\\n        ?recname\\n        ?gene\\n        ?chebi\\n        ?uniprotName\\nWHERE {\\n  SERVICE <https://sparql.rhea-db.org/sparql> {\\n     VALUES (?chebi) { (CHEBI:32395) }\\n     ?rhea rh:side/rh:contains/rh:compound ?compound .\\n     ?compound rh:chebi ?chebi .\\n     ?chebi up:name ?uniprotName .\\n  }\\n  ?uniprot up:annotation/up:catalyticActivity/up:catalyzedReaction ?rhea .\\n  ?uniprot up:mnemonic ?uniprotID .\\n  ?uniprot up:recommendedName/up:fullName ?recname .\\n  OPTIONAL {?uniprot up:encodedBy/skos:prefLabel ?gene .}\\n}'}, {'endpoint': 'https://www.bgee.org/sparql/', 'comment': 'Bgee: What are the species present in Bgee?', 'query': 'PREFIX up: <http://purl.uniprot.org/core/>\\n\\nSELECT ?species WHERE {\\n    ?species a up:Taxon .\\n}\\n'}, {'endpoint': 'https://www.bgee.org/sparql/', 'comment': 'Bgee: What are the anatomical entities including cell types, if any, where the human gene APOC1 is expressed at the post-juvenile stage along with its expression score independently of the strain and sex?', 'query': 'PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nPREFIX genex: <http://purl.org/genex#>\\n\\nSELECT DISTINCT ?anat ?cellType ?anatName ?cellTypeName ?score ?stage WHERE {\\n    ?seq a orth:Gene ;\\n        rdfs:label \"APOC1\" ;\\n        orth:organism ?organism .\\n    ?expression a genex:Expression ;\\n        genex:hasExpressionCondition ?condition ;\\n        genex:hasExpressionLevel ?score ;\\n        genex:hasSequenceUnit ?seq .\\n    ?condition a genex:ExpressionCondition ;\\n        genex:hasAnatomicalEntity ?anat ;\\n        genex:hasAnatomicalEntity ?cellType ;\\n        genex:hasDevelopmentalStage ?stage .\\n    ?anat a genex:AnatomicalEntity ;\\n        rdfs:label ?anatName .\\n    ?cellType rdfs:label ?cellTypeName .\\n    ?stage a efo:EFO_0000399 ;\\n        rdfs:label \"post-juvenile\" .\\n    ?organism obo:RO_0002162  ?species .\\n    ?species a up:Taxon ;\\n        up:commonName \"human\" .\\n    FILTER (?anat !=  obo:GO_0005575)\\n    FILTER (?anat != ?cellType)\\n} ORDER BY DESC(?score)\\n'}, {'endpoint': 'https://www.bgee.org/sparql/', 'comment': 'Bgee: What are the developmental stages present in Bgee?', 'query': 'PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX dcterms: <http://purl.org/dc/terms/>\\n\\nSELECT DISTINCT ?stage ?stageName ?stageDescription WHERE {\\n    ?stage a efo:EFO_0000399 ;\\n        rdfs:label ?stageName ;\\n        dcterms:description ?stageDescription .\\n}\\n'}, {'endpoint': 'https://www.bgee.org/sparql/', 'comment': 'Bgee: What is the post-juvenile stage link and description?', 'query': 'PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX dcterms: <http://purl.org/dc/terms/>\\n\\nSELECT DISTINCT ?stage ?stageName ?stageDescription WHERE {\\n    ?stage a efo:EFO_0000399 ;\\n        rdfs:label ?stageName ;\\n        dcterms:description ?stageDescription .\\n    FILTER (CONTAINS(?stageName,\"post-juvenile\"))\\n}\\n'}, {'endpoint': 'https://www.bgee.org/sparql/', 'comment': 'Bgee: What are the anatomical entities where the eel gene apoc1 is expressed along with its expression score independently of the strain, sex, and cell type?', 'query': 'PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nPREFIX lscr: <http://purl.org/lscr#>\\nPREFIX genex: <http://purl.org/genex#>\\n\\nSELECT DISTINCT ?anat ?anatName ?stageIRI ?score WHERE {\\n    ?seq a orth:Gene ;\\n        lscr:xrefNCBIGene <https://www.ncbi.nlm.nih.gov/gene/118230125> .\\n    ?expression a genex:Expression ;\\n        genex:hasExpressionCondition ?condition ;\\n        genex:hasExpressionLevel ?score ;\\n        genex:hasSequenceUnit ?seq .\\n    ?anat a genex:AnatomicalEntity ;\\n        rdfs:label ?anatName .\\n    ?condition a genex:ExpressionCondition ;\\n        genex:hasAnatomicalEntity ?anat ;\\n        genex:hasDevelopmentalStage ?stageIRI ;\\n        genex:hasSex \"any\" ;\\n        genex:hasStrain ?strain .\\n    ?stageIRI a efo:EFO_0000399 .\\n    ?strain a efo:EFO_0005135 ;\\n        rdfs:label \"wild-type\" .\\n    FILTER (?anat != obo:GO_0005575)\\n} ORDER BY DESC(?score)\\n'}, {'endpoint': 'https://www.bgee.org/sparql/', 'comment': 'Bgee: What are the anatomical entities where the eel gene apoc1 is expressed along with its expression score independently of the strain, sex, and cell type?', 'query': 'PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nPREFIX genex: <http://purl.org/genex#>\\nPREFIX dcterms: <http://purl.org/dc/terms/>\\n\\nSELECT DISTINCT ?anat ?anatName ?stageIRI ?score WHERE {\\n    ?seq a orth:Gene ;\\n        dcterms:identifier \"118230125\" .\\n    ?expression a genex:Expression ;\\n        genex:hasExpressionCondition ?condition ;\\n        genex:hasExpressionLevel ?score ;\\n        genex:hasSequenceUnit ?seq .\\n    ?condition a genex:ExpressionCondition ;\\n        genex:hasAnatomicalEntity ?anat ;\\n        genex:hasDevelopmentalStage ?stageIRI ;\\n        genex:hasSex \"any\" ;\\n        genex:hasStrain ?strain .\\n    ?anat rdfs:label ?anatName .\\n    ?strain a efo:EFO_0005135 ;\\n        rdfs:label \"wild-type\" .\\nFILTER (?anat !=  obo:GO_0005575)\\n} ORDER BY DESC(?score)\\n'}, {'endpoint': 'https://www.bgee.org/sparql/', 'comment': 'Bgee: What are the anatomical entities where the P02654 gene is expressed? Note that P02654 is a UniProtKB identifier of the APOC1 human gene.', 'query': 'PREFIX uniprotkb: <http://purl.uniprot.org/uniprot/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX lscr: <http://purl.org/lscr#>\\nPREFIX genex: <http://purl.org/genex#>\\n\\nSELECT DISTINCT ?anat ?anatName WHERE {\\n    ?seq a orth:Gene ;\\n        genex:isExpressedIn ?anat ;\\n        lscr:xrefUniprot uniprotkb:P02654 .\\n    ?anat a genex:AnatomicalEntity ;\\n        rdfs:label ?anatName .\\n}\\n'}, {'endpoint': 'https://www.bgee.org/sparql/', 'comment': 'Bgee: What is all the metadata related to the ENSG00000130208 gene, where ENSG00000130208 is the identifier of the APOC1 human gene.', 'query': 'PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX lscr: <http://purl.org/lscr#>\\nPREFIX dcterms: <http://purl.org/dc/terms/>\\n\\nSELECT DISTINCT ?symbol ?description ?id\\n?links ?organism ?uniprot ?ensembl ?ncbi WHERE {\\n    ?seq a orth:Gene ;\\n        rdfs:label ?symbol ;\\n        rdfs:seeAlso ?links ;\\n        dcterms:description ?description ;\\n        dcterms:identifier ?id ;\\n        orth:organism ?organism .\\n    OPTIONAL{?seq lscr:xrefUniprot ?uniprot .}\\n    OPTIONAL{?seq lscr:xrefEnsemblGene ?ensembl .}\\n    OPTIONAL{?seq lscr:xrefNCBIGene ?ncbi .}\\n    FILTER (?id = \"ENSG00000130208\")\\n}\\n'}, {'endpoint': 'https://www.bgee.org/sparql/', 'comment': 'Bgee: What are the anatomical entities where the APOC1 Homo sapiens gene is not expressed, that is where is APOC1 absent?', 'query': 'PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nPREFIX genex: <http://purl.org/genex#>\\n\\nSELECT DISTINCT ?anat ?anatName WHERE {\\n    ?seq a orth:Gene ;\\n        rdfs:label \"APOC1\" ;\\n        genex:isAbsentIn ?anat ;\\n        orth:organism ?organism .\\n    ?anat a genex:AnatomicalEntity ;\\n        rdfs:label ?anatName .\\n    ?organism obo:RO_0002162  ?species .\\n    ?species a up:Taxon ;\\n        up:scientificName \"Homo sapiens\" .\\n}\\n'}, {'endpoint': 'https://www.bgee.org/sparql/', 'comment': 'Bgee: What are the species present in Bgee and their scientific and common names?', 'query': 'PREFIX up: <http://purl.uniprot.org/core/>\\n\\nSELECT ?species ?sci_name ?common_name WHERE {\\n    ?species a up:Taxon ;\\n        up:scientificName ?sci_name ;\\n        up:rank up:Species .\\n    OPTIONAL { ?species up:commonName ?common_name . }\\n}\\n'}, {'endpoint': 'https://www.bgee.org/sparql/', 'comment': 'Bgee: What are the anatomical entities where the APOC1 gene is expressed?', 'query': 'PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX genex: <http://purl.org/genex#>\\n\\nSELECT DISTINCT ?anat ?anatName WHERE {\\n    ?seq a orth:Gene ;\\n        genex:isExpressedIn ?anat ;\\n        rdfs:label \"APOC1\" .\\n    ?anat a genex:AnatomicalEntity ;\\n        rdfs:label ?anatName .\\n}\\n'}, {'endpoint': 'https://www.bgee.org/sparql/', 'comment': 'Bgee: What are the anatomical entities where the APOC1 Homo sapiens gene is expressed?', 'query': 'PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nPREFIX genex: <http://purl.org/genex#>\\n\\nSELECT DISTINCT ?anat ?anatName WHERE {\\n    ?seq a orth:Gene ;\\n        genex:isExpressedIn ?anat ;\\n        rdfs:label \"APOC1\" .\\n    ?anat a genex:AnatomicalEntity ;\\n        rdfs:label ?anatName .\\n    ?seq orth:organism ?organism .\\n    ?organism obo:RO_0002162  ?species .\\n    ?species a up:Taxon ;\\n        up:scientificName \"Homo sapiens\" .\\n}\\n'}, {'endpoint': 'https://www.bgee.org/sparql/', 'comment': 'Bgee: What are the anatomical entities where the APOC1 gene is expressed independently of the developmental stage, sex, strain and cell type?', 'query': 'PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nPREFIX genex: <http://purl.org/genex#>\\n\\nSELECT DISTINCT ?anat ?anatName {\\n    ?seq a orth:Gene ;\\n        genex:isExpressedIn ?condition ;\\n        rdfs:label \"APOC1\" .\\n    ?condition a genex:ExpressionCondition ;\\n        genex:hasAnatomicalEntity ?anat ;\\n        genex:hasAnatomicalEntity obo:GO_0005575 ;\\n        genex:hasDevelopmentalStage ?stage ;\\n        genex:hasSex \"any\" ;\\n        genex:hasStrain ?strain .\\n    ?anat a genex:AnatomicalEntity ;\\n        rdfs:label ?anatName .\\n    ?stage a efo:EFO_0000399 ;\\n        rdfs:label \"life cycle\" .\\n    ?strain rdfs:label \"wild-type\" .\\n}\\n'}, {'endpoint': 'https://www.bgee.org/sparql/', 'comment': 'Bgee: What are the anatomical entities where the human gene APOC1 is expressed in the post-juvenile stage?', 'query': 'PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nPREFIX genex: <http://purl.org/genex#>\\n\\nSELECT DISTINCT ?anat ?anatName ?stage WHERE {\\n    ?seq a orth:Gene ;\\n        rdfs:label \"APOC1\" ;\\n        genex:isExpressedIn ?condition ;\\n        orth:organism ?organism .\\n    ?condition a genex:ExpressionCondition ;\\n        genex:hasAnatomicalEntity ?anat ;\\n        genex:hasAnatomicalEntity obo:GO_0005575 ;\\n        genex:hasDevelopmentalStage ?stage ;\\n        genex:hasSex \"any\" ;\\n        genex:hasStrain ?strain .\\n    ?anat rdfs:label ?anatName .\\n    ?stage a efo:EFO_0000399 ;\\n        rdfs:label \"post-juvenile\" .\\n    ?strain a efo:EFO_0005135 ;\\n        rdfs:label \"wild-type\" .\\n    ?organism obo:RO_0002162  ?species .\\n    ?species a up:Taxon ;\\n        up:commonName \"human\" .\\n}\\n'}, {'endpoint': 'https://www.bgee.org/sparql/', 'comment': 'Bgee: What are the anatomical entities where the human gene APOC1 is expressed in the post-juvenile stage?', 'query': 'PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nPREFIX genex: <http://purl.org/genex#>\\n\\nSELECT DISTINCT ?anat ?anatName ?stage WHERE {\\n    ?seq a orth:Gene ;\\n        rdfs:label \"APOC1\" ;\\n        genex:isExpressedIn ?condition ;\\n        orth:organism ?organism .\\n    ?condition a genex:ExpressionCondition ;\\n        genex:hasAnatomicalEntity ?anat ;\\n        genex:hasAnatomicalEntity obo:GO_0005575 ;\\n        genex:hasDevelopmentalStage ?stage .\\n    ?anat a genex:AnatomicalEntity ;\\n        rdfs:label ?anatName .\\n    ?stage a efo:EFO_0000399 ;\\n        rdfs:label \"post-juvenile\" .\\n    ?organism obo:RO_0002162  ?species .\\n    ?species a up:Taxon ;\\n        up:commonName \"human\" .\\n}\\n'}, {'endpoint': 'https://www.bgee.org/sparql/', 'comment': 'Bgee: What are the anatomical entities where the human gene APOC1 is expressed in the post-juvenile stage along with its expression score independently of the strain, sex, and cell type?', 'query': 'PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nPREFIX genex: <http://purl.org/genex#>\\n\\nSELECT DISTINCT ?anat ?anatName ?score ?stage WHERE {\\n    ?seq a orth:Gene ;\\n        rdfs:label \"APOC1\" ;\\n        orth:organism ?organism .\\n    ?expression a genex:Expression ;\\n        genex:hasExpressionCondition ?condition ;\\n        genex:hasExpressionLevel ?score ;\\n        genex:hasSequenceUnit ?seq .\\n    ?condition a genex:ExpressionCondition ;\\n        genex:hasAnatomicalEntity ?anat ;\\n        genex:hasAnatomicalEntity obo:GO_0005575 ;\\n        genex:hasDevelopmentalStage ?stage ;\\n        genex:hasSex \"any\" ;\\n        genex:hasStrain ?strain .\\n    ?anat a genex:AnatomicalEntity ;\\n        rdfs:label ?anatName .\\n    ?stage rdfs:label \"post-juvenile\" .\\n    ?strain rdfs:label \"wild-type\" .\\n    ?organism obo:RO_0002162  ?species .\\n    ?species a up:Taxon ;\\n        up:commonName \"human\" .\\nFILTER (?anat !=  obo:GO_0005575)\\n} ORDER BY DESC(?score)\\n'}, {'endpoint': 'https://www.bgee.org/sparql/', 'comment': 'Bgee: What are the anatomical entities where the human gene APOC1 is expressed at the post-juvenile stage along with its expression score independently of the strain, sex, and cell type?', 'query': 'PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nPREFIX lscr: <http://purl.org/lscr#>\\nPREFIX genex: <http://purl.org/genex#>\\nPREFIX ensembl: <http://rdf.ebi.ac.uk/resource/ensembl/>\\n\\nSELECT DISTINCT ?anat ?anatName ?score WHERE {\\n    ?seq a orth:Gene .\\n    ?expression a genex:Expression ;\\n        genex:hasExpressionCondition ?condition ;\\n        genex:hasExpressionLevel ?score ;\\n        genex:hasSequenceUnit ?seq .\\n    ?seq lscr:xrefEnsemblGene ensembl:ENSG00000130208 .\\n    ?condition a genex:ExpressionCondition ;\\n        genex:hasAnatomicalEntity ?anat .\\n    ?anat a genex:AnatomicalEntity ;\\n        rdfs:label ?anatName .\\n    ?condition genex:hasDevelopmentalStage obo:UBERON_0000113 ;\\n        genex:hasSex \"any\" ;\\n        genex:hasStrain ?strain .\\n    ?strain a efo:EFO_0005135 ;\\n        rdfs:label \"wild-type\" .\\n    ?seq orth:organism ?organism .\\n    ?organism obo:RO_0002162  taxon:9606 .\\n    FILTER (?anat !=  obo:GO_0005575)\\n} ORDER BY DESC(?score)\\n'}, {'endpoint': 'https://www.bgee.org/sparql/', 'comment': 'Bgee: What are the anatomical entities where the human gene APOC1 is expressed at the post-juvenile stage along with its expression score independently of the strain, sex, and cell type?', 'query': 'PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nPREFIX lscr: <http://purl.org/lscr#>\\nPREFIX genex: <http://purl.org/genex#>\\nPREFIX ensembl: <http://rdf.ebi.ac.uk/resource/ensembl/>\\n\\nSELECT DISTINCT ?anat ?anatName ?score  {\\n    ?seq a orth:Gene ;\\n        lscr:xrefEnsemblGene ensembl:ENSG00000130208 .\\n    ?expression a genex:Expression ;\\n        genex:hasExpressionCondition ?condition ;\\n        genex:hasExpressionLevel ?score ;\\n        genex:hasSequenceUnit ?seq .\\n    ?condition a genex:ExpressionCondition ;\\n        genex:hasAnatomicalEntity ?anat ;\\n        genex:hasDevelopmentalStage obo:UBERON_0000113 ;\\n        genex:hasSex \"any\" ;\\n        genex:hasStrain ?strain .\\n    ?anat a genex:AnatomicalEntity ;\\n        rdfs:label ?anatName .\\n    ?strain a efo:EFO_0005135 ;\\n        rdfs:label \"wild-type\" .\\n    FILTER (?anat !=  obo:GO_0005575)\\n} ORDER BY DESC(?score)\\n'}, {'endpoint': 'https://www.bgee.org/sparql/', 'comment': 'Bgee: What are the anatomical entities including cell types, if any, where the human gene APOC1 is expressed at the post-juvenile stage along with its expression score independently of the strain and sex?', 'query': 'PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nPREFIX genex: <http://purl.org/genex#>\\n\\nSELECT DISTINCT ?anat ?cellType ?anatName ?cellTypeName ?score ?stage WHERE {\\n    ?seq a orth:Gene ;\\n        rdfs:label \"APOC1\" ;\\n        orth:organism ?organism .\\n    ?expression a genex:Expression ;\\n        genex:hasExpressionCondition ?condition ;\\n        genex:hasExpressionLevel ?score ;\\n        genex:hasSequenceUnit ?seq .\\n    ?condition a genex:ExpressionCondition ;\\n        genex:hasAnatomicalEntity ?anat ;\\n        genex:hasAnatomicalEntity ?cellType ;\\n        genex:hasDevelopmentalStage ?stage ;\\n        genex:hasSex \"any\" ;\\n        genex:hasStrain ?strain .\\n    ?anat a genex:AnatomicalEntity ;\\n        rdfs:label ?anatName .\\n    ?cellType rdfs:label ?cellTypeName .\\n    ?stage rdfs:label \"post-juvenile\" .\\n    ?strain a efo:EFO_0005135 ;\\n        rdfs:label \"wild-type\" .\\n    ?organism obo:RO_0002162  ?species .\\n    ?species a up:Taxon ;\\n        up:commonName \"human\" .\\n    FILTER (?anat != obo:GO_0005575)\\n    FILTER (?anat != ?cellType)\\n} ORDER BY DESC(?score)\\n'}]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_54064/1366725701.py:38: MarkupResemblesLocatorWarning: The input looks more like a filename than markup. You may want to open this file and pass the filehandle into Beautiful Soup.\n",
      "  soup = BeautifulSoup(html_text, \"html.parser\")\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "\n",
    "from bs4 import BeautifulSoup\n",
    "from SPARQLWrapper import JSON, SPARQLWrapper\n",
    "\n",
    "# system_prompt = \"\"\"You are Expasy, an assistant that helps users to query the databases from the Swiss Institute of Bioinformatics, such as UniProt or Bgee.\n",
    "# When writing the SPARQL query try to factorize the predicates/objects of a subject as much as possible, so that the user can understand the query and the results.\n",
    "# \"\"\"\n",
    "# examples_prompt: str = \"Here are a list of questions and queries that Expasy has learned to answer, use them as base when answering the question from the user:\"\n",
    "\n",
    "endpoints = {\n",
    "    \"UniProt\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    \"Bgee\": \"https://www.bgee.org/sparql/\",\n",
    "}\n",
    "\n",
    "get_queries = \"\"\"PREFIX sh: <http://www.w3.org/ns/shacl#>\n",
    "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
    "\n",
    "SELECT DISTINCT ?comment ?query\n",
    "WHERE\n",
    "{\n",
    "    ?sq a sh:SPARQLExecutable ;\n",
    "        rdfs:label|rdfs:comment ?comment ;\n",
    "        sh:select|sh:ask|sh:construct|sh:describe ?query .\n",
    "}\"\"\"\n",
    "\n",
    "get_prefixes = \"\"\"PREFIX sh: <http://www.w3.org/ns/shacl#>\n",
    "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
    "\n",
    "SELECT ?prefix ?namespace\n",
    "WHERE {\n",
    "    [] sh:namespace ?namespace ;\n",
    "        sh:prefix ?prefix .\n",
    "} ORDER BY ?prefix\"\"\"\n",
    "\n",
    "def remove_a_tags(html_text: str) -> str:\n",
    "    \"\"\"Remove all <a> tags from the queries descriptions\"\"\"\n",
    "    soup = BeautifulSoup(html_text, \"html.parser\")\n",
    "    for a_tag in soup.find_all(\"a\"):\n",
    "        a_tag.replace_with(a_tag.text)\n",
    "    return soup.get_text()\n",
    "\n",
    "queries = []\n",
    "for endpoint_name, endpoint_url in endpoints.items():\n",
    "    sparql_endpoint = SPARQLWrapper(endpoint_url)\n",
    "    sparql_endpoint.setReturnFormat(JSON)\n",
    "\n",
    "    sparql_endpoint.setQuery(get_prefixes)\n",
    "    results = sparql_endpoint.query().convert()\n",
    "    prefix_map = {}\n",
    "    for row in results[\"results\"][\"bindings\"]:\n",
    "        prefix_map[row[\"prefix\"][\"value\"]] = row[\"namespace\"][\"value\"]\n",
    "\n",
    "    sparql_endpoint.setQuery(get_queries)\n",
    "    results = sparql_endpoint.query().convert()\n",
    "    print(f\"Found {len(results['results']['bindings'])} queries for {endpoint_url}\")\n",
    "\n",
    "    for row in results[\"results\"][\"bindings\"]:\n",
    "        query = row[\"query\"][\"value\"]\n",
    "        # Add prefixes to queries\n",
    "        for prefix, namespace in prefix_map.items():\n",
    "            prefix_str = f\"PREFIX {prefix}: <{namespace}>\"\n",
    "            if not re.search(prefix_str, query) and re.search(f\"[(| |\\u00a0|/]{prefix}:\", query):\n",
    "                query = f\"{prefix_str}\\n{query}\"\n",
    "        queries.append({\n",
    "            \"endpoint\": endpoint_url,\n",
    "            \"comment\": f\"{endpoint_name}: {remove_a_tags(row['comment']['value'])}\",\n",
    "            \"query\": query,\n",
    "        })\n",
    "\n",
    "# print(queries)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fetching 5 files: 100%|██████████| 5/5 [00:00<00:00, 14383.76it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "70\n",
      "Done generating embeddings\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "UpdateResult(operation_id=2, status=<UpdateStatus.COMPLETED: 'completed'>)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from fastembed import TextEmbedding\n",
    "from qdrant_client import QdrantClient, models\n",
    "from qdrant_client.http.models import (\n",
    "    Distance,\n",
    "    VectorParams,\n",
    ")\n",
    "\n",
    "# https://qdrant.github.io/fastembed/examples/Supported_Models/\n",
    "# TextEmbedding.list_supported_models()\n",
    "\n",
    "embedding_model = TextEmbedding(\"BAAI/bge-large-en-v1.5\")\n",
    "embed_dimensions = 1024\n",
    "\n",
    "vectordb = QdrantClient(\n",
    "    host=\"qdrant\", # Running on the same docker network with compose\n",
    "    prefer_grpc=True,\n",
    ")\n",
    "\n",
    "collection_name=\"expasy-queries\"\n",
    "if not vectordb.collection_exists(collection_name):\n",
    "    vectordb.create_collection(\n",
    "        collection_name=collection_name,\n",
    "        vectors_config=VectorParams(size=embed_dimensions, distance=Distance.COSINE),\n",
    "    )\n",
    "\n",
    "questions = [q[\"comment\"] for q in queries]\n",
    "print(len(questions))\n",
    "output = embedding_model.embed(questions)\n",
    "print(\"Done generating embeddings\")\n",
    "\n",
    "vectordb.upsert(\n",
    "    collection_name=collection_name,\n",
    "    points=models.Batch(\n",
    "        ids=list(range(1, len(queries) + 1)),\n",
    "        vectors=[embeddings.tolist() for embeddings in output],\n",
    "        payloads=queries,\n",
    "    ),\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "expasy-scripts",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
