{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Testing Expasy chat\n",
    "\n",
    "These tests requires to start the expasy chat stack locally with `docker compose -f compose.dev.yml up`\n",
    "\n",
    "Testing our Expasy query helper reusing the example queries that succeeded retrieved from `test_example_queries.ipynb`\n",
    "\n",
    "Potential interesting questions to use for evaluation:\n",
    "- Which are the human genes associated with cancer and their orthologs expressed in the rat brain?\n",
    "- Select the number of proteins for each of the subclasses of ec:1.1.1.- EC (Enzyme Commission) class (from biosoda)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "üß™ Testing 14 queries\n",
      "Reference query 0 'What is the accession number in uniprot of the human gene LCT? Return only unique protein URIs' took 163.92 seconds\n",
      "Reference query 1 'How do I filter for reviewed mouse proteins which carry an N-terminal glycine? Return protein URI and AA sequence' took 47.62 seconds\n",
      "Reference query 2 'How could I download a table that only includes the Rhea reactions for which there is experimental evidence? Return only the rhea URI' took 27.36 seconds\n",
      "Reference query 3 'Which human proteins are enzymes catalyzing a reaction involving sterols? Return the protein, sterol and reaction URI' took 9.93 seconds\n",
      "Reference query 4 'Which are the human proteins associated with cancer? Return distinct ?diseaseLabel and ?hgncSymbol' took 1.73 seconds\n",
      "Reference query 5 'In bgee how can I retrieve the confidence level and false discovery rate of a gene expression? Return distinct ?gene, ?confidence and ?fdr, limit to 10' took 0.07 seconds\n",
      "Reference query 6 'How can I get the cross-reference to the ensembl protein for the LCT protein in OMA? Return only the distinct ?ensemblURI' took 0.13 seconds\n",
      "Reference query 7 'How can I get the URI of a dataset to which an ortholog cluster belongs in OMA? Return orthologCluster, datasetURI and limit to 20' took 0.08 seconds\n",
      "Reference query 8 'Give me the list of strains associated to the Escherichia coli taxon and their name. Return ?taxon, ?strain, ?name, limit to 20' took 0.10 seconds\n",
      "Reference query 9 'Retrieve all proteins involved in pathways involving glycolysis. Return ?proteinURI, ?proteinLabel, ?pathwayLabel, limit to 20' took 0.24 seconds\n",
      "Reference query 10 'What are the orthologs in rat for protein Q9Y2T1 ? Return ?ratProtein ?ratUniProtXref' took 0.13 seconds\n",
      "Reference query 11 'Retrieve all proteins that are associated with Alzheimer disease (http://purl.uniprot.org/diseases/3832) and where they are known to be located in the cell. Return ?proteinURI, ?locationInsideCellLabel, ?locationInsideCellUri, limit to 20' took 0.28 seconds\n",
      "Reference query 12 'Retrieve all proteins in OMA that are encoded by the TP53 gene and their mnemonics and evidence types from the UniProt database. Return ?proteinOMA ?speciesLabel ?mnemonic ?evidenceType ?uniprotURI' took 21.69 seconds\n",
      "Reference query 13 'What is the function of APOC1? Return ?function' took 0.06 seconds\n",
      "\n",
      "üß† Testing model: o3-mini\n",
      "\n",
      "Missing from generated: {'protein': {'type': 'uri', 'value': 'http://purl.uniprot.org/uniprot/P09848'}}\n",
      "Missing from generated: {'protein': {'type': 'uri', 'value': 'http://purl.uniprot.org/uniprot/P09848'}}\n",
      "Missing from generated: {'protein': {'type': 'uri', 'value': 'http://purl.uniprot.org/uniprot/P09848'}}\n",
      "‚úÖ 1/3 What is the accession number in uniprot of the human gene LCT? Return only unique protein URIs = 4\n",
      "‚úÖ 2/3 What is the accession number in uniprot of the human gene LCT? Return only unique protein URIs = 4\n",
      "‚úÖ 3/3 What is the accession number in uniprot of the human gene LCT? Return only unique protein URIs = 4\n",
      "üéØ No RAG - Success: 0, Different results: 0, No results: 3, Error: 0\n",
      "üéØ RAG with validation - Success: 3, Different results: 0, No results: 0, Error: 0\n",
      "Missing from generated: {'sequence': {'type': 'literal', 'value': 'GPGLGDQERTLSLNNVKGHGRTDDECDRARHREDRLLGTRLDRDQEKLLRESKELADLARLHPTSCAPNGLNPNLMVTGGPTLAGSGRWSADPAAHLATNPWLPRSGSTSMWLAGHPYGLGPPSLHQGMAPAFPPGLGGSLPSAYQFVRDPQSGQLVVIPSDHLPHFAELMERAAVPPLWPALYPPGRSPLHHAQQLQLFSQQHFLRQQELLYLQQQAAQALELQRSAQLVERLKAQEHRTEMEEKISKRSLETTGKAGLSAAGPGLLPRKSAGLANGPAGSHGKAVSPPPSPRASPVTSLKAKVIQKVEDVSKPPAYTYPATPSSHPSSPPPASPPPTPGLTRKEEAPENVVEKKDLELEKETPSPFQALFTDIPPRYPFQALPPHYGRPYPFLLQPAAASDADGLAPDVPLPADGPERLALSPEDKPICLSPSKIPEPPRDSPEEEQLADREVKAEVEDIEEGPTELPPLESPLALPVPETMVAVSPAGGCGGSPLEAQALSTAGPGCREPSEVSDFAQVAEPQIELPSKTEHRMTALELGTQLTPEPLVETKEEPVEVPLDVPMEEPTTEAGPEDSLPQPSLTEPQPSLELSDCDLPVPEGQCLNLEAQEAVPAPASTCYLEETHSESLLPGLDDPLAGMNALAAAAELPQARPLPSLGPGVPAGEKLDTAPSLVLEHSFLQGITLLSEIAELELDRRGQEAADPEPNLVVRPSLESLLAASSHMLKEVLESPFSDPLKNLRLPRELNSNKKYSWMQKKEERMFAMKSSLEDMDALELDFRMRLAEVQRRYKEKQRELVKLQRRRDSGLSSKSLLTSDDYDLGAGIRKRHKGPEEEQEALMGMGKARSRNQSWDDHDSSSDFMSQLKIKKKKMASDQEQLASKLDRALSLKGKARKLFYKAIVRGKEMIRIGDCAVFLSAGRPNLPYIGRIQSMWESWGNNMVVRVKWFYHPEETSPGKQFHEGQHWDQKSGHSLPAALRASSQRKDFMERALYQSSHVDENDVQTVSHKCLVVGLEQYEQMLKTKKYQDSEGLYYLAGTYEPTTGMIFSTDGVPVLC'}, 'protein': {'type': 'uri', 'value': 'http://purl.uniprot.org/uniprot/Q80WC3'}}\n",
      "‚úÖ 2/3 How do I filter for reviewed mouse proteins which carry an N-terminal glycine? Return protein URI and AA sequence = 5\n",
      "‚úÖ 1/3 How do I filter for reviewed mouse proteins which carry an N-terminal glycine? Return protein URI and AA sequence = 5\n",
      "Missing from reference: {'protein': {'type': 'uri', 'value': 'http://purl.uniprot.org/uniprot/Q80WC3'}, 'aa_seq': {'type': 'literal', 'value': 'GPGLGDQERTLSLNNVKGHGRTDDECDRARHREDRLLGTRLDRDQEKLLRESKELADLARLHPTSCAPNGLNPNLMVTGGPTLAGSGRWSADPAAHLATNPWLPRSGSTSMWLAGHPYGLGPPSLHQGMAPAFPPGLGGSLPSAYQFVRDPQSGQLVVIPSDHLPHFAELMERAAVPPLWPALYPPGRSPLHHAQQLQLFSQQHFLRQQELLYLQQQAAQALELQRSAQLVERLKAQEHRTEMEEKISKRSLETTGKAGLSAAGPGLLPRKSAGLANGPAGSHGKAVSPPPSPRASPVTSLKAKVIQKVEDVSKPPAYTYPATPSSHPSSPPPASPPPTPGLTRKEEAPENVVEKKDLELEKETPSPFQALFTDIPPRYPFQALPPHYGRPYPFLLQPAAASDADGLAPDVPLPADGPERLALSPEDKPICLSPSKIPEPPRDSPEEEQLADREVKAEVEDIEEGPTELPPLESPLALPVPETMVAVSPAGGCGGSPLEAQALSTAGPGCREPSEVSDFAQVAEPQIELPSKTEHRMTALELGTQLTPEPLVETKEEPVEVPLDVPMEEPTTEAGPEDSLPQPSLTEPQPSLELSDCDLPVPEGQCLNLEAQEAVPAPASTCYLEETHSESLLPGLDDPLAGMNALAAAAELPQARPLPSLGPGVPAGEKLDTAPSLVLEHSFLQGITLLSEIAELELDRRGQEAADPEPNLVVRPSLESLLAASSHMLKEVLESPFSDPLKNLRLPRELNSNKKYSWMQKKEERMFAMKSSLEDMDALELDFRMRLAEVQRRYKEKQRELVKLQRRRDSGLSSKSLLTSDDYDLGAGIRKRHKGPEEEQEALMGMGKARSRNQSWDDHDSSSDFMSQLKIKKKKMASDQEQLASKLDRALSLKGKARKLFYKAIVRGKEMIRIGDCAVFLSAGRPNLPYIGRIQSMWESWGNNMVVRVKWFYHPEETSPGKQFHEGQHWDQKSGHSLPAALRASSQRKDFMERALYQSSHVDENDVQTVSHKCLVVGLEQYEQMLKTKKYQDSEGLYYLAGTYEPTTGMIFSTDGVPVLC'}}\n",
      "‚ùå 2/3 How do I filter for reviewed mouse proteins which carry an N-terminal glycine? Return protein URI and AA sequence\n",
      "\n",
      "Results mismatch. Ref: 5 != gen: 5\n",
      "\n",
      "#https://sparql.uniprot.org/sparql/\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\n",
      "\n",
      "SELECT ?protein ?aa_seq\n",
      "WHERE {\n",
      "  ?protein a up:Protein ;\n",
      "           up:reviewed true ;\n",
      "           up:organism ?organism ;\n",
      "           up:sequence ?seq .\n",
      "  ?organism up:scientificName \"Mus musculus\" .\n",
      "  ?seq rdf:value ?aa_seq .\n",
      "  FILTER(STRSTARTS(?aa_seq, \"G\"))\n",
      "}\n",
      "Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "    PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
      "    PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\n",
      "    SELECT ?protein ?sequence\n",
      "    WHERE\n",
      "    {\n",
      "        ?protein a up:Protein ;\n",
      "            up:organism taxon:10090 ;  # Taxonomy ID for Mus musculus (Mouse)\n",
      "            up:reviewed true ;\n",
      "            up:sequence ?isoform .\n",
      "        ?isoform rdf:value ?sequence .\n",
      "        # Ensure the N-terminal amino acid is Glycine (G)\n",
      "        FILTER (STRSTARTS(?sequence, \"G\"))\n",
      "    }\n",
      "\n",
      "‚úÖ 3/3 How do I filter for reviewed mouse proteins which carry an N-terminal glycine? Return protein URI and AA sequence = 5\n",
      "üéØ No RAG - Success: 1, Different results: 0, No results: 4, Error: 1\n",
      "üéØ RAG with validation - Success: 5, Different results: 1, No results: 0, Error: 0\n",
      "Missing from generated: {'rhea': {'type': 'uri', 'value': 'http://rdf.rhea-db.org/21408'}}\n",
      "Missing from generated: {'rhea': {'type': 'uri', 'value': 'http://rdf.rhea-db.org/21408'}}\n",
      "Missing from generated: {'rhea': {'type': 'uri', 'value': 'http://rdf.rhea-db.org/21408'}}\n",
      "Missing from generated: {'rhea': {'type': 'uri', 'value': 'http://rdf.rhea-db.org/21408'}}\n",
      "‚ùå 1/3 How could I download a table that only includes the Rhea reactions for which there is experimental evidence? Return only the rhea URI\n",
      "\n",
      "Results mismatch. Ref: 11861 != gen: 4880\n",
      "\n",
      "# https://sparql.uniprot.org/sparql\n",
      "PREFIX ECO: <http://purl.obolibrary.org/obo/ECO_>\n",
      "PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX rh: <http://rdf.rhea-db.org/>\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "\n",
      "SELECT DISTINCT ?rhea\n",
      "WHERE {\n",
      "  {  \n",
      "    SELECT DISTINCT ?rhea\n",
      "    WHERE {\n",
      "      GRAPH <https://sparql.rhea-db.org/rhea> {\n",
      "        ?rhea rdfs:subClassOf rh:Reaction .\n",
      "        ?rhea rh:side/rh:contains/rh:compound ?compound2 .\n",
      "        ?uc rdfs:subClassOf rh:Compound .\n",
      "      }\n",
      "      ?compound2 rdfs:subClassOf ?uc .\n",
      "      BIND(IF(?uc = rh:SmallMolecule, 0, 1) AS ?c)\n",
      "    } \n",
      "    GROUP BY ?rhea  \n",
      "    HAVING (SUM(?c) = 0)\n",
      "  }\n",
      "  ?catalyzedReaction up:catalyzedReaction ?rhea .\n",
      "  ?reif rdf:object ?catalyzedReaction ;\n",
      "        up:attribution ?attr .\n",
      "  ?attr up:evidence ECO:0000269 ;\n",
      "        up:source ?source .\n",
      "  ?source a up:Citation .\n",
      "}\n",
      "Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "    PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\n",
      "    SELECT DISTINCT ?rhea\n",
      "    WHERE {\n",
      "    # ECO 269 is experimental evidence\n",
      "    BIND (<http://purl.obolibrary.org/obo/ECO_0000269> as ?evidence)\n",
      "    #GRAPH <http://sparql.uniprot.org/uniprot> {\n",
      "        ?protein up:reviewed true ;\n",
      "        up:annotation ?a ;\n",
      "        up:attribution ?attribution  .\n",
      "\n",
      "        ?a a up:Catalytic_Activity_Annotation ;\n",
      "        up:catalyticActivity ?ca .\n",
      "        ?ca up:catalyzedReaction ?rhea .\n",
      "\n",
      "        [] rdf:subject ?a ;\n",
      "        rdf:predicate up:catalyticActivity ;\n",
      "        rdf:object ?ca ;\n",
      "        up:attribution ?attribution .\n",
      "\n",
      "        ?attribution up:evidence ?evidence .\n",
      "    #}\n",
      "    }\n",
      "\n",
      "Missing from generated: {'rhea': {'type': 'uri', 'value': 'http://rdf.rhea-db.org/21408'}}\n",
      "‚ùå 2/3 How could I download a table that only includes the Rhea reactions for which there is experimental evidence? Return only the rhea URI\n",
      "\n",
      "Results mismatch. Ref: 11861 != gen: 4880\n",
      "\n",
      "# https://sparql.uniprot.org/sparql\n",
      "PREFIX ECO: <http://purl.obolibrary.org/obo/ECO_>\n",
      "PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX rh: <http://rdf.rhea-db.org/>\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "\n",
      "SELECT DISTINCT ?rhea\n",
      "WHERE {\n",
      "  {  \n",
      "    SELECT DISTINCT ?rhea\n",
      "    WHERE {\n",
      "      GRAPH <https://sparql.rhea-db.org/rhea> {\n",
      "        ?rhea rdfs:subClassOf rh:Reaction .\n",
      "        ?rhea rh:side/rh:contains/rh:compound ?compound2 .\n",
      "        ?uc rdfs:subClassOf rh:Compound .\n",
      "      }\n",
      "      ?compound2 rdfs:subClassOf ?uc .\n",
      "      BIND(IF(?uc = rh:SmallMolecule, 0, 1) AS ?c)\n",
      "    } \n",
      "    GROUP BY ?rhea  \n",
      "    HAVING (SUM(?c) = 0)\n",
      "  }\n",
      "  ?catalyzedReaction up:catalyzedReaction ?rhea .\n",
      "  ?reif rdf:object ?catalyzedReaction ;\n",
      "        up:attribution ?attr .\n",
      "  ?attr up:evidence ECO:0000269 ;\n",
      "        up:source ?source .\n",
      "  ?source a up:Citation .\n",
      "}\n",
      "Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "    PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\n",
      "    SELECT DISTINCT ?rhea\n",
      "    WHERE {\n",
      "    # ECO 269 is experimental evidence\n",
      "    BIND (<http://purl.obolibrary.org/obo/ECO_0000269> as ?evidence)\n",
      "    #GRAPH <http://sparql.uniprot.org/uniprot> {\n",
      "        ?protein up:reviewed true ;\n",
      "        up:annotation ?a ;\n",
      "        up:attribution ?attribution  .\n",
      "\n",
      "        ?a a up:Catalytic_Activity_Annotation ;\n",
      "        up:catalyticActivity ?ca .\n",
      "        ?ca up:catalyzedReaction ?rhea .\n",
      "\n",
      "        [] rdf:subject ?a ;\n",
      "        rdf:predicate up:catalyticActivity ;\n",
      "        rdf:object ?ca ;\n",
      "        up:attribution ?attribution .\n",
      "\n",
      "        ?attribution up:evidence ?evidence .\n",
      "    #}\n",
      "    }\n",
      "\n",
      "Missing from generated: {'rhea': {'type': 'uri', 'value': 'http://rdf.rhea-db.org/21408'}}\n",
      "‚ùå 3/3 How could I download a table that only includes the Rhea reactions for which there is experimental evidence? Return only the rhea URI\n",
      "\n",
      "Results mismatch. Ref: 11861 != gen: 4880\n",
      "\n",
      "# https://sparql.uniprot.org/sparql\n",
      "PREFIX ECO: <http://purl.obolibrary.org/obo/ECO_>\n",
      "PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX rh: <http://rdf.rhea-db.org/>\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "\n",
      "SELECT DISTINCT ?rhea\n",
      "WHERE {\n",
      "  {\n",
      "    SELECT DISTINCT ?rhea\n",
      "    WHERE {\n",
      "      GRAPH <https://sparql.rhea-db.org/rhea> {\n",
      "        ?rhea rdfs:subClassOf rh:Reaction .\n",
      "        ?rhea rh:side/rh:contains/rh:compound ?compound2 .\n",
      "        ?uc rdfs:subClassOf rh:Compound .\n",
      "      }\n",
      "      ?compound2 rdfs:subClassOf ?uc .\n",
      "      BIND(IF(?uc = rh:SmallMolecule, 0, 1) AS ?c)\n",
      "    }\n",
      "    GROUP BY ?rhea\n",
      "    HAVING (SUM(?c) = 0)\n",
      "  }\n",
      "  ?catalyzedReaction up:catalyzedReaction ?rhea .\n",
      "  ?reif rdf:object ?catalyzedReaction ;\n",
      "        up:attribution ?attr .\n",
      "  ?attr up:evidence ECO:0000269 ;\n",
      "        up:source ?source .\n",
      "  ?source a up:Citation .\n",
      "}\n",
      "Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "    PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\n",
      "    SELECT DISTINCT ?rhea\n",
      "    WHERE {\n",
      "    # ECO 269 is experimental evidence\n",
      "    BIND (<http://purl.obolibrary.org/obo/ECO_0000269> as ?evidence)\n",
      "    #GRAPH <http://sparql.uniprot.org/uniprot> {\n",
      "        ?protein up:reviewed true ;\n",
      "        up:annotation ?a ;\n",
      "        up:attribution ?attribution  .\n",
      "\n",
      "        ?a a up:Catalytic_Activity_Annotation ;\n",
      "        up:catalyticActivity ?ca .\n",
      "        ?ca up:catalyzedReaction ?rhea .\n",
      "\n",
      "        [] rdf:subject ?a ;\n",
      "        rdf:predicate up:catalyticActivity ;\n",
      "        rdf:object ?ca ;\n",
      "        up:attribution ?attribution .\n",
      "\n",
      "        ?attribution up:evidence ?evidence .\n",
      "    #}\n",
      "    }\n",
      "\n",
      "üéØ No RAG - Success: 1, Different results: 0, No results: 7, Error: 1\n",
      "üéØ RAG with validation - Success: 5, Different results: 4, No results: 0, Error: 0\n",
      "Missing from generated: {'reaction': {'type': 'uri', 'value': 'http://rdf.rhea-db.org/21104'}, 'sterol': {'type': 'uri', 'value': 'http://purl.obolibrary.org/obo/CHEBI_16113'}, 'protein': {'type': 'uri', 'value': 'http://purl.uniprot.org/uniprot/O95992'}}\n",
      "Missing from generated: {'reaction': {'type': 'uri', 'value': 'http://rdf.rhea-db.org/21104'}, 'sterol': {'type': 'uri', 'value': 'http://purl.obolibrary.org/obo/CHEBI_16113'}, 'protein': {'type': 'uri', 'value': 'http://purl.uniprot.org/uniprot/O95992'}}\n",
      "Missing from reference: {'reaction': {'type': 'uri', 'value': 'http://rdf.rhea-db.org/21104'}, 'chebi': {'type': 'uri', 'value': 'http://purl.obolibrary.org/obo/CHEBI_16113'}, 'protein': {'type': 'uri', 'value': 'http://purl.uniprot.org/uniprot/O95992'}}\n",
      "‚ùå 1/3 Which human proteins are enzymes catalyzing a reaction involving sterols? Return the protein, sterol and reaction URI\n",
      "\n",
      "Results mismatch. Ref: 721 != gen: 721\n",
      "\n",
      "# https://sparql.uniprot.org/sparql/\n",
      "PREFIX CHEBI: <http://purl.obolibrary.org/obo/CHEBI_>\n",
      "PREFIX rdfs:  <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX rh:    <http://rdf.rhea-db.org/>\n",
      "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX up:    <http://purl.uniprot.org/core/>\n",
      "\n",
      "SELECT DISTINCT ?protein ?chebi ?reaction WHERE {\n",
      "  SERVICE <https://sparql.rhea-db.org/sparql> {\n",
      "    ?reaction rdfs:subClassOf rh:Reaction .\n",
      "    ?reaction rh:side/rh:contains/rh:compound ?compound .\n",
      "    ?compound rh:chebi ?chebi .\n",
      "    ?chebi rdfs:subClassOf* CHEBI:15889 .\n",
      "  }\n",
      "  ?protein up:organism taxon:9606 .\n",
      "  ?protein up:annotation ?a .\n",
      "  ?a a up:Catalytic_Activity_Annotation .\n",
      "  ?a up:catalyticActivity ?ca .\n",
      "  ?ca up:catalyzedReaction ?reaction .\n",
      "}\n",
      "Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "    PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
      "    PREFIX rh: <http://rdf.rhea-db.org/>\n",
      "    PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "    PREFIX CHEBI: <http://purl.obolibrary.org/obo/CHEBI_>\n",
      "    SELECT DISTINCT ?protein ?sterol ?reaction\n",
      "    WHERE {\n",
      "    SERVICE <https://sparql.rhea-db.org/sparql> {\n",
      "        ?reaction rdfs:subClassOf rh:Reaction .\n",
      "        ?reaction rh:side/rh:contains/rh:compound ?compound .\n",
      "        ?compound rh:chebi ?sterol .\n",
      "        ?sterol rdfs:subClassOf* CHEBI:15889 .\n",
      "    }\n",
      "    ?protein a up:Protein ;\n",
      "        up:organism taxon:9606 ;\n",
      "        up:annotation/up:catalyticActivity/up:catalyzedReaction ?reaction .\n",
      "    }\n",
      "\n",
      "‚úÖ 2/3 Which human proteins are enzymes catalyzing a reaction involving sterols? Return the protein, sterol and reaction URI = 721\n",
      "‚úÖ 3/3 Which human proteins are enzymes catalyzing a reaction involving sterols? Return the protein, sterol and reaction URI = 721\n",
      "üéØ No RAG - Success: 1, Different results: 0, No results: 9, Error: 2\n",
      "üéØ RAG with validation - Success: 7, Different results: 5, No results: 0, Error: 0\n",
      "Missing from generated: {'hgncSymbol': {'type': 'literal', 'value': 'MLH1'}, 'diseaseLabel': {'type': 'literal', 'value': 'Colorectal cancer'}}\n",
      "Missing from generated: {'hgncSymbol': {'type': 'literal', 'value': 'MLH1'}, 'diseaseLabel': {'type': 'literal', 'value': 'Colorectal cancer'}}\n",
      "Missing from generated: {'hgncSymbol': {'type': 'literal', 'value': 'MLH1'}, 'diseaseLabel': {'type': 'literal', 'value': 'Colorectal cancer'}}\n",
      "‚úÖ 1/3 Which are the human proteins associated with cancer? Return distinct ?diseaseLabel and ?hgncSymbol = 122\n",
      "Missing from reference: {'hgncSymbol': {'type': 'literal', 'value': 'AKT1'}, 'diseaseLabel': {'type': 'literal', 'value': 'Genetic variations in AKT1 may play a role in susceptibility to ovarian cancer.'}}\n",
      "‚ùå 2/3 Which are the human proteins associated with cancer? Return distinct ?diseaseLabel and ?hgncSymbol\n",
      "\n",
      "Results mismatch. Ref: 122 != gen: 98\n",
      "\n",
      "# https://sparql.uniprot.org/sparql/\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
      "\n",
      "SELECT DISTINCT ?diseaseLabel ?hgncSymbol\n",
      "WHERE {\n",
      "  ?protein a up:Protein ;\n",
      "           up:organism taxon:9606 ;\n",
      "           up:encodedBy ?gene ;\n",
      "           up:annotation ?diseaseAnnotation .\n",
      "  ?diseaseAnnotation a up:Disease_Annotation ;\n",
      "                     rdfs:comment ?diseaseLabel .\n",
      "  ?gene skos:prefLabel ?hgncSymbol .\n",
      "  FILTER(regex(?diseaseLabel, \"cancer\", \"i\"))\n",
      "}\n",
      "Correct query:\n",
      "PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
      "    PREFIX up:<http://purl.uniprot.org/core/>\n",
      "    PREFIX taxon:<http://purl.uniprot.org/taxonomy/>\n",
      "    PREFIX rdfs:<http://www.w3.org/2000/01/rdf-schema#>\n",
      "    PREFIX orth:<http://purl.org/net/orth#>\n",
      "    PREFIX dcterms:<http://purl.org/dc/terms/>\n",
      "    PREFIX obo:<http://purl.obolibrary.org/obo/>\n",
      "    PREFIX lscr:<http://purl.org/lscr#>\n",
      "    PREFIX genex:<http://purl.org/genex#>\n",
      "    PREFIX sio: <http://semanticscience.org/resource/>\n",
      "    SELECT DISTINCT ?diseaseLabel ?hgncSymbol\n",
      "    WHERE {\n",
      "        ?humanProtein a up:Protein ;\n",
      "            # up:organism/up:scientificName 'Homo sapiens' ;\n",
      "            up:organism taxon:9606 ;\n",
      "            up:annotation ?annotation ;\n",
      "            rdfs:seeAlso ?hgnc .\n",
      "        ?hgnc up:database <http://purl.uniprot.org/database/HGNC> ;\n",
      "            rdfs:comment ?hgncSymbol .\n",
      "        ?annotation a up:Disease_Annotation ;\n",
      "            up:disease ?disease .\n",
      "        ?disease skos:prefLabel ?diseaseLabel.\n",
      "        FILTER CONTAINS (LCASE(?diseaseLabel), \"cancer\")\n",
      "    }\n",
      "\n",
      "‚ùå 3/3 Which are the human proteins associated with cancer? Return distinct ?diseaseLabel and ?hgncSymbol\n",
      "400 Client Error:  for url: https://sparql.uniprot.org/sparql/?query=%23+https%3A%2F%2Fsparql.uniprot.org%2Fsparql%2F%0APREFIX+rdfs%3A++%3Chttp%3A%2F%2Fwww.w3.org%2F2000%2F01%2Frdf-schema%23%3E%0APREFIX+skos%3A++%3Chttp%3A%2F%2Fwww.w3.org%2F2004%2F02%2Fskos%2Fcore%23%3E%0APREFIX+up%3A++++%3Chttp%3A%2F%2Fpurl.uniprot.org%2Fcore%2F%3E%0APREFIX+taxon%3A+%3Chttp%3A%2F%2Fpurl.uniprot.org%2Ftaxonomy%2F%3E%0A%0ASELECT+DISTINCT+%3FdiseaseLabel+%3FhgncSymbol%0AWHERE+%7B%0A++%23+Use+the+MeSH+service+to+relate+diseases+to+cancer+%28MeSH+%22Neoplasms%22%3A+D009369%29%0A++SERVICE+%3Chttps%3A%2F%2Fid.nlm.nih.gov%2Fmesh%2Fsparql%3E+%7B%0A++++GRAPH+%3Chttp%3A%2F%2Fid.nlm.nih.gov%2Fmesh%3E+%7B%0A++++++%3Fmesh+%3Chttp%3A%2F%2Fid.nlm.nih.gov%2Fmesh%2Fvocab%23broaderDescriptor%3E%2A+%3Fbroader+.%0A++++++%3Fbroader+%3Chttp%3A%2F%2Fid.nlm.nih.gov%2Fmesh%2Fvocab%23preferredConcept%3E+%3Chttp%3A%2F%2Fid.nlm.nih.gov%2Fmesh%2FD009369%3E+.%0A++++%7D%0A++%7D%0A++%23+Match+disease+entities+linked+to+MeSH+cancer+terms%0A++GRAPH+%3Chttp%3A%2F%2Fsparql.uniprot.org%2Fdiseases%3E+%7B%0A++++%3Fdisease+a+up%3ADisease+%3B%0A++++++rdfs%3AseeAlso+%3Fmesh+%3B%0A++++++skos%3AprefLabel+%3FdiseaseLabel+.%0A++++%3Fmesh+up%3Adatabase+%3Chttp%3A%2F%2Fpurl.uniprot.org%2Fdatabase%2FMeSH%3E+.%0A++%7D%0A++%23+Link+to+human+proteins+and+retrieve+their+gene+HGNC+symbol+using+the+preferred+gene+name%0A++GRAPH+%3Chttp%3A%2F%2Fsparql.uniprot.org%2Funiprot%3E+%7B%0A++++%3Fprotein+a+up%3AProtein+%3B%0A+++++++++++++up%3Aorganism+taxon%3A9606+%3B%0A+++++++++++++up%3Aannotation%2Fup%3Adisease+%3Fdisease+%3B%0A+++++++++++++up%3AencodedBy+%3Fgene+.%0A++++%3Fgene+skos%3AprefLabel+%3FhgncSymbol+.%0A++%7D%0A%7D\n",
      "# https://sparql.uniprot.org/sparql/\n",
      "PREFIX rdfs:  <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX skos:  <http://www.w3.org/2004/02/skos/core#>\n",
      "PREFIX up:    <http://purl.uniprot.org/core/>\n",
      "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
      "\n",
      "SELECT DISTINCT ?diseaseLabel ?hgncSymbol\n",
      "WHERE {\n",
      "  # Use the MeSH service to relate diseases to cancer (MeSH \"Neoplasms\": D009369)\n",
      "  SERVICE <https://id.nlm.nih.gov/mesh/sparql> {\n",
      "    GRAPH <http://id.nlm.nih.gov/mesh> {\n",
      "      ?mesh <http://id.nlm.nih.gov/mesh/vocab#broaderDescriptor>* ?broader .\n",
      "      ?broader <http://id.nlm.nih.gov/mesh/vocab#preferredConcept> <http://id.nlm.nih.gov/mesh/D009369> .\n",
      "    }\n",
      "  }\n",
      "  # Match disease entities linked to MeSH cancer terms\n",
      "  GRAPH <http://sparql.uniprot.org/diseases> {\n",
      "    ?disease a up:Disease ;\n",
      "      rdfs:seeAlso ?mesh ;\n",
      "      skos:prefLabel ?diseaseLabel .\n",
      "    ?mesh up:database <http://purl.uniprot.org/database/MeSH> .\n",
      "  }\n",
      "  # Link to human proteins and retrieve their gene HGNC symbol using the preferred gene name\n",
      "  GRAPH <http://sparql.uniprot.org/uniprot> {\n",
      "    ?protein a up:Protein ;\n",
      "             up:organism taxon:9606 ;\n",
      "             up:annotation/up:disease ?disease ;\n",
      "             up:encodedBy ?gene .\n",
      "    ?gene skos:prefLabel ?hgncSymbol .\n",
      "  }\n",
      "}\n",
      "Correct query:\n",
      "PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
      "    PREFIX up:<http://purl.uniprot.org/core/>\n",
      "    PREFIX taxon:<http://purl.uniprot.org/taxonomy/>\n",
      "    PREFIX rdfs:<http://www.w3.org/2000/01/rdf-schema#>\n",
      "    PREFIX orth:<http://purl.org/net/orth#>\n",
      "    PREFIX dcterms:<http://purl.org/dc/terms/>\n",
      "    PREFIX obo:<http://purl.obolibrary.org/obo/>\n",
      "    PREFIX lscr:<http://purl.org/lscr#>\n",
      "    PREFIX genex:<http://purl.org/genex#>\n",
      "    PREFIX sio: <http://semanticscience.org/resource/>\n",
      "    SELECT DISTINCT ?diseaseLabel ?hgncSymbol\n",
      "    WHERE {\n",
      "        ?humanProtein a up:Protein ;\n",
      "            # up:organism/up:scientificName 'Homo sapiens' ;\n",
      "            up:organism taxon:9606 ;\n",
      "            up:annotation ?annotation ;\n",
      "            rdfs:seeAlso ?hgnc .\n",
      "        ?hgnc up:database <http://purl.uniprot.org/database/HGNC> ;\n",
      "            rdfs:comment ?hgncSymbol .\n",
      "        ?annotation a up:Disease_Annotation ;\n",
      "            up:disease ?disease .\n",
      "        ?disease skos:prefLabel ?diseaseLabel.\n",
      "        FILTER CONTAINS (LCASE(?diseaseLabel), \"cancer\")\n",
      "    }\n",
      "\n",
      "üéØ No RAG - Success: 1, Different results: 0, No results: 12, Error: 2\n",
      "üéØ RAG with validation - Success: 8, Different results: 6, No results: 0, Error: 1\n",
      "Missing from generated: {'gene': {'type': 'uri', 'value': 'http://omabrowser.org/ontology/oma#GENE_WBGene00001030'}, 'confidence': {'type': 'uri', 'value': 'http://purl.obolibrary.org/obo/CIO_0000029'}, 'fdr': {'type': 'typed-literal', 'datatype': 'http://www.w3.org/2001/XMLSchema#double', 'value': '0.0016728772206653400815'}}\n",
      "Missing from generated: {'gene': {'type': 'uri', 'value': 'http://omabrowser.org/ontology/oma#GENE_WBGene00001030'}, 'confidence': {'type': 'uri', 'value': 'http://purl.obolibrary.org/obo/CIO_0000029'}, 'fdr': {'type': 'typed-literal', 'datatype': 'http://www.w3.org/2001/XMLSchema#double', 'value': '0.0016728772206653400815'}}\n",
      "Missing from generated: {'gene': {'type': 'uri', 'value': 'http://omabrowser.org/ontology/oma#GENE_WBGene00001030'}, 'confidence': {'type': 'uri', 'value': 'http://purl.obolibrary.org/obo/CIO_0000029'}, 'fdr': {'type': 'typed-literal', 'datatype': 'http://www.w3.org/2001/XMLSchema#double', 'value': '0.0016728772206653400815'}}\n",
      "‚úÖ 1/3 In bgee how can I retrieve the confidence level and false discovery rate of a gene expression? Return distinct ?gene, ?confidence and ?fdr, limit to 10 = 10\n",
      "‚úÖ 2/3 In bgee how can I retrieve the confidence level and false discovery rate of a gene expression? Return distinct ?gene, ?confidence and ?fdr, limit to 10 = 10\n",
      "‚úÖ 3/3 In bgee how can I retrieve the confidence level and false discovery rate of a gene expression? Return distinct ?gene, ?confidence and ?fdr, limit to 10 = 10\n",
      "üéØ No RAG - Success: 1, Different results: 0, No results: 15, Error: 2\n",
      "üéØ RAG with validation - Success: 11, Different results: 6, No results: 0, Error: 1\n",
      "Missing from generated: {'ensemblURI': {'type': 'uri', 'value': 'http://rdf.ebi.ac.uk/resource/ensembl.protein/ENSCVAP00000023326'}}\n",
      "Missing from reference: {'ensemblURI': {'type': 'literal', 'value': 'ENSSDUG00000009394'}}\n",
      "‚ùå 1/3 How can I get the cross-reference to the ensembl protein for the LCT protein in OMA? Return only the distinct ?ensemblURI\n",
      "\n",
      "Results mismatch. Ref: 102 != gen: 102\n",
      "\n",
      "# https://sparql.omabrowser.org/sparql/\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "PREFIX sio: <http://semanticscience.org/resource/>\n",
      "PREFIX dc: <http://purl.org/dc/terms/>\n",
      "\n",
      "SELECT DISTINCT ?ensemblURI\n",
      "WHERE {\n",
      "  ?protein a orth:Protein ;\n",
      "           rdfs:label \"LCT\" ;\n",
      "           sio:SIO_010079/lscr:xrefEnsemblGene ?ensemblGene .\n",
      "  ?ensemblGene dc:identifier ?ensemblURI .\n",
      "}\n",
      "Correct query:\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "    PREFIX lscr: <http://purl.org/lscr#>\n",
      "    PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "    SELECT DISTINCT ?ensemblURI\n",
      "    WHERE {\n",
      "        ?protein a orth:Protein ;\n",
      "            rdfs:label 'LCT' ;\n",
      "            lscr:xrefEnsemblProtein ?ensemblURI .\n",
      "    }\n",
      "\n",
      "‚ùå 2/3 How can I get the cross-reference to the ensembl protein for the LCT protein in OMA? Return only the distinct ?ensemblURI\n",
      "list index out of range\n",
      "# https://sparql.omabrowser.org/sparql/\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "PREFIX sio: <http://semanticscience.org/resource/>\n",
      "PREFIX dc: <http://purl.org/dc/terms/>\n",
      "\n",
      "SELECT DISTINCT ?ensemblURI\n",
      "WHERE {\n",
      "  ?protein a orth:Protein ;\n",
      "           rdfs:label \"LCT\" ;\n",
      "           sio:SIO_010079/lscr:xrefEnsemblGene ?ensemblGene .\n",
      "  ?ensemblGene dc:identifier ?ensemblURI .\n",
      "}\n",
      "Correct query:\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "    PREFIX lscr: <http://purl.org/lscr#>\n",
      "    PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "    SELECT DISTINCT ?ensemblURI\n",
      "    WHERE {\n",
      "        ?protein a orth:Protein ;\n",
      "            rdfs:label 'LCT' ;\n",
      "            lscr:xrefEnsemblProtein ?ensemblURI .\n",
      "    }\n",
      "\n",
      "Missing from reference: {'ensemblURI': {'type': 'literal', 'value': 'ENSSDUG00000009394'}}\n",
      "‚ùå 3/3 How can I get the cross-reference to the ensembl protein for the LCT protein in OMA? Return only the distinct ?ensemblURI\n",
      "\n",
      "Results mismatch. Ref: 102 != gen: 102\n",
      "\n",
      "# https://sparql.omabrowser.org/sparql/\n",
      "PREFIX rdfs:  <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX orth:  <http://purl.org/net/orth#>\n",
      "PREFIX lscr:  <http://purl.org/lscr#>\n",
      "PREFIX sio:   <http://semanticscience.org/resource/>\n",
      "PREFIX dc:    <http://purl.org/dc/terms/>\n",
      "\n",
      "SELECT DISTINCT ?ensemblURI\n",
      "WHERE {\n",
      "  ?protein a orth:Protein ;\n",
      "           rdfs:label \"LCT\" ;\n",
      "           sio:SIO_010079/lscr:xrefEnsemblGene/dc:identifier ?ensemblURI .\n",
      "}\n",
      "Correct query:\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "    PREFIX lscr: <http://purl.org/lscr#>\n",
      "    PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "    SELECT DISTINCT ?ensemblURI\n",
      "    WHERE {\n",
      "        ?protein a orth:Protein ;\n",
      "            rdfs:label 'LCT' ;\n",
      "            lscr:xrefEnsemblProtein ?ensemblURI .\n",
      "    }\n",
      "\n",
      "üéØ No RAG - Success: 1, Different results: 0, No results: 16, Error: 4\n",
      "üéØ RAG with validation - Success: 11, Different results: 8, No results: 0, Error: 2\n",
      "Missing from generated: {'orthologCluster': {'type': 'uri', 'value': 'https://omabrowser.org/oma/hog/resolve/HOG:E0000012_-1701473419'}, 'datasetURI': {'type': 'uri', 'value': 'https://omabrowser.org/oma/current/#DATASET_OMA'}}\n",
      "Missing from generated: {'orthologCluster': {'type': 'uri', 'value': 'https://omabrowser.org/oma/hog/resolve/HOG:E0000012_-1701473419'}, 'datasetURI': {'type': 'uri', 'value': 'https://omabrowser.org/oma/current/#DATASET_OMA'}}\n",
      "‚úÖ 1/3 How can I get the URI of a dataset to which an ortholog cluster belongs in OMA? Return orthologCluster, datasetURI and limit to 20 = 20\n",
      "‚úÖ 2/3 How can I get the URI of a dataset to which an ortholog cluster belongs in OMA? Return orthologCluster, datasetURI and limit to 20 = 20\n",
      "‚úÖ 3/3 How can I get the URI of a dataset to which an ortholog cluster belongs in OMA? Return orthologCluster, datasetURI and limit to 20 = 20\n",
      "üéØ No RAG - Success: 1, Different results: 0, No results: 18, Error: 5\n",
      "üéØ RAG with validation - Success: 14, Different results: 8, No results: 0, Error: 2\n",
      "Missing from generated: {'strain': {'type': 'uri', 'value': 'http://purl.uniprot.org/taxonomy/573235#strain-11368%20%2F%20EHEC'}, 'name': {'type': 'literal', 'value': '11368 / EHEC'}, 'taxon': {'type': 'uri', 'value': 'http://purl.uniprot.org/taxonomy/573235'}}\n",
      "Missing from generated: {'strain': {'type': 'uri', 'value': 'http://purl.uniprot.org/taxonomy/573235#strain-11368%20%2F%20EHEC'}, 'name': {'type': 'literal', 'value': '11368 / EHEC'}, 'taxon': {'type': 'uri', 'value': 'http://purl.uniprot.org/taxonomy/573235'}}\n",
      "Missing from generated: {'strain': {'type': 'uri', 'value': 'http://purl.uniprot.org/taxonomy/573235#strain-11368%20%2F%20EHEC'}, 'name': {'type': 'literal', 'value': '11368 / EHEC'}, 'taxon': {'type': 'uri', 'value': 'http://purl.uniprot.org/taxonomy/573235'}}\n",
      "‚ùå 1/3 Give me the list of strains associated to the Escherichia coli taxon and their name. Return ?taxon, ?strain, ?name, limit to 20\n",
      "\n",
      "Results mismatch. Ref: 20 != gen: 0\n",
      "\n",
      "# https://sparql.uniprot.org/sparql/\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "\n",
      "SELECT ?taxon ?strain ?name\n",
      "WHERE {\n",
      "  BIND(taxon:83333 AS ?taxon)\n",
      "  ?strain a up:Taxon ;\n",
      "          rdfs:subClassOf ?taxon ;\n",
      "          up:rank ?rank ;\n",
      "          up:scientificName ?name .\n",
      "  ?rank rdfs:label ?rankLabel .\n",
      "  FILTER(lcase(str(?rankLabel)) = \"strain\")\n",
      "}\n",
      "LIMIT 20\n",
      "Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "    PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
      "    PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "    SELECT ?taxon ?strain ?name\n",
      "    WHERE {\n",
      "        ?taxon a up:Taxon .\n",
      "        ?taxon rdfs:subClassOf taxon:562 . # Escherichia coli taxon\n",
      "        ?taxon up:strain ?strain .\n",
      "        ?strain up:name ?name .\n",
      "    } LIMIT 20\n",
      "\n",
      "Missing from generated: {'strain': {'type': 'uri', 'value': 'http://purl.uniprot.org/taxonomy/573235#strain-11368%20%2F%20EHEC'}, 'name': {'type': 'literal', 'value': '11368 / EHEC'}, 'taxon': {'type': 'uri', 'value': 'http://purl.uniprot.org/taxonomy/573235'}}\n",
      "‚ùå 2/3 Give me the list of strains associated to the Escherichia coli taxon and their name. Return ?taxon, ?strain, ?name, limit to 20\n",
      "\n",
      "Results mismatch. Ref: 20 != gen: 0\n",
      "\n",
      "# https://sparql.uniprot.org/sparql/\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "\n",
      "SELECT ?taxon ?strain ?name\n",
      "WHERE {\n",
      "  ?taxon a up:Taxon ;\n",
      "         up:rank up:Strain ;\n",
      "         up:scientificName ?name ;\n",
      "         rdfs:subClassOf* taxon:562 .\n",
      "  BIND(?taxon AS ?strain)\n",
      "}\n",
      "LIMIT 20\n",
      "Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "    PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
      "    PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "    SELECT ?taxon ?strain ?name\n",
      "    WHERE {\n",
      "        ?taxon a up:Taxon .\n",
      "        ?taxon rdfs:subClassOf taxon:562 . # Escherichia coli taxon\n",
      "        ?taxon up:strain ?strain .\n",
      "        ?strain up:name ?name .\n",
      "    } LIMIT 20\n",
      "\n",
      "Missing from reference: {'strain': {'type': 'uri', 'value': 'http://purl.uniprot.org/taxonomy/83333#strain-K12%20%2F%20KS650'}, 'name': {'type': 'literal', 'value': 'K12 / KS650'}, 'taxon': {'type': 'uri', 'value': 'http://purl.uniprot.org/taxonomy/83333'}}\n",
      "‚ùå 3/3 Give me the list of strains associated to the Escherichia coli taxon and their name. Return ?taxon, ?strain, ?name, limit to 20\n",
      "\n",
      "Results mismatch. Ref: 20 != gen: 20\n",
      "\n",
      "# https://sparql.uniprot.org/sparql/\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
      "\n",
      "SELECT ?taxon ?strain ?name\n",
      "WHERE {\n",
      "  BIND(taxon:83333 AS ?taxon) .\n",
      "  ?taxon up:strain ?strain .\n",
      "  ?strain up:name ?name .\n",
      "}\n",
      "LIMIT 20\n",
      "Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "    PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
      "    PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "    SELECT ?taxon ?strain ?name\n",
      "    WHERE {\n",
      "        ?taxon a up:Taxon .\n",
      "        ?taxon rdfs:subClassOf taxon:562 . # Escherichia coli taxon\n",
      "        ?taxon up:strain ?strain .\n",
      "        ?strain up:name ?name .\n",
      "    } LIMIT 20\n",
      "\n",
      "üéØ No RAG - Success: 1, Different results: 0, No results: 20, Error: 6\n",
      "üéØ RAG with validation - Success: 14, Different results: 9, No results: 2, Error: 2\n",
      "Missing from generated: {'proteinLabel': {'type': 'literal', 'value': 'Enolase'}, 'pathwayLabel': {'type': 'literal', 'value': 'Carbohydrate degradation; glycolysis; pyruvate from D-glyceraldehyde 3-phosphate: step 4/5'}, 'proteinURI': {'type': 'uri', 'value': 'http://purl.uniprot.org/uniprot/A0A497JFR8'}}\n",
      "Missing from reference: {'proteinLabel': {'type': 'literal', 'value': 'ATP-dependent 6-phosphofructokinase'}, 'pathwayLabel': {'type': 'literal', 'value': 'canonical glycolysis'}, 'proteinURI': {'type': 'uri', 'value': 'http://purl.uniprot.org/uniprot/A0A059MPQ6'}}\n",
      "‚úÖ 1/3 Retrieve all proteins involved in pathways involving glycolysis. Return ?proteinURI, ?proteinLabel, ?pathwayLabel, limit to 20 = 20\n",
      "‚úÖ 2/3 Retrieve all proteins involved in pathways involving glycolysis. Return ?proteinURI, ?proteinLabel, ?pathwayLabel, limit to 20 = 20\n",
      "‚úÖ 3/3 Retrieve all proteins involved in pathways involving glycolysis. Return ?proteinURI, ?proteinLabel, ?pathwayLabel, limit to 20 = 20\n",
      "üéØ No RAG - Success: 1, Different results: 1, No results: 21, Error: 7\n",
      "üéØ RAG with validation - Success: 17, Different results: 9, No results: 2, Error: 2\n",
      "Missing from generated: {'ratProtein': {'type': 'uri', 'value': 'https://omabrowser.org/oma/info/RATNO00407'}, 'ratUniProtXref': {'type': 'uri', 'value': 'http://purl.uniprot.org/uniprot/A0A0G2K4L4'}}\n",
      "Missing from generated: {'ratProtein': {'type': 'uri', 'value': 'https://omabrowser.org/oma/info/RATNO00407'}, 'ratUniProtXref': {'type': 'uri', 'value': 'http://purl.uniprot.org/uniprot/A0A0G2K4L4'}}\n",
      "Missing from generated: {'ratProtein': {'type': 'uri', 'value': 'https://omabrowser.org/oma/info/RATNO00407'}, 'ratUniProtXref': {'type': 'uri', 'value': 'http://purl.uniprot.org/uniprot/A0A0G2K4L4'}}\n",
      "Missing from generated: {'ratProtein': {'type': 'uri', 'value': 'https://omabrowser.org/oma/info/RATNO00407'}, 'ratUniProtXref': {'type': 'uri', 'value': 'http://purl.uniprot.org/uniprot/A0A0G2K4L4'}}\n",
      "‚ùå 1/3 What are the orthologs in rat for protein Q9Y2T1 ? Return ?ratProtein ?ratUniProtXref\n",
      "\n",
      "Results mismatch. Ref: 3 != gen: 0\n",
      "\n",
      "# https://sparql.omabrowser.org/sparql/\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
      "\n",
      "SELECT DISTINCT ?ratProtein ?ratUniProtXref\n",
      "WHERE {\n",
      "  ?cluster a orth:OrthologsCluster ;\n",
      "           orth:hasHomologousMember ?node1 ;\n",
      "           orth:hasHomologousMember ?node2 .\n",
      "           \n",
      "  ?node1 orth:hasHomologousMember* ?protein1 .\n",
      "  ?protein1 a orth:Protein ;\n",
      "            rdfs:seeAlso <http://purl.uniprot.org/uniprot/Q9Y2T1> .\n",
      "            \n",
      "  ?node2 orth:hasHomologousMember* ?ratProtein .\n",
      "  ?ratProtein a orth:Protein ;\n",
      "              orth:organism/obo:RO_0002162/up:scientificName \"Rattus norvegicus\" ;\n",
      "              lscr:xrefUniprot ?ratUniProtXref .\n",
      "              \n",
      "  FILTER(?node1 != ?node2)\n",
      "}\n",
      "Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
      "SELECT DISTINCT ?ratProtein ?ratUniProtXref\n",
      "WHERE {\n",
      "    ?cluster a orth:OrthologsCluster.\n",
      "    ?cluster orth:hasHomologousMember ?node1.\n",
      "    ?cluster orth:hasHomologousMember ?node2.\n",
      "    ?node1 orth:hasHomologousMember* ?humanProtein.\n",
      "    ?node2 orth:hasHomologousMember* ?ratProtein.\n",
      "    ?humanProtein a orth:Protein;\n",
      "        lscr:xrefUniprot <http://purl.uniprot.org/uniprot/Q9Y2T1>.\n",
      "    ?ratProtein a orth:Protein;\n",
      "        orth:organism/obo:RO_0002162/up:scientificName 'Rattus norvegicus';\n",
      "        lscr:xrefUniprot ?ratUniProtXref.\n",
      "    FILTER(?node1 != ?node2)\n",
      "}\n",
      "\n",
      "Missing from generated: {'ratProtein': {'type': 'uri', 'value': 'https://omabrowser.org/oma/info/RATNO00407'}, 'ratUniProtXref': {'type': 'uri', 'value': 'http://purl.uniprot.org/uniprot/A0A0G2K4L4'}}\n",
      "‚ùå 2/3 What are the orthologs in rat for protein Q9Y2T1 ? Return ?ratProtein ?ratUniProtXref\n",
      "\n",
      "Results mismatch. Ref: 3 != gen: 0\n",
      "\n",
      "# https://sparql.omabrowser.org/sparql/\n",
      "PREFIX dc: <http://purl.org/dc/terms/>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
      "\n",
      "SELECT ?ratProtein ?ratUniProtXref\n",
      "WHERE {\n",
      "  ?cluster a orth:OrthologsCluster ;\n",
      "           orth:hasHomologousMember ?node1 ;\n",
      "           orth:hasHomologousMember ?node2 .\n",
      "  ?node1 orth:hasHomologousMember* ?protein1 .\n",
      "  ?node2 orth:hasHomologousMember* ?ratProtein .\n",
      "  \n",
      "  ?protein1 a orth:Protein ;\n",
      "            lscr:xrefUniprot \"Q9Y2T1\" .\n",
      "  \n",
      "  ?ratProtein a orth:Protein ;\n",
      "            lscr:xrefUniprot ?ratUniProtXref ;\n",
      "            orth:organism/obo:RO_0002162/up:scientificName \"Rattus norvegicus\" .\n",
      "  \n",
      "  FILTER(?node1 != ?node2)\n",
      "}\n",
      "Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
      "SELECT DISTINCT ?ratProtein ?ratUniProtXref\n",
      "WHERE {\n",
      "    ?cluster a orth:OrthologsCluster.\n",
      "    ?cluster orth:hasHomologousMember ?node1.\n",
      "    ?cluster orth:hasHomologousMember ?node2.\n",
      "    ?node1 orth:hasHomologousMember* ?humanProtein.\n",
      "    ?node2 orth:hasHomologousMember* ?ratProtein.\n",
      "    ?humanProtein a orth:Protein;\n",
      "        lscr:xrefUniprot <http://purl.uniprot.org/uniprot/Q9Y2T1>.\n",
      "    ?ratProtein a orth:Protein;\n",
      "        orth:organism/obo:RO_0002162/up:scientificName 'Rattus norvegicus';\n",
      "        lscr:xrefUniprot ?ratUniProtXref.\n",
      "    FILTER(?node1 != ?node2)\n",
      "}\n",
      "\n",
      "‚úÖ 3/3 What are the orthologs in rat for protein Q9Y2T1 ? Return ?ratProtein ?ratUniProtXref = 3\n",
      "üéØ No RAG - Success: 1, Different results: 1, No results: 24, Error: 7\n",
      "üéØ RAG with validation - Success: 18, Different results: 9, No results: 4, Error: 2\n",
      "Missing from generated: {'locationInsideCellUri': {'type': 'uri', 'value': 'http://purl.uniprot.org/locations/138'}, 'proteinURI': {'type': 'uri', 'value': 'http://purl.uniprot.org/uniprot/Q16643'}, 'locationInsideCellLabel': {'type': 'literal', 'value': 'Cell cortex'}}\n",
      "Missing from generated: {'locationInsideCellUri': {'type': 'uri', 'value': 'http://purl.uniprot.org/locations/138'}, 'proteinURI': {'type': 'uri', 'value': 'http://purl.uniprot.org/uniprot/Q16643'}, 'locationInsideCellLabel': {'type': 'literal', 'value': 'Cell cortex'}}\n",
      "Missing from generated: {'locationInsideCellUri': {'type': 'uri', 'value': 'http://purl.uniprot.org/locations/138'}, 'proteinURI': {'type': 'uri', 'value': 'http://purl.uniprot.org/uniprot/Q16643'}, 'locationInsideCellLabel': {'type': 'literal', 'value': 'Cell cortex'}}\n",
      "‚úÖ 1/3 Retrieve all proteins that are associated with Alzheimer disease (http://purl.uniprot.org/diseases/3832) and where they are known to be located in the cell. Return ?proteinURI, ?locationInsideCellLabel, ?locationInsideCellUri, limit to 20 = 20\n",
      "‚úÖ 2/3 Retrieve all proteins that are associated with Alzheimer disease (http://purl.uniprot.org/diseases/3832) and where they are known to be located in the cell. Return ?proteinURI, ?locationInsideCellLabel, ?locationInsideCellUri, limit to 20 = 20\n",
      "‚úÖ 3/3 Retrieve all proteins that are associated with Alzheimer disease (http://purl.uniprot.org/diseases/3832) and where they are known to be located in the cell. Return ?proteinURI, ?locationInsideCellLabel, ?locationInsideCellUri, limit to 20 = 20\n",
      "üéØ No RAG - Success: 1, Different results: 1, No results: 27, Error: 7\n",
      "üéØ RAG with validation - Success: 21, Different results: 9, No results: 4, Error: 2\n",
      "‚úÖ 1/3 Retrieve all proteins in OMA that are encoded by the TP53 gene and their mnemonics and evidence types from the UniProt database. Return ?proteinOMA ?speciesLabel ?mnemonic ?evidenceType ?uniprotURI = 154\n",
      "‚úÖ 2/3 Retrieve all proteins in OMA that are encoded by the TP53 gene and their mnemonics and evidence types from the UniProt database. Return ?proteinOMA ?speciesLabel ?mnemonic ?evidenceType ?uniprotURI = 154\n",
      "‚úÖ 3/3 Retrieve all proteins in OMA that are encoded by the TP53 gene and their mnemonics and evidence types from the UniProt database. Return ?proteinOMA ?speciesLabel ?mnemonic ?evidenceType ?uniprotURI = 154\n",
      "üéØ No RAG - Success: 1, Different results: 1, No results: 27, Error: 10\n",
      "üéØ RAG with validation - Success: 24, Different results: 9, No results: 4, Error: 2\n",
      "Missing from generated: {'function': {'type': 'literal', 'value': 'Inhibitor of lipoprotein binding to the low density lipoprotein (LDL) receptor, LDL receptor-related protein, and very low density lipoprotein (VLDL) receptor. Associates with high density lipoproteins (HDL) and the triacylglycerol-rich lipoproteins in the plasma and makes up about 10% of the protein of the VLDL and 2% of that of HDL. Appears to interfere directly with fatty acid uptake and is also the major plasma inhibitor of cholesteryl ester transfer protein (CETP). Binds free fatty acids and reduces their intracellular esterification. Modulates the interaction of APOE with beta-migrating VLDL and inhibits binding of beta-VLDL to the LDL receptor-related protein.'}}\n",
      "Missing from generated: {'function': {'type': 'literal', 'value': 'Inhibitor of lipoprotein binding to the low density lipoprotein (LDL) receptor, LDL receptor-related protein, and very low density lipoprotein (VLDL) receptor. Associates with high density lipoproteins (HDL) and the triacylglycerol-rich lipoproteins in the plasma and makes up about 10% of the protein of the VLDL and 2% of that of HDL. Appears to interfere directly with fatty acid uptake and is also the major plasma inhibitor of cholesteryl ester transfer protein (CETP). Binds free fatty acids and reduces their intracellular esterification. Modulates the interaction of APOE with beta-migrating VLDL and inhibits binding of beta-VLDL to the LDL receptor-related protein.'}}\n",
      "‚úÖ 1/3 What is the function of APOC1? Return ?function = 1\n",
      "‚úÖ 2/3 What is the function of APOC1? Return ?function = 1\n",
      "‚úÖ 3/3 What is the function of APOC1? Return ?function = 1\n",
      "üéØ No RAG - Success: 1, Different results: 1, No results: 29, Error: 11\n",
      "üéØ RAG with validation - Success: 27, Different results: 9, No results: 4, Error: 2\n",
      "     Model         RAG Approach  Success  Different Results  No Results  \\\n",
      "0  o3-mini               No RAG        1                  1          29   \n",
      "1  o3-mini  RAG with validation       27                  9           4   \n",
      "\n",
      "   Errors    Price    F1  \n",
      "0      11  0.00719  0.02  \n",
      "1       2  0.01421  0.72  \n",
      "\n",
      "\n",
      "\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "from collections import defaultdict\n",
    "import time\n",
    "\n",
    "import pandas as pd\n",
    "import requests\n",
    "from dotenv import load_dotenv\n",
    "\n",
    "from langchain_core.messages import HumanMessage, SystemMessage\n",
    "\n",
    "from expasy_agent.prompts import SYSTEM_PROMPT\n",
    "from expasy_agent.nodes.llm import load_chat_model\n",
    "from expasy_agent.config import Configuration\n",
    "from sparql_llm.utils import query_sparql\n",
    "from sparql_llm.validate_sparql import extract_sparql_queries\n",
    "\n",
    "load_dotenv()\n",
    "chat_api_key = os.getenv(\"CHAT_API_KEY\")\n",
    "\n",
    "example_queries = [\n",
    "    {\n",
    "        \"question\": \"What is the accession number in uniprot of the human gene LCT? Return only unique protein URIs\",\n",
    "        \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "        \"query\": \"\"\"PREFIX up: <http://purl.uniprot.org/core/>\n",
    "    PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
    "    PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
    "    SELECT DISTINCT ?protein\n",
    "    WHERE{\n",
    "        ?protein a up:Protein .\n",
    "        ?protein up:organism taxon:9606 .\n",
    "        ?protein up:encodedBy ?gene .\n",
    "        ?gene skos:prefLabel \"LCT\" .\n",
    "    }\"\"\",\n",
    "    },\n",
    "    {\n",
    "        # NOTE: The \"mature\" part in the question makes it harder to answer\n",
    "        # \"question\": \"How do I filter for reviewed (mouse) proteins whose mature form carries an N-terminal glycine? Return protein URI and AA sequence\",\n",
    "        \"question\": \"How do I filter for reviewed mouse proteins which carry an N-terminal glycine? Return protein URI and AA sequence\",\n",
    "        \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "        \"query\": \"\"\"PREFIX up: <http://purl.uniprot.org/core/>\n",
    "    PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
    "    PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\n",
    "    SELECT ?protein ?sequence\n",
    "    WHERE\n",
    "    {\n",
    "        ?protein a up:Protein ;\n",
    "            up:organism taxon:10090 ;  # Taxonomy ID for Mus musculus (Mouse)\n",
    "            up:reviewed true ;\n",
    "            up:sequence ?isoform .\n",
    "        ?isoform rdf:value ?sequence .\n",
    "        # Ensure the N-terminal amino acid is Glycine (G)\n",
    "        FILTER (STRSTARTS(?sequence, \"G\"))\n",
    "    }\"\"\",\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"How could I download a table that only includes the Rhea reactions for which there is experimental evidence? Return only the rhea URI\",\n",
    "        \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "        \"query\": \"\"\"PREFIX up: <http://purl.uniprot.org/core/>\n",
    "    PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\n",
    "    SELECT DISTINCT ?rhea\n",
    "    WHERE {\n",
    "    # ECO 269 is experimental evidence\n",
    "    BIND (<http://purl.obolibrary.org/obo/ECO_0000269> as ?evidence)\n",
    "    #GRAPH <http://sparql.uniprot.org/uniprot> {\n",
    "        ?protein up:reviewed true ;\n",
    "        up:annotation ?a ;\n",
    "        up:attribution ?attribution  .\n",
    "\n",
    "        ?a a up:Catalytic_Activity_Annotation ;\n",
    "        up:catalyticActivity ?ca .\n",
    "        ?ca up:catalyzedReaction ?rhea .\n",
    "\n",
    "        [] rdf:subject ?a ;\n",
    "        rdf:predicate up:catalyticActivity ;\n",
    "        rdf:object ?ca ;\n",
    "        up:attribution ?attribution .\n",
    "\n",
    "        ?attribution up:evidence ?evidence .\n",
    "    #}\n",
    "    }\"\"\",\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"Which human proteins are enzymes catalyzing a reaction involving sterols? Return the protein, sterol and reaction URI\",\n",
    "        \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "        \"query\": \"\"\"PREFIX up: <http://purl.uniprot.org/core/>\n",
    "    PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
    "    PREFIX rh: <http://rdf.rhea-db.org/>\n",
    "    PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
    "    PREFIX CHEBI: <http://purl.obolibrary.org/obo/CHEBI_>\n",
    "    SELECT DISTINCT ?protein ?sterol ?reaction\n",
    "    WHERE {\n",
    "    SERVICE <https://sparql.rhea-db.org/sparql> {\n",
    "        ?reaction rdfs:subClassOf rh:Reaction .\n",
    "        ?reaction rh:side/rh:contains/rh:compound ?compound .\n",
    "        ?compound rh:chebi ?sterol .\n",
    "        ?sterol rdfs:subClassOf* CHEBI:15889 .\n",
    "    }\n",
    "    ?protein a up:Protein ;\n",
    "        up:organism taxon:9606 ;\n",
    "        up:annotation/up:catalyticActivity/up:catalyzedReaction ?reaction .\n",
    "    }\"\"\",\n",
    "    },\n",
    "    {\n",
    "        # \"question\": \"Which are the human proteins associated with cancer (which have cancer in their disease label)? Return the unique disease label (?diseaseLabel), and HGNC symbol (?hgncSymbol)\",\n",
    "        \"question\": \"Which are the human proteins associated with cancer? Return distinct ?diseaseLabel and ?hgncSymbol\",\n",
    "        \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "        \"query\": \"\"\"PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
    "    PREFIX up:<http://purl.uniprot.org/core/>\n",
    "    PREFIX taxon:<http://purl.uniprot.org/taxonomy/>\n",
    "    PREFIX rdfs:<http://www.w3.org/2000/01/rdf-schema#>\n",
    "    PREFIX orth:<http://purl.org/net/orth#>\n",
    "    PREFIX dcterms:<http://purl.org/dc/terms/>\n",
    "    PREFIX obo:<http://purl.obolibrary.org/obo/>\n",
    "    PREFIX lscr:<http://purl.org/lscr#>\n",
    "    PREFIX genex:<http://purl.org/genex#>\n",
    "    PREFIX sio: <http://semanticscience.org/resource/>\n",
    "    SELECT DISTINCT ?diseaseLabel ?hgncSymbol\n",
    "    WHERE {\n",
    "        ?humanProtein a up:Protein ;\n",
    "            # up:organism/up:scientificName 'Homo sapiens' ;\n",
    "            up:organism taxon:9606 ;\n",
    "            up:annotation ?annotation ;\n",
    "            rdfs:seeAlso ?hgnc .\n",
    "        ?hgnc up:database <http://purl.uniprot.org/database/HGNC> ;\n",
    "            rdfs:comment ?hgncSymbol .\n",
    "        ?annotation a up:Disease_Annotation ;\n",
    "            up:disease ?disease .\n",
    "        ?disease skos:prefLabel ?diseaseLabel.\n",
    "        FILTER CONTAINS (LCASE(?diseaseLabel), \"cancer\")\n",
    "    }\"\"\",\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"In bgee how can I retrieve the confidence level and false discovery rate of a gene expression? Return distinct ?gene, ?confidence and ?fdr, limit to 10\",\n",
    "        \"endpoint\": \"https://www.bgee.org/sparql/\",\n",
    "        \"query\": \"\"\"PREFIX genex: <http://purl.org/genex#>\n",
    "    PREFIX bgee: <http://bgee.org/#>\n",
    "    PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
    "    SELECT DISTINCT ?gene ?confidence ?fdr\n",
    "    WHERE {\n",
    "        ?expression a genex:Expression ;\n",
    "            genex:hasConfidenceLevel ?confidence ;\n",
    "            genex:hasFDRpvalue ?fdr ;\n",
    "            genex:hasSequenceUnit ?gene .\n",
    "    } LIMIT 10\"\"\",\n",
    "    },\n",
    "    {\n",
    "        # There are no example with xrefEnsembl, so the RAG without validation usually fails\n",
    "        \"question\": \"How can I get the cross-reference to the ensembl protein for the LCT protein in OMA? Return only the distinct ?ensemblURI\",\n",
    "        \"endpoint\": \"https://sparql.omabrowser.org/sparql/\",\n",
    "        \"query\": \"\"\"PREFIX orth: <http://purl.org/net/orth#>\n",
    "    PREFIX lscr: <http://purl.org/lscr#>\n",
    "    PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
    "    SELECT DISTINCT ?ensemblURI\n",
    "    WHERE {\n",
    "        ?protein a orth:Protein ;\n",
    "            rdfs:label 'LCT' ;\n",
    "            lscr:xrefEnsemblProtein ?ensemblURI .\n",
    "    }\"\"\",\n",
    "    },\n",
    "    {\n",
    "        # There are no example with inDataset\n",
    "        \"question\": \"How can I get the URI of a dataset to which an ortholog cluster belongs in OMA? Return orthologCluster, datasetURI and limit to 20\",\n",
    "        \"endpoint\": \"https://sparql.omabrowser.org/sparql/\",\n",
    "        \"query\": \"\"\"PREFIX orth: <http://purl.org/net/orth#>\n",
    "    SELECT DISTINCT ?orthologCluster ?datasetURI\n",
    "    WHERE {\n",
    "        ?orthologCluster a orth:OrthologsCluster ;\n",
    "            orth:inDataset ?datasetURI .\n",
    "    } LIMIT 20\"\"\",\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"Give me the list of strains associated to the Escherichia coli taxon and their name. Return ?taxon, ?strain, ?name, limit to 20\",\n",
    "        \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "        \"query\": \"\"\"PREFIX up: <http://purl.uniprot.org/core/>\n",
    "    PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
    "    PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
    "    SELECT ?taxon ?strain ?name\n",
    "    WHERE {\n",
    "        ?taxon a up:Taxon .\n",
    "        ?taxon rdfs:subClassOf taxon:562 . # Escherichia coli taxon\n",
    "        ?taxon up:strain ?strain .\n",
    "        ?strain up:name ?name .\n",
    "    } LIMIT 20\"\"\",\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"Retrieve all proteins involved in pathways involving glycolysis. Return ?proteinURI, ?proteinLabel, ?pathwayLabel, limit to 20\",\n",
    "        \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "        \"query\": \"\"\"PREFIX up: <http://purl.uniprot.org/core/>\n",
    "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
    "SELECT DISTINCT ?proteinURI ?proteinLabel ?pathwayLabel\n",
    "WHERE {\n",
    "    ?proteinURI a up:Protein ;\n",
    "        up:recommendedName/up:fullName ?proteinLabel ;\n",
    "        up:annotation ?annotation .\n",
    "    ?annotation a up:Pathway_Annotation ;\n",
    "        rdfs:seeAlso ?pathway .\n",
    "    ?pathway rdfs:label ?pathwayLabel .\n",
    "    FILTER(CONTAINS(LCASE(?pathwayLabel), \"glycolysis\"))\n",
    "} LIMIT 20\"\"\",\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"What are the orthologs in rat for protein Q9Y2T1 ? Return ?ratProtein ?ratUniProtXref\",\n",
    "        \"endpoint\": \"https://sparql.omabrowser.org/sparql/\",\n",
    "        \"query\": \"\"\"PREFIX up: <http://purl.uniprot.org/core/>\n",
    "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
    "PREFIX orth: <http://purl.org/net/orth#>\n",
    "PREFIX lscr: <http://purl.org/lscr#>\n",
    "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
    "SELECT DISTINCT ?ratProtein ?ratUniProtXref\n",
    "WHERE {\n",
    "    ?cluster a orth:OrthologsCluster.\n",
    "    ?cluster orth:hasHomologousMember ?node1.\n",
    "    ?cluster orth:hasHomologousMember ?node2.\n",
    "    ?node1 orth:hasHomologousMember* ?humanProtein.\n",
    "    ?node2 orth:hasHomologousMember* ?ratProtein.\n",
    "    ?humanProtein a orth:Protein;\n",
    "        lscr:xrefUniprot <http://purl.uniprot.org/uniprot/Q9Y2T1>.\n",
    "    ?ratProtein a orth:Protein;\n",
    "        orth:organism/obo:RO_0002162/up:scientificName 'Rattus norvegicus';\n",
    "        lscr:xrefUniprot ?ratUniProtXref.\n",
    "    FILTER(?node1 != ?node2)\n",
    "}\"\"\",\n",
    "    },\n",
    "    # 2 of the queries used as examples in the web UI:\n",
    "    {\n",
    "        \"question\": \"What are the rat orthologs of the human TP53? Return ?ratProteinUri ?ratUniprotLink\",\n",
    "        \"endpoint\": \"https://sparql.omabrowser.org/sparql/\",\n",
    "        \"query\": \"\"\"PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
    "PREFIX orth: <http://purl.org/net/orth#>\n",
    "PREFIX up: <http://purl.uniprot.org/core/>\n",
    "PREFIX sio: <http://semanticscience.org/resource/>\n",
    "PREFIX lscr: <http://purl.org/lscr#>\n",
    "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
    "\n",
    "SELECT DISTINCT ?ratProteinUri ?ratUniprotLink WHERE {\n",
    "  ?humanProtein a orth:Protein ;\n",
    "    rdfs:label 'TP53' ;\n",
    "    orth:organism/obo:RO_0002162/up:scientificName 'Homo sapiens' .\n",
    "  ?cluster a orth:OrthologsCluster ;\n",
    "    orth:hasHomologousMember ?node1 ;\n",
    "    orth:hasHomologousMember ?node2 .\n",
    "  ?node1 orth:hasHomologousMember* ?humanProtein .\n",
    "  ?node2 orth:hasHomologousMember* ?ratProteinUri .\n",
    "  ?ratProteinUri a orth:Protein ;\n",
    "    orth:organism/obo:RO_0002162/up:scientificName 'Rattus norvegicus' ;\n",
    "    lscr:xrefUniprot ?ratUniprotLink .\n",
    "  FILTER(?node1 != ?node2)\n",
    "}\"\"\",\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"Where is expressed the gene ACE2 in human? Return ?anatUri ?anatName\",\n",
    "        \"endpoint\": \"https://www.bgee.org/sparql/\",\n",
    "        \"query\": \"\"\"PREFIX genex: <http://purl.org/genex#>\n",
    "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
    "PREFIX orth: <http://purl.org/net/orth#>\n",
    "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
    "PREFIX up: <http://purl.uniprot.org/core/>\n",
    "\n",
    "SELECT DISTINCT ?anatUri ?anatName WHERE {\n",
    "    ?seq a orth:Gene ;\n",
    "        genex:isExpressedIn ?anatUri ;\n",
    "        rdfs:label \"ACE2\" ;\n",
    "        orth:organism ?organism .\n",
    "    ?anatUri a genex:AnatomicalEntity ;\n",
    "        rdfs:label ?anatName .\n",
    "    ?organism obo:RO_0002162 ?species .\n",
    "    ?species a up:Taxon ;\n",
    "        up:scientificName \"Homo sapiens\" .\n",
    "}\"\"\",\n",
    "    },\n",
    "\n",
    "    {\n",
    "        \"question\": \"\"\"Retrieve all proteins that are associated with Alzheimer disease (http://purl.uniprot.org/diseases/3832) and where they are known to be located in the cell. Return ?proteinURI, ?locationInsideCellLabel, ?locationInsideCellUri, limit to 20\"\"\",\n",
    "        \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "        \"query\": \"\"\"PREFIX up: <http://purl.uniprot.org/core/>\n",
    "PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
    "SELECT ?proteinURI ?locationInsideCellLabel ?locationInsideCellUri\n",
    "WHERE {\n",
    "    ?proteinURI a up:Protein ;\n",
    "        up:annotation ?diseaseAnnotation , ?subcellAnnotation .\n",
    "    ?diseaseAnnotation up:disease <http://purl.uniprot.org/diseases/3832> .\n",
    "    ?subcellAnnotation up:locatedIn/up:cellularComponent ?locationInsideCellUri .\n",
    "    ?locationInsideCellUri skos:prefLabel ?locationInsideCellLabel .\n",
    "} LIMIT 20\"\"\",\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"Retrieve all proteins in OMA that are encoded by the TP53 gene and their mnemonics and evidence types from the UniProt database. Return ?proteinOMA ?speciesLabel ?mnemonic ?evidenceType ?uniprotURI\",\n",
    "        \"endpoint\": \"https://sparql.omabrowser.org/sparql/\",\n",
    "        \"query\": \"\"\"PREFIX up: <http://purl.uniprot.org/core/>\n",
    "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
    "PREFIX orth: <http://purl.org/net/orth#>\n",
    "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
    "PREFIX lscr: <http://purl.org/lscr#>\n",
    "SELECT DISTINCT ?proteinOMA ?speciesLabel ?mnemonic ?evidenceType ?uniprotURI\n",
    "WHERE {\n",
    "    ?proteinOMA a orth:Protein ;\n",
    "        orth:organism/obo:RO_0002162/up:scientificName ?speciesLabel ;\n",
    "        rdfs:label 'TP53' .\n",
    "    ?proteinOMA lscr:xrefUniprot ?uniprotURI.\n",
    "    SERVICE <http://sparql.uniprot.org/sparql> {\n",
    "        ?uniprotURI up:mnemonic ?mnemonic ;\n",
    "            up:existence/rdfs:label ?evidenceType .\n",
    "    }\n",
    "}\n",
    "\"\"\",\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"What is the function of APOC1? Return ?function\",\n",
    "        \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "        \"query\": \"\"\"PREFIX up: <http://purl.uniprot.org/core/>\n",
    "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
    "PREFIX dcterms: <http://purl.org/dc/terms/>\n",
    "SELECT DISTINCT ?function WHERE {\n",
    "    ?protein a up:Protein ;\n",
    "        up:mnemonic \"APOC1_HUMAN\" ;\n",
    "        up:annotation ?annotation .\n",
    "    ?annotation a up:Function_Annotation ;\n",
    "        rdfs:comment ?function .\n",
    "}\n",
    "\"\"\",\n",
    "    },\n",
    "\n",
    "    #################################################################\n",
    "    # New queries to test:\n",
    "    # What are the genes expressed in the human brain?\n",
    "    # FAILS to add filter for human\n",
    "    # What are the human genes expressed in the brain? WORKS as expected\n",
    "\n",
    "    # Which are the human genes associated with cancer and their orthologs?\n",
    "    # This one does not work because in the query generated the variable in the UniProt block ?protein does not match the one used in the OMA block, ?humanUniprot...\n",
    "    # TODO: when we parse the query check there is a link between the two blocks (2 block are using the same variable)\n",
    "    # https://sibkru.atlassian.net/jira/software/projects/E4/boards/6?selectedIssue=E4-34\n",
    "#     {\n",
    "#         \"question\": \"Which are the human genes associated with cancer and their orthologs? Return ?humanGeneName ?orthologUniprot, and limit to 10\",\n",
    "#         \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "#         \"query\": \"\"\"PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
    "# PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
    "# PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
    "# PREFIX up: <http://purl.uniprot.org/core/>\n",
    "# PREFIX orth: <http://purl.org/net/orth#>\n",
    "# PREFIX lscr: <http://purl.org/lscr#>\n",
    "# SELECT DISTINCT ?humanGeneName ?orthologProtein ?orthologUniprot\n",
    "# WHERE {\n",
    "#   # Retrieve human genes associated with cancer from UniProt\n",
    "#   ?humanUniprot a up:Protein ;\n",
    "#               up:organism taxon:9606 ;\n",
    "#               up:encodedBy ?gene ;\n",
    "#               up:annotation ?annotation .\n",
    "#   ?annotation a up:Disease_Annotation ;\n",
    "#               rdfs:comment ?diseaseComment .\n",
    "#   FILTER(CONTAINS(LCASE(?diseaseComment), \"cancer\"))\n",
    "#   ?gene skos:prefLabel ?humanGeneName .\n",
    "\n",
    "#   # Find orthologs of these genes using OMA\n",
    "#   SERVICE <https://sparql.omabrowser.org/sparql> {\n",
    "#     ?cluster a orth:OrthologsCluster ;\n",
    "#       orth:hasHomologousMember ?node1 ;\n",
    "#       orth:hasHomologousMember ?node2 .\n",
    "#     ?node1 orth:hasHomologousMember* ?humanProtein .\n",
    "#     ?node2 orth:hasHomologousMember* ?orthologProtein .\n",
    "#     ?humanProtein lscr:xrefUniprot ?humanUniprot .\n",
    "#     ?orthologProtein lscr:xrefUniprot ?orthologUniprot .\n",
    "#     FILTER(?node1 != ?node2)\n",
    "#   } } LIMIT 10\"\"\",\n",
    "#     },\n",
    "\n",
    "\n",
    "    # List human genes that have known orthologs in the rat and are expressed in the brain?\n",
    "\n",
    "\n",
    "    # Which are the human genes associated with cancer and their orthologs expressed in the rat brain?\n",
    "    # Find all proteins linked to arachidonate (CHEBI:32395) and their associated pathways\n",
    "    # List all enzymes that have been experimentally validated and are involved in DNA repair\n",
    "    # Find all proteins that have a mutagenesis annotation affecting their active site\n",
    "\n",
    "\n",
    "    #     {\n",
    "    #         # Way too slow for some reason\n",
    "    #         \"question\": \"Retrieve all proteins that are associated with Alzheimer disease. Return ?proteinURI, ?proteinLabel, ?diseaseURI, ?diseaseLabel, limit to 20\",\n",
    "    #         \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    #         \"query\": \"\"\"PREFIX up: <http://purl.uniprot.org/core/>\n",
    "    # PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
    "    # PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
    "    # SELECT ?proteinURI ?proteinLabel ?diseaseURI ?diseaseLabel\n",
    "    # WHERE {\n",
    "    #     ?proteinURI a up:Protein ;\n",
    "    #             up:annotation ?diseaseAnnotation ;\n",
    "    #             up:recommendedName/up:fullName ?proteinLabel .\n",
    "    #     ?diseaseAnnotation up:disease ?diseaseURI .\n",
    "    #     ?diseaseURI skos:prefLabel ?diseaseLabel .\n",
    "    #     FILTER(CONTAINS(LCASE(?diseaseLabel), \"alzheimer\"))\n",
    "    # } LIMIT 10\"\"\",\n",
    "    #     },\n",
    "\n",
    "    #     {\n",
    "    #         # Always failing the connection between OMA and Bgee\n",
    "    #         \"question\": \"For protein Q9Y2T1, what are the orthologs in rats for genes expressed in the brain? Return ?ratProtein ?ratUniProtXref\",\n",
    "    #         \"endpoint\": \"https://sparql.omabrowser.org/sparql/\",\n",
    "    #         \"query\": \"\"\"PREFIX up: <http://purl.uniprot.org/core/>\n",
    "    # PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
    "    # PREFIX orth: <http://purl.org/net/orth#>\n",
    "    # PREFIX lscr: <http://purl.org/lscr#>\n",
    "    # PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
    "    # SELECT DISTINCT ?ratProtein ?ratUniProtXref\n",
    "    # WHERE {\n",
    "    #     ?cluster a orth:OrthologsCluster.\n",
    "    #     ?cluster orth:hasHomologousMember ?node1.\n",
    "    #     ?cluster orth:hasHomologousMember ?node2.\n",
    "    #     ?node1 orth:hasHomologousMember* ?humanProtein.\n",
    "    #     ?node2 orth:hasHomologousMember* ?ratProtein.\n",
    "    #     ?humanProtein a orth:Protein;\n",
    "    #         lscr:xrefUniprot <http://purl.uniprot.org/uniprot/Q9Y2T1>.\n",
    "    #     ?ratProtein a orth:Protein;\n",
    "    #         orth:organism/obo:RO_0002162/up:scientificName 'Rattus norvegicus';\n",
    "    #         lscr:xrefUniprot ?ratUniProtXref.\n",
    "    #     FILTER(?node1 != ?node2)\n",
    "    # }\"\"\",\n",
    "    #     },\n",
    "\n",
    "    #     {\n",
    "    #         # It's making up URIs for photosynthesis instead of filtering on label\n",
    "    #         \"question\": \"Find all proteins that have a known 3D structure and are involved in photosynthesis. Return ?protein, ?structure, ?keywordLabel, limit to 20\",\n",
    "    #         \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    #         \"query\": \"\"\"PREFIX up: <http://purl.uniprot.org/core/>\n",
    "    # PREFIX keywords: <http://purl.uniprot.org/keywords/>\n",
    "    # PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
    "    # SELECT DISTINCT ?protein ?structure ?keywordLabel\n",
    "    # WHERE {\n",
    "    #     ?protein a up:Protein ;\n",
    "    #         up:classifiedWith/rdfs:label ?keywordLabel ;\n",
    "    #         rdfs:seeAlso ?structure .\n",
    "    #     ?structure up:database ?db .\n",
    "    #     ?db up:category '3D structure databases' .\n",
    "    #     FILTER(LCASE(?keywordLabel) = \"photosynthesis\")\n",
    "    # } LIMIT 20\"\"\",\n",
    "    #     },\n",
    "\n",
    "    # {\n",
    "    #     # Asking for \"full name\" will cause the RAG to fail, but validation don't catch it yet because VoID don't register the class when we have\n",
    "    #     # we need to wait for LinkSet to be added to VoID in UniProt\n",
    "    #     # Why not using \"up:Protein up:enzyme ??\"... Also up:enzymeClass never shows up in the relevant documents\n",
    "    #         \"question\": \"How can I get a list of proteins that are enzymes with their enzyme class? Return the ?proteinURI, ?enzymeClassURI, ?enzymeClassName, limit to 20\",\n",
    "    #         \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    #         \"query\": \"\"\"PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
    "    #     PREFIX up: <http://purl.uniprot.org/core/>\n",
    "    #     PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
    "    #     SELECT ?proteinURI ?enzymeClassURI ?enzymeClassName\n",
    "    #     WHERE {\n",
    "    #        ?proteinURI a up:Protein ;\n",
    "    #                    up:enzyme ?enzymeClassURI .\n",
    "    #        ?enzymeClassURI skos:prefLabel ?enzymeClassName .\n",
    "    #     #   ?protein a up:Protein ;\n",
    "    #     #            up:annotation ?enzymeAnnotation .\n",
    "    #     #   ?enzymeAnnotation a up:Catalytic_Activity_Annotation ;\n",
    "    #     #                     up:catalyticActivity ?enzymeActivity .\n",
    "    #     #   ?enzymeActivity up:enzymeClass ?enzymeClassURI .\n",
    "    #     #   ?enzymeClassURI skos:prefLabel ?enzymeClassName .\n",
    "    #     } LIMIT 20\"\"\",\n",
    "    # },\n",
    "\n",
    "    # Which are the human genes associated with lung cancer and their orthologs expressed in the rat brain?\n",
    "\n",
    "    # {\n",
    "    #     \"question\": \"Which are the human genes associated with cancer (which have cancer in their disease label) and their orthologs expressed in the rat brain? Return the disease label, human gene URI, human gene HGNC symbol, ortholog rat gene URI\",\n",
    "    #     \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    #     \"query\": \"\"\"PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
    "    # PREFIX up:<http://purl.uniprot.org/core/>\n",
    "    # PREFIX taxon:<http://purl.uniprot.org/taxonomy/>\n",
    "    # PREFIX rdfs:<http://www.w3.org/2000/01/rdf-schema#>\n",
    "    # PREFIX orth:<http://purl.org/net/orth#>\n",
    "    # PREFIX dcterms:<http://purl.org/dc/terms/>\n",
    "    # PREFIX obo:<http://purl.obolibrary.org/obo/>\n",
    "    # PREFIX lscr:<http://purl.org/lscr#>\n",
    "    # PREFIX genex:<http://purl.org/genex#>\n",
    "    # PREFIX sio: <http://semanticscience.org/resource/>\n",
    "    # SELECT DISTINCT ?diseaseLabel ?humanProtein ?hgncSymbol ?orthologRatProtein ?orthologRatGene\n",
    "    # WHERE {\n",
    "    # SERVICE <https://sparql.uniprot.org/sparql> {\n",
    "    #     SELECT DISTINCT * WHERE {\n",
    "    #     ?humanProtein a up:Protein ;\n",
    "    #         up:organism/up:scientificName 'Homo sapiens' ;\n",
    "    #         up:annotation ?annotation ;\n",
    "    #         rdfs:seeAlso ?hgnc .\n",
    "    #     ?hgnc up:database <http://purl.uniprot.org/database/HGNC> ;\n",
    "    #         rdfs:comment ?hgncSymbol .\n",
    "    #     ?annotation a up:Disease_Annotation ;\n",
    "    #             up:disease ?disease .\n",
    "    #     ?disease skos:prefLabel ?diseaseLabel.\n",
    "    #     FILTER CONTAINS (?diseaseLabel, \"cancer\")\n",
    "    #     }\n",
    "    # }\n",
    "    # SERVICE <https://sparql.omabrowser.org/sparql/> {\n",
    "    #     SELECT ?humanProtein ?orthologRatProtein ?orthologRatGene WHERE {\n",
    "    #     ?humanProteinOma a orth:Protein ;\n",
    "    #             lscr:xrefUniprot ?humanProtein .\n",
    "    #     ?orthologRatProtein a orth:Protein ;\n",
    "    #         sio:SIO_010079 ?orthologRatGene ;\n",
    "    #         orth:organism/obo:RO_0002162/up:scientificName 'Rattus norvegicus' .\n",
    "    #     ?cluster a orth:OrthologsCluster .\n",
    "    #     ?cluster orth:hasHomologousMember ?node1 .\n",
    "    #     ?cluster orth:hasHomologousMember ?node2 .\n",
    "    #     ?node1 orth:hasHomologousMember* ?humanProteinOma .\n",
    "    #     ?node2 orth:hasHomologousMember* ?orthologRatProtein .\n",
    "    #     FILTER(?node1 != ?node2)\n",
    "    #     }\n",
    "    # }\n",
    "    # SERVICE <https://www.bgee.org/sparql/> {\n",
    "    #     ?orthologRatGene genex:isExpressedIn ?anatEntity ;\n",
    "    #         orth:organism ?ratOrganism .\n",
    "    #     ?anatEntity rdfs:label 'brain' .\n",
    "    #     ?ratOrganism obo:RO_0002162 taxon:10116 .\n",
    "    # }\n",
    "    # }\"\"\",\n",
    "    # },\n",
    "]\n",
    "\n",
    "\n",
    "def result_sets_are_same(gen_set, ref_set) -> bool:\n",
    "    gen_set, ref_set = list(gen_set), list(ref_set)\n",
    "    for item in gen_set:\n",
    "        if item not in ref_set:\n",
    "            print(f\"Missing from reference: {item}\")\n",
    "            return False\n",
    "    for item in ref_set:\n",
    "        if item not in gen_set:\n",
    "            print(f\"Missing from generated: {item}\")\n",
    "            return False\n",
    "    return True\n",
    "\n",
    "\n",
    "# QLEVER_UNIPROT = \"https://qlever.cs.uni-freiburg.de/api/uniprot\"\n",
    "\n",
    "# Price per million tokens, open source models based on fireworks.io pricing\n",
    "# https://openai.com/api/pricing/\n",
    "# https://fireworks.ai/pricing\n",
    "models = {\n",
    "    # \"Llama3.1 8B\": {\n",
    "    #     \"id\": \"hf:meta-llama/Meta-Llama-3.1-8B-Instruct\",\n",
    "    #     \"price_input\": 0.2,\n",
    "    #     \"price_output\": 0.2,\n",
    "    # },\n",
    "    # \"Mixtral 8x22B\": {\n",
    "    #     \"id\": \"hf:mistralai/Mixtral-8x22B-Instruct-v0.1\",\n",
    "    #     \"price_input\": 1.20,\n",
    "    #     \"price_output\": 1.20,\n",
    "    # },\n",
    "    \"o3-mini\": {\n",
    "        \"id\": \"openai/o3-mini\",\n",
    "        \"price_input\": 1.1,\n",
    "        \"price_output\": 4.4,\n",
    "    },\n",
    "    # üéØ RAG with validation - Success: 27, Different results: 9, No results: 4, Error: 2\n",
    "    # Price before fixing the token_usage gathering: 0.01421\n",
    "    # \"gpt-4o\": {\n",
    "    #     \"id\": \"openai/gpt-4o\",\n",
    "    #     \"price_input\": 5,\n",
    "    #     \"price_output\": 15,\n",
    "    # },\n",
    "    # # üéØ RAG with validation - Success: 32, Different results: 7, No results: 3, Error: 0\n",
    "    # \"gpt-4o-mini\": {\n",
    "    #     \"id\": \"openai/gpt-4o-mini\",\n",
    "    #     \"price_input\": 0.15,\n",
    "    #     \"price_output\": 0.6,\n",
    "    # },\n",
    "}\n",
    "\n",
    "\n",
    "def answer_no_rag(question: str, model: str):\n",
    "    client = load_chat_model(Configuration(model=model))\n",
    "    response = client.invoke([\n",
    "      SystemMessage(content=SYSTEM_PROMPT),\n",
    "      HumanMessage(content=question),\n",
    "    ])\n",
    "    response = response.model_dump()\n",
    "    response[\"messages\"] = [{\"content\": response[\"content\"], \"response_metadata\": response[\"response_metadata\"]}]\n",
    "    return response\n",
    "\n",
    "\n",
    "def answer_rag_without_validation(question: str, model: str):\n",
    "    response = requests.post(\n",
    "        \"http://localhost:8000/chat\",\n",
    "        headers={\"Authorization\": f\"Bearer {chat_api_key}\"},\n",
    "        json={\n",
    "            \"model\": model,\n",
    "            \"messages\": [{\"role\": \"user\", \"content\": question}],\n",
    "            \"stream\": False,\n",
    "            \"validate_output\": False,\n",
    "        },\n",
    "        timeout=120,\n",
    "    )\n",
    "    return response.json()\n",
    "\n",
    "\n",
    "def answer_rag_with_validation(question: str, model: str):\n",
    "    response = requests.post(\n",
    "        \"http://localhost:8000/chat\",\n",
    "        headers={\"Authorization\": f\"Bearer {chat_api_key}\"},\n",
    "        json={\n",
    "            \"model\": model,\n",
    "            \"messages\": [{\"role\": \"user\", \"content\": question}],\n",
    "            \"stream\": False,\n",
    "        },\n",
    "        timeout=120,\n",
    "    )\n",
    "    return response.json()\n",
    "\n",
    "\n",
    "list_of_approaches = {\n",
    "    \"No RAG\": answer_no_rag,\n",
    "    # \"RAG without validation\": answer_rag_without_validation,\n",
    "    \"RAG with validation\": answer_rag_with_validation,\n",
    "}\n",
    "\n",
    "results_data = {\n",
    "    \"Model\": [],\n",
    "    \"RAG Approach\": [],\n",
    "    \"Success\": [],\n",
    "    \"Different Results\": [],\n",
    "    \"No Results\": [],\n",
    "    \"Errors\": [],\n",
    "    \"Price\": [],\n",
    "    # 'Precision': [],\n",
    "    # \"Recall\": [],\n",
    "    \"F1\": [],\n",
    "}\n",
    "\n",
    "number_of_tries = 3\n",
    "\n",
    "print(f\"üß™ Testing {len(example_queries)} queries\")\n",
    "\n",
    "# Get results for the reference queries first\n",
    "ref_results = []\n",
    "for i, test_query in enumerate(example_queries):\n",
    "    res_ref_finally_pass = False\n",
    "    while not res_ref_finally_pass:\n",
    "        try:\n",
    "            query_start_time = time.time()\n",
    "            res_from_ref = query_sparql(test_query[\"query\"], test_query[\"endpoint\"], timeout=300)[\"results\"][\"bindings\"]\n",
    "            print(f\"Reference query {i} '{test_query['question']}' took {time.time() - query_start_time:.2f} seconds\")\n",
    "            ref_results.append(res_from_ref)\n",
    "            res_ref_finally_pass = True\n",
    "        except Exception as e:\n",
    "            print(f\"Timeout for reference query {i}: {e}, Trying again because we know it should work.\")\n",
    "            res_ref_finally_pass = False\n",
    "    # res_from_ref = query_sparql(test_query[\"query\"], QLEVER_UNIPROT)[\"results\"][\"bindings\"]\n",
    "\n",
    "\n",
    "for model_label, model in models.items():\n",
    "    print(f\"\\nüß† Testing model: {model_label}\\n\")\n",
    "    res = defaultdict(dict)\n",
    "    # e.g. res[\"No RAG\"][\"success\"] += 1\n",
    "    for approach in list_of_approaches:\n",
    "        res[approach] = defaultdict(int)\n",
    "\n",
    "    for query_num, test_query in enumerate(example_queries):\n",
    "\n",
    "        for approach, approach_func in list_of_approaches.items():\n",
    "            # print(f\"Approach {approach}\")\n",
    "            for t in range(number_of_tries):\n",
    "                response = approach_func(test_query[\"question\"], model[\"id\"])\n",
    "                # print(response)\n",
    "                chat_resp_md = response[\"messages\"][-1][\"content\"]\n",
    "                # chat_resp_md = response[\"choices\"][0][\"message\"][\"content\"]\n",
    "                # TODO: loop over all messages to get the total token usage in case of multiple messages (fix by calling LLM)\n",
    "                for msg in response[\"messages\"]:\n",
    "                  if \"response_metadata\" in msg and \"token_usage\" in msg[\"response_metadata\"]:\n",
    "                    res[approach][\"input_tokens\"] += msg[\"response_metadata\"][\"token_usage\"][\"prompt_tokens\"]\n",
    "                    res[approach][\"output_tokens\"] += msg[\"response_metadata\"][\"token_usage\"][\"completion_tokens\"]\n",
    "                # res[approach][\"input_tokens\"] += response[\"messages\"][-1][\"response_metadata\"][\"token_usage\"][\"prompt_tokens\"]\n",
    "                # res[approach][\"output_tokens\"] += response[\"messages\"][-1][\"response_metadata\"][\"token_usage\"][\"completion_tokens\"]\n",
    "                # print(chat_resp_md)\n",
    "                try:\n",
    "                    generated_sparql = extract_sparql_queries(chat_resp_md)[-1]\n",
    "                    if generated_sparql[\"query\"].strip() == test_query[\"query\"].strip():\n",
    "                        print(f\"‚úÖ {t+1}/{number_of_tries} {test_query['question']}. EXACT MATCH\")\n",
    "                        res[approach][\"success\"] += 1\n",
    "                        continue\n",
    "\n",
    "                    # Execute the generated query\n",
    "                    res_from_generated = query_sparql(\n",
    "                        generated_sparql[\"query\"], generated_sparql[\"endpoint_url\"], timeout=300\n",
    "                    )[\"results\"][\"bindings\"]\n",
    "                    # res_from_generated = query_sparql(generated_sparql[\"query\"], QLEVER_UNIPROT)[\"results\"][\"bindings\"]\n",
    "\n",
    "                    if not result_sets_are_same(res_from_generated, ref_results[query_num]):\n",
    "                        if len(res_from_generated) == 0:\n",
    "                            res[approach][\"no_results\"] += 1\n",
    "                        else:\n",
    "                            res[approach][\"different_results\"] += 1\n",
    "                        raise Exception(\n",
    "                            f\"\\nResults mismatch. Ref: {len(ref_results[query_num])} != gen: {len(res_from_generated)}\\n\"\n",
    "                        )\n",
    "                    else:\n",
    "                        print(f\"‚úÖ {t+1}/{number_of_tries} {test_query['question']} = {len(res_from_generated)}\")\n",
    "                        res[approach][\"success\"] += 1\n",
    "\n",
    "                except Exception as e:\n",
    "                    res[approach][\"fail\"] += 1\n",
    "                    if approach != \"No RAG\":\n",
    "                        fail_emoji = \"‚ùå\" if approach == \"RAG with validation\" else \"‚õîÔ∏è\"\n",
    "                        print(f\"{fail_emoji} {t+1}/{number_of_tries} {test_query['question']}\\n{e}\")\n",
    "                        print(generated_sparql[\"query\"])\n",
    "                        print(\"Correct query:\")\n",
    "                        print(test_query[\"query\"])\n",
    "                        print(\"\")\n",
    "\n",
    "        for approach in list_of_approaches:\n",
    "            print(\n",
    "                f\"üéØ {approach} - Success: {res[approach]['success']}, Different results: {res[approach]['different_results']}, No results: {res[approach]['no_results']}, Error: {res[approach]['fail']-res[approach]['no_results']-res[approach]['different_results']}\"\n",
    "            )\n",
    "\n",
    "    for approach, result_row in res.items():\n",
    "        mean_price = (\n",
    "            (result_row[\"input_tokens\"] * model[\"price_input\"] / 1000000)\n",
    "            + (result_row[\"output_tokens\"] * model[\"price_output\"] / 1000000)\n",
    "        ) / (len(example_queries) * number_of_tries)\n",
    "        precision = result_row[\"success\"] / (result_row[\"success\"] + result_row[\"fail\"])\n",
    "        recall = result_row[\"success\"] / (result_row[\"success\"] + result_row[\"fail\"] - result_row[\"different_results\"])\n",
    "        results_data[\"Model\"].append(model_label)\n",
    "        results_data[\"RAG Approach\"].append(approach)\n",
    "        results_data[\"Success\"].append(result_row[\"success\"])\n",
    "        results_data[\"Different Results\"].append(result_row[\"different_results\"])\n",
    "        results_data[\"No Results\"].append(result_row[\"no_results\"])\n",
    "        results_data[\"Errors\"].append(result_row[\"fail\"] - result_row[\"no_results\"] - result_row[\"different_results\"])\n",
    "        results_data[\"Price\"].append(round(mean_price, 5))\n",
    "        # results_data['Precision'].append(precision)\n",
    "        # results_data['Recall'].append(recall)\n",
    "        if precision + recall == 0:\n",
    "            results_data[\"F1\"].append(0)\n",
    "        else:\n",
    "            results_data[\"F1\"].append(round(2 * (precision * recall) / (precision + recall), 2))\n",
    "\n",
    "\n",
    "df = pd.DataFrame(results_data)\n",
    "print(df)\n",
    "print(\"\\n\\n\")\n",
    "print(df.to_csv(\"tests_results.csv\", index=False))\n",
    "\n",
    "# Output Latex table\n",
    "latex_str = \"\"\n",
    "prev_model = list(models.keys())[0]\n",
    "for index, row in df.iterrows():\n",
    "    row_str = \" & \".join([str(item) for item in row])  # Join all values in the row with \" & \"\n",
    "    row_str += \" \\\\\\\\\"\n",
    "    if row[\"Model\"] != prev_model:\n",
    "        latex_str += \"\\\\midrule\\n\"\n",
    "        prev_model = row[\"Model\"]\n",
    "    latex_str += row_str + \"\\n\"\n",
    "with open(\"tests_results_latex.txt\", \"w\") as f:\n",
    "    f.write(latex_str)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
