{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Testing Expasy query helper\n",
    "\n",
    "Testing our Expasy query helper reusing the example queries that succeeded retrieved from `test_example_queries.ipynb`\n",
    "\n",
    "Potential interesting questions to use for evaluation:\n",
    "- Which are the human genes associated with cancer and their orthologs expressed in the rat brain?\n",
    "- Select the number of proteins for each of the subclasses of ec:1.1.1.- EC (Enzyme Commission) class (from biosoda)\n",
    "- "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "üß™ Testing 2 queries\n",
      "Missing from generated: {'ensemblURI': {'type': 'uri', 'value': 'http://rdf.ebi.ac.uk/resource/ensembl.protein/ENSBIXP00000024889'}}\n",
      "‚ùå 1/3 How can I get the cross ref to the protein in ensembl for the LCT protein in OMA? Return only the distinct ensembl URI \n",
      "Results mismatch. Ref: 96 != gen: 0\n",
      ". Generated query:\n",
      "# https://sparql.omabrowser.org/sparql\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX xref: <http://omabrowser.org/ontology/xref#>\n",
      "PREFIX dcterms: <http://purl.org/dc/terms/>\n",
      "\n",
      "SELECT DISTINCT ?ensemblURI WHERE {\n",
      "  ?protein rdfs:label \"LCT\" .\n",
      "  ?protein xref:ensembl ?ensemblURI .\n",
      "}\n",
      "Correct query:\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "SELECT DISTINCT ?ensemblURI\n",
      "WHERE {\n",
      "    ?protein a orth:Protein ;\n",
      "        rdfs:label 'LCT' ;\n",
      "        lscr:xrefEnsemblProtein ?ensemblURI .\n",
      "}\n",
      "\n",
      "Missing from generated: {'ensemblURI': {'type': 'uri', 'value': 'http://rdf.ebi.ac.uk/resource/ensembl.protein/ENSBIXP00000024889'}}\n",
      "‚ùå 2/3 How can I get the cross ref to the protein in ensembl for the LCT protein in OMA? Return only the distinct ensembl URI \n",
      "Results mismatch. Ref: 96 != gen: 0\n",
      ". Generated query:\n",
      "# https://sparql.omabrowser.org/sparql\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX dc: <http://purl.org/dc/elements/1.1/>\n",
      "PREFIX xref: <http://omabrowser.org/ontology/xref#>\n",
      "\n",
      "SELECT DISTINCT ?ensemblURI WHERE {\n",
      "  ?protein rdfs:label \"LCT\" .\n",
      "  ?protein xref:Ensembl ?ensemblURI .\n",
      "}\n",
      "Correct query:\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "SELECT DISTINCT ?ensemblURI\n",
      "WHERE {\n",
      "    ?protein a orth:Protein ;\n",
      "        rdfs:label 'LCT' ;\n",
      "        lscr:xrefEnsemblProtein ?ensemblURI .\n",
      "}\n",
      "\n",
      "Missing from generated: {'ensemblURI': {'type': 'uri', 'value': 'http://rdf.ebi.ac.uk/resource/ensembl.protein/ENSBIXP00000024889'}}\n",
      "‚ùå 3/3 How can I get the cross ref to the protein in ensembl for the LCT protein in OMA? Return only the distinct ensembl URI \n",
      "Results mismatch. Ref: 96 != gen: 0\n",
      ". Generated query:\n",
      "# https://sparql.omabrowser.org/sparql\n",
      "\n",
      "SELECT DISTINCT ?ensemblURI WHERE {\n",
      "  ?protein a up:Protein ;\n",
      "           up:recommendedName ?name ;\n",
      "           rdfs:seeAlso ?ensemblURI .\n",
      "  ?name rdfs:label \"Lactase\"@en .\n",
      "  FILTER(CONTAINS(STR(?ensemblURI), \"ensembl\"))\n",
      "}\n",
      "Correct query:\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "SELECT DISTINCT ?ensemblURI\n",
      "WHERE {\n",
      "    ?protein a orth:Protein ;\n",
      "        rdfs:label 'LCT' ;\n",
      "        lscr:xrefEnsemblProtein ?ensemblURI .\n",
      "}\n",
      "\n",
      "Missing from generated: {'ensemblURI': {'type': 'uri', 'value': 'http://rdf.ebi.ac.uk/resource/ensembl.protein/ENSBIXP00000024889'}}\n",
      "‚ùå 1/3 How can I get the cross ref to the protein in ensembl for the LCT protein in OMA? Return only the distinct ensembl URI \n",
      "Results mismatch. Ref: 96 != gen: 0\n",
      ". Generated query:\n",
      "#https://sparql.omabrowser.org/sparql/\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX dc: <http://purl.org/dc/terms/>\n",
      "\n",
      "SELECT DISTINCT ?ensemblURI\n",
      "WHERE {\n",
      "    ?protein a orth:Protein.\n",
      "    ?protein rdfs:label 'LCT'.\n",
      "    ?protein lscr:xrefEnsembl ?ensemblURI.\n",
      "}\n",
      "Correct query:\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "SELECT DISTINCT ?ensemblURI\n",
      "WHERE {\n",
      "    ?protein a orth:Protein ;\n",
      "        rdfs:label 'LCT' ;\n",
      "        lscr:xrefEnsemblProtein ?ensemblURI .\n",
      "}\n",
      "\n",
      "Missing from generated: {'ensemblURI': {'type': 'uri', 'value': 'http://rdf.ebi.ac.uk/resource/ensembl.protein/ENSBIXP00000024889'}}\n",
      "‚ùå 2/3 How can I get the cross ref to the protein in ensembl for the LCT protein in OMA? Return only the distinct ensembl URI \n",
      "Results mismatch. Ref: 96 != gen: 0\n",
      ". Generated query:\n",
      "#https://sparql.omabrowser.org/sparql/\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX dc: <http://purl.org/dc/terms/>\n",
      "\n",
      "SELECT DISTINCT ?ensemblURI\n",
      "WHERE {\n",
      "    ?protein a orth:Protein.\n",
      "    ?protein rdfs:label 'LCT'.\n",
      "    ?protein lscr:xrefEnsembl ?ensemblURI.\n",
      "}\n",
      "Correct query:\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "SELECT DISTINCT ?ensemblURI\n",
      "WHERE {\n",
      "    ?protein a orth:Protein ;\n",
      "        rdfs:label 'LCT' ;\n",
      "        lscr:xrefEnsemblProtein ?ensemblURI .\n",
      "}\n",
      "\n",
      "Missing from generated: {'ensemblURI': {'type': 'uri', 'value': 'http://rdf.ebi.ac.uk/resource/ensembl.protein/ENSBIXP00000024889'}}\n",
      "‚ùå 3/3 How can I get the cross ref to the protein in ensembl for the LCT protein in OMA? Return only the distinct ensembl URI \n",
      "Results mismatch. Ref: 96 != gen: 0\n",
      ". Generated query:\n",
      "#https://sparql.omabrowser.org/sparql/\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX dc: <http://purl.org/dc/terms/>\n",
      "\n",
      "SELECT DISTINCT ?ensemblURI\n",
      "WHERE {\n",
      "    ?protein a orth:Protein.\n",
      "    ?protein rdfs:label 'LCT'.\n",
      "    ?protein lscr:xrefEnsembl ?ensemblURI.\n",
      "}\n",
      "Correct query:\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "SELECT DISTINCT ?ensemblURI\n",
      "WHERE {\n",
      "    ?protein a orth:Protein ;\n",
      "        rdfs:label 'LCT' ;\n",
      "        lscr:xrefEnsemblProtein ?ensemblURI .\n",
      "}\n",
      "\n",
      "‚úÖ 1/3 How can I get the cross ref to the protein in ensembl for the LCT protein in OMA? Return only the distinct ensembl URI = 96\n",
      "‚úÖ 2/3 How can I get the cross ref to the protein in ensembl for the LCT protein in OMA? Return only the distinct ensembl URI = 96\n",
      "Missing from reference: {'ensemblURI': {'type': 'uri', 'value': 'http://rdf.ebi.ac.uk/resource/ensembl.transcript/ENSBIXT00000051865'}}\n",
      "‚ùå 3/3 How can I get the cross ref to the protein in ensembl for the LCT protein in OMA? Return only the distinct ensembl URI \n",
      "Results mismatch. Ref: 96 != gen: 99\n",
      ". Generated query:\n",
      "#https://sparql.omabrowser.org/sparql/\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX dc: <http://purl.org/dc/terms/>\n",
      "\n",
      "SELECT DISTINCT ?ensemblURI\n",
      "WHERE {\n",
      "    ?protein a orth:Protein;\n",
      "             rdfs:label 'LCT'.\n",
      "    ?protein lscr:xrefEnsemblTranscript ?ensemblURI.\n",
      "}\n",
      "Correct query:\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "SELECT DISTINCT ?ensemblURI\n",
      "WHERE {\n",
      "    ?protein a orth:Protein ;\n",
      "        rdfs:label 'LCT' ;\n",
      "        lscr:xrefEnsemblProtein ?ensemblURI .\n",
      "}\n",
      "\n",
      "‚öñÔ∏è No RAG - Success: 0, Mismatch: 3, Error: 0\n",
      "‚öñÔ∏è RAG without validation - Success: 0, Mismatch: 3, Error: 0\n",
      "‚öñÔ∏è RAG with validation - Success: 2, Mismatch: 1, Error: 0\n",
      "Missing from generated: {'orthologCluster': {'type': 'uri', 'value': 'https://omabrowser.org/oma/hog/resolve/HOG:D0000193_-2035759834'}, 'datasetURI': {'type': 'uri', 'value': 'https://omabrowser.org/oma/current/#DATASET_OMA'}}\n",
      "‚ùå 1/3 How can I get the URI of a dataset to which an ortholog cluster belongs in OMA? Return orthologCluster, datasetURI and limit to 20 \n",
      "Results mismatch. Ref: 20 != gen: 0\n",
      ". Generated query:\n",
      "# https://sparql.omabrowser.org/sparql\n",
      "SELECT ?orthologCluster ?datasetURI\n",
      "WHERE {\n",
      "  ?orthologCluster a oma:OrthologCluster .\n",
      "  ?orthologCluster oma:inDataset ?datasetURI .\n",
      "}\n",
      "LIMIT 20\n",
      "Correct query:\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "SELECT DISTINCT ?orthologCluster ?datasetURI\n",
      "WHERE {\n",
      "    ?orthologCluster a orth:OrthologsCluster ;\n",
      "        orth:inDataset ?datasetURI .\n",
      "} LIMIT 20\n",
      "\n",
      "Missing from generated: {'orthologCluster': {'type': 'uri', 'value': 'https://omabrowser.org/oma/hog/resolve/HOG:D0000193_-2035759834'}, 'datasetURI': {'type': 'uri', 'value': 'https://omabrowser.org/oma/current/#DATASET_OMA'}}\n",
      "‚ùå 2/3 How can I get the URI of a dataset to which an ortholog cluster belongs in OMA? Return orthologCluster, datasetURI and limit to 20 \n",
      "Results mismatch. Ref: 20 != gen: 0\n",
      ". Generated query:\n",
      "# https://sparql.omabrowser.org/sparql\n",
      "SELECT ?orthologCluster ?datasetURI\n",
      "WHERE {\n",
      "  ?orthologCluster a oma:OrthologCluster ;\n",
      "                   oma:inDataset ?datasetURI .\n",
      "}\n",
      "LIMIT 20\n",
      "Correct query:\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "SELECT DISTINCT ?orthologCluster ?datasetURI\n",
      "WHERE {\n",
      "    ?orthologCluster a orth:OrthologsCluster ;\n",
      "        orth:inDataset ?datasetURI .\n",
      "} LIMIT 20\n",
      "\n",
      "Missing from generated: {'orthologCluster': {'type': 'uri', 'value': 'https://omabrowser.org/oma/hog/resolve/HOG:D0000193_-2035759834'}, 'datasetURI': {'type': 'uri', 'value': 'https://omabrowser.org/oma/current/#DATASET_OMA'}}\n",
      "‚ùå 3/3 How can I get the URI of a dataset to which an ortholog cluster belongs in OMA? Return orthologCluster, datasetURI and limit to 20 \n",
      "Results mismatch. Ref: 20 != gen: 0\n",
      ". Generated query:\n",
      "# https://sparql.omabrowser.org/sparql\n",
      "SELECT ?orthologCluster ?datasetURI\n",
      "WHERE {\n",
      "  ?orthologCluster a oma:OrthologCluster .\n",
      "  ?orthologCluster oma:inDataset ?datasetURI .\n",
      "}\n",
      "LIMIT 20\n",
      "Correct query:\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "SELECT DISTINCT ?orthologCluster ?datasetURI\n",
      "WHERE {\n",
      "    ?orthologCluster a orth:OrthologsCluster ;\n",
      "        orth:inDataset ?datasetURI .\n",
      "} LIMIT 20\n",
      "\n",
      "Missing from generated: {'orthologCluster': {'type': 'uri', 'value': 'https://omabrowser.org/oma/hog/resolve/HOG:D0000193_-2035759834'}, 'datasetURI': {'type': 'uri', 'value': 'https://omabrowser.org/oma/current/#DATASET_OMA'}}\n",
      "‚ùå 1/3 How can I get the URI of a dataset to which an ortholog cluster belongs in OMA? Return orthologCluster, datasetURI and limit to 20 \n",
      "Results mismatch. Ref: 20 != gen: 0\n",
      ". Generated query:\n",
      "# https://sparql.omabrowser.org/sparql/\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX dcterms: <http://purl.org/dc/terms/>\n",
      "\n",
      "SELECT ?orthologCluster ?datasetURI\n",
      "WHERE {\n",
      "    ?orthologCluster a orth:OrthologsCluster.\n",
      "    ?orthologCluster dcterms:isPartOf ?datasetURI.\n",
      "}\n",
      "LIMIT 20\n",
      "Correct query:\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "SELECT DISTINCT ?orthologCluster ?datasetURI\n",
      "WHERE {\n",
      "    ?orthologCluster a orth:OrthologsCluster ;\n",
      "        orth:inDataset ?datasetURI .\n",
      "} LIMIT 20\n",
      "\n",
      "Missing from generated: {'orthologCluster': {'type': 'uri', 'value': 'https://omabrowser.org/oma/hog/resolve/HOG:D0000193_-2035759834'}, 'datasetURI': {'type': 'uri', 'value': 'https://omabrowser.org/oma/current/#DATASET_OMA'}}\n",
      "‚ùå 2/3 How can I get the URI of a dataset to which an ortholog cluster belongs in OMA? Return orthologCluster, datasetURI and limit to 20 \n",
      "Results mismatch. Ref: 20 != gen: 0\n",
      ". Generated query:\n",
      "# https://sparql.omabrowser.org/sparql/\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX dcterms: <http://purl.org/dc/terms/>\n",
      "\n",
      "SELECT ?orthologCluster ?datasetURI\n",
      "WHERE {\n",
      "    ?orthologCluster a orth:OrthologsCluster.\n",
      "    ?orthologCluster dcterms:isPartOf ?datasetURI.\n",
      "}\n",
      "LIMIT 20\n",
      "Correct query:\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "SELECT DISTINCT ?orthologCluster ?datasetURI\n",
      "WHERE {\n",
      "    ?orthologCluster a orth:OrthologsCluster ;\n",
      "        orth:inDataset ?datasetURI .\n",
      "} LIMIT 20\n",
      "\n",
      "Missing from generated: {'orthologCluster': {'type': 'uri', 'value': 'https://omabrowser.org/oma/hog/resolve/HOG:D0000193_-2035759834'}, 'datasetURI': {'type': 'uri', 'value': 'https://omabrowser.org/oma/current/#DATASET_OMA'}}\n",
      "‚ùå 3/3 How can I get the URI of a dataset to which an ortholog cluster belongs in OMA? Return orthologCluster, datasetURI and limit to 20 \n",
      "Results mismatch. Ref: 20 != gen: 0\n",
      ". Generated query:\n",
      "#https://sparql.omabrowser.org/sparql/\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX dcterms: <http://purl.org/dc/terms/>\n",
      "\n",
      "SELECT ?orthologCluster ?datasetURI\n",
      "WHERE {\n",
      "    ?orthologCluster a orth:OrthologsCluster.\n",
      "    ?orthologCluster dcterms:isPartOf ?datasetURI.\n",
      "}\n",
      "LIMIT 20\n",
      "Correct query:\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "SELECT DISTINCT ?orthologCluster ?datasetURI\n",
      "WHERE {\n",
      "    ?orthologCluster a orth:OrthologsCluster ;\n",
      "        orth:inDataset ?datasetURI .\n",
      "} LIMIT 20\n",
      "\n",
      "‚úÖ 1/3 How can I get the URI of a dataset to which an ortholog cluster belongs in OMA? Return orthologCluster, datasetURI and limit to 20 = 20\n",
      "‚úÖ 2/3 How can I get the URI of a dataset to which an ortholog cluster belongs in OMA? Return orthologCluster, datasetURI and limit to 20 = 20\n",
      "‚úÖ 3/3 How can I get the URI of a dataset to which an ortholog cluster belongs in OMA? Return orthologCluster, datasetURI and limit to 20 = 20\n",
      "‚öñÔ∏è No RAG - Success: 0, Mismatch: 6, Error: 0\n",
      "‚öñÔ∏è RAG without validation - Success: 0, Mismatch: 6, Error: 0\n",
      "‚öñÔ∏è RAG with validation - Success: 5, Mismatch: 1, Error: 0\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "import os\n",
    "import requests\n",
    "\n",
    "from openai import OpenAI\n",
    "from qdrant_client.models import FieldCondition, Filter, MatchValue, ScoredPoint\n",
    "from rdflib.plugins.sparql import prepareQuery, prepareUpdate\n",
    "from rdflib.plugins.sparql.algebra import translateQuery\n",
    "from SPARQLWrapper import JSON, TURTLE, XML, SPARQLWrapper\n",
    "from dotenv import load_dotenv\n",
    "\n",
    "from expasy_chat.validate_sparql import extract_sparql_queries\n",
    "from expasy_chat.config import settings\n",
    "\n",
    "load_dotenv()\n",
    "expasy_api_key = os.getenv('EXPASY_API_KEY')\n",
    "\n",
    "example_queries = [\n",
    "  {\n",
    "    \"question\": \"What is the accession number in uniprot of the human gene LCT? Return only unique protein URIs\",\n",
    "    \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    \"query\": \"\"\"PREFIX up: <http://purl.uniprot.org/core/>\n",
    "PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
    "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
    "SELECT DISTINCT ?protein\n",
    "WHERE{\n",
    "    ?protein a up:Protein .\n",
    "    ?protein up:organism taxon:9606 .\n",
    "    ?protein up:encodedBy ?gene .\n",
    "    ?gene skos:prefLabel \"LCT\" .\n",
    "}\"\"\",\n",
    "  },\n",
    "  {\n",
    "    # NOTE: The \"mature\" part in the question makes it harder to answer\n",
    "    # \"question\": \"How do I filter for reviewed (mouse) proteins whose mature form carries an N-terminal glycine? Return protein URI and AA sequence\",\n",
    "    \"question\": \"How do I filter for reviewed mouse proteins which carry an N-terminal glycine? Return protein URI and AA sequence\",\n",
    "    \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    \"query\": \"\"\"PREFIX up: <http://purl.uniprot.org/core/>\n",
    "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
    "PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\n",
    "SELECT ?protein ?sequence\n",
    "WHERE\n",
    "{\n",
    "    ?protein a up:Protein ;\n",
    "        up:organism taxon:10090 ;  # Taxonomy ID for Mus musculus (Mouse)\n",
    "        up:reviewed true ;\n",
    "        up:sequence ?isoform .\n",
    "    ?isoform rdf:value ?sequence .\n",
    "    # Ensure the N-terminal amino acid is Glycine (G)\n",
    "    FILTER (STRSTARTS(?sequence, \"G\"))\n",
    "}\"\"\",\n",
    "  },\n",
    "\n",
    "  {\n",
    "    \"question\": \"How could I download a table that only includes the Rhea reactions for which there is experimental evidence? Return only the rhea URI\",\n",
    "    \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    \"query\": \"\"\"PREFIX up: <http://purl.uniprot.org/core/>\n",
    "PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\n",
    "SELECT DISTINCT ?rhea\n",
    "WHERE {\n",
    "  # ECO 269 is experimental evidence\n",
    "  BIND (<http://purl.obolibrary.org/obo/ECO_0000269> as ?evidence)\n",
    "  #GRAPH <http://sparql.uniprot.org/uniprot> {\n",
    "    ?protein up:reviewed true ;\n",
    "      up:annotation ?a ;\n",
    "      up:attribution ?attribution  .\n",
    "\n",
    "    ?a a up:Catalytic_Activity_Annotation ;\n",
    "      up:catalyticActivity ?ca .\n",
    "    ?ca up:catalyzedReaction ?rhea .\n",
    "\n",
    "    [] rdf:subject ?a ;\n",
    "      rdf:predicate up:catalyticActivity ;\n",
    "      rdf:object ?ca ;\n",
    "      up:attribution ?attribution .\n",
    "\n",
    "    ?attribution up:evidence ?evidence .\n",
    "  #}\n",
    "}\"\"\",\n",
    "  },\n",
    "\n",
    "  {\n",
    "    \"question\": \"Which human proteins are enzymes catalyzing a reaction involving sterols? Return the protein, sterol and reaction URI\",\n",
    "    \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    \"query\": \"\"\"PREFIX up: <http://purl.uniprot.org/core/>\n",
    "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
    "PREFIX rh: <http://rdf.rhea-db.org/>\n",
    "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
    "PREFIX CHEBI: <http://purl.obolibrary.org/obo/CHEBI_>\n",
    "SELECT DISTINCT ?protein ?sterol ?reaction\n",
    "WHERE {\n",
    "  SERVICE <https://sparql.rhea-db.org/sparql> {\n",
    "    ?reaction rdfs:subClassOf rh:Reaction .\n",
    "    ?reaction rh:side/rh:contains/rh:compound ?compound .\n",
    "    ?compound rh:chebi ?sterol .\n",
    "    ?sterol rdfs:subClassOf* CHEBI:15889 .\n",
    "  }\n",
    "  ?protein a up:Protein ;\n",
    "    up:organism taxon:9606 ;\n",
    "    up:annotation/up:catalyticActivity/up:catalyzedReaction ?reaction .\n",
    "}\"\"\",\n",
    "  },\n",
    "\n",
    "  {\n",
    "    \"question\": \"Which are the human proteins associated with cancer (which have cancer in their disease label)? Return the unique disease label, and HGNC symbol\",\n",
    "    \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    \"query\": \"\"\"PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
    "PREFIX up:<http://purl.uniprot.org/core/>\n",
    "PREFIX taxon:<http://purl.uniprot.org/taxonomy/>\n",
    "PREFIX rdfs:<http://www.w3.org/2000/01/rdf-schema#>\n",
    "PREFIX orth:<http://purl.org/net/orth#>\n",
    "PREFIX dcterms:<http://purl.org/dc/terms/>\n",
    "PREFIX obo:<http://purl.obolibrary.org/obo/>\n",
    "PREFIX lscr:<http://purl.org/lscr#>\n",
    "PREFIX genex:<http://purl.org/genex#>\n",
    "PREFIX sio: <http://semanticscience.org/resource/>\n",
    "SELECT DISTINCT ?diseaseLabel ?hgncSymbol\n",
    "WHERE {\n",
    "    ?humanProtein a up:Protein ;\n",
    "        # up:organism/up:scientificName 'Homo sapiens' ;\n",
    "        up:organism taxon:9606 ;\n",
    "        up:annotation ?annotation ;\n",
    "        rdfs:seeAlso ?hgnc .\n",
    "    ?hgnc up:database <http://purl.uniprot.org/database/HGNC> ;\n",
    "        rdfs:comment ?hgncSymbol .\n",
    "    ?annotation a up:Disease_Annotation ;\n",
    "        up:disease ?disease .\n",
    "    ?disease skos:prefLabel ?diseaseLabel.\n",
    "    FILTER CONTAINS (LCASE(?diseaseLabel), \"cancer\")\n",
    "}\"\"\",\n",
    "  },\n",
    "\n",
    "  {\n",
    "    # There are no example with xrefEnsembl, so the RAG without validation usually fails\n",
    "    \"question\": \"How can I get the cross ref to the protein in ensembl for the LCT protein in OMA? Return only the distinct ensembl URI\",\n",
    "    \"endpoint\": \"https://sparql.omabrowser.org/sparql/\",\n",
    "    \"query\": \"\"\"PREFIX orth: <http://purl.org/net/orth#>\n",
    "PREFIX lscr: <http://purl.org/lscr#>\n",
    "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
    "SELECT DISTINCT ?ensemblURI\n",
    "WHERE {\n",
    "    ?protein a orth:Protein ;\n",
    "        rdfs:label 'LCT' ;\n",
    "        lscr:xrefEnsemblProtein ?ensemblURI .\n",
    "}\"\"\",\n",
    "  },\n",
    "\n",
    "  {\n",
    "    # There are no example with inDataset\n",
    "    \"question\": \"How can I get the URI of a dataset to which an ortholog cluster belongs in OMA? Return orthologCluster, datasetURI and limit to 20\",\n",
    "    \"endpoint\": \"https://sparql.omabrowser.org/sparql/\",\n",
    "    \"query\": \"\"\"PREFIX orth: <http://purl.org/net/orth#>\n",
    "SELECT DISTINCT ?orthologCluster ?datasetURI\n",
    "WHERE {\n",
    "    ?orthologCluster a orth:OrthologsCluster ;\n",
    "        orth:inDataset ?datasetURI .\n",
    "} LIMIT 20\"\"\",\n",
    "  },\n",
    "\n",
    "    # In bgee how can I retrieve the confidence and false discovery rate of a gene expression? Limit to 10\n",
    "\n",
    "    # Give me 20 proteins that are enzymes, with their enzyme class and the name for their class from uniprot (make sure to define the up:Enzyme type for the subject of the enzyme class)\n",
    "    # Give me the list of strains associated to Escherichia coli from uniprot and their scientific name (don't use up:Taxon, just the up:Strain)\n",
    "\n",
    "    # Give me the query to retrieve the common name, scientific name of taxons, with the scientific name of their strains limit to 100 results\n",
    "    # Retrieve strains of taxons. Return scientific and common name of taxons, and scientific name of the strains, limit to 100 results\n",
    "\n",
    "\n",
    "#   {\n",
    "#     \"question\": \"Which are the human genes associated with cancer (which have cancer in their disease label) and their orthologs expressed in the rat brain? Return the disease label, human gene URI, human gene HGNC symbol, ortholog rat gene URI\",\n",
    "#     \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "#     \"query\": \"\"\"PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
    "# PREFIX up:<http://purl.uniprot.org/core/>\n",
    "# PREFIX taxon:<http://purl.uniprot.org/taxonomy/>\n",
    "# PREFIX rdfs:<http://www.w3.org/2000/01/rdf-schema#>\n",
    "# PREFIX orth:<http://purl.org/net/orth#>\n",
    "# PREFIX dcterms:<http://purl.org/dc/terms/>\n",
    "# PREFIX obo:<http://purl.obolibrary.org/obo/>\n",
    "# PREFIX lscr:<http://purl.org/lscr#>\n",
    "# PREFIX genex:<http://purl.org/genex#>\n",
    "# PREFIX sio: <http://semanticscience.org/resource/>\n",
    "# SELECT DISTINCT ?diseaseLabel ?humanProtein ?hgncSymbol ?orthologRatProtein ?orthologRatGene\n",
    "# WHERE {\n",
    "#   SERVICE <https://sparql.uniprot.org/sparql> {\n",
    "# \tSELECT DISTINCT * WHERE {\n",
    "#   \t?humanProtein a up:Protein ;\n",
    "#       \tup:organism/up:scientificName 'Homo sapiens' ;\n",
    "#       \tup:annotation ?annotation ;\n",
    "#       \trdfs:seeAlso ?hgnc .\n",
    "#   \t?hgnc up:database <http://purl.uniprot.org/database/HGNC> ;\n",
    "#       \trdfs:comment ?hgncSymbol .\n",
    "#   \t?annotation a up:Disease_Annotation ;\n",
    "#       \t\tup:disease ?disease .\n",
    "#   \t?disease skos:prefLabel ?diseaseLabel.\n",
    "#   \tFILTER CONTAINS (?diseaseLabel, \"cancer\")\n",
    "# \t}\n",
    "#   }\n",
    "#   SERVICE <https://sparql.omabrowser.org/sparql/> {\n",
    "# \tSELECT ?humanProtein ?orthologRatProtein ?orthologRatGene WHERE {\n",
    "#   \t?humanProteinOma a orth:Protein ;\n",
    "#       \t\tlscr:xrefUniprot ?humanProtein .\n",
    "#   \t?orthologRatProtein a orth:Protein ;\n",
    "#       \tsio:SIO_010079 ?orthologRatGene ;\n",
    "#       \torth:organism/obo:RO_0002162/up:scientificName 'Rattus norvegicus' .\n",
    "#   \t?cluster a orth:OrthologsCluster .\n",
    "#   \t?cluster orth:hasHomologousMember ?node1 .\n",
    "#   \t?cluster orth:hasHomologousMember ?node2 .\n",
    "#   \t?node1 orth:hasHomologousMember* ?humanProteinOma .\n",
    "#   \t?node2 orth:hasHomologousMember* ?orthologRatProtein .\n",
    "#   \tFILTER(?node1 != ?node2)\n",
    "# \t}\n",
    "#   }\n",
    "#   SERVICE <https://www.bgee.org/sparql/> {\n",
    "# \t?orthologRatGene genex:isExpressedIn ?anatEntity ;\n",
    "#   \t\torth:organism ?ratOrganism .\n",
    "# \t?anatEntity rdfs:label 'brain' .\n",
    "# \t?ratOrganism obo:RO_0002162 taxon:10116 .\n",
    "#   }\n",
    "# }\"\"\",\n",
    "#   },\n",
    "\n",
    "]\n",
    "\n",
    "def result_sets_are_same(gen_set, ref_set) -> bool:\n",
    "    gen_set, ref_set = list(gen_set), list(ref_set)\n",
    "    for item in gen_set:\n",
    "        if item not in ref_set:\n",
    "            print(f\"Missing from reference: {item}\")\n",
    "            return False\n",
    "    for item in ref_set:\n",
    "        if item not in gen_set:\n",
    "            print(f\"Missing from generated: {item}\")\n",
    "            return False\n",
    "    return True\n",
    "\n",
    "QLEVER_UNIPROT = \"https://qlever.cs.uni-freiburg.de/api/uniprot\"\n",
    "\n",
    "client = OpenAI()\n",
    "def answer_no_rag(question: str):\n",
    "    response = client.chat.completions.create(\n",
    "        model=settings.llm_model,\n",
    "        messages=[\n",
    "            {\"role\": \"system\", \"content\": settings.system_prompt},\n",
    "            {\"role\": \"user\", \"content\": question},\n",
    "        ],\n",
    "        stream=False,\n",
    "        temperature=0,\n",
    "        #   response_format={ \"type\": \"json_object\" },\n",
    "    )\n",
    "    return response.choices[0].message.content\n",
    "\n",
    "def answer_rag_without_validation(question: str):\n",
    "    response = requests.post(\"http://localhost:8000/chat\",\n",
    "        json={\n",
    "            \"messages\": [{\"role\": \"user\", \"content\": question}],\n",
    "            \"stream\": False,\n",
    "            \"api_key\": expasy_api_key,\n",
    "            \"validate\": False,\n",
    "        },\n",
    "        timeout=60,\n",
    "    )\n",
    "    return response.json()[\"choices\"][0][\"message\"][\"content\"]\n",
    "\n",
    "def answer_rag_with_validation(question: str):\n",
    "    response = requests.post(\"http://localhost:8000/chat\",\n",
    "        json={\n",
    "            \"messages\": [{\"role\": \"user\", \"content\": question}],\n",
    "            \"stream\": False,\n",
    "            \"api_key\": expasy_api_key,\n",
    "        },\n",
    "        timeout=60,\n",
    "    )\n",
    "    return response.json()[\"choices\"][0][\"message\"][\"content\"]\n",
    "\n",
    "\n",
    "list_of_approaches = [\n",
    "    answer_no_rag,\n",
    "    answer_rag_without_validation,\n",
    "    answer_rag_with_validation,\n",
    "]\n",
    "\n",
    "number_of_tries = 3\n",
    "\n",
    "success = [0, 0, 0]\n",
    "mismatch = [0, 0, 0]\n",
    "fail = [0, 0, 0]\n",
    "\n",
    "\n",
    "print(f\"üß™ Testing {len(example_queries)} queries\")\n",
    "for i, test_query in enumerate(example_queries):\n",
    "    # if i < 1:\n",
    "    #     continue\n",
    "    # Execute the reference query\n",
    "    ref_endpoint = SPARQLWrapper(test_query[\"endpoint\"])\n",
    "    # ref_endpoint = SPARQLWrapper(QLEVER_UNIPROT)\n",
    "    ref_endpoint.setReturnFormat(JSON)\n",
    "    # ref_endpoint.setTimeout(200)\n",
    "    ref_endpoint.setQuery(test_query[\"query\"])\n",
    "    results = ref_endpoint.query().convert()\n",
    "    res_from_ref = results[\"results\"][\"bindings\"]\n",
    "\n",
    "\n",
    "    for approach_index, approach in enumerate(list_of_approaches):\n",
    "        for t in range(number_of_tries):\n",
    "            # resp = requests.post(\"http://localhost:8000/chat\",\n",
    "            #     json={\n",
    "            #         \"messages\": [{\"role\": \"user\", \"content\": test_query[\"question\"]}],\n",
    "            #         \"stream\": False,\n",
    "            #         \"api_key\": expasy_api_key,\n",
    "            #     },\n",
    "            #     timeout=60,\n",
    "            # )\n",
    "            # chat_resp_md = resp.json()[\"choices\"][0][\"message\"][\"content\"]\n",
    "\n",
    "            chat_resp_md = approach(test_query[\"question\"])\n",
    "\n",
    "            try:\n",
    "                generated_sparql = extract_sparql_queries(chat_resp_md)[-1]\n",
    "                if generated_sparql[\"query\"].strip() == test_query[\"query\"].strip():\n",
    "                    print(f\"‚úÖ {t+1}/{number_of_tries} {test_query['question']}. EXACT MATCH\")\n",
    "                    success[approach_index] += 1\n",
    "                    continue\n",
    "\n",
    "                # Execute the generated query\n",
    "                sparql_endpoint = SPARQLWrapper(generated_sparql[\"endpoint\"])\n",
    "                # sparql_endpoint = SPARQLWrapper(QLEVER_UNIPROT)\n",
    "                sparql_endpoint.setReturnFormat(JSON)\n",
    "                sparql_endpoint.setTimeout(200)\n",
    "                sparql_endpoint.setQuery(generated_sparql[\"query\"])\n",
    "                results = sparql_endpoint.query().convert()\n",
    "                res_from_generated = results[\"results\"][\"bindings\"]\n",
    "\n",
    "                if not result_sets_are_same(res_from_generated, res_from_ref):\n",
    "                    mismatch[approach_index] += 1\n",
    "                    raise Exception(f\"\\nResults mismatch. Ref: {len(res_from_ref)} != gen: {len(res_from_generated)}\\n\")\n",
    "                else:\n",
    "                    print(f\"‚úÖ {t+1}/{number_of_tries} {test_query['question']} = {len(res_from_generated)}\")\n",
    "                    success[approach_index] += 1\n",
    "\n",
    "            except Exception as e:\n",
    "                fail[approach_index] += 1\n",
    "                print(f\"‚ùå {t+1}/{number_of_tries} {test_query['question']} {e}. Generated query:\")\n",
    "                print(generated_sparql[\"query\"])\n",
    "                print(\"Correct query:\")\n",
    "                print(test_query[\"query\"])\n",
    "                print(\"\")\n",
    "            # print(f\"Results: {len(results['results']['bindings'])}\")\n",
    "\n",
    "    print(f\"‚öñÔ∏è No RAG - Success: {success[0]}, Mismatch: {mismatch[0]}, Error: {fail[0]-mismatch[0]}\")\n",
    "    print(f\"‚öñÔ∏è RAG without validation - Success: {success[1]}, Mismatch: {mismatch[1]}, Error: {fail[1]-mismatch[1]}\")\n",
    "    print(f\"‚öñÔ∏è RAG with validation - Success: {success[2]}, Mismatch: {mismatch[2]}, Error: {fail[2]-mismatch[2]}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "expasy-scripts",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
