{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Testing Expasy query helper\n",
    "\n",
    "Testing our Expasy query helper reusing the example queries that succeeded retrieved from `test_example_queries.ipynb`\n",
    "\n",
    "Potential interesting questions to use for evaluation:\n",
    "- Which are the human genes associated with cancer and their orthologs expressed in the rat brain?\n",
    "- Select the number of proteins for each of the subclasses of ec:1.1.1.- EC (Enzyme Commission) class (from biosoda)\n",
    "- "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "üß™ Testing 4 queries\n",
      "‚úÖ what is the accession number in uniprot of the human gene LCT? Return only the distinct protein URIs = 4\n",
      "‚úÖ what is the accession number in uniprot of the human gene LCT? Return only the distinct protein URIs = 4\n",
      "‚úÖ what is the accession number in uniprot of the human gene LCT? Return only the distinct protein URIs = 4\n",
      "‚öñÔ∏è Success: 3, Mismatch: 0, Error: 0\n",
      "‚úÖ How do I filter for reviewed mouse proteins which carry an N-terminal glycine? Return protein URI and AA sequence = 20\n",
      "‚úÖ How do I filter for reviewed mouse proteins which carry an N-terminal glycine? Return protein URI and AA sequence = 20\n",
      "‚úÖ How do I filter for reviewed mouse proteins which carry an N-terminal glycine? Return protein URI and AA sequence = 20\n",
      "‚öñÔ∏è Success: 6, Mismatch: 0, Error: 0\n",
      "‚úÖ How could I download a table that only includes the Rhea reactions for which there is experimental evidence? Return only the rhea URI = 47936\n",
      "‚úÖ How could I download a table that only includes the Rhea reactions for which there is experimental evidence? Return only the rhea URI = 47936\n",
      "‚úÖ How could I download a table that only includes the Rhea reactions for which there is experimental evidence? Return only the rhea URI = 11569\n",
      "‚öñÔ∏è Success: 9, Mismatch: 0, Error: 0\n",
      "‚úÖ Which human proteins are enzymes catalyzing a reaction involving lipids? Return the protein, lipid and reaction URI = 14840\n",
      "Missing from reference: {'reaction': {'type': 'uri', 'value': 'http://rdf.rhea-db.org/17469'}, 'protein': {'type': 'uri', 'value': 'http://purl.uniprot.org/uniprot/O95237'}, 'lipid': {'type': 'uri', 'value': 'http://purl.obolibrary.org/obo/CHEBI_17336'}}\n",
      "Ref: 14840 != gen: 14851\n",
      "‚ùå 1/3 Which human proteins are enzymes catalyzing a reaction involving lipids? Return the protein, lipid and reaction URI \n",
      "Results mismatch. Generated query:\n",
      "# https://sparql.uniprot.org/sparql/\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX rh: <http://rdf.rhea-db.org/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX CHEBI: <http://purl.obolibrary.org/obo/CHEBI_>\n",
      "\n",
      "SELECT DISTINCT ?protein ?lipid ?reaction\n",
      "WHERE {\n",
      "  # Rhea: retrieve reactions involving lipids (CHEBI:18059 and its subclasses)\n",
      "  SERVICE <https://sparql.rhea-db.org/sparql> {\n",
      "    ?reaction rdfs:subClassOf rh:Reaction .\n",
      "    ?reaction rh:side/rh:contains/rh:compound ?compound .\n",
      "    ?compound (rh:chebi|(rh:reactivePart/rh:chebi)|(rh:underlyingChebi/rh:chebi)) ?lipid .\n",
      "    ?lipid rdfs:subClassOf* CHEBI:18059 . # lipid\n",
      "  }\n",
      " \n",
      "  # UniProt: retrieve human enzymes catalyzing these reactions\n",
      "  ?protein a up:Protein ;\n",
      "           up:organism taxon:9606 ;\n",
      "           up:annotation ?caa .\n",
      "  ?caa a up:Catalytic_Activity_Annotation ;\n",
      "       up:catalyticActivity ?ca.\n",
      "  ?ca up:catalyzedReaction ?reaction .\n",
      "}\n",
      "Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX rh: <http://rdf.rhea-db.org/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX CHEBI: <http://purl.obolibrary.org/obo/CHEBI_>\n",
      "SELECT DISTINCT ?protein ?lipid ?reaction\n",
      "WHERE {\n",
      "  SERVICE <https://sparql.rhea-db.org/sparql> {\n",
      "    ?reaction rdfs:subClassOf rh:Reaction .\n",
      "    ?reaction rh:side/rh:contains/rh:compound ?compound .\n",
      "    ?compound rh:chebi ?lipid .\n",
      "    ?lipid rdfs:subClassOf* CHEBI:18059 .\n",
      "  }\n",
      "  ?protein a up:Protein ;\n",
      "    up:organism taxon:9606 ;\n",
      "    up:annotation/up:catalyticActivity/up:catalyzedReaction ?reaction .\n",
      "}\n",
      "\n",
      "‚úÖ Which human proteins are enzymes catalyzing a reaction involving lipids? Return the protein, lipid and reaction URI = 14840\n",
      "‚öñÔ∏è Success: 11, Mismatch: 1, Error: 0\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "import os\n",
    "import requests\n",
    "\n",
    "from qdrant_client.models import FieldCondition, Filter, MatchValue, ScoredPoint\n",
    "from rdflib.plugins.sparql import prepareQuery, prepareUpdate\n",
    "from rdflib.plugins.sparql.algebra import translateQuery\n",
    "from SPARQLWrapper import JSON, TURTLE, XML, SPARQLWrapper\n",
    "from dotenv import load_dotenv\n",
    "\n",
    "from expasy_chat.utils import extract_sparql_queries\n",
    "\n",
    "load_dotenv()\n",
    "expasy_api_key = os.getenv('EXPASY_API_KEY')\n",
    "\n",
    "example_queries = [\n",
    "  {\n",
    "    \"question\": \"what is the accession number in uniprot of the human gene LCT? Return only the distinct protein URIs\",\n",
    "    \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    \"query\": \"\"\"PREFIX up: <http://purl.uniprot.org/core/>\n",
    "PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
    "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
    "\n",
    "SELECT ?protein\n",
    "WHERE\n",
    "{\n",
    "    ?protein a up:Protein .\n",
    "    ?protein up:organism taxon:9606 .\n",
    "    ?protein up:encodedBy ?gene .\n",
    "    ?gene skos:prefLabel \"LCT\" .\n",
    "}\"\"\",\n",
    "  },\n",
    "  {\n",
    "    # NOTE: The mature part of the question makes it harder to answer\n",
    "    # \"question\": \"How do I filter for reviewed (mouse) proteins whose mature form carries an N-terminal glycine? Return protein URI and AA sequence\",\n",
    "    \"question\": \"How do I filter for reviewed mouse proteins which carry an N-terminal glycine? Return protein URI and AA sequence\",\n",
    "    \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    \"query\": \"\"\"PREFIX up: <http://purl.uniprot.org/core/>\n",
    "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
    "PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\n",
    "SELECT ?protein ?sequence\n",
    "WHERE\n",
    "{\n",
    "    ?protein a up:Protein ;\n",
    "        up:organism taxon:10090 ;  # Taxonomy ID for Mus musculus (Mouse)\n",
    "        up:reviewed true ;\n",
    "        up:sequence ?isoform .\n",
    "    ?isoform rdf:value ?sequence .\n",
    "    # Ensure the N-terminal amino acid is Glycine (G)\n",
    "    FILTER (STRSTARTS(?sequence, \"G\"))\n",
    "}\"\"\",\n",
    "  },\n",
    "\n",
    "  {\n",
    "    \"question\": \"How could I download a table that only includes the Rhea reactions for which there is experimental evidence? Return only the rhea URI\",\n",
    "    \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    \"query\": \"\"\"PREFIX up: <http://purl.uniprot.org/core/>\n",
    "PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\n",
    "SELECT DISTINCT ?rhea\n",
    "WHERE {\n",
    "  # ECO 269 is experimental evidence\n",
    "  BIND (<http://purl.obolibrary.org/obo/ECO_0000269> as ?evidence)\n",
    "  GRAPH <http://sparql.uniprot.org/uniprot> {\n",
    "    ?protein up:reviewed true ;\n",
    "      up:annotation ?a ;\n",
    "      up:attribution ?attribution  .\n",
    "\n",
    "    ?a a up:Catalytic_Activity_Annotation ;\n",
    "      up:catalyticActivity ?ca .\n",
    "    ?ca up:catalyzedReaction ?rhea .\n",
    "\n",
    "    [] rdf:subject ?a ;\n",
    "      rdf:predicate up:catalyticActivity ;\n",
    "      rdf:object ?ca ;\n",
    "      up:attribution ?attribution .\n",
    "\n",
    "    ?attribution up:evidence ?evidence .\n",
    "  }\n",
    "}\"\"\",\n",
    "  },\n",
    "\n",
    "  {\n",
    "    \"question\": \"Which human proteins are enzymes catalyzing a reaction involving lipids? Return the protein, lipid and reaction URI\",\n",
    "    \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    \"query\": \"\"\"PREFIX up: <http://purl.uniprot.org/core/>\n",
    "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
    "PREFIX rh: <http://rdf.rhea-db.org/>\n",
    "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
    "PREFIX CHEBI: <http://purl.obolibrary.org/obo/CHEBI_>\n",
    "SELECT DISTINCT ?protein ?lipid ?reaction\n",
    "WHERE {\n",
    "  SERVICE <https://sparql.rhea-db.org/sparql> {\n",
    "    ?reaction rdfs:subClassOf rh:Reaction .\n",
    "    ?reaction rh:side/rh:contains/rh:compound ?compound .\n",
    "    ?compound rh:chebi ?lipid .\n",
    "    ?lipid rdfs:subClassOf* CHEBI:18059 .\n",
    "  }\n",
    "  ?protein a up:Protein ;\n",
    "    up:organism taxon:9606 ;\n",
    "    up:annotation/up:catalyticActivity/up:catalyzedReaction ?reaction .\n",
    "}\"\"\",\n",
    "  },\n",
    "\n",
    "]\n",
    "\n",
    "number_of_tries = 3\n",
    "\n",
    "success = 0\n",
    "mismatch = 0\n",
    "fail = 0\n",
    "\n",
    "def result_sets_are_same(gen_set, ref_set):\n",
    "    gen_set, ref_set = list(gen_set), list(ref_set)\n",
    "    for item in gen_set:\n",
    "        if item not in ref_set:\n",
    "            print(f\"Missing from reference: {item}\")\n",
    "            return False\n",
    "    for item in ref_set:\n",
    "        if item not in gen_set:\n",
    "            print(f\"Missing from generated: {item}\")\n",
    "            return False\n",
    "    return True\n",
    "\n",
    "\n",
    "print(f\"üß™ Testing {len(example_queries)} queries\")\n",
    "for i, test_query in enumerate(example_queries):\n",
    "    # if i < 1:\n",
    "    #     continue\n",
    "    # Execute the reference query\n",
    "    sparql_endpoint.setQuery(test_query[\"query\"])\n",
    "    results = sparql_endpoint.query().convert()\n",
    "    res_from_ref = results[\"results\"][\"bindings\"]\n",
    "\n",
    "    for t in range(number_of_tries):\n",
    "        resp = requests.post(\"http://localhost:8000/chat\",\n",
    "            json={\n",
    "                \"messages\": [{\"role\": \"user\", \"content\": test_query[\"question\"]}],\n",
    "                \"stream\": False,\n",
    "                \"api_key\": expasy_api_key,\n",
    "            },\n",
    "            timeout=60,\n",
    "        )\n",
    "        chat_resp_md = resp.json()[\"choices\"][0][\"message\"][\"content\"]\n",
    "\n",
    "        try:\n",
    "            generated_sparql = extract_sparql_queries(chat_resp_md)[-1]\n",
    "            if generated_sparql[\"query\"].strip() == test_query[\"query\"].strip():\n",
    "                print(f\"‚úÖ {t+1}/{number_of_tries} {test_query['question']}. EXACT MATCH\")\n",
    "                success += 1\n",
    "                continue\n",
    "\n",
    "            # Execute the generated query\n",
    "            sparql_endpoint = SPARQLWrapper(generated_sparql[\"endpoint\"])\n",
    "            sparql_endpoint.setReturnFormat(JSON)\n",
    "            sparql_endpoint.setTimeout(200)\n",
    "            sparql_endpoint.setQuery(generated_sparql[\"query\"])\n",
    "            results = sparql_endpoint.query().convert()\n",
    "            res_from_generated = results[\"results\"][\"bindings\"]\n",
    "\n",
    "            if not result_sets_are_same(res_from_generated, res_from_ref):\n",
    "                mismatch += 1\n",
    "                print(f\"Ref: {len(res_from_ref)} != gen: {len(res_from_generated)}\")\n",
    "                raise Exception(f\"\\nResults mismatch\")\n",
    "            else:\n",
    "                print(f\"‚úÖ {t+1}/{number_of_tries} {test_query['question']} = {len(res_from_generated)}\")\n",
    "                success += 1\n",
    "\n",
    "        except Exception as e:\n",
    "            fail += 1\n",
    "            print(f\"‚ùå {t+1}/{number_of_tries} {test_query['question']} {e}. Generated query:\")\n",
    "            print(generated_sparql[\"query\"])\n",
    "            print(\"Correct query:\")\n",
    "            print(test_query[\"query\"])\n",
    "            print(\"\")\n",
    "        # print(f\"Results: {len(results['results']['bindings'])}\")\n",
    "\n",
    "    print(f\"‚öñÔ∏è Success: {success}, Mismatch: {mismatch}, Error: {fail-mismatch}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "expasy-scripts",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
