{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Testing Expasy query helper\n",
    "\n",
    "Testing our Expasy query helper reusing the example queries that succeeded retrieved from `test_example_queries.ipynb`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "SELECT ?taxon\n",
      "FROM <http://sparql.uniprot.org/taxonomy>\n",
      "WHERE\n",
      "{\n",
      "    ?taxon a up:Taxon .\n",
      "}\n",
      "✅ Select all taxa from the UniProt taxonomy. EXACT MATCH\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX keywords: <http://purl.uniprot.org/keywords/>\n",
      "SELECT (COUNT(DISTINCT ?taxon) AS ?numExtinctOrganisms)\n",
      "WHERE\n",
      "{\n",
      "  GRAPH <http://sparql.uniprot.org/taxonomy> {\n",
      "    ?taxon a up:Taxon .\n",
      "  }\n",
      "  ?protein up:organism ?taxon ;\n",
      "           up:classifiedWith keywords:952 .\n",
      "}\n",
      "❌ How many distinct extinct organisms are represented in UniProtKB 29 != 1. Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX keywords: <http://purl.uniprot.org/keywords/>\n",
      "SELECT \n",
      "  ?taxon\n",
      "  (SAMPLE(?name) AS ?anName)\n",
      "  (COUNT(DISTINCT ?protein) AS ?entriesPerExtinctTaxon)\n",
      "WHERE\n",
      "{\n",
      "  GRAPH<http://sparql.uniprot.org/taxonomy>{\n",
      "    ?taxon a up:Taxon ;\n",
      "           up:scientificName ?name .\n",
      "  }\n",
      "  ?protein up:organism ?taxon ;\n",
      "           up:classifiedWith keywords:952 .\n",
      "} GROUP BY ?taxon ORDER BY ?taxon\n",
      "\n",
      "⚖️ Success: 1, Mismatch: 1, Fail: 1\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\n",
      "PREFIX faldo: <http://biohackathon.org/resource/faldo#>\n",
      "SELECT ?protein ?annotation ?begin ?text\n",
      "WHERE\n",
      "{\n",
      "        ?protein a up:Protein ;\n",
      "            up:organism taxon:9606 ; \n",
      "            up:annotation ?annotation .\n",
      "        ?annotation a up:Natural_Variant_Annotation ;\n",
      "            rdfs:comment ?text ;\n",
      "            up:substitution ?substitution ;\n",
      "            up:range/faldo:begin\n",
      "                [ faldo:position ?begin ;\n",
      "                  faldo:reference ?sequence ] .\n",
      "        ?sequence rdf:value ?value .\n",
      "        BIND (substr(?value, ?begin, 1) as ?original) .\n",
      "        FILTER(?original = 'Y' && ?substitution = 'F') .\n",
      "}\n",
      "✅ Select all human UniProt entries with a sequence variant that leads to a tyrosine to phenylalanine substitution. EXACT MATCH\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "SELECT ?protein\n",
      "WHERE\n",
      "{\n",
      "\t?protein a up:Protein . \n",
      "\t?protein up:created '2010-11-30'^^xsd:date\n",
      "}\n",
      "✅ Select all UniProt entries that were integrated on the 30th of November 2010. EXACT MATCH\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "SELECT (AVG(?linksToPdbPerEntry) AS ?avgLinksToPdbPerEntry)\n",
      "WHERE\n",
      "{\n",
      "\tSELECT ?protein (COUNT(DISTINCT ?db) AS ?linksToPdbPerEntry)\n",
      "\tWHERE\n",
      "\t{\n",
      "\t\t?protein a up:Protein .\n",
      "\t\t?protein rdfs:seeAlso ?db .\n",
      "\t\t?db up:database <http://purl.uniprot.org/database/PDB> .\n",
      "\t}\n",
      "\tGROUP BY ?protein\n",
      "}\n",
      "✅ Select the average number of cross-references to the PDB database of UniProt entries that have at least one cross-reference to the PDB database 1 == 1\n",
      "⚖️ Success: 4, Mismatch: 1, Fail: 1\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX ec: <http://purl.uniprot.org/enzyme/>\n",
      "SELECT ?ecClass (COUNT(?protein) as ?size)\n",
      "WHERE\n",
      "{\n",
      "    VALUES (?ecClass) {(ec:1.-.-.-) (ec:2.-.-.-) (ec:3.-.-.-) (ec:4.-.-.-) (ec:5.-.-.-) (ec:6.-.-.-) (ec:7.-.-.-)} .\n",
      "    ?protein ( up:enzyme | up:domain/up:enzyme | up:component/up:enzyme ) ?enzyme .\n",
      "    # Enzyme subclasses are materialized, do not use rdfs:subClassOf+\n",
      "    ?enzyme rdfs:subClassOf ?ecClass .\n",
      "}\n",
      "GROUP BY ?ecClass ORDER BY ?ecClass\n",
      "✅ Select the number of UniProt entries for each of the EC (Enzyme Commission) top level categories. EXACT MATCH\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\n",
      "SELECT \n",
      "    ?accession\n",
      "    ?annotation_acc \n",
      "    ?pubmed\n",
      "WHERE\n",
      "{\n",
      "        ?protein a up:Protein ;\n",
      "            up:annotation ?annotation .\n",
      "        ?annotation a up:Natural_Variant_Annotation .\n",
      "        ?linkToEvidence rdf:object ?annotation ;\n",
      "                        up:attribution ?attribution .\n",
      "        ?attribution up:source ?source .\n",
      "        ?source a up:Journal_Citation .\n",
      "  BIND(SUBSTR(STR(?protein),33) AS ?accession)\n",
      "  BIND(IF(CONTAINS(STR(?annotation), \"#SIP\"), SUBSTR(STR(?annotation),33), SUBSTR(STR(?annotation),36))AS?annotation_acc)\n",
      "  BIND(SUBSTR(STR(?source),35) AS ?pubmed)\n",
      "}\n",
      "✅ Find all Natural Variant Annotations if associated via an evidence tag to an article with a pubmed identifier. EXACT MATCH\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "SELECT ?taxon ?name\n",
      "WHERE\n",
      "{\n",
      "    ?taxon a up:Taxon .\n",
      "    ?taxon up:scientificName ?name .\n",
      "    # Taxon subclasses are materialized, do not use rdfs:subClassOf+\n",
      "    ?taxon rdfs:subClassOf taxon:2 .\n",
      "}\n",
      "✅ Select all bacterial taxa and their scientific name from the UniProt taxonomy. EXACT MATCH\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\n",
      "SELECT \n",
      "    ?pubmed \n",
      "    (COUNT(?attribution) AS ?citations)\n",
      "WHERE\n",
      "{\n",
      "        ?protein a up:Protein ;\n",
      "            up:organism taxon:9606 ;\n",
      "            up:annotation ?annotation .\n",
      "        ?linkToEvidence rdf:object ?annotation ;\n",
      "                        up:attribution ?attribution .\n",
      "        ?attribution up:source ?source .\n",
      "        ?source a up:Journal_Citation .\n",
      "        BIND(SUBSTR(STR(?source), 35) AS ?pubmed)\n",
      "} \n",
      "GROUP BY ?pubmed \n",
      "ORDER BY DESC(?citations)\n",
      "✅ Find how often an article in pubmed was used in an evidence tag in a human protein (ordered by most used to least) 43889 == 43889\n",
      "⚖️ Success: 8, Mismatch: 1, Fail: 1\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "\n",
      "SELECT \n",
      "    (CONCAT(SUBSTR(STR(?protein), 33)) AS ?uniprot)\n",
      "    (GROUP_CONCAT(DISTINCT ?biotype; separator=\"; \") AS ?biological_processes)\n",
      "    (GROUP_CONCAT(DISTINCT ?celtype; separator=\"; \") AS ?cellular_components)\n",
      "    (GROUP_CONCAT(DISTINCT ?moltype; separator=\"; \") AS ?molecular_functions)\n",
      "WHERE\n",
      "{\n",
      "    VALUES (?ac) {(\"P12345\") (\"Q67890\")}  # Replace with the accessions you're interested in\n",
      "    BIND (IRI(CONCAT(\"http://purl.uniprot.org/uniprot/\",?ac)) AS ?protein)\n",
      "    ?protein a up:Protein .\n",
      "    ?protein up:classifiedWith ?goTerm .\n",
      "    # Determine if the type is biological process\n",
      "    OPTIONAL {\n",
      "        ?goTerm rdfs:subClassOf <http://purl.obolibrary.org/obo/GO_0008150>.\n",
      "        ?goTerm rdfs:label ?biotype .\n",
      "    }\n",
      "    # Determine if the type is cellular component\n",
      "    OPTIONAL {\n",
      "        ?goTerm rdfs:subClassOf <http://purl.obolibrary.org/obo/GO_0005575>.\n",
      "        ?goTerm rdfs:label ?celtype .\n",
      "    }\n",
      "    # Determine if the type is molecular function\n",
      "    OPTIONAL {\n",
      "        ?goTerm rdfs:subClassOf <http://purl.obolibrary.org/obo/GO_0003674> .\n",
      "        ?goTerm rdfs:label ?moltype .\n",
      "    }\n",
      "    # Filter out the uniprot keywords\n",
      "    FILTER(bound(?biotype) || bound(?celtype) || bound(?moltype))\n",
      "}\n",
      "GROUP BY ?protein\n",
      "✅ For two accessions find the GO term labels and group them into GO process,function and component 2 == 2\n",
      "⚖️ Success: 9, Mismatch: 1, Fail: 1\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\n",
      "PREFIX GO: <http://purl.obolibrary.org/obo/GO_>\n",
      "\n",
      "SELECT\n",
      "    (COUNT(DISTINCT(?protein)) AS ?proteinCount)\n",
      "WHERE\n",
      "{   \n",
      "    ?protein rdf:type up:Protein ;\n",
      "             up:reviewed true ;\n",
      "             up:classifiedWith|(up:classifiedWith/rdfs:subClassOf) GO:0016301 . # GO:0016301 corresponds to kinase activity\n",
      "}\n",
      "✅ Number of reviewed entries (Swiss-Prot) that are related to kinase activity 1 == 1\n",
      "⚖️ Success: 10, Mismatch: 1, Fail: 1\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "\n",
      "SELECT\n",
      "  ?protein\n",
      "  ?anyKindOfName \n",
      "  ?names \n",
      "  ?partType\n",
      "WHERE\n",
      "{\n",
      "  BIND(<http://purl.uniprot.org/uniprot/P05067> AS ?protein)\n",
      "  \n",
      "  ?protein a up:Protein .\n",
      "  \n",
      "  {\n",
      "    ?protein (up:recommendedName|up:alternativeName) ?structuredName .\n",
      "  }\n",
      "  UNION\n",
      "  {\n",
      "    VALUES(?partType){(up:domain) (up:component)}\n",
      "    ?protein ?partType ?part .\n",
      "    ?part (up:recommendedName|up:alternativeName) ?structuredName .\n",
      "  }\n",
      "  \n",
      "  ?structuredName ?anyKindOfName  ?names .\n",
      "  ?anyKindOfName rdfs:subPropertyOf up:structuredNameType .\n",
      "}\n",
      "❌ Find all names associated with uniprot entry P05067, and if the name is associated with the entry it's domains or its components The read operation timed out. Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "SELECT\n",
      "  ?protein\n",
      "  ?anyKindOfName \n",
      "  ?names \n",
      "  ?partType\n",
      "WHERE\n",
      "{\n",
      "  BIND(<http://purl.uniprot.org/uniprot/P05067> AS ?protein)\n",
      "  ?protein a up:Protein .\n",
      "  {\n",
      "    ?protein (up:recommendedName|up:alternativeName) ?structuredName .\n",
      "  }\n",
      "    UNION\n",
      "  {\n",
      "    VALUES(?partType){(up:domain) (up:component)}\n",
      "    ?protein ?partType ?part .\n",
      "    ?part (up:recommendedName|up:alternativeName) ?structuredName .\n",
      "  }\n",
      "  ?structuredName ?anyKindOfName  ?names .\n",
      "  ?anyKindOfName rdfs:subPropertyOf up:structuredNameType .\n",
      "}\n",
      "\n",
      "⚖️ Success: 10, Mismatch: 1, Fail: 2\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "SELECT \n",
      "  ?protein\n",
      "WHERE\n",
      "{\n",
      "  ?protein a up:Protein ;\n",
      "           up:proteome <http://purl.uniprot.org/proteomes/UP000000625#Chromosome> .\n",
      "}\n",
      "❌ Get the list of uniprot entries for the chromosome of proteome UP000000625 8802 != 4404. Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "SELECT \n",
      "  ?protein\n",
      "  ?proteome \n",
      "WHERE\n",
      "{\n",
      "  ?protein a up:Protein ;\n",
      "           up:reviewed true ;\n",
      "           up:proteome ?proteome .\n",
      "  VALUES (?proteome) {(<http://purl.uniprot.org/proteomes/UP000000625#Chromosome>)}\n",
      "}\n",
      "\n",
      "⚖️ Success: 10, Mismatch: 2, Fail: 3\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\n",
      "\n",
      "SELECT ?protein ?organism ?isoform ?sequence\n",
      "WHERE\n",
      "{\n",
      "    ?protein a up:Protein .\n",
      "    ?protein up:organism ?organism .\n",
      "    # Taxon subclasses are materialized, do not use rdfs:subClassOf+\n",
      "    ?organism rdfs:subClassOf taxon:83333 . # E. coli K12 and all its strains\n",
      "    ?protein up:sequence ?isoform .\n",
      "    ?isoform rdf:value ?sequence .\n",
      "}\n",
      "✅ Select all UniProt entries, and their organism and amino acid sequences (including isoforms), for E. coli K12 and all its strains 3746 == 3746\n",
      "⚖️ Success: 11, Mismatch: 2, Fail: 3\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
      "SELECT\n",
      "  ?protein \n",
      "  (GROUP_CONCAT(?locusName; separator=',') AS ?locusNames)\n",
      "WHERE \n",
      "{ \n",
      "  ?protein a up:Protein ;\n",
      "    up:organism taxon:360910 ;\n",
      "    up:encodedBy ?gene .\n",
      "  ?gene up:locusName ?locusName .\n",
      "} \n",
      "GROUP BY ?protein \n",
      "HAVING (COUNT(?locusName) > 1)\n",
      "✅ Find UniProt entries with merged loci in Bordetella avium. EXACT MATCH\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\n",
      "SELECT \n",
      "    ?protein \n",
      "    (GROUP_CONCAT(?comment; separator=\", \") AS ?comments)\n",
      "WHERE\n",
      "{\n",
      "    ?protein a up:Protein ;\n",
      "             up:annotation ?annotation . \n",
      "    ?annotation rdf:type up:Topological_Domain_Annotation ;\n",
      "                rdfs:comment ?comment .\n",
      "} \n",
      "GROUP BY ?protein \n",
      "HAVING (COUNT(?annotation) > 1)\n",
      "✅ Find UniProtKB entries with more than 1 Topological domain annotation 175067 == 175067\n",
      "⚖️ Success: 13, Mismatch: 2, Fail: 3\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "SELECT \n",
      "    ?annotation ?comment\n",
      "WHERE {\n",
      "    ?annotation a up:Natural_Variant_Annotation ;\n",
      "        rdfs:comment ?comment . \n",
      "} \n",
      "ORDER BY DESC(STRLEN(?comment))\n",
      "LIMIT 1\n",
      "❌ Find longest comment text associated with a UniProtKB Natural Variant Annotation 66010 != 1. Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "SELECT \n",
      "    ?annotation ?comment\n",
      "WHERE {\n",
      "    ?annotation a up:Natural_Variant_Annotation ;\n",
      "        rdfs:comment ?comment . \n",
      "} \n",
      "ORDER BY DESC(STRLEN(?comment))\n",
      "\n",
      "\n",
      "⚖️ Success: 13, Mismatch: 3, Fail: 4\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\n",
      "SELECT \n",
      "    ?comment1 \n",
      "    ?comment2 \n",
      "    (COUNT(?comment1) AS ?count1)\n",
      "WHERE\n",
      "{\n",
      "    ?protein a up:Protein ;\n",
      "               up:annotation ?annotation1 , \n",
      "                             ?annotation2 . \n",
      "    ?annotation1 rdf:type up:Topological_Domain_Annotation ;\n",
      "        rdfs:comment ?rawComment1 .\n",
      "    ?annotation2 rdf:type up:Topological_Domain_Annotation ;\n",
      "        rdfs:comment ?rawComment2 . \n",
      "    BIND(IF(contains(?rawComment1, ';'), \n",
      "            STRBEFORE(?rawComment1,';'), \n",
      "            ?rawComment1) AS ?comment1)\n",
      "    BIND(IF(contains(?rawComment2, ';'), \n",
      "            STRBEFORE(?rawComment2,';'), \n",
      "            ?rawComment2) AS ?comment2)\n",
      "    FILTER(?annotation1 != ?annotation2)\n",
      "} \n",
      "GROUP BY ?comment1 ?comment2 \n",
      "ORDER BY DESC(COUNT(?comment1))\n",
      "✅ Find the co-occurence count of Topological Domain comment text in UniProtKB entries. EXACT MATCH\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX uniprotkb: <http://purl.uniprot.org/uniprot/>\n",
      "SELECT \n",
      "    ?similar ?identity\n",
      "FROM <http://sparql.uniprot.org/uniref>\n",
      "FROM <http://sparql.uniprot.org/uniprot>\n",
      "WHERE\n",
      "{\n",
      "    BIND (uniprotkb:P05067 AS ?protein)\n",
      "    ?cluster up:member ?member ;\n",
      "             up:member/up:sequenceFor ?protein;\n",
      "             up:identity ?identity .\n",
      "    ?member up:sequenceFor ?similar .\n",
      "    FILTER(!sameTerm(?similar, ?protein))\n",
      "} \n",
      "ORDER BY DESC(?identity)\n",
      "❌ Find the similar proteins for UniProtKB entry P05067 sorted by UniRef cluser identity 79 != 857. Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX uniprotkb: <http://purl.uniprot.org/uniprot/>\n",
      "SELECT \n",
      "    ?similar ?identity\n",
      "FROM <http://sparql.uniprot.org/uniref>\n",
      "FROM <http://sparql.uniprot.org/uniprot>\n",
      "WHERE\n",
      "{\n",
      "    BIND (uniprotkb:P05607 AS ?protein)\n",
      "    ?cluster up:member ?member ;\n",
      "             up:member/up:sequenceFor ?protein;\n",
      "             up:identity ?identity .\n",
      "    ?member up:sequenceFor ?similar .\n",
      "    FILTER(!sameTerm(?similar, ?protein))\n",
      "} \n",
      "ORDER BY DESC(?identity)\n",
      "\n",
      "\n",
      "⚖️ Success: 14, Mismatch: 4, Fail: 5\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\n",
      "PREFIX faldo: <http://biohackathon.org/resource/faldo#>\n",
      "\n",
      "SELECT \n",
      "  ?protein \n",
      "  ?comment\n",
      "  ?begin\n",
      "  ?end \n",
      "WHERE\n",
      "{\n",
      "  ?protein a up:Protein ;\n",
      "    up:organism taxon:9606 ; \n",
      "    up:sequence ?sequence ;\n",
      "    up:annotation ?annotation .\n",
      "  ?annotation a up:Modified_Residue_Annotation ;\n",
      "    rdfs:comment ?comment ;\n",
      "    up:range ?range .\n",
      "  ?range \n",
      "    faldo:begin [ faldo:position ?begin ; faldo:reference ?sequence ] ;\n",
      "    faldo:end [ faldo:position ?end ; faldo:reference ?sequence ] .\n",
      "  ?sequence rdf:value ?aaSequence .\n",
      "  FILTER (SUBSTR(?aaSequence, ?begin -2, 4) = \"VSTQ\")     \n",
      "  FILTER (CONTAINS(?comment, \"Phosphothreonine\"))\n",
      "}\n",
      "✅ Find the human protein which contains an Epitope VSTQ, where T is a phosporylated threonine 2 == 2\n",
      "⚖️ Success: 15, Mismatch: 4, Fail: 5\n",
      "PREFIX wdt: <http://www.wikidata.org/prop/direct/>\n",
      "PREFIX uniprotkb: <http://purl.uniprot.org/uniprot/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX ps: <http://www.wikidata.org/prop/statement/>\n",
      "PREFIX pq: <http://www.wikidata.org/prop/qualifier/>\n",
      "PREFIX p: <http://www.wikidata.org/prop/>\n",
      "\n",
      "SELECT \n",
      "  ?protein \n",
      "  ?begin\n",
      "  ?end\n",
      "  ?chromosome\n",
      "  ?assembly\n",
      "WHERE {\n",
      "    {\n",
      "        BIND(uniprotkb:P05067 AS ?proteinIRI)\n",
      "        BIND (SUBSTR(STR(?proteinIRI), STRLEN(STR(uniprotkb:))+1) AS ?protein)\n",
      "    }\n",
      "    SERVICE <https://query.wikidata.org/sparql> {\n",
      "        ?wp wdt:P352 ?protein ;\n",
      "            wdt:P702 ?wg . \n",
      "        ?wg p:P644   ?wgss .\n",
      "        ?wgss ps:P644        ?begin ;\n",
      "              pq:P1057/wdt:P1813 ?chromosome ;\n",
      "              pq:P659/rdfs:label ?assembly .\n",
      "        ?wg p:P645 ?wgse .\n",
      "        ?wgse ps:P645         ?end ;\n",
      "              pq:P1057/wdt:P1813 ?chromosome ;\n",
      "              pq:P659/rdfs:label ?assembly .\n",
      "        FILTER(lang(?assembly) = \"en\")\n",
      "    } \n",
      "}\n",
      "✅ For the human entry P05067 (Amyloid-beta precursor protein) find the gene start ends in WikiData 2 == 2\n",
      "⚖️ Success: 16, Mismatch: 4, Fail: 5\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\n",
      "SELECT  \n",
      "  ?protein\n",
      "  ?rhea \n",
      "WHERE {\n",
      "  # ECO 269 is experimental evidence\n",
      "  BIND (<http://purl.obolibrary.org/obo/ECO_0000269> as ?evidence)\n",
      "  GRAPH <http://sparql.uniprot.org/uniprot> {\n",
      "    ?protein up:reviewed true ;\n",
      "      up:annotation ?a ;\n",
      "      up:attribution ?attribution  .\n",
      "\n",
      "    ?a a up:Catalytic_Activity_Annotation ;\n",
      "      up:catalyticActivity ?ca .\n",
      "    ?ca up:catalyzedReaction ?rhea .\n",
      "  \n",
      "    [] rdf:subject ?a ;\n",
      "      rdf:predicate up:catalyticActivity ;\n",
      "      rdf:object ?ca ;\n",
      "      up:attribution ?attribution .\n",
      "\n",
      "    ?attribution up:evidence ?evidence .\n",
      "  }\n",
      "}\n",
      "✅ Retrieve entries and Catalytic activities in the reviewed (Swiss-Prot) section that have experimental evidences,. EXACT MATCH\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "SELECT ?protein\n",
      "WHERE\n",
      "{\n",
      "    ?protein a up:Protein .\n",
      "    ?protein up:mnemonic 'A4_HUMAN'\n",
      "}\n",
      "✅ Select the UniProt entry with the mnemonic 'A4_HUMAN'. EXACT MATCH\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
      "PREFIX patent: <http://data.epo.org/linked-data/def/patent/>\n",
      "\n",
      "SELECT ?citation ?patent ?application ?applicationNo\n",
      "WHERE\n",
      "{\n",
      "  ?citation a up:Patent_Citation ;\n",
      "    skos:exactMatch ?patent .\n",
      "  FILTER(CONTAINS(STR(?patent), 'EP'))\n",
      "  BIND(SUBSTR(STR(?patent), 35) AS ?applicationNo)\n",
      "  \n",
      "  SERVICE <https://data.epo.org/linked-data/query> {\n",
      "    ?application patent:publicationNumber ?applicationNo\n",
      "  }\n",
      "}\n",
      "✅ Connect patents cited in UniProtKB with those in the patent database at EPO via publication number. 66 == 66\n",
      "⚖️ Success: 19, Mismatch: 4, Fail: 5\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
      "PREFIX patent: <http://data.epo.org/linked-data/def/patent/>\n",
      "\n",
      "SELECT ?patent ?applicationNo ?grantDate\n",
      "WHERE\n",
      "{\n",
      "    # Retrieve Patents from UniProtKB\n",
      "    ?citation a up:Patent_Citation ;\n",
      "              skos:exactMatch ?patent .\n",
      "    \n",
      "    # Extract the application number\n",
      "    BIND(SUBSTR(STR(?patent), 35) AS ?applicationNo)\n",
      "\n",
      "    # Connect to the EPO Linked Data for details on patents\n",
      "    SERVICE <https://data.epo.org/linked-data/query>\n",
      "    {\n",
      "        ?publication patent:publicationNumber ?applicationNo ;\n",
      "                     patent:application ?application .\n",
      "        ?application patent:grantDate ?grantDate.\n",
      "    }\n",
      "\n",
      "    # Identify patents with grant dates > 20 years ago\n",
      "    BIND((year(now()) - 20) AS ?thisYearMinusTwenty)\n",
      "    BIND(year(?grantDate) AS ?grantYear)\n",
      "    FILTER(?grantYear < ?thisYearMinusTwenty)\n",
      "}\n",
      "ORDER BY ?grantYear\n",
      "✅ Connect patents cited in UniProtKB with those in the patent database at EPO via publication number, whose grant date is more than twenty years in the past. 71 == 71\n",
      "⚖️ Success: 20, Mismatch: 4, Fail: 5\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "SELECT ?protein ?pdb\n",
      "WHERE\n",
      "{\n",
      "    ?protein a up:Protein .\n",
      "    ?protein rdfs:seeAlso ?pdb .\n",
      "    ?pdb up:database <http://purl.uniprot.org/database/PDB>\n",
      "}\n",
      "✅ Select a mapping of UniProt to PDB entries using the UniProt cross-references to the PDB database 377848 == 377848\n",
      "⚖️ Success: 21, Mismatch: 4, Fail: 5\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX rh: <http://rdf.rhea-db.org/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "PREFIX CHEBI: <http://purl.obolibrary.org/obo/CHEBI_>\n",
      "\n",
      "SELECT DISTINCT ?chebi ?reaction ?humanProtein ?humanProteinLabel ?mouseProtein ?mouseProteinLabel ?cluster \n",
      "WHERE {\n",
      "  # Rhea: Retrieve reactions involving sterols (CHEBI:15889)\n",
      "  SERVICE <https://sparql.rhea-db.org/sparql> {\n",
      "    ?reaction rdfs:subClassOf rh:Reaction .\n",
      "    ?reaction rh:side/rh:contains/rh:compound ?compound .\n",
      "    ?compound rh:chebi ?chebi .\n",
      "    ?chebi rdfs:subClassOf* CHEBI:15889\n",
      "  }\n",
      "\n",
      "  # UniProt: Retrieve human enzymes catalyzing these reactions\n",
      "  ?humanProtein up:organism taxon:9606 ;\n",
      "                up:annotation ?a ;\n",
      "                rdfs:label ?humanProteinLabel .\n",
      "  ?a a up:Catalytic_Activity_Annotation ;\n",
      "     up:catalyticActivity ?ca .\n",
      "  ?ca up:catalyzedReaction ?reaction .\n",
      "\n",
      "  # OMABrowser: Retrieve mouse homologs of these human enzymes\n",
      "  SERVICE <https://sparql.omabrowser.org/sparql/> {\n",
      "    ?cluster a orth:OrthologousGroup ;\n",
      "             orth:hasHomologousMember ?node1, ?node2 .\n",
      "    ?node1 orth:hasOrthologousMember ?orthoProtein1 .\n",
      "    ?node2 orth:hasOrthologousMember ?orthoProtein2 .\n",
      "    ?orthoProtein1 lscr:xrefUniprot ?humanProtein .\n",
      "    ?orthoProtein2 lscr:xrefUniprot ?mouseProtein ;\n",
      "                  rdfs:label ?mouseProteinLabel .\n",
      "    # Ensure we have mouse proteins\n",
      "    ?orthoProtein2 orth:organism/<http://purl.obolibrary.org/obo/RO_0002162> taxon:10090 .\n",
      "  }\n",
      "}\n",
      "❌ Find mouse homologs in OMABrowser of human enzymes that catalyze reactions involving Sterols (CHEBI:15889). Federating with Rhea-DB and OMABrowser. EndPointInternalError: The endpoint returned the HTTP status code 500. \n",
      "\n",
      "Response:\n",
      "b'<!DOCTYPE html SYSTEM \"about:legacy-compat\">\\n<html xmlns=\"http://www.w3.org/1999/xhtml\" lang=\"en\" xml:lang=\"en\"><head><title>Error</title><meta content=\"text/html; charset=UTF-8\" http-equiv=\"Content-Type\"/><link href=\"/\" rel=\"home\"/><link href=\"/base.css\" type=\"text/css\" rel=\"stylesheet\"/><link type=\"image/vnd.microsoft.icon\" href=\"/uniprot-favicon.ico\" rel=\"shortcut icon\"/><link href=\"/uniprot-sparql.css\" type=\"text/css\" rel=\"stylesheet\"/><link href=\"http://creativecommons.org/licenses/by/4.0/\" rel=\"license\"/><script src=\"https://cdn.jsdelivr.net/npm/yasgui-yasqe@2.11.22/dist/yasqe.bundled.min.js\"></script><link type=\"text/css\" rel=\"stylesheet\" href=\"https://cdn.jsdelivr.net/yasqe/2.2/yasqe.min.css\"/><meta content=\"width=device-width, initial-scale=1\" name=\"viewport\"/><meta content=\"no-referrer\" name=\"referrer\"/></head><body typeof=\"schema:WebPage\"><nav id=\"nav-header\"><a accesskey=\"1\" href=\"https://www.uniprot.org/\" id=\"www\"><img alt=\"\" src=\"data:image/svg+xml;base64,PHN2ZyB2aWV3Qm94PSIwIDAgNjQwIDI5MyIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj48ZyBmaWxsPSIjZmZmIiBmaWxsLXJ1bGU9Im5vbnplcm8iPjxwYXRoIGQ9Ik0xOTkuNjU5LDEwNi4xMDMgQzE5OS42NTksMTEyLjM5OCAxOTQuNTU2LDExNy41MDEgMTg4LjI2LDExNy41MDEgQzE4MS45NjQsMTE3LjUwMSAxNzYuODYxLDExMi4zOTcgMTc2Ljg2MSwxMDYuMTAzIEMxNzYuODYxLDk5LjgwNyAxODEuOTY1LDk0LjcwNCAxODguMjYsOTQuNzA0IEMxOTQuNTU1LDk0LjcwNCAxOTkuNjU5LDk5LjgwNyAxOTkuNjU5LDEwNi4xMDMgWiI+PC9wYXRoPjxwYXRoIGQ9Ik0zMzQuMzgzLDE0NC40NTkgQzMyNS4yNDIsMTQ0LjQ1OSAzMTcuODI5LDE1MS44NzIgMzE3LjgyOSwxNjEuMDE1IEwzMTcuNzk3LDE2NS4xODUgTDMxNy43OTcsMTk0LjIxIEMzMTcuNzk3LDIwMC4zMTYgMzEyLjg0OCwyMDUuMjY1IDMwNi43NDIsMjA1LjI2NSBDMzAwLjYzOCwyMDUuMjY1IDI5NS42ODksMjAwLjMxNyAyOTUuNjg5LDE5NC4yMSBMMjk1LjY4OSwxNjUuMTg1IEwyOTUuNzIsMTYxLjAxNSBDMjk1LjcyLDEzOS42NjEgMzEzLjAzLDEyMi4zNTEgMzM0LjM4MywxMjIuMzUxIEwzNDEuMjg1LDEyMi4zNTEgQzM0Ny4zOSwxMjIuMzUxIDM1Mi4zNCwxMjcuMyAzNTIuMzQsMTMzLjQwNSBDMzUyLjM0LDEzOS41MDkgMzQ3LjM5MSwxNDQuNDU5IDM0MS4yODUsMTQ0LjQ1OSBMMzM0LjM4MywxNDQuNDU5IFoiPjwvcGF0aD48cGF0aCBkPSJNNDMzLjg2MSwxNjMuODEzIEM0MzMuODYxLDE4Ny40MTcgNDE0LjcyNCwyMDYuNTU1IDM5MS4xMTksMjA2LjU1NSBDMzY3LjUxNCwyMDYuNTU1IDM0OC4zNzUsMTg3LjQxNyAzNDguMzc1LDE2My44MTMgQzM0OC4zNzUsMTQwLjIwNiAzNjcuNTE0LDEyMS4wNyAzOTEuMTE5LDEyMS4wNyBDNDE0LjcyNCwxMjEuMDcgNDMzLjg2MSwxNDAuMjA2IDQzMy44NjEsMTYzLjgxMyBaIE0zOTEuMDI5LDE0MS42OTQgQzM3OC44MTgsMTQxLjY5NCAzNjguOTIsMTUxLjU5MiAzNjguOTIsMTYzLjgwMSBDMzY4LjkyLDE3Ni4wMTEgMzc4LjgxOCwxODUuOTA5IDM5MS4wMjksMTg1LjkwOSBDNDAzLjIzOCwxODUuOTA5IDQxMy4xMzYsMTc2LjAxMSA0MTMuMTM2LDE2My44MDEgQzQxMy4xMzcsMTUxLjU5MyA0MDMuMjM4LDE0MS42OTQgMzkxLjAyOSwxNDEuNjk0IFoiPjwvcGF0aD48cGF0aCBkPSJNNDgzLjcwMywxODMuMTU2IEw0NzkuNDY3LDE4My4xNTUgQzQ3MC4zMjIsMTgzLjE1NSA0NjIuOTEsMTc1Ljc0MyA0NjIuOTEsMTY2LjU5OCBMNDYyLjg3NywxNjIuNDMgTDQ2Mi44NzcsMTQ0LjQ1NyBMNDgyLjIyMSwxNDQuNDU4IEM0ODguMzI2LDE0NC40NTggNDkzLjI3NiwxMzkuNTEgNDkzLjI3NiwxMzMuNDA0IEM0OTMuMjc2LDEyNy4yOTggNDg4LjMyNywxMjIuMzQ5IDQ4Mi4yMjEsMTIyLjM0OSBMNDYyLjg3NywxMjIuMzQ5IEw0NjIuODc3LDExNi44MjQgQzQ2Mi44NzcsMTEwLjcxOSA0NTcuOTMsMTA1Ljc2OSA0NTEuODI0LDEwNS43NjkgQzQ0NS43MTgsMTA1Ljc2OSA0NDAuNzY5LDExMC43MTggNDQwLjc2OSwxMTYuODI0IEw0NDAuNzY5LDE2Mi40MjkgTDQ0MC44LDE2Ni41OTkgQzQ0MC44LDE4Ny45NTIgNDU4LjExMywyMDUuMjYyIDQ3OS40NjYsMjA1LjI2MiBMNDgzLjcwMiwyMDUuMjYzIEM0ODkuODA3LDIwNS4yNjMgNDk0Ljc1NywyMDAuMzE1IDQ5NC43NTcsMTk0LjIwOCBDNDk0Ljc1OCwxODguMTA1IDQ4OS44MDksMTgzLjE1NiA0ODMuNzAzLDE4My4xNTYgWiI+PC9wYXRoPjxwYXRoIGQ9Ik0yOTIuOTgxLDEzNS4wNDIgQzI5Mi45ODEsMTEzLjY4OCAyNzUuNjcsOTYuMzc4IDI1NC4zMTcsOTYuMzc4IEwyNTAuMTQ3LDk2LjM0NiBMMjIyLjQ1OCw5Ni4zNDYgQzIxNi4zNTMsOTYuMzQ2IDIxMS40MDQsMTAxLjI5NSAyMTEuNDA0LDEwNy40IEwyMTEuNDA0LDE5NC4yIEMyMTEuNDA0LDIwMC4zMDYgMjE2LjM1MywyMDUuMjU2IDIyMi40NTgsMjA1LjI1NiBDMjI4LjU2MiwyMDUuMjU2IDIzMy41MTIsMjAwLjMwNyAyMzMuNTEyLDE5NC4yIEwyMzMuNTEyLDExOC40NTUgTDI1MC4xNDgsMTE4LjQ1NSBMMjU0LjMxNywxMTguNDg3IEMyNjMuNDYxLDExOC40ODcgMjcwLjg3NCwxMjUuODk5IDI3MC44NzQsMTM1LjA0MyBDMjcwLjg3NCwxNDQuMTg3IDI2My40NjEsMTUxLjYgMjU0LjMxNywxNTEuNiBMMjU1LjY3NCwxNTEuNjE2IEwyNTEuNTIsMTUxLjYxNiBDMjQ1LjQxNiwxNTEuNjE2IDI0MC40NjYsMTU2LjU2MyAyNDAuNDY2LDE2Mi42NyBDMjQwLjQ2NiwxNjguNzc0IDI0NS40MTUsMTczLjcyNCAyNTEuNTIsMTczLjcyNCBMMjU1LjY3NCwxNzMuNzI0IEwyNTQuMzE4LDE3My43MDggQzI3NS42NzEsMTczLjcwNiAyOTIuOTgxLDE1Ni4zOTYgMjkyLjk4MSwxMzUuMDQyIFoiPjwvcGF0aD48cGF0aCBkPSJNNDYyLjYwNywyNi4zMTkgQzQ2Mi42MDcsMzEuNDk0IDQ1OC40MSwzNS42ODkgNDUzLjIzNiwzNS42ODkgQzQ0OC4wNiwzNS42ODkgNDQzLjg2NSwzMS40OTQgNDQzLjg2NSwyNi4zMTkgQzQ0My44NjUsMjEuMTQ0IDQ0OC4wNiwxNi45NDggNDUzLjIzNiwxNi45NDggQzQ1OC40MSwxNi45NDggNDYyLjYwNywyMS4xNDUgNDYyLjYwNywyNi4zMTkgWiI+PC9wYXRoPjxwYXRoIGQ9Ik00MzAuNTE0LDU0LjY4OCBDNDMwLjUxNCw1OC44MjggNDI3LjE1Nyw2Mi4xODUgNDIzLjAxNiw2Mi4xODUgQzQxOC44NzcsNjIuMTg1IDQxNS41Miw1OC44MjggNDE1LjUyLDU0LjY4OCBDNDE1LjUyLDUwLjU0OCA0MTguODc3LDQ3LjE5MDk5OTggNDIzLjAxNiw0Ny4xOTA5OTk4IEM0MjcuMTU2LDQ3LjE5IDQzMC41MTQsNTAuNTQ4IDQzMC41MTQsNTQuNjg4IFoiPjwvcGF0aD48cGF0aCBkPSJNNTA1LjU1NywxMS43MTMgQzUwNS41NTcsMTguMTgzIDUwMC4zMTEsMjMuNDI2IDQ5My44NDIsMjMuNDI2IEM0ODcuMzczLDIzLjQyNiA0ODIuMTI5LDE4LjE4MyA0ODIuMTI5LDExLjcxMyBDNDgyLjEyOSw1LjI0MyA0ODcuMzczLDAgNDkzLjg0MiwwIEM1MDAuMzExLDAgNTA1LjU1Nyw1LjI0MyA1MDUuNTU3LDExLjcxMyBaIj48L3BhdGg+PHBhdGggZD0iTTU1NS4xMTksMTYuMzE4IEM1NTUuMTE5LDI0LjQwNCA1NDguNTY0LDMwLjk2IDU0MC40NzYsMzAuOTYgQzUzMi4zOTIsMzAuOTYgNTI1LjgzNSwyNC40MDQgNTI1LjgzNSwxNi4zMTggQzUyNS44MzUsOC4yMzEgNTMyLjM5MiwxLjY3NiA1NDAuNDc2LDEuNjc2IEM1NDguNTY0LDEuNjc3IDU1NS4xMTksOC4yMzEgNTU1LjExOSwxNi4zMTggWiI+PC9wYXRoPjxwYXRoIGQ9Ik02MDIuMTAyLDQyLjA0NiBDNjAyLjEwMiw1Mi4xNTMgNTkzLjkwOSw2MC4zNDcgNTgzLjgwMSw2MC4zNDcgQzU3My42OTIsNjAuMzQ3IDU2NS40OTgsNTIuMTU0IDU2NS40OTgsNDIuMDQ2IEM1NjUuNDk4LDMxLjkzNyA1NzMuNjkxLDIzLjc0NCA1ODMuODAxLDIzLjc0NCBDNTkzLjkwOCwyMy43NDQgNjAyLjEwMiwzMS45MzcgNjAyLjEwMiw0Mi4wNDYgWiI+PC9wYXRoPjxwYXRoIGQ9Ik02MzUuNDM2LDkwLjI4MiBDNjM1LjQzNiwxMDEuOTA3IDYyNi4wMTIsMTExLjMyOSA2MTQuMzg5LDExMS4zMjkgQzYwMi43NjYsMTExLjMyOSA1OTMuMzQyLDEwMS45MDcgNTkzLjM0Miw5MC4yODIgQzU5My4zNDIsNzguNjU3IDYwMi43NjYsNjkuMjM1IDYxNC4zODksNjkuMjM1IEM2MjYuMDEyLDY5LjIzNSA2MzUuNDM2LDc4LjY1NyA2MzUuNDM2LDkwLjI4MiBaIj48L3BhdGg+PHBhdGggZD0iTTYzOS4zODcsMTUzLjU5NSBDNjM5LjM4NywxNjYuOTYgNjI4LjU1MSwxNzcuNzk2IDYxNS4xODQsMTc3Ljc5NiBDNjAxLjgxNSwxNzcuNzk2IDU5MC45NzksMTY2Ljk2IDU5MC45NzksMTUzLjU5NSBDNTkwLjk3OSwxNDAuMjMgNjAxLjgxNSwxMjkuMzk1IDYxNS4xODQsMTI5LjM5NSBDNjI4LjU1MSwxMjkuMzk1IDYzOS4zODcsMTQwLjIyOSA2MzkuMzg3LDE1My41OTUgWiI+PC9wYXRoPjxwYXRoIGQ9Ik02MTIuNzY2LDIxNS44ODYgQzYxMi43NjYsMjMxLjMyNiA2MDAuMjUsMjQzLjg0MSA1ODQuODExLDI0My44NDEgQzU2OS4zNzIsMjQzLjg0MSA1NTYuODU2LDIzMS4zMjYgNTU2Ljg1NiwyMTUuODg2IEM1NTYuODU2LDIwMC40NDUgNTY5LjM3MiwxODcuOTMxIDU4NC44MTEsMTg3LjkzMSBDNjAwLjI1LDE4Ny45MzEgNjEyLjc2NiwyMDAuNDQ0IDYxMi43NjYsMjE1Ljg4NiBaIj48L3BhdGg+PHBhdGggZD0iTTU1MS43NzMsMjU5LjE5NCBDNTUxLjc3MywyNzYuNTAyIDUzNy43NDQsMjkwLjUzMSA1MjAuNDM3LDI5MC41MzEgQzUwMy4xMywyOTAuNTMxIDQ4OS4xMDEsMjc2LjUwMiA0ODkuMTAxLDI1OS4xOTQgQzQ4OS4xMDEsMjQxLjg4NiA1MDMuMTMsMjI3Ljg1OCA1MjAuNDM3LDIyNy44NTggQzUzNy43NDQsMjI3Ljg1OCA1NTEuNzczLDI0MS44ODcgNTUxLjc3MywyNTkuMTk0IFoiPjwvcGF0aD48cGF0aCBkPSJNNDcyLjEwNywyNTcuMzggQzQ3Mi4xMDcsMjc2Ljc3NCA0NTYuMzg0LDI5Mi40OTUgNDM2Ljk5MiwyOTIuNDk1IEM0MTcuNiwyOTIuNDk1IDQwMS44NzcsMjc2Ljc3MyA0MDEuODc3LDI1Ny4zOCBDNDAxLjg3NywyMzcuOTg0IDQxNy42LDIyMi4yNjcgNDM2Ljk5MiwyMjIuMjY3IEM0NTYuMzg0LDIyMi4yNjcgNDcyLjEwNywyMzcuOTg0IDQ3Mi4xMDcsMjU3LjM4IFoiPjwvcGF0aD48cGF0aCBkPSJNMTY1LjgwNSwxNTkuNjU4IEwxNjUuODA1LDE5NC4yMSBDMTY1LjgwNSwyMDAuMzE2IDE2MC44NTYsMjA1LjI2NSAxNTQuNzUxLDIwNS4yNjUgQzE0OC42NDcsMjA1LjI2NSAxNDMuNjk3LDIwMC4zMTcgMTQzLjY5NywxOTQuMjEgTDE0My42OTcsMTU5LjY1OCBMMTQzLjY4MSwxNjEuMDE0IEMxNDMuNjgxLDE1MS44NzEgMTM2LjI2OSwxNDQuNDU4IDEyNy4xMjUsMTQ0LjQ1OCBDMTE3Ljk4MSwxNDQuNDU4IDExMC41NjksMTUxLjg3MSAxMTAuNTY5LDE2MS4wMTQgTDExMC41MzYsMTY1LjE4NCBMMTEwLjUzNiwxOTQuMjA5IEMxMTAuNTM2LDIwMC4zMTUgMTA1LjU4NywyMDUuMjY0IDk5LjQ4MiwyMDUuMjY0IEM5My4zNzcsMjA1LjI2NCA4OC40MjgsMjAwLjMxNiA4OC40MjgsMTk0LjIwOSBMODguNDI4LDE2NS4xODQgTDg4LjQ2LDE2MS4wMTQgQzg4LjQ2LDEzOS42NiAxMDUuNzcsMTIyLjM1IDEyNy4xMjQsMTIyLjM1IEMxNDguNDc4LDEyMi4zNSAxNjUuNzg4LDEzOS42NjEgMTY1Ljc4OCwxNjEuMDE0IEwxNjUuODA1LDE1OS42NTggWiI+PC9wYXRoPjxwYXRoIGQ9Ik0wLDE2Ny45NTcgTDAsMTA3LjM5OSBDMCwxMDEuMjk1IDQuOTQ5LDk2LjM0NSAxMS4wNTQsOTYuMzQ1IEMxNy4xNTksOTYuMzQ1IDIyLjEwOCwxMDEuMjk0IDIyLjEwOCwxMDcuMzk5IEwyMi4xMDgsMTY3Ljk1NyBMMjIuMTIzLDE2Ni42MDEgQzIyLjEyMywxNzUuNzQ0IDI5LjUzNiwxODMuMTU3IDM4LjY3OSwxODMuMTU3IEM0Ny44MjIsMTgzLjE1NyA1NS4yMzUsMTc1Ljc0NCA1NS4yMzUsMTY2LjYwMSBMNTUuMjY4LDE2Mi40MzEgTDU1LjI2OCwxMDcuNCBDNTUuMjY4LDEwMS4yOTYgNjAuMjE3LDk2LjM0NSA2Ni4zMjIsOTYuMzQ1IEM3Mi40MjcsOTYuMzQ1IDc3LjM3NiwxMDEuMjk1IDc3LjM3NiwxMDcuNCBMNzcuMzc2LDE2Mi40MzEgTDc3LjM0NCwxNjYuNjAyIEM3Ny4zNDQsMTg3Ljk1NiA2MC4wMzQsMjA1LjI2NSAzOC42OCwyMDUuMjY1IEMxNy4zMjYsMjA1LjI2NSAwLjAxNiwxODcuOTU1IDAuMDE2LDE2Ni42MDIgTDAsMTY3Ljk1NyBaIj48L3BhdGg+PHBhdGggZD0iTTE5OS4zMTQsMTk0LjIxIEMxOTkuMzE0LDIwMC4zMTYgMTk0LjM2NSwyMDUuMjY1IDE4OC4yNiwyMDUuMjY1IEMxODIuMTU2LDIwNS4yNjUgMTc3LjIwNiwyMDAuMzE3IDE3Ny4yMDYsMTk0LjIxIEwxNzcuMjA2LDEzMy40NzggQzE3Ny4yMDYsMTI3LjM3MyAxODIuMTU1LDEyMi40MjMgMTg4LjI2LDEyMi40MjMgQzE5NC4zNjUsMTIyLjQyMyAxOTkuMzE0LDEyNy4zNzIgMTk5LjMxNCwxMzMuNDc4IEwxOTkuMzE0LDE5NC4yMSBaIj48L3BhdGg+PC9nPjwvc3ZnPgo=\" title=\"UniProt home\"/></a><a href=\"/sparql\">SPARQL</a><a href=\"https://www.uniprot.org/downloads\" id=\"download\">Downloads</a><a accesskey=\"9\" href=\"https://www.uniprot.org/contact\" id=\"contact\"><img alt=\"Contact\" src=\"data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZpZXdCb3g9IjAgMCAxMyAxMCIgZmlsbD0iI2ZmZiIgd2lkdGg9IjEuNGVtIj48cGF0aCBmaWxsLXJ1bGU9ImV2ZW5vZGQiIGQ9Ik0xMiAxLjI1QzEyIC41NiAxMS40NiAwIDEwLjggMEgxLjJDLjU0IDAgMCAuNTYgMCAxLjI1djcuNUMwIDkuNDQuNTQgMTAgMS4yIDEwaDkuNmMuNjYgMCAxLjItLjU2IDEuMi0xLjI1di03LjV6bS0xLjIgMEw2IDQuMzggMS4yIDEuMjVoOS42ek02IDUuNjNMMS4yIDIuNXY2LjI1aDkuNlYyLjVMNiA1LjYzeiI+PC9wYXRoPjwvc3ZnPgo=\"/></a><a rel=\"help\" href=\"https://www.uniprot.org/help/\" id=\"help\">Documentation/Help</a></nav><a accesskey=\"2\" href=\"#content\" hidden=\"hidden\">Skip Header</a><main class=\"sparql error\" id=\"content\"><h1>500 Internal Server Error</h1><p>The server was not able to handle your request.</p><p> If this error occurs repeatedly, \\n\\t\\tplease <a id=\"contactLink\">contact us</a>\\n\\t\\tand indicate the page you were trying to access.</p><script language=\"javascript\">\\n\\t\\t\\tvar part2 = document.domain.replace(/sparql\\\\./,\\'\\');\\n\\t\\t\\tdocument.getElementById(\\'contactLink\\').href=part2 + \\'/contact\\';\\n\\t\\t</script></main><footer id=\"page-footer\"><section class=\"small\" id=\"copyright\">\\n\\t\\t\\t&copy; 2002&ndash; <span property=\"schema:copyrightYear\">2024</span>\\n\\t\\t\\t<a accesskey=\"9\" href=\"https://www.uniprot.org/help/about\">UniProt Consortium</a> |\\n\\t\\t\\t<a href=\"https://www.uniprot.org/help/license\" property=\"schema:license\">License\\n\\t\\t\\t\\t&amp; Disclaimer</a> | <a href=\"https://www.uniprot.org/help/privacy\">Privacy\\n\\t\\t\\t\\tNotice</a></section><section><div resource=\"/help/about#UniProt\" id=\"logos\" typeof=\"NGO\" property=\"creator\"><span resource=\"/help/about#EBI\" typeof=\"Organization\" class=\"ebi\" property=\"member\"><span hidden=\"hidden\" property=\"name\">European Bioinformatics Institute</span><a href=\"https://www.ebi.ac.uk/\" property=\"url\"></a></span><span resource=\"/help/about#PIR\" typeof=\"Organization\" class=\"pir\" property=\"member\"><span hidden=\"hidden\" property=\"name\">Protein Information Resource</span><a href=\"http://pir.georgetown.edu/\" property=\"url\"></a></span><span resource=\"/help/about#SIB\" typeof=\"Organization\" class=\"sib\" property=\"member\"><span hidden=\"hidden\" property=\"name\">SIB Swiss Institute of Bioinformatics</span><a href=\"https://www.sib.swiss/\" property=\"url\"></a></span></div></section><section id=\"poweredby\"><h3>Powered by</h3><span class=\"vitalit\" resource=\"https://www.sib.swiss/alan-bridge-group/\" typeof=\"schema:NGO\" property=\"schema:sponsor\"><h4>Hardware</h4><a href=\"https://www.sib.swiss/alan-bridge-group\" property=\"schema:url\"><img alt=\"Swiss-Prot\" src=\"https://www.sib.swiss//templates/sib/images/SIB_LogoQ_GBv.svg\"/><h5 property=\"schema:name\">Swiss-Prot</h5></a></span><span class=\"openlinksw\" resource=\"http://www.openlinksw.com/#this\" typeof=\"schema:Corporation\" property=\"schema:sponsor\"><h4>Software</h4><a href=\"http://virtuoso.openlinksw.com/\" property=\"schema:url\"><img alt=\"Virtuoso\" src=\"/images/logos/virtuoso.png\"/><h5 property=\"schema:name\">Virtuoso</h5></a></span></section></footer><script src=\"/sparql.js\" type=\"text/javascript\"> </script></body></html>'. Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX rh: <http://rdf.rhea-db.org/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "PREFIX CHEBI: <http://purl.obolibrary.org/obo/CHEBI_>\n",
      "SELECT\n",
      "  DISTINCT\n",
      "    ?chebi\n",
      "    ?reaction\n",
      "    ?humanProtein\n",
      "    ?mouseProtein\n",
      "    ?cluster \n",
      "WHERE {\n",
      "  SERVICE <https://sparql.rhea-db.org/sparql> {\n",
      "    ?reaction rdfs:subClassOf rh:Reaction .\n",
      "    ?reaction rh:side/rh:contains/rh:compound ?compound .\n",
      "    ?compound rh:chebi ?chebi .\n",
      "    ?chebi rdfs:subClassOf* CHEBI:15889\n",
      "  }\n",
      "\n",
      "  ?humanProtein up:organism taxon:9606 .\n",
      "  ?humanProtein up:annotation ?a .\n",
      "  ?a a up:Catalytic_Activity_Annotation .\n",
      "  ?a up:catalyticActivity ?ca .\n",
      "  ?ca up:catalyzedReaction ?reaction .\n",
      "\n",
      "  SERVICE  <https://sparql.omabrowser.org/sparql> {\n",
      "    ?cluster a orth:ParalogsCluster .\n",
      "    ?cluster orth:hasHomologousMember ?node1 , ?node2 .\n",
      "    ?node1 orth:hasHomologousMember* ?orthoProtein1 .\n",
      "    ?node2 orth:hasHomologousMember* ?orthoProtein2 .\n",
      "    ?orthoProtein1 lscr:xrefUniprot ?mouseProtein .\n",
      "    ?orthoProtein2 lscr:xrefUniprot ?humanProtein .\n",
      "    # inTaxon mouse\n",
      "    ?orthoProtein1 orth:organism/<http://purl.obolibrary.org/obo/RO_0002162> taxon:10090 . \n",
      "  }\n",
      "}\n",
      "\n",
      "⚖️ Success: 21, Mismatch: 4, Fail: 6\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX sachem: <http://bioinfo.uochb.cas.cz/rdf/v1.0/sachem#>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "SELECT \n",
      "  ?protein\n",
      "  ?mnemonic\n",
      "  ?proteinName\n",
      "  ?ligandSimilarityScore\n",
      "  ?ligand\n",
      "WHERE {\n",
      "  SERVICE <https://idsm.elixir-czech.cz/sparql/endpoint/chebi> {\n",
      "    ?ssc sachem:compound ?ligand; \n",
      "      sachem:score ?ligandSimilarityScore ;\n",
      "      sachem:similaritySearch ?sss .\n",
      "        # Smiles of Heme: C34H32FeN4O4\n",
      "    ?sss    sachem:query \"CC1=C(CCC([O-])=O)C2=[N+]3C1=Cc1c(C)c(C=C)c4C=C5C(C)=C(C=C)C6=[N+]5[Fe-]3(n14)n1c(=C6)c(C)c(CCC([O-])=O)c1=C2\";\n",
      "      sachem:cutoff \"8e-1\"^^xsd:double ;\n",
      "      sachem:aromaticityMode sachem:aromaticityDetect ;\n",
      "      sachem:similarityRadius 1 ;\n",
      "      sachem:tautomerMode sachem:ignoreTautomers .\n",
      "  }\n",
      "  ?protein up:mnemonic ?mnemonic ;\n",
      "    up:recommendedName/up:fullName ?proteinName ;\n",
      "    up:annotation ?annotation .\n",
      "  ?annotation a up:Binding_Site_Annotation ;\n",
      "      up:ligand/rdfs:subClassOf ?ligand .\n",
      "}\n",
      "ORDER BY DESC(?ligandSimilarityScore)\n",
      "✅ Proteins with annotated binding sites for ligands similar to heme 967082 == 967082\n",
      "⚖️ Success: 22, Mismatch: 4, Fail: 6\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\n",
      "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
      "PREFIX CHEBI: <http://purl.obolibrary.org/obo/CHEBI_>\n",
      "SELECT\n",
      "  ?ligand\n",
      "  ?ligandName \n",
      "  (COUNT(DISTINCT ?protein) as ?entries)\n",
      "WHERE {\n",
      "   ?protein up:annotation ?annotation .\n",
      "   \n",
      "   VALUES ?evs { obo:ECO_0000269 obo:ECO_0007744 } .\n",
      "   VALUES ?chebids { CHEBI:25213 CHEBI:25214 } .\n",
      "   ?st rdf:subject ?protein ; \n",
      "       rdf:predicate up:annotation ; \n",
      "       rdf:object ?annotation ;\n",
      "       up:attribution/up:evidence ?evs .\n",
      "\n",
      "   ?annotation up:ligand/rdfs:subClassOf ?ligand .\n",
      "   ?ligand rdfs:subClassOf+ ?chebids ;\n",
      "     rdfs:label ?ligandName .\n",
      "}\n",
      "GROUP BY ?ligand ?ligandName\n",
      "ORDER BY DESC(?entries)\n",
      "✅ Number of proteins with annotated binding sites for metals or metal sulfur clusters (and experimental evidence for the binding). EXACT MATCH\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "SELECT\n",
      "  ?protein\n",
      "  ?ligandName \n",
      "  ?ligandNote\n",
      "  ?chebi\n",
      "WHERE {\n",
      "   ?protein a up:Protein ;\n",
      "            up:annotation ?annotation .\n",
      "   ?annotation a up:Binding_Site_Annotation ;\n",
      "               up:ligand ?ligand .\n",
      "   ?ligand rdfs:comment ?ligandNote ;\n",
      "           rdfs:subClassOf/up:identifier ?chebi ;\n",
      "           rdfs:label ?ligandName .\n",
      "   FILTER(REGEX(?ligandNote, \"allosteric\", \"i\"))\n",
      "}\n",
      "❌ Select Enzymes that have Ligands that have a known allosteric effect 730 != 0. Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "SELECT\n",
      "  ?protein\n",
      "  ?ligandName \n",
      "  ?ligandNote\n",
      "  ?chebi\n",
      "WHERE {\n",
      "   ?protein up:annotation ?annotation .\n",
      "   ?annotation a up:Binding_Site_Annotation . \n",
      "   ?annotation up:ligand ?ligand .\n",
      "   ?ligand rdfs:comment ?ligandNote ;\n",
      "     rdfs:subClassOf ?chebi ;\n",
      "     rdfs:label ?ligandName .\n",
      "   FILTER(REGEX(?ligandNote, \"allosteric\", \"i\"))\n",
      "}\n",
      "\n",
      "\n",
      "⚖️ Success: 23, Mismatch: 4, Fail: 7\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "SELECT\n",
      "  ?pdbId ?chain ?pdbChain ?uniprot\n",
      "WHERE\n",
      "{\n",
      "  # Replace the values with your list of PDB identifiers and chains.\n",
      "  VALUES (?pdbId ?pdbChain) { ('6VXC' 'A') ('1BG3' 'B') /* Add more pairs as needed */ }\n",
      "\n",
      "  # Make an IRI out of the pdbId\n",
      "  BIND(iri(concat('http://rdf.wwpdb.org/pdb/', ?pdbId)) AS ?pdb)\n",
      "\n",
      "  # Map to UniProt entries\n",
      "  ?uniprot rdfs:seeAlso ?pdb .\n",
      "  ?pdb up:database <http://purl.uniprot.org/database/PDB> ;\n",
      "       up:chainSequenceMapping ?chainSm .\n",
      "  ?chainSm up:chain ?chainsPlusRange .\n",
      "\n",
      "  # Extract the list of chains from the text representation.\n",
      "  BIND(STRBEFORE(?chainsPlusRange, '=') AS ?chain)\n",
      "\n",
      "  # Filter those that match.\n",
      "  FILTER(CONTAINS(?chain, ?pdbChain))\n",
      "}\n",
      "❌ Map PDB identifiers plus chains to UniProt QueryBadFormed: A bad request has been sent to the endpoint: probably the SPARQL query is badly formed. \n",
      "\n",
      "Response:\n",
      "b'<!DOCTYPE html SYSTEM \"about:legacy-compat\">\\n<html xmlns=\"http://www.w3.org/1999/xhtml\" lang=\"en\" xml:lang=\"en\"><head><title>UniProt</title><meta content=\"text/html; charset=UTF-8\" http-equiv=\"Content-Type\"/><link href=\"/\" rel=\"home\"/><link href=\"/base.css\" type=\"text/css\" rel=\"stylesheet\"/><link type=\"image/vnd.microsoft.icon\" href=\"/uniprot-favicon.ico\" rel=\"shortcut icon\"/><link href=\"/uniprot-sparql.css\" type=\"text/css\" rel=\"stylesheet\"/><link href=\"http://creativecommons.org/licenses/by/4.0/\" rel=\"license\"/><script src=\"https://cdn.jsdelivr.net/npm/yasgui-yasqe@2.11.22/dist/yasqe.bundled.min.js\"></script><link type=\"text/css\" rel=\"stylesheet\" href=\"https://cdn.jsdelivr.net/yasqe/2.2/yasqe.min.css\"/><meta content=\"width=device-width, initial-scale=1\" name=\"viewport\"/><meta content=\"no-referrer\" name=\"referrer\"/></head><body typeof=\"schema:WebPage\"><nav id=\"nav-header\"><a accesskey=\"1\" href=\"https://www.uniprot.org/\" id=\"www\"><img alt=\"\" src=\"data:image/svg+xml;base64,PHN2ZyB2aWV3Qm94PSIwIDAgNjQwIDI5MyIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj48ZyBmaWxsPSIjZmZmIiBmaWxsLXJ1bGU9Im5vbnplcm8iPjxwYXRoIGQ9Ik0xOTkuNjU5LDEwNi4xMDMgQzE5OS42NTksMTEyLjM5OCAxOTQuNTU2LDExNy41MDEgMTg4LjI2LDExNy41MDEgQzE4MS45NjQsMTE3LjUwMSAxNzYuODYxLDExMi4zOTcgMTc2Ljg2MSwxMDYuMTAzIEMxNzYuODYxLDk5LjgwNyAxODEuOTY1LDk0LjcwNCAxODguMjYsOTQuNzA0IEMxOTQuNTU1LDk0LjcwNCAxOTkuNjU5LDk5LjgwNyAxOTkuNjU5LDEwNi4xMDMgWiI+PC9wYXRoPjxwYXRoIGQ9Ik0zMzQuMzgzLDE0NC40NTkgQzMyNS4yNDIsMTQ0LjQ1OSAzMTcuODI5LDE1MS44NzIgMzE3LjgyOSwxNjEuMDE1IEwzMTcuNzk3LDE2NS4xODUgTDMxNy43OTcsMTk0LjIxIEMzMTcuNzk3LDIwMC4zMTYgMzEyLjg0OCwyMDUuMjY1IDMwNi43NDIsMjA1LjI2NSBDMzAwLjYzOCwyMDUuMjY1IDI5NS42ODksMjAwLjMxNyAyOTUuNjg5LDE5NC4yMSBMMjk1LjY4OSwxNjUuMTg1IEwyOTUuNzIsMTYxLjAxNSBDMjk1LjcyLDEzOS42NjEgMzEzLjAzLDEyMi4zNTEgMzM0LjM4MywxMjIuMzUxIEwzNDEuMjg1LDEyMi4zNTEgQzM0Ny4zOSwxMjIuMzUxIDM1Mi4zNCwxMjcuMyAzNTIuMzQsMTMzLjQwNSBDMzUyLjM0LDEzOS41MDkgMzQ3LjM5MSwxNDQuNDU5IDM0MS4yODUsMTQ0LjQ1OSBMMzM0LjM4MywxNDQuNDU5IFoiPjwvcGF0aD48cGF0aCBkPSJNNDMzLjg2MSwxNjMuODEzIEM0MzMuODYxLDE4Ny40MTcgNDE0LjcyNCwyMDYuNTU1IDM5MS4xMTksMjA2LjU1NSBDMzY3LjUxNCwyMDYuNTU1IDM0OC4zNzUsMTg3LjQxNyAzNDguMzc1LDE2My44MTMgQzM0OC4zNzUsMTQwLjIwNiAzNjcuNTE0LDEyMS4wNyAzOTEuMTE5LDEyMS4wNyBDNDE0LjcyNCwxMjEuMDcgNDMzLjg2MSwxNDAuMjA2IDQzMy44NjEsMTYzLjgxMyBaIE0zOTEuMDI5LDE0MS42OTQgQzM3OC44MTgsMTQxLjY5NCAzNjguOTIsMTUxLjU5MiAzNjguOTIsMTYzLjgwMSBDMzY4LjkyLDE3Ni4wMTEgMzc4LjgxOCwxODUuOTA5IDM5MS4wMjksMTg1LjkwOSBDNDAzLjIzOCwxODUuOTA5IDQxMy4xMzYsMTc2LjAxMSA0MTMuMTM2LDE2My44MDEgQzQxMy4xMzcsMTUxLjU5MyA0MDMuMjM4LDE0MS42OTQgMzkxLjAyOSwxNDEuNjk0IFoiPjwvcGF0aD48cGF0aCBkPSJNNDgzLjcwMywxODMuMTU2IEw0NzkuNDY3LDE4My4xNTUgQzQ3MC4zMjIsMTgzLjE1NSA0NjIuOTEsMTc1Ljc0MyA0NjIuOTEsMTY2LjU5OCBMNDYyLjg3NywxNjIuNDMgTDQ2Mi44NzcsMTQ0LjQ1NyBMNDgyLjIyMSwxNDQuNDU4IEM0ODguMzI2LDE0NC40NTggNDkzLjI3NiwxMzkuNTEgNDkzLjI3NiwxMzMuNDA0IEM0OTMuMjc2LDEyNy4yOTggNDg4LjMyNywxMjIuMzQ5IDQ4Mi4yMjEsMTIyLjM0OSBMNDYyLjg3NywxMjIuMzQ5IEw0NjIuODc3LDExNi44MjQgQzQ2Mi44NzcsMTEwLjcxOSA0NTcuOTMsMTA1Ljc2OSA0NTEuODI0LDEwNS43NjkgQzQ0NS43MTgsMTA1Ljc2OSA0NDAuNzY5LDExMC43MTggNDQwLjc2OSwxMTYuODI0IEw0NDAuNzY5LDE2Mi40MjkgTDQ0MC44LDE2Ni41OTkgQzQ0MC44LDE4Ny45NTIgNDU4LjExMywyMDUuMjYyIDQ3OS40NjYsMjA1LjI2MiBMNDgzLjcwMiwyMDUuMjYzIEM0ODkuODA3LDIwNS4yNjMgNDk0Ljc1NywyMDAuMzE1IDQ5NC43NTcsMTk0LjIwOCBDNDk0Ljc1OCwxODguMTA1IDQ4OS44MDksMTgzLjE1NiA0ODMuNzAzLDE4My4xNTYgWiI+PC9wYXRoPjxwYXRoIGQ9Ik0yOTIuOTgxLDEzNS4wNDIgQzI5Mi45ODEsMTEzLjY4OCAyNzUuNjcsOTYuMzc4IDI1NC4zMTcsOTYuMzc4IEwyNTAuMTQ3LDk2LjM0NiBMMjIyLjQ1OCw5Ni4zNDYgQzIxNi4zNTMsOTYuMzQ2IDIxMS40MDQsMTAxLjI5NSAyMTEuNDA0LDEwNy40IEwyMTEuNDA0LDE5NC4yIEMyMTEuNDA0LDIwMC4zMDYgMjE2LjM1MywyMDUuMjU2IDIyMi40NTgsMjA1LjI1NiBDMjI4LjU2MiwyMDUuMjU2IDIzMy41MTIsMjAwLjMwNyAyMzMuNTEyLDE5NC4yIEwyMzMuNTEyLDExOC40NTUgTDI1MC4xNDgsMTE4LjQ1NSBMMjU0LjMxNywxMTguNDg3IEMyNjMuNDYxLDExOC40ODcgMjcwLjg3NCwxMjUuODk5IDI3MC44NzQsMTM1LjA0MyBDMjcwLjg3NCwxNDQuMTg3IDI2My40NjEsMTUxLjYgMjU0LjMxNywxNTEuNiBMMjU1LjY3NCwxNTEuNjE2IEwyNTEuNTIsMTUxLjYxNiBDMjQ1LjQxNiwxNTEuNjE2IDI0MC40NjYsMTU2LjU2MyAyNDAuNDY2LDE2Mi42NyBDMjQwLjQ2NiwxNjguNzc0IDI0NS40MTUsMTczLjcyNCAyNTEuNTIsMTczLjcyNCBMMjU1LjY3NCwxNzMuNzI0IEwyNTQuMzE4LDE3My43MDggQzI3NS42NzEsMTczLjcwNiAyOTIuOTgxLDE1Ni4zOTYgMjkyLjk4MSwxMzUuMDQyIFoiPjwvcGF0aD48cGF0aCBkPSJNNDYyLjYwNywyNi4zMTkgQzQ2Mi42MDcsMzEuNDk0IDQ1OC40MSwzNS42ODkgNDUzLjIzNiwzNS42ODkgQzQ0OC4wNiwzNS42ODkgNDQzLjg2NSwzMS40OTQgNDQzLjg2NSwyNi4zMTkgQzQ0My44NjUsMjEuMTQ0IDQ0OC4wNiwxNi45NDggNDUzLjIzNiwxNi45NDggQzQ1OC40MSwxNi45NDggNDYyLjYwNywyMS4xNDUgNDYyLjYwNywyNi4zMTkgWiI+PC9wYXRoPjxwYXRoIGQ9Ik00MzAuNTE0LDU0LjY4OCBDNDMwLjUxNCw1OC44MjggNDI3LjE1Nyw2Mi4xODUgNDIzLjAxNiw2Mi4xODUgQzQxOC44NzcsNjIuMTg1IDQxNS41Miw1OC44MjggNDE1LjUyLDU0LjY4OCBDNDE1LjUyLDUwLjU0OCA0MTguODc3LDQ3LjE5MDk5OTggNDIzLjAxNiw0Ny4xOTA5OTk4IEM0MjcuMTU2LDQ3LjE5IDQzMC41MTQsNTAuNTQ4IDQzMC41MTQsNTQuNjg4IFoiPjwvcGF0aD48cGF0aCBkPSJNNTA1LjU1NywxMS43MTMgQzUwNS41NTcsMTguMTgzIDUwMC4zMTEsMjMuNDI2IDQ5My44NDIsMjMuNDI2IEM0ODcuMzczLDIzLjQyNiA0ODIuMTI5LDE4LjE4MyA0ODIuMTI5LDExLjcxMyBDNDgyLjEyOSw1LjI0MyA0ODcuMzczLDAgNDkzLjg0MiwwIEM1MDAuMzExLDAgNTA1LjU1Nyw1LjI0MyA1MDUuNTU3LDExLjcxMyBaIj48L3BhdGg+PHBhdGggZD0iTTU1NS4xMTksMTYuMzE4IEM1NTUuMTE5LDI0LjQwNCA1NDguNTY0LDMwLjk2IDU0MC40NzYsMzAuOTYgQzUzMi4zOTIsMzAuOTYgNTI1LjgzNSwyNC40MDQgNTI1LjgzNSwxNi4zMTggQzUyNS44MzUsOC4yMzEgNTMyLjM5MiwxLjY3NiA1NDAuNDc2LDEuNjc2IEM1NDguNTY0LDEuNjc3IDU1NS4xMTksOC4yMzEgNTU1LjExOSwxNi4zMTggWiI+PC9wYXRoPjxwYXRoIGQ9Ik02MDIuMTAyLDQyLjA0NiBDNjAyLjEwMiw1Mi4xNTMgNTkzLjkwOSw2MC4zNDcgNTgzLjgwMSw2MC4zNDcgQzU3My42OTIsNjAuMzQ3IDU2NS40OTgsNTIuMTU0IDU2NS40OTgsNDIuMDQ2IEM1NjUuNDk4LDMxLjkzNyA1NzMuNjkxLDIzLjc0NCA1ODMuODAxLDIzLjc0NCBDNTkzLjkwOCwyMy43NDQgNjAyLjEwMiwzMS45MzcgNjAyLjEwMiw0Mi4wNDYgWiI+PC9wYXRoPjxwYXRoIGQ9Ik02MzUuNDM2LDkwLjI4MiBDNjM1LjQzNiwxMDEuOTA3IDYyNi4wMTIsMTExLjMyOSA2MTQuMzg5LDExMS4zMjkgQzYwMi43NjYsMTExLjMyOSA1OTMuMzQyLDEwMS45MDcgNTkzLjM0Miw5MC4yODIgQzU5My4zNDIsNzguNjU3IDYwMi43NjYsNjkuMjM1IDYxNC4zODksNjkuMjM1IEM2MjYuMDEyLDY5LjIzNSA2MzUuNDM2LDc4LjY1NyA2MzUuNDM2LDkwLjI4MiBaIj48L3BhdGg+PHBhdGggZD0iTTYzOS4zODcsMTUzLjU5NSBDNjM5LjM4NywxNjYuOTYgNjI4LjU1MSwxNzcuNzk2IDYxNS4xODQsMTc3Ljc5NiBDNjAxLjgxNSwxNzcuNzk2IDU5MC45NzksMTY2Ljk2IDU5MC45NzksMTUzLjU5NSBDNTkwLjk3OSwxNDAuMjMgNjAxLjgxNSwxMjkuMzk1IDYxNS4xODQsMTI5LjM5NSBDNjI4LjU1MSwxMjkuMzk1IDYzOS4zODcsMTQwLjIyOSA2MzkuMzg3LDE1My41OTUgWiI+PC9wYXRoPjxwYXRoIGQ9Ik02MTIuNzY2LDIxNS44ODYgQzYxMi43NjYsMjMxLjMyNiA2MDAuMjUsMjQzLjg0MSA1ODQuODExLDI0My44NDEgQzU2OS4zNzIsMjQzLjg0MSA1NTYuODU2LDIzMS4zMjYgNTU2Ljg1NiwyMTUuODg2IEM1NTYuODU2LDIwMC40NDUgNTY5LjM3MiwxODcuOTMxIDU4NC44MTEsMTg3LjkzMSBDNjAwLjI1LDE4Ny45MzEgNjEyLjc2NiwyMDAuNDQ0IDYxMi43NjYsMjE1Ljg4NiBaIj48L3BhdGg+PHBhdGggZD0iTTU1MS43NzMsMjU5LjE5NCBDNTUxLjc3MywyNzYuNTAyIDUzNy43NDQsMjkwLjUzMSA1MjAuNDM3LDI5MC41MzEgQzUwMy4xMywyOTAuNTMxIDQ4OS4xMDEsMjc2LjUwMiA0ODkuMTAxLDI1OS4xOTQgQzQ4OS4xMDEsMjQxLjg4NiA1MDMuMTMsMjI3Ljg1OCA1MjAuNDM3LDIyNy44NTggQzUzNy43NDQsMjI3Ljg1OCA1NTEuNzczLDI0MS44ODcgNTUxLjc3MywyNTkuMTk0IFoiPjwvcGF0aD48cGF0aCBkPSJNNDcyLjEwNywyNTcuMzggQzQ3Mi4xMDcsMjc2Ljc3NCA0NTYuMzg0LDI5Mi40OTUgNDM2Ljk5MiwyOTIuNDk1IEM0MTcuNiwyOTIuNDk1IDQwMS44NzcsMjc2Ljc3MyA0MDEuODc3LDI1Ny4zOCBDNDAxLjg3NywyMzcuOTg0IDQxNy42LDIyMi4yNjcgNDM2Ljk5MiwyMjIuMjY3IEM0NTYuMzg0LDIyMi4yNjcgNDcyLjEwNywyMzcuOTg0IDQ3Mi4xMDcsMjU3LjM4IFoiPjwvcGF0aD48cGF0aCBkPSJNMTY1LjgwNSwxNTkuNjU4IEwxNjUuODA1LDE5NC4yMSBDMTY1LjgwNSwyMDAuMzE2IDE2MC44NTYsMjA1LjI2NSAxNTQuNzUxLDIwNS4yNjUgQzE0OC42NDcsMjA1LjI2NSAxNDMuNjk3LDIwMC4zMTcgMTQzLjY5NywxOTQuMjEgTDE0My42OTcsMTU5LjY1OCBMMTQzLjY4MSwxNjEuMDE0IEMxNDMuNjgxLDE1MS44NzEgMTM2LjI2OSwxNDQuNDU4IDEyNy4xMjUsMTQ0LjQ1OCBDMTE3Ljk4MSwxNDQuNDU4IDExMC41NjksMTUxLjg3MSAxMTAuNTY5LDE2MS4wMTQgTDExMC41MzYsMTY1LjE4NCBMMTEwLjUzNiwxOTQuMjA5IEMxMTAuNTM2LDIwMC4zMTUgMTA1LjU4NywyMDUuMjY0IDk5LjQ4MiwyMDUuMjY0IEM5My4zNzcsMjA1LjI2NCA4OC40MjgsMjAwLjMxNiA4OC40MjgsMTk0LjIwOSBMODguNDI4LDE2NS4xODQgTDg4LjQ2LDE2MS4wMTQgQzg4LjQ2LDEzOS42NiAxMDUuNzcsMTIyLjM1IDEyNy4xMjQsMTIyLjM1IEMxNDguNDc4LDEyMi4zNSAxNjUuNzg4LDEzOS42NjEgMTY1Ljc4OCwxNjEuMDE0IEwxNjUuODA1LDE1OS42NTggWiI+PC9wYXRoPjxwYXRoIGQ9Ik0wLDE2Ny45NTcgTDAsMTA3LjM5OSBDMCwxMDEuMjk1IDQuOTQ5LDk2LjM0NSAxMS4wNTQsOTYuMzQ1IEMxNy4xNTksOTYuMzQ1IDIyLjEwOCwxMDEuMjk0IDIyLjEwOCwxMDcuMzk5IEwyMi4xMDgsMTY3Ljk1NyBMMjIuMTIzLDE2Ni42MDEgQzIyLjEyMywxNzUuNzQ0IDI5LjUzNiwxODMuMTU3IDM4LjY3OSwxODMuMTU3IEM0Ny44MjIsMTgzLjE1NyA1NS4yMzUsMTc1Ljc0NCA1NS4yMzUsMTY2LjYwMSBMNTUuMjY4LDE2Mi40MzEgTDU1LjI2OCwxMDcuNCBDNTUuMjY4LDEwMS4yOTYgNjAuMjE3LDk2LjM0NSA2Ni4zMjIsOTYuMzQ1IEM3Mi40MjcsOTYuMzQ1IDc3LjM3NiwxMDEuMjk1IDc3LjM3NiwxMDcuNCBMNzcuMzc2LDE2Mi40MzEgTDc3LjM0NCwxNjYuNjAyIEM3Ny4zNDQsMTg3Ljk1NiA2MC4wMzQsMjA1LjI2NSAzOC42OCwyMDUuMjY1IEMxNy4zMjYsMjA1LjI2NSAwLjAxNiwxODcuOTU1IDAuMDE2LDE2Ni42MDIgTDAsMTY3Ljk1NyBaIj48L3BhdGg+PHBhdGggZD0iTTE5OS4zMTQsMTk0LjIxIEMxOTkuMzE0LDIwMC4zMTYgMTk0LjM2NSwyMDUuMjY1IDE4OC4yNiwyMDUuMjY1IEMxODIuMTU2LDIwNS4yNjUgMTc3LjIwNiwyMDAuMzE3IDE3Ny4yMDYsMTk0LjIxIEwxNzcuMjA2LDEzMy40NzggQzE3Ny4yMDYsMTI3LjM3MyAxODIuMTU1LDEyMi40MjMgMTg4LjI2LDEyMi40MjMgQzE5NC4zNjUsMTIyLjQyMyAxOTkuMzE0LDEyNy4zNzIgMTk5LjMxNCwxMzMuNDc4IEwxOTkuMzE0LDE5NC4yMSBaIj48L3BhdGg+PC9nPjwvc3ZnPgo=\" title=\"UniProt home\"/></a><a href=\"/sparql\">SPARQL</a><a href=\"https://www.uniprot.org/downloads\" id=\"download\">Downloads</a><a accesskey=\"9\" href=\"https://www.uniprot.org/contact\" id=\"contact\"><img alt=\"Contact\" src=\"data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZpZXdCb3g9IjAgMCAxMyAxMCIgZmlsbD0iI2ZmZiIgd2lkdGg9IjEuNGVtIj48cGF0aCBmaWxsLXJ1bGU9ImV2ZW5vZGQiIGQ9Ik0xMiAxLjI1QzEyIC41NiAxMS40NiAwIDEwLjggMEgxLjJDLjU0IDAgMCAuNTYgMCAxLjI1djcuNUMwIDkuNDQuNTQgMTAgMS4yIDEwaDkuNmMuNjYgMCAxLjItLjU2IDEuMi0xLjI1di03LjV6bS0xLjIgMEw2IDQuMzggMS4yIDEuMjVoOS42ek02IDUuNjNMMS4yIDIuNXY2LjI1aDkuNlYyLjVMNiA1LjYzeiI+PC9wYXRoPjwvc3ZnPgo=\"/></a><a rel=\"help\" href=\"https://www.uniprot.org/help/\" id=\"help\">Documentation/Help</a></nav><a accesskey=\"2\" href=\"#content\" hidden=\"hidden\">Skip Header</a><main class=\"sparql results\" id=\"content\"><section class=\"error\" id=\"page-header\"><h2 class=\"page-title\">Malformed query.</h2><p>Encountered &quot; &quot;/&quot; &quot;/ &quot;&quot; at line 9, column 57.\\nWas expecting one of:\\n    &quot;(&quot; ...\\n    &quot;}&quot; ...\\n    &lt;NIL&gt; ...\\n    </p></section><div class=\"main-aside\"><div id=\"queryform\"><form method=\"get\" action=\"/sparql/\" focus=\"query\" name=\"sparql-form\" id=\"sparql-form\"><button id=\"addPrefix\" type=\"button\">Add common prefixes</button><textarea accesskey=\"1\" rows=\"10\" cols=\"80\" name=\"query\" id=\"query\">PREFIX up: &lt;http://purl.uniprot.org/core/&gt;\\nPREFIX rdfs: &lt;http://www.w3.org/2000/01/rdf-schema#&gt;\\nSELECT\\n  ?pdbId ?chain ?pdbChain ?uniprot\\nWHERE\\n{\\n  # Replace the values with your list of PDB identifiers and chains.\\n  VALUES (?pdbId ?pdbChain) { (\\'6VXC\\' \\'A\\') (\\'1BG3\\' \\'B\\') /* Add more pairs as needed */ }\\n\\n  # Make an IRI out of the pdbId\\n  BIND(iri(concat(\\'http://rdf.wwpdb.org/pdb/\\', ?pdbId)) AS ?pdb)\\n\\n  # Map to UniProt entries\\n  ?uniprot rdfs:seeAlso ?pdb .\\n  ?pdb up:database &lt;http://purl.uniprot.org/database/PDB&gt; ;\\n       up:chainSequenceMapping ?chainSm .\\n  ?chainSm up:chain ?chainsPlusRange .\\n\\n  # Extract the list of chains from the text representation.\\n  BIND(STRBEFORE(?chainsPlusRange, \\'=\\') AS ?chain)\\n\\n  # Filter those that match.\\n  FILTER(CONTAINS(?chain, ?pdbChain))\\n}</textarea><input type=\"hidden\" name=\"format\" value=\"html\"><button id=\"sparqlFormSubmitter\" onclick=\"return false;\">Submit Query</button></form></div></div></main><footer id=\"page-footer\"><section class=\"small\" id=\"copyright\">\\n\\t\\t\\t&copy; 2002&ndash; <span property=\"schema:copyrightYear\">2024</span>\\n\\t\\t\\t<a accesskey=\"9\" href=\"https://www.uniprot.org/help/about\">UniProt Consortium</a> |\\n\\t\\t\\t<a href=\"https://www.uniprot.org/help/license\" property=\"schema:license\">License\\n\\t\\t\\t\\t&amp; Disclaimer</a> | <a href=\"https://www.uniprot.org/help/privacy\">Privacy\\n\\t\\t\\t\\tNotice</a></section><section><div resource=\"/help/about#UniProt\" id=\"logos\" typeof=\"NGO\" property=\"creator\"><span resource=\"/help/about#EBI\" typeof=\"Organization\" class=\"ebi\" property=\"member\"><span hidden=\"hidden\" property=\"name\">European Bioinformatics Institute</span><a href=\"https://www.ebi.ac.uk/\" property=\"url\"></a></span><span resource=\"/help/about#PIR\" typeof=\"Organization\" class=\"pir\" property=\"member\"><span hidden=\"hidden\" property=\"name\">Protein Information Resource</span><a href=\"http://pir.georgetown.edu/\" property=\"url\"></a></span><span resource=\"/help/about#SIB\" typeof=\"Organization\" class=\"sib\" property=\"member\"><span hidden=\"hidden\" property=\"name\">SIB Swiss Institute of Bioinformatics</span><a href=\"https://www.sib.swiss/\" property=\"url\"></a></span></div></section><section id=\"poweredby\"><h3>Powered by</h3><span class=\"vitalit\" resource=\"https://www.sib.swiss/alan-bridge-group/\" typeof=\"schema:NGO\" property=\"schema:sponsor\"><h4>Hardware</h4><a href=\"https://www.sib.swiss/alan-bridge-group\" property=\"schema:url\"><img alt=\"Swiss-Prot\" src=\"https://www.sib.swiss//templates/sib/images/SIB_LogoQ_GBv.svg\"/><h5 property=\"schema:name\">Swiss-Prot</h5></a></span><span class=\"openlinksw\" resource=\"http://www.openlinksw.com/#this\" typeof=\"schema:Corporation\" property=\"schema:sponsor\"><h4>Software</h4><a href=\"http://virtuoso.openlinksw.com/\" property=\"schema:url\"><img alt=\"Virtuoso\" src=\"/images/logos/virtuoso.png\"/><h5 property=\"schema:name\">Virtuoso</h5></a></span></section></footer><script src=\"/sparql.js\" type=\"text/javascript\"> </script></body></html>'. Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "SELECT\n",
      "  ?pdbId ?chain ?pdbChain ?uniprot\n",
      "WHERE\n",
      "{\n",
      "  # A space separated list of pairs of PDB identifiers and the chain code.\n",
      "  VALUES(?pdbId ?pdbChain) { ('6VXC' 'A') ('1BG3' 'B') }\n",
      "\n",
      "  # Make an IRI out of the pdbId\n",
      "  BIND(iri(concat('http://rdf.wwpdb.org/pdb/', ?pdbId)) AS ?pdb)\n",
      "\n",
      "  # Map to UniProt entries\n",
      "  ?uniprot rdfs:seeAlso ?pdb .\n",
      "  ?pdb up:database <http://purl.uniprot.org/database/PDB> ;\n",
      "       up:chainSequenceMapping ?chainSm .\n",
      "  ?chainSm up:chain ?chainsPlusRange .\n",
      "\n",
      "  # Extract the list of chains from the text representation.\n",
      "  BIND(STRBEFORE(?chainsPlusRange, '=') AS ?chain)\n",
      "\n",
      "  # Filter those that match.\n",
      "  FILTER(CONTAINS(?chain, ?pdbChain))\n",
      "}\n",
      "\n",
      "⚖️ Success: 23, Mismatch: 4, Fail: 8\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX uniprotkb: <http://purl.uniprot.org/uniprot/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "\n",
      "SELECT ?uniprot ?hgnc ?hgncSymbol\n",
      "WHERE\n",
      "{\n",
      "  # Optional: A specific list of UniProt primary accessions for filtering\n",
      "  VALUES (?acc) { ('YOUR_1ST_UNIPROT_ACC') ('YOUR_2ND_UNIPROT_ACC') }  # Replace with needed accessions or remove line to query all\n",
      "\n",
      "  BIND(iri(concat(str(uniprotkb:), ?acc)) AS ?uniprot)\n",
      "  \n",
      "  # Query mapping to HGNC entries\n",
      "  ?uniprot rdfs:seeAlso ?hgnc .\n",
      "  ?hgnc up:database <http://purl.uniprot.org/database/HGNC> ;\n",
      "       rdfs:comment ?hgncSymbol .\n",
      "}\n",
      "❌ Map UniProt to HGNC identifiers and Symbols 2 != 0. Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX uniprotkb: <http://purl.uniprot.org/uniprot/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "SELECT\n",
      "  ?uniprot\n",
      "  ?hgnc\n",
      "  ?hgncSymbol\n",
      "WHERE\n",
      "{\n",
      "  # A space separated list of UniProt primary accessions.\n",
      "  VALUES (?acc) {('P05067') ('P00750')}\n",
      "  BIND(iri(concat(str(uniprotkb:), ?acc)) AS ?uniprot)\n",
      "  ?uniprot rdfs:seeAlso ?hgnc .\n",
      "  ?hgnc up:database <http://purl.uniprot.org/database/HGNC> ;\n",
      "       rdfs:comment ?hgncSymbol .\n",
      "}\n",
      "\n",
      "⚖️ Success: 23, Mismatch: 4, Fail: 9\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX keywords: <http://purl.uniprot.org/keywords/>\n",
      "SELECT DISTINCT ?link\n",
      "WHERE\n",
      "{\n",
      "    ?protein a up:Protein . \n",
      "    ?protein up:classifiedWith keywords:KW-0005 .\n",
      "    ?protein rdfs:seeAlso ?link .\n",
      "    ?link up:database ?db .\n",
      "    ?db up:category '3D structure databases'\n",
      "}\n",
      "❌ Select all cross-references to external databases of the category '3D structure databases' of UniProt entries that are classified with the keyword 'Acetoin biosynthesis (KW-0005)' 4303 != 0. Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX keywords: <http://purl.uniprot.org/keywords/>\n",
      "SELECT DISTINCT ?link\n",
      "WHERE\n",
      "{\n",
      "    ?protein a up:Protein . \n",
      "    ?protein up:classifiedWith keywords:5 .\n",
      "    ?protein rdfs:seeAlso ?link .\n",
      "    ?link up:database ?db .\n",
      "    ?db up:category '3D structure databases'\n",
      "}\n",
      "\n",
      "⚖️ Success: 23, Mismatch: 4, Fail: 10\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX sachem: <http://bioinfo.uochb.cas.cz/rdf/v1.0/sachem#>\n",
      "PREFIX rh: <http://rdf.rhea-db.org/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX chebihash: <http://purl.obolibrary.org/obo/chebi#>\n",
      "PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
      "SELECT DISTINCT ?protein ?proteinLabel ?reaction ?chebi \n",
      "WHERE {\n",
      "  # Find complete ChEBIs with a Cholestane skeleton, via the Czech Elixir node IDSM Sachem chemical substructure search\n",
      "  SERVICE <https://idsm.elixir-czech.cz/sparql/endpoint/chebi> {\n",
      "    ?chebi sachem:substructureSearch [\n",
      "      sachem:query \"[C@]12(CCC3CCCC[C@]3(C)[C@@]1([H])CC[C@]1(C)[C@@]([H])([C@@](C)([H])CCCC(C)C)CC[C@@]21[H])[H]\"\n",
      "    ] .\n",
      "  }\n",
      "\n",
      "  # Use Rhea to map the found ChEBIs to reactions\n",
      "  SERVICE <https://sparql.rhea-db.org/sparql> {\n",
      "    ?reaction rdfs:subClassOf rh:Reaction ;\n",
      "              rh:side/rh:contains/rh:compound ?compound .\n",
      "    ?compound rh:chebi ?chebi .\n",
      "  }\n",
      "\n",
      "  # Retrieve human proteins that catalyze those reactions from UniProt\n",
      "  ?ca up:catalyzedReaction ?reaction .\n",
      "  ?protein up:annotation / up:catalyticActivity ?ca ;\n",
      "           up:organism taxon:9606 ;\n",
      "           skos:prefLabel ?proteinLabel .\n",
      "}\n",
      "❌ Find Human proteins that catalyze reactions where substrates or product have a Cholestane skeleton 27 != 0. Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX sachem: <http://bioinfo.uochb.cas.cz/rdf/v1.0/sachem#>\n",
      "PREFIX rh: <http://rdf.rhea-db.org/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "SELECT\n",
      "  DISTINCT\n",
      "    ?protein\n",
      "    ?disease\n",
      "    ?rhea\n",
      "    ?chebi\n",
      "    ?omim\n",
      "WHERE {\n",
      "    # Find complete ChEBIs with a Cholestane skeleton, via the Czech Elixir node IDSM Sachem chemical substructure search.\n",
      "    SERVICE <https://idsm.elixir-czech.cz/sparql/endpoint/chebi> {\n",
      "      ?chebi sachem:substructureSearch [\n",
      "        sachem:query\n",
      "\"[C@]12(CCC3CCCC[C@]3(C)[C@@]1([H])CC[C@]1(C)[C@@]([H])([C@@](C)([H])CCCC(C)C)CC[C@@]21[H])[H]\"\n",
      "].\n",
      "   }\n",
      "   # Use the fact that UniProt catalytic activities are annotated using Rhea \n",
      "   # Mapping the found ChEBIs to Rhea reactions\n",
      "   SERVICE <https://sparql.rhea-db.org/sparql>{\n",
      "     ?rhea rh:side/rh:contains/rh:compound/rdfs:subClassOf ?chebi .\n",
      "   }\n",
      "   # Match the found Rhea reactions with human UniProtKB proteins\n",
      "   ?protein up:annotation/up:catalyticActivity/up:catalyzedReaction ?rhea .\n",
      "   ?protein up:organism taxon:9606 .\n",
      "   # Find only those human entries that have an annotated related disease, and optionaly map these to OMIM\n",
      "   ?protein up:annotation/up:disease ?disease .\n",
      "   OPTIONAL {\n",
      "     ?disease rdfs:seeAlso ?omim .\n",
      "     ?omim up:database <http://purl.uniprot.org/database/MIM>\n",
      "   }\n",
      "}\n",
      "\n",
      "⚖️ Success: 23, Mismatch: 4, Fail: 11\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "SELECT\n",
      "  ?ensemblGene ?protein ?rhea\n",
      "WHERE {\n",
      "  ?protein up:reviewed true ;\n",
      "           up:organism taxon:9606 .\n",
      "  ?protein up:annotation ?caa ;\n",
      "           rdfs:seeAlso ?ensemblTranscript .\n",
      "  ?ensemblTranscript up:database <http://purl.uniprot.org/database/Ensembl> .\n",
      "  ?caa up:catalyticActivity ?ca .\n",
      "  ?ca up:catalyzedReaction ?rhea .\n",
      "  ?ensemblTranscript up:transcribedFrom ?ensemblGene .\n",
      "}\n",
      "✅ Select the Gene Protein Reaction sets for Human (Ensembl Gene, Human UniProtKB, Catalyzed Rhea reactions) 122340 == 122340\n",
      "⚖️ Success: 24, Mismatch: 4, Fail: 11\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
      "SELECT ?protein ?name\n",
      "WHERE\n",
      "{\n",
      "        ?protein a up:Protein .\n",
      "        ?protein up:reviewed true .\n",
      "        ?protein up:recommendedName ?recommended .\n",
      "        ?recommended up:fullName ?name .\n",
      "        ?protein up:encodedBy ?gene .\n",
      "        ?gene skos:prefLabel ?text .\n",
      "        FILTER CONTAINS(?text, 'DNA')\n",
      "}\n",
      "✅ Select reviewed UniProt entries (Swiss-Prot), and their recommended protein name, that have a preferred gene name that contains the text 'DNA'. EXACT MATCH\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "SELECT ?geneName ?diseaseAnnotation\n",
      "WHERE\n",
      "{\n",
      "    ?protein a up:Protein .\n",
      "    ?protein up:organism taxon:9606 .  # Restrict to human proteins\n",
      "    ?protein up:encodedBy ?gene .\n",
      "    ?gene skos:prefLabel ?geneName .\n",
      "    ?protein up:annotation ?diseaseAnnotation .\n",
      "    ?diseaseAnnotation a up:Disease_Annotation .\n",
      "    ?diseaseAnnotation rdfs:comment ?diseaseAnnotationText .\n",
      "}\n",
      "✅ Select the preferred gene name and disease annotation of all human UniProt entries that are known to be involved in a disease 31652 == 31652\n",
      "⚖️ Success: 26, Mismatch: 4, Fail: 11\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
      "PREFIX rh: <http://rdf.rhea-db.org/>\n",
      "PREFIX CHEBI: <http://purl.obolibrary.org/obo/CHEBI_>\n",
      "\n",
      "SELECT \n",
      "    DISTINCT\n",
      "        ?uniprot\n",
      "        ?uniprotID\n",
      "        ?recname\n",
      "        ?gene\n",
      "        ?chebi\n",
      "        ?uniprotName\n",
      "WHERE {\n",
      "  SERVICE <https://sparql.rhea-db.org/sparql> {\n",
      "     VALUES (?chebi) { (CHEBI:32395) }\n",
      "     ?rhea rh:side/rh:contains/rh:compound ?compound .\n",
      "     ?compound rh:chebi ?chebi .\n",
      "     ?chebi up:name ?uniprotName .\n",
      "  }\n",
      "  ?uniprot up:annotation/up:catalyticActivity/up:catalyzedReaction ?rhea .\n",
      "  ?uniprot up:mnemonic ?uniprotID .\n",
      "  ?uniprot up:recommendedName/up:fullName ?recname .\n",
      "  OPTIONAL {?uniprot up:encodedBy/skos:prefLabel ?gene .}\n",
      "}\n",
      "✅ Find all proteins linked to arachidonate (CHEBI:32395). EXACT MATCH\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "\n",
      "SELECT ?species ?sci_name ?common_name WHERE {\n",
      "    ?species a up:Taxon ;\n",
      "        up:scientificName ?sci_name ;\n",
      "        up:rank up:Species .\n",
      "    OPTIONAL { ?species up:commonName ?common_name . }\n",
      "}\n",
      "✅ What are the species present in Bgee? 52 == 52\n",
      "⚖️ Success: 28, Mismatch: 4, Fail: 11\n",
      "PREFIX uniprotkb: <http://purl.uniprot.org/uniprot/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "PREFIX genex: <http://purl.org/genex#>\n",
      "\n",
      "SELECT DISTINCT ?anat ?anatName WHERE {\n",
      "    ?seq a orth:Gene ;\n",
      "        genex:isExpressedIn ?anat ;\n",
      "        lscr:xrefUniprot uniprotkb:P02654 .\n",
      "    ?anat a genex:AnatomicalEntity ;\n",
      "        rdfs:label ?anatName .\n",
      "}\n",
      "✅ What are the anatomical entities where the P02654 gene is expressed? Note that P02654 is a UniProtKB identifier of the APOC1 human gene.. EXACT MATCH\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "PREFIX dcterms: <http://purl.org/dc/terms/>\n",
      "\n",
      "SELECT DISTINCT ?symbol ?description ?id\n",
      "?links ?organism ?uniprot ?ensembl ?ncbi WHERE {\n",
      "    ?seq a orth:Gene ;\n",
      "        rdfs:label ?symbol ;\n",
      "        rdfs:seeAlso ?links ;\n",
      "        dcterms:description ?description ;\n",
      "        dcterms:identifier ?id ;\n",
      "        orth:organism ?organism .\n",
      "    OPTIONAL{?seq lscr:xrefUniprot ?uniprot .}\n",
      "    OPTIONAL{?seq lscr:xrefEnsemblGene ?ensembl .}\n",
      "    OPTIONAL{?seq lscr:xrefNCBIGene ?ncbi .}\n",
      "    FILTER (?id = \"ENSG00000130208\")\n",
      "}\n",
      "✅ What is all the metadata related to the ENSG00000130208 gene, where ENSG00000130208 is the identifier of the APOC1 human gene.. EXACT MATCH\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
      "PREFIX genex: <http://purl.org/genex#>\n",
      "\n",
      "SELECT DISTINCT ?anat ?anatName WHERE {\n",
      "    ?seq a orth:Gene ;\n",
      "        rdfs:label \"APOC1\" ;\n",
      "        genex:isAbsentIn ?anat ;\n",
      "        orth:organism ?organism .\n",
      "    ?anat a genex:AnatomicalEntity ;\n",
      "        rdfs:label ?anatName .\n",
      "    ?organism obo:RO_0002162  ?species .\n",
      "    ?species a up:Taxon ;\n",
      "        up:scientificName \"Homo sapiens\" .\n",
      "}\n",
      "✅ What are the anatomical entities where the APOC1 Homo sapiens gene is not expressed, that is where is APOC1 absent?. EXACT MATCH\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "\n",
      "SELECT ?species ?sci_name ?common_name WHERE {\n",
      "    ?species a up:Taxon ;\n",
      "        up:scientificName ?sci_name ;\n",
      "        up:rank up:Species .\n",
      "    OPTIONAL { ?species up:commonName ?common_name . }\n",
      "}\n",
      "✅ What are the species present in Bgee and their scientific and common names?. EXACT MATCH\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX genex: <http://purl.org/genex#>\n",
      "\n",
      "SELECT DISTINCT ?anat ?anatName WHERE {\n",
      "    ?seq a orth:Gene ;\n",
      "        genex:isExpressedIn ?anat ;\n",
      "        rdfs:label \"APOC1\" .\n",
      "    ?anat a genex:AnatomicalEntity ;\n",
      "        rdfs:label ?anatName .\n",
      "}\n",
      "✅ What are the anatomical entities where the APOC1 gene is expressed?. EXACT MATCH\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
      "PREFIX genex: <http://purl.org/genex#>\n",
      "\n",
      "SELECT DISTINCT ?anat ?anatName WHERE {\n",
      "    ?seq a orth:Gene ;\n",
      "        genex:isExpressedIn ?anat ;\n",
      "        rdfs:label \"APOC1\" .\n",
      "    ?anat a genex:AnatomicalEntity ;\n",
      "        rdfs:label ?anatName .\n",
      "    ?seq orth:organism ?organism .\n",
      "    ?organism obo:RO_0002162  ?species .\n",
      "    ?species a up:Taxon ;\n",
      "        up:scientificName \"Homo sapiens\" .\n",
      "}\n",
      "✅ What are the anatomical entities where the APOC1 Homo sapiens gene is expressed?. EXACT MATCH\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "PREFIX genex: <http://purl.org/genex#>\n",
      "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX ensembl: <http://rdf.ebi.ac.uk/resource/ensembl/>\n",
      "PREFIX efo: <http://www.ebi.ac.uk/efo/>\n",
      "\n",
      "SELECT DISTINCT ?anat ?anatName ?score WHERE {\n",
      "    ?seq a orth:Gene ;\n",
      "        lscr:xrefEnsemblGene ensembl:ENSG00000130208 ;\n",
      "        orth:organism ?organism .\n",
      "    ?organism obo:RO_0002162 taxon:9606 .\n",
      "    \n",
      "    ?expression a genex:Expression ;\n",
      "        genex:hasExpressionCondition ?condition ;\n",
      "        genex:hasExpressionLevel ?score ;\n",
      "        genex:hasSequenceUnit ?seq .\n",
      "    \n",
      "    ?condition a genex:ExpressionCondition ;\n",
      "        genex:hasAnatomicalEntity ?anat ;\n",
      "        genex:hasDevelopmentalStage obo:UBERON_0000113 ;\n",
      "        genex:hasSex \"any\" ;\n",
      "        genex:hasStrain ?strain .\n",
      "    \n",
      "    ?anat rdfs:label ?anatName .\n",
      "    \n",
      "    ?strain a efo:EFO_0005135 ;\n",
      "        rdfs:label \"wild-type\" .\n",
      "    \n",
      "    FILTER (?anat != obo:GO_0005575)\n",
      "} ORDER BY DESC(?score)\n",
      "✅ What are the anatomical entities where the human gene APOC1 is expressed in the post-juvenile stage along with its expression score independently of the strain, sex, and cell type? 456 == 456\n",
      "⚖️ Success: 35, Mismatch: 4, Fail: 11\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "SELECT ?protein ?OMA_link\n",
      "WHERE\n",
      "{\n",
      "    ?protein a orth:Protein.\n",
      "    ?protein orth:organism ?organism.\n",
      "    ?organism up:scientificName 'Rattus norvegicus'.\n",
      "    ?protein rdfs:seeAlso ?OMA_link.\n",
      "}\n",
      "❌ Find all Rattus norvegicus' proteins present in OMA RDF database. <urlopen error [Errno 110] Connection timed out>. Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "SELECT ?protein ?OMA_link\n",
      "WHERE\n",
      "{\n",
      "    ?protein a orth:Protein.\n",
      "    ?protein  orth:organism ?organism.\n",
      "    ?inTaxon rdfs:label 'in taxon'@en.\n",
      "    ?organism ?inTaxon ?taxon.\n",
      "    ?taxon  up:scientificName 'Rattus norvegicus'.\n",
      "    ?protein rdfs:seeAlso ?OMA_link.\n",
      "}\n",
      "\n",
      "⚖️ Success: 35, Mismatch: 4, Fail: 12\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
      "select distinct ?HOG ?MEMBER ?GENE_LABEL\n",
      "where {\n",
      "    ?HOG a orth:HOG;\n",
      "         orth:hasHomologousMember ?node1;\n",
      "         orth:hasTaxonomicRange ?taxRange .\n",
      "    ?taxRange orth:taxRange 'Vertebrata' .\n",
      "    ?node1 orth:hasHomologousMember* ?query .\n",
      "    ?query a orth:Protein ;\n",
      "           orth:organism/obo:RO_0002162/up:scientificName 'Homo sapiens';\n",
      "           rdfs:label 'CDIN1' .\n",
      "    ?HOG orth:hasHomologousMember ?nodeMember .\n",
      "    ?nodeMember orth:hasHomologousMember* ?MEMBER .\n",
      "    ?MEMBER a orth:Protein .\n",
      "    OPTIONAL {\n",
      "        ?MEMBER rdfs:label ?GENE_LABEL .\n",
      "    }\n",
      "}\n",
      "❌ Retrieve all proteins belongong to the Hierarchical Orthologous Group (HOG) at the level 'Vertebrata' to which humans' CDIN1 gene belong, together with their gene name symbol if available. <urlopen error [Errno 110] Connection timed out>. Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
      "select distinct ?HOG ?MEMBER ?GENE_LABEL\n",
      "where {\n",
      "    ?HOG a orth:OrthologsCluster ;\n",
      "      orth:hasHomologousMember ?node1 ;\n",
      "      orth:hasTaxonomicRange ?taxRange .\n",
      "    ?taxRange orth:taxRange 'Vertebrata' .\n",
      "    ?node1 orth:hasHomologousMember* ?query ;\n",
      "      orth:hasHomologousMember* ?MEMBER .\n",
      "    ?MEMBER a orth:Protein .\n",
      "    OPTIONAL {\n",
      "        ?MEMBER rdfs:label ?GENE_LABEL .\n",
      "    }\n",
      "    ?query a orth:Protein ;\n",
      "      orth:organism/obo:RO_0002162/up:scientificName 'Homo sapiens';\n",
      "      rdfs:label 'CDIN1'.\n",
      "}\n",
      "\n",
      "⚖️ Success: 35, Mismatch: 4, Fail: 13\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "SELECT ?species ?sciname\n",
      "WHERE {\n",
      "    ?species a up:Taxon.\n",
      "    ?species up:scientificName ?sciname.\n",
      "    ?species up:rank up:Species.\n",
      "}\n",
      "❌ Which species are available on OMA database and their scientific names? <urlopen error [Errno 110] Connection timed out>. Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "SELECT ?species ?sciname WHERE\n",
      "{\n",
      "    ?species a up:Taxon.\n",
      "    ?species up:scientificName ?sciname.\n",
      "    ?species up:rank up:Species.\n",
      "}\n",
      "\n",
      "⚖️ Success: 35, Mismatch: 4, Fail: 14\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "SELECT DISTINCT ?proteinOMA ?species ?mnemonic ?evidenceType ?UniProt_URI\n",
      "WHERE {\n",
      "    ?proteinOMA a orth:Protein;\n",
      "    orth:organism/obo:RO_0002162/up:scientificName ?species;\n",
      "    rdfs:label 'INS';\n",
      "    lscr:xrefUniprot ?UniProt_URI.\n",
      "  \n",
      "    # Search the INS gene mnemonics and evidence types from UniProt database.\n",
      "    SERVICE <http://sparql.uniprot.org/sparql> { \n",
      "        ?UniProt_URI up:mnemonic ?mnemonic;\n",
      "        up:existence/rdfs:label ?evidenceType.\n",
      "    }\n",
      "}\n",
      "❌ Retrieve all proteins in OMA that is encoded by the INS gene and their mnemonics and evidence types from Uniprot database (federated query). <urlopen error [Errno 110] Connection timed out>. Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "SELECT DISTINCT ?proteinOMA ?species  ?mnemonic  ?evidenceType ?UniProt_URI\n",
      "WHERE {\n",
      "    ?proteinOMA a orth:Protein;\n",
      "    orth:organism/obo:RO_0002162/up:scientificName ?species;\n",
      "    rdfs:label 'INS'.\n",
      "    ?proteinOMA lscr:xrefUniprot ?UniProt_URI.\n",
      "    #Search the INS gene mnemonics and evidence types from Uniprot database.\n",
      "    service <http://sparql.uniprot.org/sparql> { \n",
      "    ?UniProt_URI up:mnemonic ?mnemonic;\n",
      "    up:existence/rdfs:label ?evidenceType. }\n",
      "}\n",
      "\n",
      "⚖️ Success: 35, Mismatch: 4, Fail: 15\n",
      "PREFIX sio: <http://semanticscience.org/resource/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "PREFIX ensembl: <http://rdf.ebi.ac.uk/resource/ensembl/>\n",
      "SELECT ?protein2 ?OMA_LINK\n",
      "WHERE {\n",
      "    # The tree that contains orthologs. The leaves are proteins.\n",
      "    # This graph pattern defines the relationship protein1 is orthologous to protein2\n",
      "    ?cluster a orth:OrthologsCluster.\n",
      "    ?cluster orth:hasHomologousMember ?node1.\n",
      "    ?cluster orth:hasHomologousMember ?node2.\n",
      "    ?node2 orth:hasHomologousMember* ?protein2.\n",
      "    ?node1 orth:hasHomologousMember* ?protein1.\n",
      "    \n",
      "    # Specify the protein to look for its orthologs\n",
      "    ?protein1 sio:SIO_010079/lscr:xrefEnsemblGene ensembl:ENSLACG00000002497.\n",
      "    \n",
      "    # The OMA link to the second protein\n",
      "    ?protein2 rdfs:seeAlso ?OMA_LINK.\n",
      "    \n",
      "    FILTER(?node1 != ?node2)\n",
      "}\n",
      "❌ Retrieve all genes that are orthologous to ENSLACG00000002497 Ensembl gene (identifier) <urlopen error [Errno 110] Connection timed out>. Correct query:\n",
      "PREFIX sio: <http://semanticscience.org/resource/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "PREFIX ensembl: <http://rdf.ebi.ac.uk/resource/ensembl/>\n",
      "select ?protein2 ?OMA_LINK \n",
      "where {\n",
      "    #The three that contains Orthologs. The leafs are proteins.\n",
      "    #This graph pattern defines the relationship protein1 is Orthologs to protein2\n",
      "    ?cluster a orth:OrthologsCluster.\n",
      "    ?cluster orth:hasHomologousMember ?node1.\n",
      "    ?cluster orth:hasHomologousMember ?node2. \n",
      "    ?node2 orth:hasHomologousMember* ?protein2. \n",
      "    ?node1 orth:hasHomologousMember* ?protein1.\n",
      "    ########\n",
      "     \n",
      "    #Specify the protein to look for its orthologs\n",
      "    ?protein1 sio:SIO_010079/lscr:xrefEnsemblGene  ensembl:ENSLACG00000002497.\n",
      "    ########\n",
      "     \n",
      "    #The OMA link to the second protein\n",
      "    ?protein2 rdfs:seeAlso ?OMA_LINK. \n",
      "    ########\n",
      "     \n",
      "    filter(?node1 != ?node2) \n",
      "}\n",
      "\n",
      "⚖️ Success: 35, Mismatch: 4, Fail: 16\n",
      "PREFIX sio: <http://semanticscience.org/resource/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "PREFIX ensembl: <http://rdf.ebi.ac.uk/resource/ensembl/>\n",
      "\n",
      "select ?protein2 ?OMA_LINK \n",
      "where {\n",
      "    # The tree that contains paralogs. The leaves are proteins.\n",
      "    # This graph pattern defines the relationship protein1 is paralogous to protein2\n",
      "    ?cluster a orth:ParalogsCluster.\n",
      "    ?cluster orth:hasHomologousMember ?node1.\n",
      "    ?cluster orth:hasHomologousMember ?node2.\n",
      "    ?node2 orth:hasHomologousMember* ?protein2.\n",
      "    ?node1 orth:hasHomologousMember* ?protein1.\n",
      "\n",
      "    # Specify the protein to look for its paralogs\n",
      "    ?protein1 sio:SIO_010079 / lscr:xrefEnsemblGene ensembl:ENSG00000244734.\n",
      "\n",
      "    # The OMA link to the second protein\n",
      "    ?protein2 rdfs:seeAlso ?OMA_LINK. \n",
      "    \n",
      "    filter(?node1 != ?node2) \n",
      "}\n",
      "❌ Retrieve all genes that are paralogous to ENSG00000244734 Ensembl gene (identifier). <urlopen error [Errno 110] Connection timed out>. Correct query:\n",
      "PREFIX sio: <http://semanticscience.org/resource/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "PREFIX ensembl: <http://rdf.ebi.ac.uk/resource/ensembl/>\n",
      "select ?protein2 ?OMA_LINK \n",
      "where {\n",
      "    #The three that contains paralogs. The leafs are proteins.\n",
      "    #This graph pattern defines the relationship protein1 is paralogous to protein2\n",
      "    ?cluster a orth:ParalogsCluster.\n",
      "    ?cluster orth:hasHomologousMember ?node1.\n",
      "    ?cluster orth:hasHomologousMember ?node2.\n",
      "    ?node2 orth:hasHomologousMember* ?protein2.\n",
      "    ?node1 orth:hasHomologousMember* ?protein1.\n",
      "    ########\n",
      "\n",
      "    #Specify the protein to look for its paralogs\n",
      "    ?protein1 sio:SIO_010079/lscr:xrefEnsemblGene  ensembl:ENSG00000244734.\n",
      "    ########\n",
      "\n",
      "    #The OMA link to the second protein\n",
      "    ?protein2 rdfs:seeAlso ?OMA_LINK. \n",
      "    ########\n",
      "     \n",
      "    filter(?node1 != ?node2) \n",
      "}\n",
      "\n",
      "⚖️ Success: 35, Mismatch: 4, Fail: 17\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "PREFIX dc: <http://purl.org/dc/terms/>\n",
      "SELECT ?protein2 ?OMA_LINK ?Uniprot_Link\n",
      "WHERE {\n",
      "    ?cluster a orth:ParalogsCluster.\n",
      "    ?cluster orth:hasHomologousMember ?node1.\n",
      "    ?cluster orth:hasHomologousMember ?node2.\n",
      "    ?node2 orth:hasHomologousMember* ?protein2.\n",
      "    ?node1 orth:hasHomologousMember* ?protein1.\n",
      "    ?protein1 a orth:Protein.\n",
      "    ?protein1 dc:identifier 'HUMAN00529'.\n",
      "    ?protein2 a orth:Protein.\n",
      "    ?protein2 rdfs:seeAlso ?OMA_LINK.\n",
      "    ?protein2 lscr:xrefUniprot ?Uniprot_Link.\n",
      "    FILTER(?node1 != ?node2)\n",
      "}\n",
      "❌ Retrieve all genes that are paralogous to HUMAN00529 OMA protein (identifier) and their cross-reference links to OMA and Uniprot. <urlopen error [Errno 110] Connection timed out>. Correct query:\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "PREFIX dc: <http://purl.org/dc/terms/>\n",
      "select ?protein2 ?Uniprot_link\n",
      "where {\n",
      "    ?cluster a orth:ParalogsCluster.\n",
      "    ?cluster orth:hasHomologousMember ?node1.\n",
      "    ?cluster orth:hasHomologousMember ?node2.\n",
      "    ?node2 orth:hasHomologousMember* ?protein2.\n",
      "    ?node1 orth:hasHomologousMember* ?protein1.\n",
      "    ?protein1 a orth:Protein.\n",
      "    ?protein1 dc:identifier 'HUMAN00529'.\n",
      "    ?protein2 a orth:Protein. \n",
      "    ?protein2  lscr:xrefUniprot ?Uniprot_link. \n",
      "    filter(?node1 != ?node2)\n",
      "}\n",
      "\n",
      "⚖️ Success: 35, Mismatch: 4, Fail: 18\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "PREFIX dc: <http://purl.org/dc/terms/>\n",
      "SELECT ?orthologous_protein ?OMA_LINK ?Uniprot_link\n",
      "WHERE {\n",
      "    ?cluster a orth:OrthologsCluster.\n",
      "    ?cluster orth:hasHomologousMember ?node1.\n",
      "    ?cluster orth:hasHomologousMember ?node2.\n",
      "    ?node2 orth:hasHomologousMember* ?orthologous_protein.\n",
      "    ?node1 orth:hasHomologousMember* ?HUMAN22169_protein.\n",
      "    ?HUMAN22169_protein a orth:Protein.\n",
      "    ?HUMAN22169_protein dc:identifier 'HUMAN22169'.\n",
      "    ?orthologous_protein a orth:Protein;\n",
      "                      rdfs:seeAlso ?OMA_LINK;\n",
      "                      lscr:xrefUniprot ?Uniprot_link.\n",
      "    FILTER(?node1 != ?node2)\n",
      "}\n",
      "❌ Retrieve all genes that are orthologous to HUMAN22169 OMA protein (identifier) and their cross-reference links to OMA and Uniprot. <urlopen error [Errno 110] Connection timed out>. Correct query:\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "PREFIX dc: <http://purl.org/dc/terms/>\n",
      "select ?protein2 ?Uniprot_link\n",
      "where {\n",
      "    ?cluster a orth:OrthologsCluster.\n",
      "    ?cluster orth:hasHomologousMember ?node1.\n",
      "    ?cluster orth:hasHomologousMember ?node2.\n",
      "    ?node2 orth:hasHomologousMember* ?protein2.\n",
      "    ?node1 orth:hasHomologousMember* ?protein1.\n",
      "    ?protein1 a orth:Protein.\n",
      "    ?protein1 dc:identifier 'HUMAN22169'.\n",
      "    ?protein2 a orth:Protein. \n",
      "    ?protein2  lscr:xrefUniprot ?Uniprot_link. \n",
      "    filter(?node1 != ?node2)\n",
      "}\n",
      "\n",
      "⚖️ Success: 35, Mismatch: 4, Fail: 19\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX sio: <http://semanticscience.org/resource/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "PREFIX dc: <http://purl.org/dc/terms/>\n",
      "SELECT ?protein1_label ?protein2 ?geneName2 ?species2 ?Prot2_uniprot ?prot2_ensemblGeneId ?OMA_LINK\n",
      "WHERE {\n",
      "    # The cluster of orthologous genes\n",
      "    ?cluster a orth:OrthologsCluster.\n",
      "    ?cluster orth:hasHomologousMember ?node1.\n",
      "    ?cluster orth:hasHomologousMember ?node2.\n",
      "    ?node2 orth:hasHomologousMember* ?protein2.\n",
      "    ?node1 orth:hasHomologousMember* ?protein1.\n",
      "    \n",
      "    # Filter for Rabbit's APOCI or APOC1 protein\n",
      "    ?protein1 a orth:Protein;\n",
      "              orth:organism/obo:RO_0002162/up:scientificName 'Oryctolagus cuniculus';\n",
      "              lscr:xrefUniprot ?Prot1_uniprot;\n",
      "              rdfs:label ?protein1_label.\n",
      "    FILTER (?protein1_label IN ('APOCI', 'APOC1'))\n",
      "    \n",
      "    # Information about orthologous proteins\n",
      "    ?protein2 a orth:Protein;\n",
      "              lscr:xrefUniprot ?Prot2_uniprot;\n",
      "              sio:SIO_010079/lscr:xrefEnsemblGene/dc:identifier ?prot2_ensemblGeneId;\n",
      "              rdfs:label ?geneName2;\n",
      "              orth:organism/obo:RO_0002162/up:scientificName ?species2;\n",
      "              rdfs:seeAlso ?OMA_LINK.\n",
      "\n",
      "    FILTER(?node1 != ?node2)\n",
      "}\n",
      "❌ Retrieve all genes per species that are orthologous to Rabbit's APOCI or APOC1 gene and their cross-reference links to OMA and Uniprot including the corresponding Ensembl gene identifier. <urlopen error [Errno 110] Connection timed out>. Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX sio: <http://semanticscience.org/resource/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "PREFIX dc: <http://purl.org/dc/terms/>\n",
      "select ?protein1    ?protein2  ?geneName2  ?species2 ?Prot2_uniprot ?prot2_ensemblGeneId  \n",
      "where {\n",
      "    ?cluster a orth:OrthologsCluster.\n",
      "    ?cluster orth:hasHomologousMember ?node1.\n",
      "    ?cluster orth:hasHomologousMember ?node2. \n",
      "    ?node2 orth:hasHomologousMember* ?protein2. \n",
      "    ?node1 orth:hasHomologousMember* ?protein1.\n",
      "    ?protein1 a orth:Protein;\n",
      "    orth:organism/obo:RO_0002162/up:scientificName 'Oryctolagus cuniculus';\n",
      "    rdfs:label 'APOCI'.\n",
      "    ?protein2 a orth:Protein; \n",
      "    lscr:xrefUniprot ?Prot2_uniprot;\n",
      "    sio:SIO_010079/lscr:xrefEnsemblGene/dc:identifier  ?prot2_ensemblGeneId;\n",
      "    rdfs:label ?geneName2;\n",
      "    orth:organism/obo:RO_0002162/up:scientificName ?species2.\n",
      "    filter(?node1 != ?node2)\n",
      "}\n",
      "\n",
      "⚖️ Success: 35, Mismatch: 4, Fail: 20\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "select distinct ?RABBIT_PROTEIN ?RABBIT_UNIPROT_XREF \n",
      "where {\n",
      "    ?cluster a orth:OrthologsCluster.\n",
      "    ?cluster orth:hasHomologousMember ?node1.\n",
      "    ?cluster orth:hasHomologousMember ?node2. \n",
      "    ?node2 orth:hasHomologousMember* ?RABBIT_PROTEIN. \n",
      "    ?node1 orth:hasHomologousMember* ?MOUSE_PROTEIN.\n",
      "    ?MOUSE_PROTEIN a orth:Protein;\n",
      "     orth:organism/obo:RO_0002162/up:scientificName 'Mus musculus';\n",
      "     rdfs:label 'Hbb-y';\n",
      "     lscr:xrefUniprot ?MOUSE_UNIPROT_XREF.\n",
      "    ?RABBIT_PROTEIN a orth:Protein;\n",
      "     orth:organism/obo:RO_0002162/up:scientificName 'Oryctolagus cuniculus';\n",
      "     lscr:xrefUniprot ?RABBIT_UNIPROT_XREF.\n",
      "    filter(?node1 != ?node2)\n",
      "}\n",
      "❌ Retrieve all Rabbit's proteins encoded by genes that are orthologous to Mouses's hemoglobin Y gene and their cross-reference links to Uniprot. <urlopen error [Errno 110] Connection timed out>. Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "select distinct ?MOUSE_PROTEIN ?RABIT_PROTEIN ?MOUSE_UNIPROT_XREF ?RABIT_UNIPROT_XREF \n",
      "where {\n",
      "    ?cluster a orth:OrthologsCluster.\n",
      "    ?cluster orth:hasHomologousMember ?node1.\n",
      "    ?cluster orth:hasHomologousMember ?node2. \n",
      "    ?node2 orth:hasHomologousMember* ?RABIT_PROTEIN. \n",
      "    ?node1 orth:hasHomologousMember* ?MOUSE_PROTEIN.\n",
      "    ?MOUSE_PROTEIN a orth:Protein.\n",
      "    ?MOUSE_PROTEIN  orth:organism/obo:RO_0002162/up:scientificName 'Mus musculus';\n",
      "    rdfs:label 'HBB-Y';\n",
      "    lscr:xrefUniprot ?MOUSE_UNIPROT_XREF.\n",
      "    ?RABIT_PROTEIN a orth:Protein.\n",
      "    ?RABIT_PROTEIN orth:organism/obo:RO_0002162/up:scientificName 'Oryctolagus cuniculus' .\n",
      "    ?RABIT_PROTEIN lscr:xrefUniprot ?RABIT_UNIPROT_XREF.\n",
      "    filter(?node1 != ?node2)\n",
      "}\n",
      "\n",
      "⚖️ Success: 35, Mismatch: 4, Fail: 21\n",
      "PREFIX rh: <http://rdf.rhea-db.org/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX CHEBI: <http://purl.obolibrary.org/obo/CHEBI_>\n",
      "# Select all Rhea reactions that have a given ChEBI ID as reaction participant\n",
      "# \n",
      "# Replace CHEBI:29985 with the desired ChEBI ID\n",
      "# This query corresponds to the Rhea website query: https://www.rhea-db.org/rhea?query=chebi:YOUR_CHEBI_ID\n",
      "SELECT distinct ?chebi ?rhea ?equation\n",
      "WHERE {\n",
      "  ?rhea rdfs:subClassOf rh:Reaction .\n",
      "  ?rhea rh:equation ?equation .\n",
      "  ?rhea rh:side/rh:contains/rh:compound ?compound .\n",
      "  #\n",
      "  # The ChEBI can be used either as a small molecule, the reactive part of a macromolecule, or as a polymer.\n",
      "  #\n",
      "  ?compound (rh:chebi|(rh:reactivePart/rh:chebi)|(rh:underlyingChebi/rh:chebi)) ?chebi .\n",
      "  VALUES (?chebi) { (CHEBI:29985) }\n",
      "}\n",
      "✅ Select all Rhea reactions that have a given ChEBI ID as reaction participant 244 == 244\n",
      "⚖️ Success: 36, Mismatch: 4, Fail: 21\n",
      "PREFIX rh: <http://rdf.rhea-db.org/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "# Query 5 \n",
      "# Select all Rhea reactions mapped to KEGG reactions\n",
      "# KEGG reactions are mapped to Rhea bidirectional reactions\n",
      "# Rhea web query: https://www.rhea-db.org/rhea?query=kegg:*\n",
      "#\n",
      "SELECT ?rhea ?kegg ?rheaDir \n",
      "WHERE {\n",
      "  ?rhea rdfs:subClassOf rh:Reaction .\n",
      "  ?rhea rh:bidirectionalReaction ?rheaDir .\n",
      "  ?rheaDir rdfs:seeAlso ?kegg .\n",
      "  FILTER (regex(str(?kegg),'kegg'))\n",
      "}\n",
      "✅ Select all Rhea reactions mapped to KEGG reactions. EXACT MATCH\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX rh: <http://rdf.rhea-db.org/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "# Query 15\n",
      "# Select all ChEBI compounds used in Rhea as reaction participant\n",
      "# \n",
      "# This query can not be expressed in the Rhea website\n",
      "SELECT ?chebi ?name (count(?rhea) as ?countRhea)\n",
      "WHERE {\n",
      "  ?rhea rdfs:subClassOf rh:Reaction .\n",
      "  ?rhea rh:side/rh:contains/rh:compound ?compound .\n",
      "  #\n",
      "  # the ChEBI can be used either as a small molecule, the reactive part of a macromolecule or as a polymer.\n",
      "  #\n",
      "  ?compound (rh:chebi|(rh:reactivePart/rh:chebi)|(rh:underlyingChebi/rh:chebi)) ?chebi .\n",
      "  ?chebi up:name ?name .\n",
      "}\n",
      "GROUP BY ?chebi ?name\n",
      "ORDER BY DESC(?countRhea)\n",
      "✅ Select all ChEBI compounds used in Rhea as reaction participant. EXACT MATCH\n",
      "PREFIX rh: <http://rdf.rhea-db.org/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "# Select all Rhea reactions mapped to MetaCyc reactions\n",
      "# MetaCyc reactions are mapped to either undirected, left-to-right, right-to-left or bidirectional reactions\n",
      "# Rhea web query: https://www.rhea-db.org/rhea?query=metacyc:*\n",
      "#\n",
      "SELECT distinct ?rhea ?rheaDir ?metacyc\n",
      "WHERE {\n",
      "  ?rhea rdfs:subClassOf rh:Reaction .\n",
      "  {\n",
      "    ?rhea rdfs:seeAlso ?metacyc .\n",
      "    FILTER CONTAINS(str(?metacyc), \"METACYC\") \n",
      "    BIND(?rhea as ?rheaDir)\n",
      "  }\n",
      "  UNION\n",
      "  {\n",
      "    ?rhea rh:directionalReaction ?directionalReaction .\n",
      "    ?directionalReaction rdfs:seeAlso ?metacyc . \n",
      "    FILTER CONTAINS(str(?metacyc), \"METACYC\") \n",
      "    BIND(?directionalReaction as ?rheaDir  )\n",
      "  }\n",
      "  UNION\n",
      "  {\n",
      "    ?rhea rh:bidirectionalReaction ?bidirectionalReaction .\n",
      "    ?bidirectionalReaction rdfs:seeAlso ?metacyc . \n",
      "    FILTER CONTAINS(str(?metacyc), \"METACYC\") \n",
      "    BIND(?bidirectionalReaction as ?rheaDir  )\n",
      "  }\n",
      "}\n",
      "✅ Select all Rhea reactions mapped to MetaCyc reactions 9293 == 9293\n",
      "⚖️ Success: 39, Mismatch: 4, Fail: 21\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX rh: <http://rdf.rhea-db.org/>\n",
      "PREFIX chebi: <http://purl.obolibrary.org/obo/chebi#>\n",
      "PREFIX chebislash: <http://purl.obolibrary.org/obo/chebi/>\n",
      "\n",
      "SELECT DISTINCT\n",
      "    ?uniprot ?mnemonic ?rhea ?chebi ?chebiLabel ?smiles ?inchiKey\n",
      "WHERE\n",
      "{\n",
      "    # Retrieve human proteins annotated with catalytic activity\n",
      "    ?uniprot up:annotation/up:catalyticActivity/up:catalyzedReaction ?rhea ;\n",
      "             up:organism taxon:9606 ;\n",
      "             up:mnemonic ?mnemonic .\n",
      "    \n",
      "    # Federated query to Rhea to get the compounds involved in these reactions\n",
      "    SERVICE <https://sparql.rhea-db.org/sparql> {\n",
      "        ?rhea rh:side/rh:contains/rh:compound ?compound .\n",
      "        ?compound rh:chebi ?chebi .\n",
      "        ?compound (rh:chebi|(rh:reactivePart/rh:chebi)|(rh:underlyingChebi/rh:chebi)) ?chebi .\n",
      "    }\n",
      "    \n",
      "    # Retrieve compound details from the ChEBI ontology\n",
      "    ?chebi rdfs:label ?chebiLabel ;\n",
      "           chebislash:smiles ?smiles ;\n",
      "           chebislash:inchikey ?inchiKey .\n",
      "}\n",
      "❌ Generate a draft human metabolome 279926 != 0. Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX rh: <http://rdf.rhea-db.org/>\n",
      "PREFIX chebislash: <http://purl.obolibrary.org/obo/chebi/>\n",
      "SELECT\n",
      "  ?uniprot ?mnemonic ?rhea ?chebi ?smiles ?inchiKey\n",
      "WHERE\n",
      "{\n",
      "  ?rhea rh:side/rh:contains/rh:compound ?compound .\n",
      "  ?compound (rh:chebi|(rh:reactivePart/rh:chebi)|(rh:underlyingChebi/rh:chebi)) ?chebi .\n",
      "  ?chebi chebislash:smiles ?smiles ;\n",
      "          chebislash:inchikey ?inchiKey .\n",
      "  SERVICE <https://sparql.uniprot.org/sparql/> {\n",
      "     ?uniprot up:annotation/up:catalyticActivity/up:catalyzedReaction ?rhea ;\n",
      "                                             up:organism taxon:9606 ;\n",
      "                                             up:mnemonic ?mnemonic .\n",
      "\n",
      "   }\n",
      "}\n",
      "\n",
      "⚖️ Success: 39, Mismatch: 4, Fail: 22\n",
      "PREFIX rh: <http://rdf.rhea-db.org/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX ec: <http://purl.uniprot.org/enzyme/>\n",
      "# \n",
      "# Select all Rhea reactions mapped to a specific enzyme classification (EC class, subclass, or sub-subclass)\n",
      "# Customize the FILTER regex to match the desired EC (sub)-class.\n",
      "#\n",
      "SELECT ?ec ?ecNumber ?rhea ?accession ?equation\n",
      "WHERE {\n",
      "  ?rhea rdfs:subClassOf rh:Reaction .\n",
      "  ?rhea rh:accession ?accession .\n",
      "  ?rhea rh:ec ?ec .\n",
      "  BIND(strafter(str(?ec),str(ec:)) as ?ecNumber)\n",
      "  # Example: class (e.g., EC 1.-.-.-)\n",
      "  FILTER (regex(?ecNumber,'^1\\\\.')) \n",
      "  # Uncomment for subclass (e.g., EC 1.1.-.-)\n",
      "  #FILTER (regex(?ecNumber,'^1\\\\.1\\\\.')) \n",
      "  # Uncomment for sub-subclass (e.g., EC 1.1.1.-)\n",
      "  #FILTER (regex(?ecNumber,'^1\\\\.1\\\\.1\\\\.')) \n",
      "  ?rhea rh:equation ?equation .\n",
      "}\n",
      "✅ Select all Rhea reactions linked to an enzyme classification (sub)-class 2658 == 2658\n",
      "⚖️ Success: 40, Mismatch: 4, Fail: 22\n",
      "PREFIX rh: <http://rdf.rhea-db.org/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "\n",
      "# Query to select all Rhea reactions (unspecified direction)\n",
      "# This query returns the identifier (id), accession, boolean attributes (isChemicallyBalanced, isTransport),\n",
      "# and chemical equation for each reaction.\n",
      "\n",
      "SELECT ?rhea ?id ?accession ?isChemicallyBalanced ?isTransport  ?equation \n",
      "WHERE {\n",
      "  ?rhea rdfs:subClassOf rh:Reaction .\n",
      "  ?rhea rh:id ?id .\n",
      "  ?rhea rh:accession ?accession .\n",
      "  ?rhea rh:equation ?equation .\n",
      "  ?rhea rh:isTransport ?isTransport .\n",
      "  ?rhea rh:isChemicallyBalanced ?isChemicallyBalanced .\n",
      "}\n",
      "✅ Select all Rhea reactions 16609 == 16609\n",
      "⚖️ Success: 41, Mismatch: 4, Fail: 22\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX rh: <http://rdf.rhea-db.org/>\n",
      "PREFIX CHEBI: <http://purl.obolibrary.org/obo/CHEBI_>\n",
      "# Query 16\n",
      "# Select all Rhea reactions that have a pair of ChEBI IDs as reaction participant and in opposite side\n",
      "# This sample queries for CHEBI:29985 (L-glutamate) and CHEBI:58359 (L-glutamine). Adjust the ChEBI IDs as needed.\n",
      "SELECT ?chebi1 ?name1 ?chebi2 ?name2 ?rhea ?equation\n",
      "WHERE {\n",
      "  VALUES (?chebi1) { (CHEBI:29985) }  # Replace with the first ChEBI ID\n",
      "  ?chebi1 up:name ?name1 .\n",
      "  ?rhea rh:side ?reactionSide1 .\n",
      "  ?reactionSide1 rh:contains / rh:compound / rh:chebi ?chebi1 .\n",
      "\n",
      "  VALUES (?chebi2) { (CHEBI:58359) }  # Replace with the second ChEBI ID\n",
      "  ?chebi2 up:name ?name2 .\n",
      "\n",
      "  ?rhea rh:side ?reactionSide2 .\n",
      "  ?reactionSide2 rh:contains / rh:compound / rh:chebi ?chebi2 .\n",
      "  \n",
      "  ?reactionSide1 rh:transformableTo ?reactionSide2 .\n",
      "  \n",
      "  ?rhea rh:equation ?equation .\n",
      "}\n",
      "✅ Select all Rhea reactions that have a pair of ChEBI IDs as reaction participant and in opposite side 32 == 32\n",
      "⚖️ Success: 42, Mismatch: 4, Fail: 22\n",
      "PREFIX rh: <http://rdf.rhea-db.org/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "# Query 7 \n",
      "# Select all Rhea reactions mapped to Reactome reactions\n",
      "# MetaCyc reactions are mapped to either undirected, left-to-right or right-to-left reactions\n",
      "# Rhea web query: https://www.rhea-db.org/rhea?query=reactome:*\n",
      "#    \n",
      "SELECT distinct ?rhea ?rheaDir ?reactome  \n",
      "WHERE {\n",
      "  ?rhea rdfs:subClassOf rh:Reaction .\n",
      "  {\n",
      "    ?rhea rdfs:seeAlso ?reactome .\n",
      "    FILTER CONTAINS(str(?reactome), \"reactome\") \n",
      "    BIND(?rhea as ?rheaDir)\n",
      "  }\n",
      "  UNION\n",
      "  {\n",
      "    ?rhea rh:directionalReaction ?directionalReaction .\n",
      "    ?directionalReaction rdfs:seeAlso ?reactome . \n",
      "    FILTER CONTAINS(str(?reactome), \"reactome\") \n",
      "    BIND(?directionalReaction as ?rheaDir  )\n",
      "  }\n",
      "}\n",
      "✅ Select all Rhea reactions mapped to Reactome reactions. EXACT MATCH\n",
      "PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
      "PREFIX rh: <http://rdf.rhea-db.org/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX ec: <http://purl.uniprot.org/enzyme/>\n",
      "\n",
      "# Retrieve the count of reactions mapped to each level (main class) of the enzyme classification\n",
      "# This query mimics the Filter section of the Rhea website (Browse all reactions)\n",
      "# https://www.rhea-db.org/rhea?query=\n",
      "SELECT ?ecClass (STR(?ecName) AS ?ecClassName) (COUNT(?rhea) AS ?rheaCount)\n",
      "WHERE {\n",
      "  SERVICE <https://sparql.uniprot.org/sparql> {\n",
      "    VALUES (?ecClass) { (ec:1.-.-.-)(ec:2.-.-.-)(ec:3.-.-.-)(ec:4.-.-.-)(ec:5.-.-.-) (ec:6.-.-.-) (ec:7.-.-.-)}\n",
      "    ?ecNumber rdfs:subClassOf ?ecClass .\n",
      "    ?ecClass skos:prefLabel ?ecName .\n",
      "  }\n",
      "  ?rhea rdfs:subClassOf rh:Reaction .\n",
      "  ?rhea rh:ec ?ecNumber .\n",
      "} \n",
      "GROUP BY ?ecClass ?ecName\n",
      "✅ Distribution of reactions according to the first class of the enzyme classification (federated query) 7 == 7\n",
      "⚖️ Success: 44, Mismatch: 4, Fail: 22\n",
      "PREFIX rh: <http://rdf.rhea-db.org/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX pubmed: <http://rdf.ncbi.nlm.nih.gov/pubmed/>\n",
      "# Query to select all Rhea reactions annotated with PubMed identifier 29867142\n",
      "SELECT ?pubmed ?rhea ?accession ?isTransport ?equation \n",
      "WHERE {\n",
      "  ?rhea rdfs:subClassOf rh:Reaction .\n",
      "  ?rhea rh:accession ?accession .\n",
      "  ?rhea rh:citation ?pubmed .\n",
      "  VALUES (?pubmed) { (pubmed:29867142) }\n",
      "  ?rhea rh:isTransport ?isTransport .\n",
      "  ?rhea rh:equation ?equation .\n",
      "} ORDER BY ?rhea\n",
      "✅ Select all Rhea reactions annotated with a given Pubmed ID 14 == 14\n",
      "⚖️ Success: 45, Mismatch: 4, Fail: 22\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX rh: <http://rdf.rhea-db.org/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX owl: <http://www.w3.org/2002/07/owl#>\n",
      "PREFIX chebihash: <http://purl.obolibrary.org/obo/chebi#>\n",
      "PREFIX CHEBI: <http://purl.obolibrary.org/obo/CHEBI_>\n",
      "\n",
      "# Query to select all Rhea reactions that involve a lipid (CHEBI:18059) or its children.\n",
      "SELECT DISTINCT ?chebi ?name ?rhea ?equation\n",
      "WHERE {\n",
      "  ?rhea rdfs:subClassOf rh:Reaction .\n",
      "  ?rhea rh:equation ?equation .\n",
      "  ?rhea rh:side/rh:contains/rh:compound ?compound .\n",
      "  \n",
      "  # The ChEBI can be used either as a small molecule, the reactive part of a macromolecule, or as a polymer.\n",
      "  { \n",
      "    ?chebi rdfs:subClassOf* CHEBI:18059 . # Lipid or its children\n",
      "    ?compound (rh:chebi | (rh:reactivePart/rh:chebi) | (rh:underlyingChebi/rh:chebi)) ?chebi .\n",
      "  }\n",
      "  UNION \n",
      "  { # Include non-pH 7.3 species\n",
      "    ?not7_3 rdfs:subClassOf* CHEBI:18059 .\n",
      "    ?not7_3 rdfs:subClassOf ?chebiRestriction .\n",
      "    ?chebiRestriction a owl:Restriction .\n",
      "    ?chebiRestriction owl:onProperty chebihash:has_major_microspecies_at_pH_7_3 .\n",
      "    ?chebiRestriction owl:someValuesFrom ?chebi .\n",
      "    ?compound (rh:chebi | (rh:reactivePart/rh:chebi) | (rh:underlyingChebi/rh:chebi)) ?chebi .\n",
      "  }\n",
      "  ?chebi up:name ?name .\n",
      "}\n",
      "✅ Select all Rhea reactions that involve a lipid, i.e. children of CHEBI:18059 in the ChEBI hierarchy. 12834 == 12834\n",
      "⚖️ Success: 46, Mismatch: 4, Fail: 22\n",
      "PREFIX rh: <http://rdf.rhea-db.org/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "# Query 8\n",
      "# Select all Rhea reactions mapped to GO molecular function\n",
      "# \n",
      "# This query corresponds to the Rhea website query:\n",
      "# https://www.rhea-db.org/rhea?query=go:*\n",
      "#\n",
      "SELECT ?go ?rhea ?equation \n",
      "WHERE {\n",
      "  ?rhea rdfs:subClassOf rh:Reaction .\n",
      "  ?rhea rdfs:seeAlso ?go .\n",
      "  FILTER CONTAINS(str(?go), \"GO_\")   \n",
      "  ?rhea rh:equation ?equation .\n",
      "}\n",
      "✅ Select all Rhea reactions mapped to GO molecular functions. EXACT MATCH\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX rh: <http://rdf.rhea-db.org/>\n",
      "# Select all Rhea reactions used to annotate enzyme sequences in UniProtKB\n",
      "# and return the number of UniProtKB entries\n",
      "# Federated query using a service to UniProt SPARQL endpoint\n",
      "#\n",
      "# This query corresponds to the Rhea website query:\n",
      "# https://www.rhea-db.org/rhea?query=uniprot:*\n",
      "#\n",
      "SELECT ?uniprotCount ?rhea ?accession ?equation \n",
      "WHERE {\n",
      "  SERVICE <https://sparql.uniprot.org/sparql> { \n",
      "    SELECT ?rhea (count(?uniprot) as ?uniprotCount) {\n",
      "      ?uniprot up:annotation/up:catalyticActivity/up:catalyzedReaction ?rhea . \n",
      "    }\n",
      "    GROUP BY ?rhea\n",
      "  }\n",
      "  ?rhea rh:accession ?accession .\n",
      "  ?rhea rh:equation ?equation .\n",
      "}\n",
      "✅ Select all Rhea reactions linked to protein sequences (enzymes and transporters) in UniProtKB 12402 == 12402\n",
      "⚖️ Success: 48, Mismatch: 4, Fail: 22\n",
      "PREFIX rh: <http://rdf.rhea-db.org/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "# Query 3\n",
      "# Select the specific form of RHEA:11628.\n",
      "# This query mimics the Related reactions sections of\n",
      "# https://www.rhea-db.org/rhea/11628\n",
      "#\n",
      "SELECT ?rhea ?equation ?specificForm ?specificEquation \n",
      "WHERE {\n",
      "  VALUES (?rhea) {(rh:11628)}\n",
      "  ?rhea rh:equation ?equation .\n",
      "  ?specificForm rdfs:subClassOf+ ?rhea .\n",
      "  ?specificForm rh:equation ?specificEquation .\n",
      "}\n",
      "✅ Select the specific form of RHEA:11628, a reaction that deals with general chemical classes (i.e. select the equivalent reactions involving instances of the chemical classes in RHEA:11628 12 == 12\n",
      "⚖️ Success: 49, Mismatch: 4, Fail: 22\n",
      "PREFIX sachem: <http://bioinfo.uochb.cas.cz/rdf/v1.0/sachem#>\n",
      "PREFIX rh: <http://rdf.rhea-db.org/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
      "\n",
      "SELECT ?rhea ?chebi ?chebiName ?equation\n",
      "WHERE {\n",
      "  SERVICE <https://idsm.elixir-czech.cz/sparql/endpoint/chebi> {\n",
      "    ?chebi sachem:substructureSearch [\n",
      "        sachem:query \"[C@]12(CCC3CCCC[C@]3(C)[C@@]1([H])CC[C@]1(C)[C@@]([H])([C@@](C)([H])CCCC(C)C)CC[C@@]21[H])[H]\" ].\n",
      "  }\n",
      "  ?chebi oboInOwl:id ?chebiID .\n",
      "  ?chebi rdfs:label ?chebiName .\n",
      "\n",
      "  ?rhea rdfs:subClassOf rh:Reaction .\n",
      "  ?rhea rh:side/rh:contains/rh:compound ?compound .\n",
      "  ?compound (rh:chebi|(rh:reactivePart/rh:chebi)|(rh:underlyingChebi/rh:chebi)) ?chebi .\n",
      "  ?rhea rh:equation ?equation .\n",
      "}\n",
      "❌ Use IDSM Sachem to find ChEBIs with a a Cholestane skeleton (in SMILES). Then match returned ChEBIs to Rhea undirected reactions. QueryBadFormed: A bad request has been sent to the endpoint: probably the SPARQL query is badly formed. \n",
      "\n",
      "Response:\n",
      "b'<!DOCTYPE html SYSTEM \"about:legacy-compat\">\\n<html xmlns=\"http://www.w3.org/1999/xhtml\" lang=\"en\" xml:lang=\"en\"><head><title>Rhea</title><meta content=\"text/html; charset=UTF-8\" http-equiv=\"Content-Type\"/><link href=\"/\" rel=\"home\"/><link href=\"/base.css\" type=\"text/css\" rel=\"stylesheet\"/><link type=\"image/vnd.microsoft.icon\" href=\"https://www.rhea-db.org//favicon.ico\" rel=\"shortcut icon\"/><link href=\"/rhea-sparql.css\" type=\"text/css\" rel=\"stylesheet\"/><link href=\"https://creativecommons.org/licenses/by/4.0/\" rel=\"license\"/><script src=\"https://cdn.jsdelivr.net/npm/yasgui-yasqe@2.11.22/dist/yasqe.bundled.min.js\"></script><link type=\"text/css\" rel=\"stylesheet\" href=\"https://cdn.jsdelivr.net/yasqe/2.2/yasqe.min.css\"/><meta content=\"width=device-width, initial-scale=1\" name=\"viewport\"/><meta content=\"no-referrer\" name=\"referrer\"/></head><body typeof=\"schema:WebPage\"><header class=\"row-container\"><div class=\"logos row-container\"><a href=\"https://www.sib.swiss\" class=\"sib_logo\"><img src=\"https://www.rhea-db.org//style/images/sib_logo_short.png\" style=\"height: 50px\" alt=\"SIB logo\"/></a><a href=\"https://www.rhea-db.org\" class=\"rhea_logo\"><img src=\"https://www.rhea-db.org//style/images/rhea_logo.png\" style=\"height: 50px\" alt=\"Rhea logo\"/></a></div><nav class=\"row-container\" id=\"menu\"><a href=\"/\" itemprop=\"url\"><svg xmlns=\"http://www.w3.org/2000/svg\" width=\"24\" viewBox=\"0 0 24 24\" height=\"24\"><path fill=\"none\" d=\"M0 0h24v24H0z\"/><path d=\"M3.9 12c0-1.71 1.39-3.1 3.1-3.1h4V7H7c-2.76 0-5 2.24-5 5s2.24 5 5 5h4v-1.9H7c-1.71 0-3.1-1.39-3.1-3.1zM8 13h8v-2H8v2zm9-6h-4v1.9h4c1.71 0 3.1 1.39 3.1 3.1s-1.39 3.1-3.1 3.1h-4V17h4c2.76 0 5-2.24 5-5s-2.24-5-5-5z\"/></svg><span itemprop=\"name\">SPARQL</span></a><a href=\"https://www.rhea-db.org//help/download\" itemprop=\"url\"><svg xmlns=\"http://www.w3.org/2000/svg\" width=\"24\" viewBox=\"0 0 24 24\" height=\"24\"><path fill=\"none\" d=\"M0 0h24v24H0z\"/><path d=\"M19.35 10.04C18.67 6.59 15.64 4 12 4 9.11 4 6.6 5.64 5.35 8.04 2.34 8.36 0 10.91 0 14c0 3.31 2.69 6 6 6h13c2.76 0 5-2.24 5-5 0-2.64-2.05-4.78-4.65-4.96zM17 13l-5 5-5-5h3V9h4v4h3z\"/></svg><span itemprop=\"name\">Download</span></a><a href=\"https://www.rhea-db.org//help\" itemprop=\"url\"><svg xmlns=\"http://www.w3.org/2000/svg\" width=\"24\" viewBox=\"0 0 24 24\" height=\"24\"><path fill=\"none\" d=\"M0 0h24v24H0V0z\"/><path d=\"M19.07 4.93l-1.41 1.41C19.1 7.79 20 9.79 20 12c0 4.42-3.58 8-8 8s-8-3.58-8-8c0-4.08 3.05-7.44 7-7.93v2.02C8.16 6.57 6 9.03 6 12c0 3.31 2.69 6 6 6s6-2.69 6-6c0-1.66-.67-3.16-1.76-4.24l-1.41 1.41C15.55 9.9 16 10.9 16 12c0 2.21-1.79 4-4 4s-4-1.79-4-4c0-1.86 1.28-3.41 3-3.86v2.14c-.6.35-1 .98-1 1.72 0 1.1.9 2 2 2s2-.9 2-2c0-.74-.4-1.38-1-1.72V2h-1C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10c0-2.76-1.12-5.26-2.93-7.07z\"/></svg><span itemprop=\"name\">Help</span></a><a href=\"https://www.rhea-db.org//feedback\" itemprop=\"url\"><svg xmlns=\"http://www.w3.org/2000/svg\" width=\"24\" viewBox=\"0 0 24 24\" height=\"24\"><path fill=\"none\" d=\"M0 0h24v24H0z\"/><path d=\"M20 4H4c-1.1 0-1.99.9-1.99 2L2 18c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2zm0 4l-8 5-8-5V6l8 5 8-5v2z\"/></svg><span itemprop=\"name\">Feedback</span></a></nav></header><a accesskey=\"2\" href=\"#content\" hidden=\"hidden\">Skip Header</a><main class=\"sparql results\" id=\"content\"><section class=\"error\" id=\"page-header\"><h2 class=\"page-title\">Malformed query.</h2><p>org.eclipse.rdf4j.query.parser.sparql.ast.VisitorException: QName \\'oboInOwl:id\\' uses an undefined prefix</p></section><div class=\"main-aside\"><div id=\"queryform\"><form method=\"get\" action=\"/sparql/\" focus=\"query\" name=\"sparql-form\" id=\"sparql-form\"><button id=\"addPrefix\" type=\"button\">Add common prefixes</button><textarea accesskey=\"1\" rows=\"10\" cols=\"80\" name=\"query\" id=\"query\">PREFIX sachem: &lt;http://bioinfo.uochb.cas.cz/rdf/v1.0/sachem#&gt;\\nPREFIX rh: &lt;http://rdf.rhea-db.org/&gt;\\nPREFIX rdfs: &lt;http://www.w3.org/2000/01/rdf-schema#&gt;\\nPREFIX obo: &lt;http://purl.obolibrary.org/obo/&gt;\\n\\nSELECT ?rhea ?chebi ?chebiName ?equation\\nWHERE {\\n  SERVICE &lt;https://idsm.elixir-czech.cz/sparql/endpoint/chebi&gt; {\\n    ?chebi sachem:substructureSearch [\\n        sachem:query &quot;[C@]12(CCC3CCCC[C@]3(C)[C@@]1([H])CC[C@]1(C)[C@@]([H])([C@@](C)([H])CCCC(C)C)CC[C@@]21[H])[H]&quot; ].\\n  }\\n  ?chebi oboInOwl:id ?chebiID .\\n  ?chebi rdfs:label ?chebiName .\\n\\n  ?rhea rdfs:subClassOf rh:Reaction .\\n  ?rhea rh:side/rh:contains/rh:compound ?compound .\\n  ?compound (rh:chebi|(rh:reactivePart/rh:chebi)|(rh:underlyingChebi/rh:chebi)) ?chebi .\\n  ?rhea rh:equation ?equation .\\n}</textarea><input type=\"hidden\" name=\"format\" value=\"html\"><button id=\"sparqlFormSubmitter\" onclick=\"return false;\">Submit Query</button></form></div></div></main><footer><a href=\"https://www.sib.swiss\" target=\"_blank\"><img src=\"https://www.rhea-db.org//style/images/sib_logo_short.png\" alt=\"Supported by SIB\"/></a><div id=\"fund\">\\n\\t\\t\\tThis SPARQL endpoint is hosted by the <a href=\"https://www.sib.swiss/alan-bridge-group/\" target=\"_blank\">SIB Swiss Institute of\\n\\t\\t\\tBioinformatics</a>, and uses <a href=\"https://rdf4j.org/\" target=\"_blank\">Eclipse rdf4j</a> software.<br/>Rhea\\n\\t\\t\\tdevelopment and curation activities at the SIB are supported by the\\n\\t\\t\\tSwiss Federal Government through the State Secretariat for\\n\\t\\t\\tEducation, Research and Innovation (SERI).<br/> The development of\\n\\t\\t\\tthis SPARQL endpoint was supported in part by ELIXIR (Implementation\\n\\t\\t\\tstudy on \"<a href=\"https://www.elixir-europe.org/platforms/data/microbial-metabolism-resource\" target=\"_blank\">A\\n\\t\\t\\t\\tmicrobial metabolism resource for Systems Biology</a>\").\\n\\t\\t</div><div id=\"legal\"><a href=\"https://www.rhea-db.org/help/license-disclaimer\">License\\n\\t\\t\\t\\t& Disclaimer</a> | <a href=\"https://www.sib.swiss/privacy-policy-2018\">Privacy\\n\\t\\t\\t\\tNotice</a></div></footer><script src=\"/sparql.js\" type=\"text/javascript\"> </script></body></html>'. Correct query:\n",
      "PREFIX sachem: <http://bioinfo.uochb.cas.cz/rdf/v1.0/sachem#>\n",
      "PREFIX rh: <http://rdf.rhea-db.org/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "\n",
      "SELECT \n",
      "  ?rhea \n",
      "  ?chebi\n",
      "WHERE {\n",
      "  SERVICE <https://idsm.elixir-czech.cz/sparql/endpoint/chebi> {\n",
      "    ?chebi sachem:substructureSearch [\n",
      "        sachem:query \"[C@]12(CCC3CCCC[C@]3(C)[C@@]1([H])CC[C@]1(C)[C@@]([H])([C@@](C)([H])CCCC(C)C)CC[C@@]21[H])[H]\" ].\n",
      "  }\n",
      "  ?rhea rh:side/rh:contains/rh:compound/rdfs:subClassOf ?chebi .\n",
      "}\n",
      "\n",
      "⚖️ Success: 49, Mismatch: 4, Fail: 23\n",
      "PREFIX rh: <http://rdf.rhea-db.org/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX ec: <http://purl.uniprot.org/enzyme/>\n",
      "# Query 9\n",
      "# Select all Rhea reactions mapped to EC numbers (enzyme classification)\n",
      "# \n",
      "# This query corresponds to the Rhea website query:\n",
      "# https://www.rhea-db.org/rhea?query=ec:*\n",
      "#\n",
      "SELECT ?ec ?ecNumber ?rhea ?accession ?equation \n",
      "WHERE {\n",
      "  ?rhea rdfs:subClassOf rh:Reaction .\n",
      "  ?rhea rh:accession ?accession .\n",
      "  ?rhea rh:ec ?ec .\n",
      "  BIND(strafter(str(?ec),str(ec:)) as ?ecNumber)\n",
      "  ?rhea rh:isTransport ?isTransport .\n",
      "  ?rhea rh:equation ?equation .\n",
      "}\n",
      "✅ Select all Rhea reactions mapped to enzyme classification (EC numbers). EXACT MATCH\n",
      "PREFIX rh: <http://rdf.rhea-db.org/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "# Generic query to select all cross-references mapped to a specified Rhea reaction ID\n",
      "SELECT DISTINCT ?rhea ?rheaDir ?xref \n",
      "WHERE {\n",
      "  ?rhea rdfs:subClassOf rh:Reaction .\n",
      "  VALUES (?rhea) {(<REPLACE_WITH_RHEA_ID>)} # Replace with dynamic Rhea reaction ID\n",
      "  {\n",
      "    ?rhea rdfs:seeAlso ?xref .\n",
      "    BIND(?rhea as ?rheaDir)\n",
      "  }\n",
      "  UNION\n",
      "  {\n",
      "    ?rhea rh:directionalReaction ?directionalReaction .\n",
      "    ?directionalReaction rdfs:seeAlso ?xref . \n",
      "    BIND(?directionalReaction as ?rheaDir)\n",
      "  }\n",
      "  UNION\n",
      "  {\n",
      "    ?rhea rh:bidirectionalReaction ?bidirectionalReaction .\n",
      "    ?bidirectionalReaction rdfs:seeAlso ?xref . \n",
      "    BIND(?bidirectionalReaction as ?rheaDir)\n",
      "  }\n",
      "}\n",
      "❌ Select all cross-references for a given reaction QueryBadFormed: A bad request has been sent to the endpoint: probably the SPARQL query is badly formed. \n",
      "\n",
      "Response:\n",
      "b'<!DOCTYPE html SYSTEM \"about:legacy-compat\">\\n<html xmlns=\"http://www.w3.org/1999/xhtml\" lang=\"en\" xml:lang=\"en\"><head><title>Rhea</title><meta content=\"text/html; charset=UTF-8\" http-equiv=\"Content-Type\"/><link href=\"/\" rel=\"home\"/><link href=\"/base.css\" type=\"text/css\" rel=\"stylesheet\"/><link type=\"image/vnd.microsoft.icon\" href=\"https://www.rhea-db.org//favicon.ico\" rel=\"shortcut icon\"/><link href=\"/rhea-sparql.css\" type=\"text/css\" rel=\"stylesheet\"/><link href=\"https://creativecommons.org/licenses/by/4.0/\" rel=\"license\"/><script src=\"https://cdn.jsdelivr.net/npm/yasgui-yasqe@2.11.22/dist/yasqe.bundled.min.js\"></script><link type=\"text/css\" rel=\"stylesheet\" href=\"https://cdn.jsdelivr.net/yasqe/2.2/yasqe.min.css\"/><meta content=\"width=device-width, initial-scale=1\" name=\"viewport\"/><meta content=\"no-referrer\" name=\"referrer\"/></head><body typeof=\"schema:WebPage\"><header class=\"row-container\"><div class=\"logos row-container\"><a href=\"https://www.sib.swiss\" class=\"sib_logo\"><img src=\"https://www.rhea-db.org//style/images/sib_logo_short.png\" style=\"height: 50px\" alt=\"SIB logo\"/></a><a href=\"https://www.rhea-db.org\" class=\"rhea_logo\"><img src=\"https://www.rhea-db.org//style/images/rhea_logo.png\" style=\"height: 50px\" alt=\"Rhea logo\"/></a></div><nav class=\"row-container\" id=\"menu\"><a href=\"/\" itemprop=\"url\"><svg xmlns=\"http://www.w3.org/2000/svg\" width=\"24\" viewBox=\"0 0 24 24\" height=\"24\"><path fill=\"none\" d=\"M0 0h24v24H0z\"/><path d=\"M3.9 12c0-1.71 1.39-3.1 3.1-3.1h4V7H7c-2.76 0-5 2.24-5 5s2.24 5 5 5h4v-1.9H7c-1.71 0-3.1-1.39-3.1-3.1zM8 13h8v-2H8v2zm9-6h-4v1.9h4c1.71 0 3.1 1.39 3.1 3.1s-1.39 3.1-3.1 3.1h-4V17h4c2.76 0 5-2.24 5-5s-2.24-5-5-5z\"/></svg><span itemprop=\"name\">SPARQL</span></a><a href=\"https://www.rhea-db.org//help/download\" itemprop=\"url\"><svg xmlns=\"http://www.w3.org/2000/svg\" width=\"24\" viewBox=\"0 0 24 24\" height=\"24\"><path fill=\"none\" d=\"M0 0h24v24H0z\"/><path d=\"M19.35 10.04C18.67 6.59 15.64 4 12 4 9.11 4 6.6 5.64 5.35 8.04 2.34 8.36 0 10.91 0 14c0 3.31 2.69 6 6 6h13c2.76 0 5-2.24 5-5 0-2.64-2.05-4.78-4.65-4.96zM17 13l-5 5-5-5h3V9h4v4h3z\"/></svg><span itemprop=\"name\">Download</span></a><a href=\"https://www.rhea-db.org//help\" itemprop=\"url\"><svg xmlns=\"http://www.w3.org/2000/svg\" width=\"24\" viewBox=\"0 0 24 24\" height=\"24\"><path fill=\"none\" d=\"M0 0h24v24H0V0z\"/><path d=\"M19.07 4.93l-1.41 1.41C19.1 7.79 20 9.79 20 12c0 4.42-3.58 8-8 8s-8-3.58-8-8c0-4.08 3.05-7.44 7-7.93v2.02C8.16 6.57 6 9.03 6 12c0 3.31 2.69 6 6 6s6-2.69 6-6c0-1.66-.67-3.16-1.76-4.24l-1.41 1.41C15.55 9.9 16 10.9 16 12c0 2.21-1.79 4-4 4s-4-1.79-4-4c0-1.86 1.28-3.41 3-3.86v2.14c-.6.35-1 .98-1 1.72 0 1.1.9 2 2 2s2-.9 2-2c0-.74-.4-1.38-1-1.72V2h-1C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10c0-2.76-1.12-5.26-2.93-7.07z\"/></svg><span itemprop=\"name\">Help</span></a><a href=\"https://www.rhea-db.org//feedback\" itemprop=\"url\"><svg xmlns=\"http://www.w3.org/2000/svg\" width=\"24\" viewBox=\"0 0 24 24\" height=\"24\"><path fill=\"none\" d=\"M0 0h24v24H0z\"/><path d=\"M20 4H4c-1.1 0-1.99.9-1.99 2L2 18c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2zm0 4l-8 5-8-5V6l8 5 8-5v2z\"/></svg><span itemprop=\"name\">Feedback</span></a></nav></header><a accesskey=\"2\" href=\"#content\" hidden=\"hidden\">Skip Header</a><main class=\"sparql results\" id=\"content\"><section class=\"error\" id=\"page-header\"><h2 class=\"page-title\">Malformed query.</h2><p>Not a valid (absolute) IRI: REPLACE_WITH_RHEA_ID</p></section><div class=\"main-aside\"><div id=\"queryform\"><form method=\"get\" action=\"/sparql/\" focus=\"query\" name=\"sparql-form\" id=\"sparql-form\"><button id=\"addPrefix\" type=\"button\">Add common prefixes</button><textarea accesskey=\"1\" rows=\"10\" cols=\"80\" name=\"query\" id=\"query\">PREFIX rh: &lt;http://rdf.rhea-db.org/&gt;\\nPREFIX rdfs: &lt;http://www.w3.org/2000/01/rdf-schema#&gt;\\n# Generic query to select all cross-references mapped to a specified Rhea reaction ID\\nSELECT DISTINCT ?rhea ?rheaDir ?xref \\nWHERE {\\n  ?rhea rdfs:subClassOf rh:Reaction .\\n  VALUES (?rhea) {(&lt;REPLACE_WITH_RHEA_ID&gt;)} # Replace with dynamic Rhea reaction ID\\n  {\\n    ?rhea rdfs:seeAlso ?xref .\\n    BIND(?rhea as ?rheaDir)\\n  }\\n  UNION\\n  {\\n    ?rhea rh:directionalReaction ?directionalReaction .\\n    ?directionalReaction rdfs:seeAlso ?xref . \\n    BIND(?directionalReaction as ?rheaDir)\\n  }\\n  UNION\\n  {\\n    ?rhea rh:bidirectionalReaction ?bidirectionalReaction .\\n    ?bidirectionalReaction rdfs:seeAlso ?xref . \\n    BIND(?bidirectionalReaction as ?rheaDir)\\n  }\\n}</textarea><input type=\"hidden\" name=\"format\" value=\"html\"><button id=\"sparqlFormSubmitter\" onclick=\"return false;\">Submit Query</button></form></div></div></main><footer><a href=\"https://www.sib.swiss\" target=\"_blank\"><img src=\"https://www.rhea-db.org//style/images/sib_logo_short.png\" alt=\"Supported by SIB\"/></a><div id=\"fund\">\\n\\t\\t\\tThis SPARQL endpoint is hosted by the <a href=\"https://www.sib.swiss/alan-bridge-group/\" target=\"_blank\">SIB Swiss Institute of\\n\\t\\t\\tBioinformatics</a>, and uses <a href=\"https://rdf4j.org/\" target=\"_blank\">Eclipse rdf4j</a> software.<br/>Rhea\\n\\t\\t\\tdevelopment and curation activities at the SIB are supported by the\\n\\t\\t\\tSwiss Federal Government through the State Secretariat for\\n\\t\\t\\tEducation, Research and Innovation (SERI).<br/> The development of\\n\\t\\t\\tthis SPARQL endpoint was supported in part by ELIXIR (Implementation\\n\\t\\t\\tstudy on \"<a href=\"https://www.elixir-europe.org/platforms/data/microbial-metabolism-resource\" target=\"_blank\">A\\n\\t\\t\\t\\tmicrobial metabolism resource for Systems Biology</a>\").\\n\\t\\t</div><div id=\"legal\"><a href=\"https://www.rhea-db.org/help/license-disclaimer\">License\\n\\t\\t\\t\\t& Disclaimer</a> | <a href=\"https://www.sib.swiss/privacy-policy-2018\">Privacy\\n\\t\\t\\t\\tNotice</a></div></footer><script src=\"/sparql.js\" type=\"text/javascript\"> </script></body></html>'. Correct query:\n",
      "PREFIX rh: <http://rdf.rhea-db.org/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "# Query 4\n",
      "# Select all cross-references mapped to RHEA:21016\n",
      "# \n",
      "# This query mimics the Cross-references section of \n",
      "# https://www.rhea-db.org/rhea/21016\n",
      "#\n",
      "SELECT distinct ?rhea ?rheaDir ?xref \n",
      "WHERE {\n",
      "  ?rhea rdfs:subClassOf rh:Reaction .\n",
      "  VALUES (?rhea) {(rh:21016)}\n",
      "  {\n",
      "    ?rhea rdfs:seeAlso ?xref .\n",
      "    BIND(?rhea as ?rheaDir)\n",
      "  }\n",
      "  UNION\n",
      "  {\n",
      "    ?rhea rh:directionalReaction ?directionalReaction .\n",
      "    ?directionalReaction rdfs:seeAlso ?xref . \n",
      "    BIND(?directionalReaction as ?rheaDir  )\n",
      "  }\n",
      "  UNION\n",
      "  {\n",
      "    ?rhea rh:bidirectionalReaction ?bidirectionalReaction .\n",
      "    ?bidirectionalReaction rdfs:seeAlso ?xref . \n",
      "    BIND(?bidirectionalReaction as ?rheaDir  )\n",
      "  }\n",
      "}\n",
      "\n",
      "⚖️ Success: 50, Mismatch: 4, Fail: 24\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "import requests\n",
    "\n",
    "from qdrant_client.models import FieldCondition, Filter, MatchValue, ScoredPoint\n",
    "from rdflib.plugins.sparql import prepareQuery, prepareUpdate\n",
    "from rdflib.plugins.sparql.algebra import translateQuery\n",
    "from SPARQLWrapper import JSON, TURTLE, XML, SPARQLWrapper\n",
    "\n",
    "from expasy_api.embed import QUERIES_COLLECTION, get_vectordb\n",
    "\n",
    "example_queries = [\n",
    "#   {\n",
    "#     \"question\": \"Was any UniProt entry integrated on the 9th of January 2013\",\n",
    "#     \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "#     \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nASK\\nWHERE\\n{\\n\\t?protein a up:Protein .\\n\\t?protein up:created '2013-01-09'^^xsd:date\\n}\",\n",
    "#     \"results\": 1,\n",
    "#     \"runtime\": 0\n",
    "#   },\n",
    "  {\n",
    "    \"question\": \"Select all taxa from the UniProt taxonomy\",\n",
    "    \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nSELECT ?taxon\\nFROM <http://sparql.uniprot.org/taxonomy>\\nWHERE\\n{\\n    ?taxon a up:Taxon .\\n}\",\n",
    "    \"results\": 2941742,\n",
    "    \"runtime\": 47\n",
    "  },\n",
    "  {\n",
    "    \"question\": \"How many distinct extinct organisms are represented in UniProtKB\",\n",
    "    \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX keywords: <http://purl.uniprot.org/keywords/>\\nSELECT \\n  ?taxon\\n  (SAMPLE(?name) AS ?anName)\\n  (COUNT(DISTINCT ?protein) AS ?entriesPerExtinctTaxon)\\nWHERE\\n{\\n  GRAPH<http://sparql.uniprot.org/taxonomy>{\\n    ?taxon a up:Taxon ;\\n           up:scientificName ?name .\\n  }\\n  ?protein up:organism ?taxon ;\\n           up:classifiedWith keywords:952 .\\n} GROUP BY ?taxon ORDER BY ?taxon\",\n",
    "    \"results\": 29,\n",
    "    \"runtime\": 0\n",
    "  },\n",
    "  {\n",
    "    \"question\": \"Select all human UniProt entries with a sequence variant that leads to a tyrosine to phenylalanine substitution\",\n",
    "    \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX taxon: <http://purl.uniprot.org/taxonomy/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\\nPREFIX faldo: <http://biohackathon.org/resource/faldo#>\\nSELECT ?protein ?annotation ?begin ?text\\nWHERE\\n{\\n        ?protein a up:Protein ;\\n            up:organism taxon:9606 ; \\n            up:annotation ?annotation .\\n        ?annotation a up:Natural_Variant_Annotation ;\\n            rdfs:comment ?text ;\\n            up:substitution ?substitution ;\\n            up:range/faldo:begin\\n                [ faldo:position ?begin ;\\n                  faldo:reference ?sequence ] .\\n        ?sequence rdf:value ?value .\\n        BIND (substr(?value, ?begin, 1) as ?original) .\\n        FILTER(?original = 'Y' && ?substitution = 'F') .\\n} \",\n",
    "    \"results\": 128,\n",
    "    \"runtime\": 0\n",
    "  },\n",
    "  {\n",
    "    \"question\": \"Select all UniProt entries that were integrated on the 30th of November 2010\",\n",
    "    \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nSELECT ?protein\\nWHERE\\n{\\n\\t?protein a up:Protein . \\n\\t?protein up:created '2010-11-30'^^xsd:date\\n} \",\n",
    "    \"results\": 643989,\n",
    "    \"runtime\": 107\n",
    "  },\n",
    "  {\n",
    "    \"question\": \"Select the average number of cross-references to the PDB database of UniProt entries that have at least one cross-reference to the PDB database\",\n",
    "    \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nSELECT (AVG(?linksToPdbPerEntry) AS ?avgLinksToPdbPerEntry)\\nWHERE\\n{\\n\\tSELECT ?protein (COUNT(DISTINCT ?db) AS ?linksToPdbPerEntry)\\n\\tWHERE\\n\\t{\\n\\t\\t?protein a up:Protein .\\n\\t\\t?protein rdfs:seeAlso ?db .\\n\\t\\t?db up:database <http://purl.uniprot.org/database/PDB> .\\n\\t}\\n\\tGROUP BY ?protein ORDER BY DESC(?linksToPdbPerEntry)\\n}\",\n",
    "    \"results\": 1,\n",
    "    \"runtime\": 2\n",
    "  },\n",
    "  {\n",
    "    \"question\": \"Select the number of UniProt entries for each of the EC (Enzyme Commission) top level categories\",\n",
    "    \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX ec: <http://purl.uniprot.org/enzyme/>\\nSELECT ?ecClass (COUNT(?protein) as ?size)\\nWHERE\\n{\\n    VALUES (?ecClass) {(ec:1.-.-.-) (ec:2.-.-.-) (ec:3.-.-.-) (ec:4.-.-.-) (ec:5.-.-.-) (ec:6.-.-.-) (ec:7.-.-.-)} .\\n    ?protein ( up:enzyme | up:domain/up:enzyme | up:component/up:enzyme ) ?enzyme .\\n    # Enzyme subclasses are materialized, do not use rdfs:subClassOf+\\n    ?enzyme rdfs:subClassOf ?ecClass .\\n}\\nGROUP BY ?ecClass ORDER BY ?ecClass\\n\",\n",
    "    \"results\": 7,\n",
    "    \"runtime\": 0\n",
    "  },\n",
    "  {\n",
    "    \"question\": \"Find all Natural Variant Annotations if associated via an evidence tag to an article with a pubmed identifier\",\n",
    "    \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\\nSELECT \\n    ?accession\\n    ?annotation_acc \\n    ?pubmed\\nWHERE\\n{\\n        ?protein a up:Protein ;\\n            up:annotation ?annotation .\\n        ?annotation a up:Natural_Variant_Annotation .\\n        ?linkToEvidence rdf:object ?annotation ;\\n                        up:attribution ?attribution .\\n        ?attribution up:source ?source .\\n        ?source a up:Journal_Citation .\\n  BIND(SUBSTR(STR(?protein),33) AS ?accession)\\n  BIND(IF(CONTAINS(STR(?annotation), \\\"#SIP\\\"), SUBSTR(STR(?annotation),33), SUBSTR(STR(?annotation),36))AS?annotation_acc)\\n  BIND(SUBSTR(STR(?source),35) AS ?pubmed)\\n}\\n\",\n",
    "    \"results\": 197124,\n",
    "    \"runtime\": 9\n",
    "  },\n",
    "  {\n",
    "    \"question\": \"Select all bacterial taxa and their scientific name from the UniProt taxonomy\",\n",
    "    \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX taxon: <http://purl.uniprot.org/taxonomy/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nSELECT ?taxon ?name\\nWHERE\\n{\\n    ?taxon a up:Taxon .\\n    ?taxon up:scientificName ?name .\\n    # Taxon subclasses are materialized, do not use rdfs:subClassOf+\\n    ?taxon rdfs:subClassOf taxon:2 .\\n}\",\n",
    "    \"results\": 556591,\n",
    "    \"runtime\": 22\n",
    "  },\n",
    "  {\n",
    "    \"question\": \"Find how often an article in pubmed was used in an evidence tag in a human protein (ordered by most used to least)\",\n",
    "    \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX taxon: <http://purl.uniprot.org/taxonomy/>\\nPREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\\nSELECT \\n    ?source \\n    (COUNT(?attribution) AS ?attribitions)\\nWHERE\\n{\\n        ?protein a up:Protein ;\\n            up:organism taxon:9606 ;\\n            up:annotation ?annotation .\\n        ?linkToEvidence rdf:object ?annotation ;\\n                        up:attribution ?attribution .\\n        ?attribution up:source ?source .\\n        ?source a up:Journal_Citation .\\n} GROUP BY ?source ORDER BY DESC(COUNT(?attribution))\\n\",\n",
    "    \"results\": 43889,\n",
    "    \"runtime\": 30\n",
    "  },\n",
    "  {\n",
    "    \"question\": \"For two accessions find the GO term labels and group them into GO process,function and component\",\n",
    "    \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\n\\nSELECT \\n    (CONCAT(SUBSTR(STR(?protein), 33)) AS ?uniprot)\\n    (GROUP_CONCAT(?celtype; separator=\\\";\\\") AS ?celtypes)\\n    (GROUP_CONCAT(?biotype; separator=\\\";\\\") AS ?biotypes)\\n    (GROUP_CONCAT(?moltype; separator=\\\";\\\") AS ?moltypes)\\nWHERE\\n{\\n    VALUES (?ac) {(\\\"Q6GZX4\\\") (\\\"Q96375\\\")}\\n    BIND (IRI(CONCAT(\\\"http://purl.uniprot.org/uniprot/\\\",?ac)) AS ?protein)\\n    ?protein a up:Protein .\\n    ?protein up:classifiedWith ?goTerm .\\n    #Determine if the type is biological_process\\n    OPTIONAL {\\n        ?goTerm rdfs:subClassOf <http://purl.obolibrary.org/obo/GO_0008150>.\\n        ?goTerm rdfs:label ?biotype .\\n    }\\n    #Determine if the type is cellular_component\\n    OPTIONAL {\\n        ?goTerm rdfs:subClassOf <http://purl.obolibrary.org/obo/GO_0005575>.\\n        ?goTerm rdfs:label ?celtype .\\n    }\\n    #Determine if the type is molecular_function\\n    OPTIONAL {\\n        ?goTerm rdfs:subClassOf <http://purl.obolibrary.org/obo/GO_0003674> .\\n        ?goTerm rdfs:label ?moltype .\\n    }\\n    #Filter out the uniprot keywords\\n    FILTER(bound(?biotype) || bound(?celtype) || bound(?moltype))\\n} GROUP BY ?protein\\n\",\n",
    "    \"results\": 2,\n",
    "    \"runtime\": 0\n",
    "  },\n",
    "  {\n",
    "    \"question\": \"Number of reviewed entries (Swiss-Prot) that are related to kinase activity\",\n",
    "    \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX taxon: <http://purl.uniprot.org/taxonomy/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\\nPREFIX GO: <http://purl.obolibrary.org/obo/GO_>\\nSELECT\\n    (COUNT(DISTINCT(?protein)) AS ?pc)\\nWHERE\\n{   \\n    ?protein rdf:type up:Protein ;\\n        up:reviewed true  ;\\n        up:organism taxon:9606 ;\\n        up:classifiedWith|(up:classifiedWith/rdfs:subClassOf) GO:0016301 .\\n}\",\n",
    "    \"results\": 1,\n",
    "    \"runtime\": 3\n",
    "  },\n",
    "  {\n",
    "    \"question\": \"Find all names associated with uniprot entry P05067, and if the name is associated with the entry it's domains or its components\",\n",
    "    \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nSELECT\\n  ?protein\\n  ?anyKindOfName \\n  ?names \\n  ?partType\\nWHERE\\n{\\n  BIND(<http://purl.uniprot.org/uniprot/P05067> AS ?protein)\\n  ?protein a up:Protein .\\n  {\\n    ?protein (up:recommendedName|up:alternativeName) ?structuredName .\\n  }\\n    UNION\\n  {\\n    VALUES(?partType){(up:domain) (up:component)}\\n    ?protein ?partType ?part .\\n    ?part (up:recommendedName|up:alternativeName) ?structuredName .\\n  }\\n  ?structuredName ?anyKindOfName  ?names .\\n  ?anyKindOfName rdfs:subPropertyOf up:structuredNameType .\\n}\",\n",
    "    \"results\": 50,\n",
    "    \"runtime\": 90\n",
    "  },\n",
    "  {\n",
    "    \"question\": \"Get the list of uniprot entries for the chromosome of proteome UP000000625\",\n",
    "    \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nSELECT \\n  ?protein\\n  ?proteome \\nWHERE\\n{\\n  ?protein a up:Protein ;\\n           up:reviewed true ;\\n           up:proteome ?proteome .\\n  VALUES (?proteome) {(<http://purl.uniprot.org/proteomes/UP000000625#Chromosome>)}\\n}\",\n",
    "    \"results\": 8802,\n",
    "    \"runtime\": 3\n",
    "  },\n",
    "  {\n",
    "    \"question\": \"Select all UniProt entries, and their organism and amino acid sequences (including isoforms), for E. coli K12 and all its strains\",\n",
    "    \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX taxon: <http://purl.uniprot.org/taxonomy/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\\nSELECT ?protein ?organism ?isoform ?sequence\\nWHERE\\n{\\n    ?protein a up:Protein .\\n    ?protein up:organism ?organism .\\n    # Taxon subclasses are materialized, do not use rdfs:subClassOf+\\n    ?organism rdfs:subClassOf taxon:83333 .\\n    ?protein up:sequence ?isoform .\\n    ?isoform rdf:value ?sequence .\\n}\",\n",
    "    \"results\": 3746,\n",
    "    \"runtime\": 0\n",
    "  },\n",
    "  {\n",
    "    \"question\": \"Find UniProt entries with merged loci in Bordetella avium\",\n",
    "    \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX taxon: <http://purl.uniprot.org/taxonomy/>\\nSELECT\\n  ?protein \\n  (GROUP_CONCAT(?locusName; separator=',') AS ?locusNames)\\nWHERE \\n{ \\n  ?protein a up:Protein ;\\n    up:organism taxon:360910 ;\\n    up:encodedBy ?gene .\\n  ?gene up:locusName ?locusName .\\n} \\nGROUP BY ?protein \\nHAVING (COUNT(?locusName) > 1)\\n\",\n",
    "    \"results\": 3349,\n",
    "    \"runtime\": 2\n",
    "  },\n",
    "  {\n",
    "    \"question\": \"Find UniProtKB entries with more than 1 Topological domain annotation\",\n",
    "    \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\\nSELECT \\n    ?protein \\n    (GROUP_CONCAT(?comment; separator=\\\", \\\") AS ?comments)\\nWHERE\\n{\\n    ?protein a up:Protein ;\\n            up:annotation ?annotation . \\n    ?annotation rdf:type up:Topological_Domain_Annotation ;\\n            rdfs:comment ?comment .\\n} \\nGROUP BY ?protein \\nHAVING (COUNT(?annotation) > 1)\\n\",\n",
    "    \"results\": 175067,\n",
    "    \"runtime\": 10\n",
    "  },\n",
    "  {\n",
    "    \"question\": \"Find longest comment text associated with a UniProtKB Natural Variant Annotation\",\n",
    "    \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nSELECT \\n    ?annotation ?comment\\nWHERE {\\n    ?annotation a up:Natural_Variant_Annotation ;\\n        rdfs:comment ?comment . \\n} \\nORDER BY DESC(STRLEN(?comment))\\n\",\n",
    "    \"results\": 66010,\n",
    "    \"runtime\": 2\n",
    "  },\n",
    "  {\n",
    "    \"question\": \"Find the co-occurence count of Topological Domain comment text in UniProtKB entries\",\n",
    "    \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\\nSELECT \\n    ?comment1 \\n    ?comment2 \\n    (COUNT(?comment1) AS ?count1)\\nWHERE\\n{\\n    ?protein a up:Protein ;\\n               up:annotation ?annotation1 , \\n                             ?annotation2 . \\n    ?annotation1 rdf:type up:Topological_Domain_Annotation ;\\n        rdfs:comment ?rawComment1 .\\n    ?annotation2 rdf:type up:Topological_Domain_Annotation ;\\n        rdfs:comment ?rawComment2 . \\n    BIND(IF(contains(?rawComment1, ';'), \\n            STRBEFORE(?rawComment1,';'), \\n            ?rawComment1) AS ?comment1)\\n    BIND(IF(contains(?rawComment2, ';'), \\n            STRBEFORE(?rawComment2,';'), \\n            ?rawComment2) AS ?comment2)\\n    FILTER(?annotation1 != ?annotation2)\\n} \\nGROUP BY ?comment1 ?comment2 \\nORDER BY DESC(COUNT(?comment1))\\n\",\n",
    "    \"results\": 118,\n",
    "    \"runtime\": 21\n",
    "  },\n",
    "  {\n",
    "    \"question\": \"Find the similar proteins for UniProtKB entry P05067 sorted by UniRef cluser identity\",\n",
    "    \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX uniprotkb: <http://purl.uniprot.org/uniprot/>\\nSELECT \\n    ?similar ?identity\\nFROM <http://sparql.uniprot.org/uniref>\\nFROM <http://sparql.uniprot.org/uniprot>\\nWHERE\\n{\\n    BIND (uniprotkb:P05607 AS ?protein)\\n    ?cluster up:member ?member ;\\n             up:member/up:sequenceFor ?protein;\\n             up:identity ?identity .\\n    ?member up:sequenceFor ?similar .\\n    FILTER(!sameTerm(?similar, ?protein))\\n} \\nORDER BY DESC(?identity)\\n\",\n",
    "    \"results\": 79,\n",
    "    \"runtime\": 0\n",
    "  },\n",
    "  {\n",
    "    \"question\": \"Find the human protein which contains an Epitope VSTQ, where T is a phosporylated threonine\",\n",
    "    \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX taxon: <http://purl.uniprot.org/taxonomy/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\\nPREFIX faldo: <http://biohackathon.org/resource/faldo#>\\nSELECT \\n  ?protein \\n  ?comment\\n  ?begin\\n  ?end \\nWHERE\\n{\\n  ?protein a up:Protein ;\\n    up:organism taxon:9606 ; \\n    up:sequence ?sequence ;\\n    up:annotation ?annotation .\\n  ?annotation a up:Modified_Residue_Annotation ;\\n    rdfs:comment ?comment ;\\n    up:range ?range .\\n  ?range \\n    faldo:begin [ faldo:position ?begin ; faldo:reference ?sequence ] ;\\n    faldo:end [ faldo:position ?end ; faldo:reference ?sequence ] .\\n  ?sequence rdf:value ?aaSequence .\\n  FILTER (SUBSTR(?aaSequence, ?begin -2 , 4) = \\\"VSTQ\\\")     \\n  FILTER (CONTAINS(?comment, \\\"Phosphothreonine\\\"))\\n}\\n\",\n",
    "    \"results\": 2,\n",
    "    \"runtime\": 13\n",
    "  },\n",
    "  {\n",
    "    \"question\": \"For the human entry P05067 (Amyloid-beta precursor protein) find the gene start ends in WikiData\",\n",
    "    \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    \"query\": \"PREFIX wdt: <http://www.wikidata.org/prop/direct/>\\nPREFIX uniprotkb: <http://purl.uniprot.org/uniprot/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX ps: <http://www.wikidata.org/prop/statement/>\\nPREFIX pq: <http://www.wikidata.org/prop/qualifier/>\\nPREFIX p: <http://www.wikidata.org/prop/>\\n\\nSELECT \\n\\t?protein \\n\\t?begin\\n\\t?end\\n\\t?chromosome\\n\\t?assembly\\nWHERE {\\n    {\\n        BIND(uniprotkb:P05067 AS ?proteinIRI)\\n        BIND (SUBSTR(STR(?proteinIRI), STRLEN(STR(uniprotkb:))+1) AS ?protein)\\n    }\\n    SERVICE <https://query.wikidata.org/sparql> {\\n        ?wp wdt:P352 ?protein ;\\n            wdt:P702 ?wg . \\n        ?wg p:P644   ?wgss .\\n        ?wgss ps:P644        ?begin ;\\n          pq:P1057/wdt:P1813 ?chromosome ;\\n          pq:P659/rdfs:label ?assembly .\\n        ?wg p:P645 ?wgse .\\n        ?wgse ps:P645        ?end ;\\n          pq:P1057/wdt:P1813 ?chromosome ;\\n          pq:P659/rdfs:label ?assembly .\\n        FILTER(lang(?assembly) = \\\"en\\\")\\n  } \\n}\\n\",\n",
    "    \"results\": 2,\n",
    "    \"runtime\": 0\n",
    "  },\n",
    "  {\n",
    "    \"question\": \"Retrieve entries and Catalytic activities in the reviewed (Swiss-Prot) section that have experimental evidences,\",\n",
    "    \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\\nSELECT  \\n  ?protein\\n  ?rhea \\nWHERE {\\n  # ECO 269 is experimental evidence\\n  BIND (<http://purl.obolibrary.org/obo/ECO_0000269> as ?evidence)\\n  GRAPH <http://sparql.uniprot.org/uniprot> {\\n    ?protein up:reviewed true ;\\n      up:annotation ?a ;\\n      up:attribution ?attribution  .\\n\\n    ?a a up:Catalytic_Activity_Annotation ;\\n      up:catalyticActivity ?ca .\\n    ?ca up:catalyzedReaction ?rhea .\\n  \\n    [] rdf:subject ?a ;\\n      rdf:predicate up:catalyticActivity ;\\n      rdf:object ?ca ;\\n      up:attribution ?attribution .\\n\\n    ?attribution up:evidence ?evidence .\\n  }\\n}\\n\",\n",
    "    \"results\": 47936,\n",
    "    \"runtime\": 4\n",
    "  },\n",
    "  {\n",
    "    \"question\": \"Select the UniProt entry with the mnemonic 'A4_HUMAN'\",\n",
    "    \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nSELECT ?protein\\nWHERE\\n{\\n    ?protein a up:Protein .\\n    ?protein up:mnemonic 'A4_HUMAN'\\n}\",\n",
    "    \"results\": 1,\n",
    "    \"runtime\": 0\n",
    "  },\n",
    "  {\n",
    "    \"question\": \"Connect patents cited in UniProtKB with those in the patent database at EPO via publication number.\",\n",
    "    \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX skos: <http://www.w3.org/2004/02/skos/core#>\\nPREFIX patent: <http://data.epo.org/linked-data/def/patent/>\\n\\nSELECT ?citation ?patent ?application ?applicationNo\\nWHERE\\n{\\n  ?citation a up:Patent_Citation ;\\n    skos:exactMatch ?patent .\\n  FILTER(CONTAINS(STR(?patent), 'EP'))\\n  BIND(SUBSTR(STR(?patent), 35) AS ?applicationNo)\\n  SERVICE<https://data.epo.org/linked-data/query>{\\n    ?application patent:publicationNumber ?applicationNo\\n  }\\n}\",\n",
    "    \"results\": 66,\n",
    "    \"runtime\": 2\n",
    "  },\n",
    "  {\n",
    "    \"question\": \"Connect patents cited in UniProtKB with those in the patent database at EPO via publication number, whose grant date is more than twenty years in the past.\",\n",
    "    \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX skos: <http://www.w3.org/2004/02/skos/core#>\\nPREFIX patent: <http://data.epo.org/linked-data/def/patent/>\\n\\nSELECT ?grantDate ?patent ?application ?applicationNo\\nWHERE\\n{\\n    ?citation a up:Patent_Citation ;\\n  skos:exactMatch ?patent .\\n  BIND(SUBSTR(STR(?patent), 35) AS ?applicationNo)\\n  BIND(SUBSTR(STR(?patent), 33, 2) AS ?countryCode)\\n  SERVICE<https://data.epo.org/linked-data/query>{\\n    ?publication patent:publicationNumber ?applicationNo ;\\n                 patent:application ?application . \\n    ?application patent:grantDate ?grantDate .\\n  }\\n  BIND((year(now()) - 20) AS ?thisYearMinusTwenty)\\n  BIND(year(?grantDate) AS ?grantYear)\\n  FILTER(?grantYear < ?thisYearMinusTwenty)\\n} ORDER BY ?grantYear\",\n",
    "    \"results\": 71,\n",
    "    \"runtime\": 12\n",
    "  },\n",
    "  {\n",
    "    \"question\": \"Select a mapping of UniProt to PDB entries using the UniProt cross-references to the PDB database\",\n",
    "    \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nSELECT ?protein ?db\\nWHERE\\n{\\n    ?protein a up:Protein .\\n    ?protein rdfs:seeAlso ?db .\\n    ?db up:database <http://purl.uniprot.org/database/PDB>\\n}\",\n",
    "    \"results\": 377848,\n",
    "    \"runtime\": 36\n",
    "  },\n",
    "  {\n",
    "    \"question\": \"Find mouse homologs in OMABrowser of human enzymes that catalyze reactions involving Sterols (CHEBI:15889). Federating with Rhea-DB and OMABrowser.\",\n",
    "    \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX taxon: <http://purl.uniprot.org/taxonomy/>\\nPREFIX rh: <http://rdf.rhea-db.org/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX lscr: <http://purl.org/lscr#>\\nPREFIX CHEBI: <http://purl.obolibrary.org/obo/CHEBI_>\\nSELECT\\n  DISTINCT\\n    ?chebi\\n    ?reaction\\n    ?humanProtein\\n    ?mouseProtein\\n    ?cluster \\nWHERE {\\n  SERVICE <https://sparql.rhea-db.org/sparql> {\\n    ?reaction rdfs:subClassOf rh:Reaction .\\n    ?reaction rh:side/rh:contains/rh:compound ?compound .\\n    ?compound rh:chebi ?chebi .\\n    ?chebi rdfs:subClassOf* CHEBI:15889\\n  }\\n\\n  ?humanProtein up:organism taxon:9606 .\\n  ?humanProtein up:annotation ?a .\\n  ?a a up:Catalytic_Activity_Annotation .\\n  ?a up:catalyticActivity ?ca .\\n  ?ca up:catalyzedReaction ?reaction .\\n\\n  SERVICE  <https://sparql.omabrowser.org/sparql> {\\n    ?cluster a orth:ParalogsCluster .\\n    ?cluster orth:hasHomologousMember ?node1 , ?node2 .\\n    ?node1 orth:hasHomologousMember* ?orthoProtein1 .\\n    ?node2 orth:hasHomologousMember* ?orthoProtein2 .\\n    ?orthoProtein1 lscr:xrefUniprot ?mouseProtein .\\n    ?orthoProtein2 lscr:xrefUniprot ?humanProtein .\\n    # inTaxon mouse\\n    ?orthoProtein1 orth:organism/<http://purl.obolibrary.org/obo/RO_0002162> taxon:10090 . \\n  }\\n}\",\n",
    "    \"results\": 31052,\n",
    "    \"runtime\": 320\n",
    "  },\n",
    "  {\n",
    "    \"question\": \"Proteins with annotated binding sites for ligands similar to heme\",\n",
    "    \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX sachem: <http://bioinfo.uochb.cas.cz/rdf/v1.0/sachem#>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nSELECT \\n  ?protein\\n  ?mnemonic\\n  ?proteinName\\n  ?ligandSimilarityScore\\n  ?ligand\\nWHERE {\\n  SERVICE <https://idsm.elixir-czech.cz/sparql/endpoint/chebi> {\\n    ?ssc sachem:compound ?ligand; \\n      sachem:score ?ligandSimilarityScore ;\\n      sachem:similaritySearch ?sss .\\n        # Smiles of Heme\\n    ?sss    sachem:query \\\"CC1=C(CCC([O-])=O)C2=[N+]3C1=Cc1c(C)c(C=C)c4C=C5C(C)=C(C=C)C6=[N+]5[Fe-]3(n14)n1c(=C6)c(C)c(CCC([O-])=O)c1=C2\\\";\\n      sachem:cutoff \\\"8e-1\\\"^^xsd:double ;\\n      sachem:aromaticityMode sachem:aromaticityDetect ;\\n      sachem:similarityRadius 1 ;\\n      sachem:tautomerMode sachem:ignoreTautomers .\\n  }\\n  ?protein up:mnemonic ?mnemonic ;\\n    up:recommendedName/up:fullName ?proteinName ;\\n    up:annotation ?annotation .\\n  ?annotation a up:Binding_Site_Annotation ;\\n      up:ligand/rdfs:subClassOf ?ligand .\\n}\\nORDER BY DESC(?ligandSimilarityScore)\",\n",
    "    \"results\": 967082,\n",
    "    \"runtime\": 181\n",
    "  },\n",
    "  {\n",
    "    \"question\": \"Number of proteins with annotated binding sites for metals or metal sulfur clusters (and experimental evidence for the binding)\",\n",
    "    \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nPREFIX CHEBI: <http://purl.obolibrary.org/obo/CHEBI_>\\nSELECT\\n  ?ligand\\n  ?ligandName \\n  (COUNT(DISTINCT ?protein) as ?entries)\\nWHERE {\\n   ?protein up:annotation ?annotation .\\n   \\n   VALUES ?evs { obo:ECO_0000269 obo:ECO_0007744 } .\\n   VALUES ?chebids { CHEBI:25213 CHEBI:25214 } .\\n   ?st rdf:subject ?protein ; \\n       rdf:predicate up:annotation ; \\n       rdf:object ?annotation ;\\n       up:attribution/up:evidence ?evs .\\n\\n   ?annotation up:ligand/rdfs:subClassOf ?ligand .\\n   ?ligand rdfs:subClassOf+ ?chebids ;\\n     rdfs:label ?ligandName .\\n}\\nGROUP BY ?ligand ?ligandName\\nORDER BY DESC(?entries)\",\n",
    "    \"results\": 23,\n",
    "    \"runtime\": 16\n",
    "  },\n",
    "  {\n",
    "    \"question\": \"Select Enzymes that have Ligands that have a known allosteric effect\",\n",
    "    \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nSELECT\\n  ?protein\\n  ?ligandName \\n  ?ligandNote\\n  ?chebi\\nWHERE {\\n   ?protein up:annotation ?annotation .\\n   ?annotation a up:Binding_Site_Annotation . \\n   ?annotation up:ligand ?ligand .\\n   ?ligand rdfs:comment ?ligandNote ;\\n     rdfs:subClassOf ?chebi ;\\n     rdfs:label ?ligandName .\\n   FILTER(REGEX(?ligandNote, \\\"allosteric\\\", \\\"i\\\"))\\n}\\n\",\n",
    "    \"results\": 730,\n",
    "    \"runtime\": 22\n",
    "  },\n",
    "  {\n",
    "    \"question\": \"Map PDB identifiers plus chains to UniProt\",\n",
    "    \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nSELECT\\n  ?pdbId ?chain ?pdbChain ?uniprot\\nWHERE\\n{\\n  # A space separated list of pairs of PDB identifiers and the chain code.\\n  VALUES(?pdbId ?pdbChain) { ('6VXC' 'A') ('1BG3' 'B') }\\n\\n  # Make an IRI out of the pdbId\\n  BIND(iri(concat('http://rdf.wwpdb.org/pdb/', ?pdbId)) AS ?pdb)\\n\\n  # Map to UniProt entries\\n  ?uniprot rdfs:seeAlso ?pdb .\\n  ?pdb up:database <http://purl.uniprot.org/database/PDB> ;\\n       up:chainSequenceMapping ?chainSm .\\n  ?chainSm up:chain ?chainsPlusRange .\\n\\n  # Extract the list of chains from the text representation.\\n  BIND(STRBEFORE(?chainsPlusRange, '=') AS ?chain)\\n\\n  # Filter those that match.\\n  FILTER(CONTAINS(?chain, ?pdbChain))\\n}\",\n",
    "    \"results\": 5,\n",
    "    \"runtime\": 0\n",
    "  },\n",
    "  {\n",
    "    \"question\": \"Map UniProt to HGNC identifiers and Symbols\",\n",
    "    \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX uniprotkb: <http://purl.uniprot.org/uniprot/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nSELECT\\n  ?uniprot\\n  ?hgnc\\n  ?hgncSymbol\\nWHERE\\n{\\n  # A space separated list of UniProt primary accessions.\\n  VALUES (?acc) {('P05067') ('P00750')}\\n  BIND(iri(concat(str(uniprotkb:), ?acc)) AS ?uniprot)\\n  ?uniprot rdfs:seeAlso ?hgnc .\\n  ?hgnc up:database <http://purl.uniprot.org/database/HGNC> ;\\n       rdfs:comment ?hgncSymbol .\\n}\",\n",
    "    \"results\": 2,\n",
    "    \"runtime\": 0\n",
    "  },\n",
    "  {\n",
    "    \"question\": \"Select all cross-references to external databases of the category '3D structure databases' of UniProt entries that are classified with the keyword 'Acetoin biosynthesis (KW-0005)'\",\n",
    "    \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX keywords: <http://purl.uniprot.org/keywords/>\\nSELECT DISTINCT ?link\\nWHERE\\n{\\n    ?protein a up:Protein . \\n    ?protein up:classifiedWith keywords:5 .\\n    ?protein rdfs:seeAlso ?link .\\n    ?link up:database ?db .\\n    ?db up:category '3D structure databases'\\n}\",\n",
    "    \"results\": 4303,\n",
    "    \"runtime\": 2\n",
    "  },\n",
    "  {\n",
    "    \"question\": \"Find Human proteins that catalyze reactions where substrates or product have a Cholestane skeleton\",\n",
    "    \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX taxon: <http://purl.uniprot.org/taxonomy/>\\nPREFIX sachem: <http://bioinfo.uochb.cas.cz/rdf/v1.0/sachem#>\\nPREFIX rh: <http://rdf.rhea-db.org/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nSELECT\\n  DISTINCT\\n    ?protein\\n    ?disease\\n    ?rhea\\n    ?chebi\\n    ?omim\\nWHERE {\\n    # Find complete ChEBIs with a Cholestane skeleton, via the Czech Elixir node IDSM Sachem chemical substructure search.\\n    SERVICE <https://idsm.elixir-czech.cz/sparql/endpoint/chebi> {\\n      ?chebi sachem:substructureSearch [\\n        sachem:query\\n\\\"[C@]12(CCC3CCCC[C@]3(C)[C@@]1([H])CC[C@]1(C)[C@@]([H])([C@@](C)([H])CCCC(C)C)CC[C@@]21[H])[H]\\\"\\n].\\n   }\\n   # Use the fact that UniProt catalytic activities are annotated using Rhea \\n   # Mapping the found ChEBIs to Rhea reactions\\n   SERVICE <https://sparql.rhea-db.org/sparql>{\\n     ?rhea rh:side/rh:contains/rh:compound/rdfs:subClassOf ?chebi .\\n   }\\n   # Match the found Rhea reactions with human UniProtKB proteins\\n   ?protein up:annotation/up:catalyticActivity/up:catalyzedReaction ?rhea .\\n   ?protein up:organism taxon:9606 .\\n   # Find only those human entries that have an annotated related disease, and optionaly map these to OMIM\\n   ?protein up:annotation/up:disease ?disease .\\n   OPTIONAL {\\n     ?disease rdfs:seeAlso ?omim .\\n     ?omim up:database <http://purl.uniprot.org/database/MIM>\\n   }\\n}\",\n",
    "    \"results\": 27,\n",
    "    \"runtime\": 19\n",
    "  },\n",
    "  {\n",
    "    \"question\": \"Select the Gene Protein Reaction sets for Human (Ensembl Gene, Human UniProtKB, Catalyzed Rhea reactions)\",\n",
    "    \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX taxon: <http://purl.uniprot.org/taxonomy/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nSELECT\\n?ensemblGene ?protein ?rhea\\nWHERE {\\n  ?protein up:reviewed true ;\\n           up:organism taxon:9606 .\\n  ?protein up:annotation ?caa ;\\n           rdfs:seeAlso ?ensemblTranscript .\\n  ?ensemblTranscript up:database <http://purl.uniprot.org/database/Ensembl> .\\n\\t?caa up:catalyticActivity ?ca .\\n  ?ca up:catalyzedReaction ?rhea .\\n  ?ensemblTranscript up:transcribedFrom ?ensemblGene \\n}\",\n",
    "    \"results\": 122340,\n",
    "    \"runtime\": 15\n",
    "  },\n",
    "  {\n",
    "    \"question\": \"Select reviewed UniProt entries (Swiss-Prot), and their recommended protein name, that have a preferred gene name that contains the text 'DNA'\",\n",
    "    \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX skos: <http://www.w3.org/2004/02/skos/core#>\\nSELECT ?protein ?name\\nWHERE\\n{\\n        ?protein a up:Protein .\\n        ?protein up:reviewed true .\\n        ?protein up:recommendedName ?recommended .\\n        ?recommended up:fullName ?name .\\n        ?protein up:encodedBy ?gene .\\n        ?gene skos:prefLabel ?text .\\n        FILTER CONTAINS(?text, 'DNA')\\n}\\n\",\n",
    "    \"results\": 876,\n",
    "    \"runtime\": 98\n",
    "  },\n",
    "  {\n",
    "    \"question\": \"Select the preferred gene name and disease annotation of all human UniProt entries that are known to be involved in a disease\",\n",
    "    \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX taxon: <http://purl.uniprot.org/taxonomy/>\\nPREFIX skos: <http://www.w3.org/2004/02/skos/core#>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nSELECT ?name ?text\\nWHERE\\n{\\n        ?protein a up:Protein . \\n        ?protein up:organism taxon:9606 .\\n        ?protein up:encodedBy ?gene . \\n        ?gene skos:prefLabel ?name .\\n        ?protein up:annotation ?annotation .\\n        ?annotation a up:Disease_Annotation .\\n        ?annotation rdfs:comment ?text\\n}\",\n",
    "    \"results\": 31652,\n",
    "    \"runtime\": 2\n",
    "  },\n",
    "      {\n",
    "        \"question\": \"Find all proteins linked to arachidonate (CHEBI:32395)\",\n",
    "        \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "        \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX skos: <http://www.w3.org/2004/02/skos/core#>\\nPREFIX rh: <http://rdf.rhea-db.org/>\\nPREFIX CHEBI: <http://purl.obolibrary.org/obo/CHEBI_>\\n\\nSELECT \\n    DISTINCT\\n        ?uniprot\\n        ?uniprotID\\n        ?recname\\n        ?gene\\n        ?chebi\\n        ?uniprotName\\nWHERE {\\n  SERVICE <https://sparql.rhea-db.org/sparql> {\\n     VALUES (?chebi) { (CHEBI:32395) }\\n     ?rhea rh:side/rh:contains/rh:compound ?compound .\\n     ?compound rh:chebi ?chebi .\\n     ?chebi up:name ?uniprotName .\\n  }\\n  ?uniprot up:annotation/up:catalyticActivity/up:catalyzedReaction ?rhea .\\n  ?uniprot up:mnemonic ?uniprotID .\\n  ?uniprot up:recommendedName/up:fullName ?recname .\\n  OPTIONAL {?uniprot up:encodedBy/skos:prefLabel ?gene .}\\n}\",\n",
    "        \"results\": 23346,\n",
    "        \"runtime\": 30\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"What are the species present in Bgee?\",\n",
    "        \"endpoint\": \"https://www.bgee.org/sparql/\",\n",
    "        \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\n\\nSELECT ?species WHERE {\\n    ?species a up:Taxon .\\n}\\n\",\n",
    "        \"results\": 52,\n",
    "        \"runtime\": 0\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"What are the anatomical entities where the P02654 gene is expressed? Note that P02654 is a UniProtKB identifier of the APOC1 human gene.\",\n",
    "        \"endpoint\": \"https://www.bgee.org/sparql/\",\n",
    "        \"query\": \"PREFIX uniprotkb: <http://purl.uniprot.org/uniprot/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX lscr: <http://purl.org/lscr#>\\nPREFIX genex: <http://purl.org/genex#>\\n\\nSELECT DISTINCT ?anat ?anatName WHERE {\\n    ?seq a orth:Gene ;\\n        genex:isExpressedIn ?anat ;\\n        lscr:xrefUniprot uniprotkb:P02654 .\\n    ?anat a genex:AnatomicalEntity ;\\n        rdfs:label ?anatName .\\n}\\n\",\n",
    "        \"results\": 526,\n",
    "        \"runtime\": 0\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"What is all the metadata related to the ENSG00000130208 gene, where ENSG00000130208 is the identifier of the APOC1 human gene.\",\n",
    "        \"endpoint\": \"https://www.bgee.org/sparql/\",\n",
    "        \"query\": \"PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX lscr: <http://purl.org/lscr#>\\nPREFIX dcterms: <http://purl.org/dc/terms/>\\n\\nSELECT DISTINCT ?symbol ?description ?id\\n?links ?organism ?uniprot ?ensembl ?ncbi WHERE {\\n    ?seq a orth:Gene ;\\n        rdfs:label ?symbol ;\\n        rdfs:seeAlso ?links ;\\n        dcterms:description ?description ;\\n        dcterms:identifier ?id ;\\n        orth:organism ?organism .\\n    OPTIONAL{?seq lscr:xrefUniprot ?uniprot .}\\n    OPTIONAL{?seq lscr:xrefEnsemblGene ?ensembl .}\\n    OPTIONAL{?seq lscr:xrefNCBIGene ?ncbi .}\\n    FILTER (?id = \\\"ENSG00000130208\\\")\\n}\\n\",\n",
    "        \"results\": 16,\n",
    "        \"runtime\": 1\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"What are the anatomical entities where the APOC1 Homo sapiens gene is not expressed, that is where is APOC1 absent?\",\n",
    "        \"endpoint\": \"https://www.bgee.org/sparql/\",\n",
    "        \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nPREFIX genex: <http://purl.org/genex#>\\n\\nSELECT DISTINCT ?anat ?anatName WHERE {\\n    ?seq a orth:Gene ;\\n        rdfs:label \\\"APOC1\\\" ;\\n        genex:isAbsentIn ?anat ;\\n        orth:organism ?organism .\\n    ?anat a genex:AnatomicalEntity ;\\n        rdfs:label ?anatName .\\n    ?organism obo:RO_0002162  ?species .\\n    ?species a up:Taxon ;\\n        up:scientificName \\\"Homo sapiens\\\" .\\n}\\n\",\n",
    "        \"results\": 6,\n",
    "        \"runtime\": 0\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"What are the species present in Bgee and their scientific and common names?\",\n",
    "        \"endpoint\": \"https://www.bgee.org/sparql/\",\n",
    "        \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\n\\nSELECT ?species ?sci_name ?common_name WHERE {\\n    ?species a up:Taxon ;\\n        up:scientificName ?sci_name ;\\n        up:rank up:Species .\\n    OPTIONAL { ?species up:commonName ?common_name . }\\n}\\n\",\n",
    "        \"results\": 52,\n",
    "        \"runtime\": 0\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"What are the anatomical entities where the APOC1 gene is expressed?\",\n",
    "        \"endpoint\": \"https://www.bgee.org/sparql/\",\n",
    "        \"query\": \"PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX genex: <http://purl.org/genex#>\\n\\nSELECT DISTINCT ?anat ?anatName WHERE {\\n    ?seq a orth:Gene ;\\n        genex:isExpressedIn ?anat ;\\n        rdfs:label \\\"APOC1\\\" .\\n    ?anat a genex:AnatomicalEntity ;\\n        rdfs:label ?anatName .\\n}\\n\",\n",
    "        \"results\": 584,\n",
    "        \"runtime\": 0\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"What are the anatomical entities where the APOC1 Homo sapiens gene is expressed?\",\n",
    "        \"endpoint\": \"https://www.bgee.org/sparql/\",\n",
    "        \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nPREFIX genex: <http://purl.org/genex#>\\n\\nSELECT DISTINCT ?anat ?anatName WHERE {\\n    ?seq a orth:Gene ;\\n        genex:isExpressedIn ?anat ;\\n        rdfs:label \\\"APOC1\\\" .\\n    ?anat a genex:AnatomicalEntity ;\\n        rdfs:label ?anatName .\\n    ?seq orth:organism ?organism .\\n    ?organism obo:RO_0002162  ?species .\\n    ?species a up:Taxon ;\\n        up:scientificName \\\"Homo sapiens\\\" .\\n}\\n\",\n",
    "        \"results\": 526,\n",
    "        \"runtime\": 0\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"What are the anatomical entities where the human gene APOC1 is expressed in the post-juvenile stage along with its expression score independently of the strain, sex, and cell type?\",\n",
    "        \"endpoint\": \"https://www.bgee.org/sparql/\",\n",
    "        \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nPREFIX genex: <http://purl.org/genex#>\\n\\nSELECT DISTINCT ?anat ?anatName ?score ?stage WHERE {\\n    ?seq a orth:Gene ;\\n        rdfs:label \\\"APOC1\\\" ;\\n        orth:organism ?organism .\\n    ?expression a genex:Expression ;\\n        genex:hasExpressionCondition ?condition ;\\n        genex:hasExpressionLevel ?score ;\\n        genex:hasSequenceUnit ?seq .\\n    ?condition a genex:ExpressionCondition ;\\n        genex:hasAnatomicalEntity ?anat ;\\n        genex:hasAnatomicalEntity obo:GO_0005575 ;\\n        genex:hasDevelopmentalStage ?stage ;\\n        genex:hasSex \\\"any\\\" ;\\n        genex:hasStrain ?strain .\\n    ?anat a genex:AnatomicalEntity ;\\n        rdfs:label ?anatName .\\n    ?stage rdfs:label \\\"post-juvenile\\\" .\\n    ?strain rdfs:label \\\"wild-type\\\" .\\n    ?organism obo:RO_0002162  ?species .\\n    ?species a up:Taxon ;\\n        up:commonName \\\"human\\\" .\\nFILTER (?anat !=  obo:GO_0005575)\\n} ORDER BY DESC(?score)\\n\",\n",
    "        \"results\": 456,\n",
    "        \"runtime\": 60\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"Find all Rattus norvegicus' proteins present in OMA RDF database.\",\n",
    "        \"endpoint\": \"https://sparql.omabrowser.org/sparql/\",\n",
    "        \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX orth: <http://purl.org/net/orth#>\\nSELECT ?protein ?OMA_link\\nWHERE\\n{\\n    ?protein a orth:Protein.\\n    ?protein  orth:organism ?organism.\\n    ?inTaxon rdfs:label 'in taxon'@en.\\n    ?organism ?inTaxon ?taxon.\\n    ?taxon  up:scientificName 'Rattus norvegicus'.\\n    ?protein rdfs:seeAlso ?OMA_link.\\n}\",\n",
    "        \"results\": 22232,\n",
    "        \"runtime\": 3\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"Retrieve all proteins belongong to the Hierarchical Orthologous Group (HOG) at the level 'Vertebrata' to which humans' CDIN1 gene belong, together with their gene name symbol if available.\",\n",
    "        \"endpoint\": \"https://sparql.omabrowser.org/sparql/\",\n",
    "        \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nselect distinct ?HOG ?MEMBER ?GENE_LABEL\\nwhere {\\n    ?HOG a orth:OrthologsCluster ;\\n      orth:hasHomologousMember ?node1 ;\\n      orth:hasTaxonomicRange ?taxRange .\\n    ?taxRange orth:taxRange 'Vertebrata' .\\n    ?node1 orth:hasHomologousMember* ?query ;\\n      orth:hasHomologousMember* ?MEMBER .\\n    ?MEMBER a orth:Protein .\\n    OPTIONAL {\\n        ?MEMBER rdfs:label ?GENE_LABEL .\\n    }\\n    ?query a orth:Protein ;\\n      orth:organism/obo:RO_0002162/up:scientificName 'Homo sapiens';\\n      rdfs:label 'CDIN1'.\\n}\",\n",
    "        \"results\": 135,\n",
    "        \"runtime\": 0\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"Which species are available on OMA database and their scientific names?\",\n",
    "        \"endpoint\": \"https://sparql.omabrowser.org/sparql/\",\n",
    "        \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nSELECT ?species ?sciname WHERE\\n{\\n    ?species a up:Taxon.\\n    ?species up:scientificName ?sciname.\\n    ?species up:rank up:Species.\\n}\",\n",
    "        \"results\": 2851,\n",
    "        \"runtime\": 0\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"Retrieve all proteins in OMA that is encoded by the INS gene and their mnemonics and evidence types from Uniprot database (federated query).\",\n",
    "        \"endpoint\": \"https://sparql.omabrowser.org/sparql/\",\n",
    "        \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nPREFIX lscr: <http://purl.org/lscr#>\\nSELECT DISTINCT ?proteinOMA ?species  ?mnemonic  ?evidenceType ?UniProt_URI\\nWHERE {\\n    ?proteinOMA a orth:Protein;\\n    orth:organism/obo:RO_0002162/up:scientificName ?species;\\n    rdfs:label 'INS'.\\n    ?proteinOMA lscr:xrefUniprot ?UniProt_URI.\\n    #Search the INS gene mnemonics and evidence types from Uniprot database.\\n    service <http://sparql.uniprot.org/sparql> { \\n    ?UniProt_URI up:mnemonic ?mnemonic;\\n    up:existence/rdfs:label ?evidenceType. }\\n}\",\n",
    "        \"results\": 76,\n",
    "        \"runtime\": 12\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"Retrieve all genes that are orthologous to ENSLACG00000002497 Ensembl gene (identifier)\",\n",
    "        \"endpoint\": \"https://sparql.omabrowser.org/sparql/\",\n",
    "        \"query\": \"PREFIX sio: <http://semanticscience.org/resource/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX lscr: <http://purl.org/lscr#>\\nPREFIX ensembl: <http://rdf.ebi.ac.uk/resource/ensembl/>\\nselect ?protein2 ?OMA_LINK \\nwhere {\\n    #The three that contains Orthologs. The leafs are proteins.\\n    #This graph pattern defines the relationship protein1 is Orthologs to protein2\\n    ?cluster a orth:OrthologsCluster.\\n    ?cluster orth:hasHomologousMember ?node1.\\n    ?cluster orth:hasHomologousMember ?node2. \\n    ?node2 orth:hasHomologousMember* ?protein2. \\n    ?node1 orth:hasHomologousMember* ?protein1.\\n    ########\\n     \\n    #Specify the protein to look for its orthologs\\n    ?protein1 sio:SIO_010079/lscr:xrefEnsemblGene  ensembl:ENSLACG00000002497.\\n    ########\\n     \\n    #The OMA link to the second protein\\n    ?protein2 rdfs:seeAlso ?OMA_LINK. \\n    ########\\n     \\n    filter(?node1 != ?node2) \\n}\",\n",
    "        \"results\": 42,\n",
    "        \"runtime\": 0\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"Retrieve all genes that are paralogous to ENSG00000244734 Ensembl gene (identifier).\",\n",
    "        \"endpoint\": \"https://sparql.omabrowser.org/sparql/\",\n",
    "        \"query\": \"PREFIX sio: <http://semanticscience.org/resource/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX lscr: <http://purl.org/lscr#>\\nPREFIX ensembl: <http://rdf.ebi.ac.uk/resource/ensembl/>\\nselect ?protein2 ?OMA_LINK \\nwhere {\\n    #The three that contains paralogs. The leafs are proteins.\\n    #This graph pattern defines the relationship protein1 is paralogous to protein2\\n    ?cluster a orth:ParalogsCluster.\\n    ?cluster orth:hasHomologousMember ?node1.\\n    ?cluster orth:hasHomologousMember ?node2.\\n    ?node2 orth:hasHomologousMember* ?protein2.\\n    ?node1 orth:hasHomologousMember* ?protein1.\\n    ########\\n\\n    #Specify the protein to look for its paralogs\\n    ?protein1 sio:SIO_010079/lscr:xrefEnsemblGene  ensembl:ENSG00000244734.\\n    ########\\n\\n    #The OMA link to the second protein\\n    ?protein2 rdfs:seeAlso ?OMA_LINK. \\n    ########\\n     \\n    filter(?node1 != ?node2) \\n}\",\n",
    "        \"results\": 473,\n",
    "        \"runtime\": 2\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"Retrieve all genes that are paralogous to HUMAN00529 OMA protein (identifier) and their cross-reference links to OMA and Uniprot.\",\n",
    "        \"endpoint\": \"https://sparql.omabrowser.org/sparql/\",\n",
    "        \"query\": \"PREFIX orth: <http://purl.org/net/orth#>\\nPREFIX lscr: <http://purl.org/lscr#>\\nPREFIX dc: <http://purl.org/dc/terms/>\\nselect ?protein2 ?Uniprot_link\\nwhere {\\n    ?cluster a orth:ParalogsCluster.\\n    ?cluster orth:hasHomologousMember ?node1.\\n    ?cluster orth:hasHomologousMember ?node2.\\n    ?node2 orth:hasHomologousMember* ?protein2.\\n    ?node1 orth:hasHomologousMember* ?protein1.\\n    ?protein1 a orth:Protein.\\n    ?protein1 dc:identifier 'HUMAN00529'.\\n    ?protein2 a orth:Protein. \\n    ?protein2  lscr:xrefUniprot ?Uniprot_link. \\n    filter(?node1 != ?node2)\\n}\",\n",
    "        \"results\": 979,\n",
    "        \"runtime\": 0\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"Retrieve all genes that are orthologous to HUMAN22169 OMA protein (identifier) and their cross-reference links to OMA and Uniprot.\",\n",
    "        \"endpoint\": \"https://sparql.omabrowser.org/sparql/\",\n",
    "        \"query\": \"PREFIX orth: <http://purl.org/net/orth#>\\nPREFIX lscr: <http://purl.org/lscr#>\\nPREFIX dc: <http://purl.org/dc/terms/>\\nselect ?protein2 ?Uniprot_link\\nwhere {\\n    ?cluster a orth:OrthologsCluster.\\n    ?cluster orth:hasHomologousMember ?node1.\\n    ?cluster orth:hasHomologousMember ?node2.\\n    ?node2 orth:hasHomologousMember* ?protein2.\\n    ?node1 orth:hasHomologousMember* ?protein1.\\n    ?protein1 a orth:Protein.\\n    ?protein1 dc:identifier 'HUMAN22169'.\\n    ?protein2 a orth:Protein. \\n    ?protein2  lscr:xrefUniprot ?Uniprot_link. \\n    filter(?node1 != ?node2)\\n}\",\n",
    "        \"results\": 161,\n",
    "        \"runtime\": 0\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"Retrieve all genes per species that are orthologous to Rabbit's APOCI or APOC1 gene and their cross-reference links to OMA and Uniprot including the corresponding Ensembl gene identifier.\",\n",
    "        \"endpoint\": \"https://sparql.omabrowser.org/sparql/\",\n",
    "        \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX sio: <http://semanticscience.org/resource/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nPREFIX lscr: <http://purl.org/lscr#>\\nPREFIX dc: <http://purl.org/dc/terms/>\\nselect ?protein1    ?protein2  ?geneName2  ?species2 ?Prot2_uniprot ?prot2_ensemblGeneId  \\nwhere {\\n    ?cluster a orth:OrthologsCluster.\\n    ?cluster orth:hasHomologousMember ?node1.\\n    ?cluster orth:hasHomologousMember ?node2. \\n    ?node2 orth:hasHomologousMember* ?protein2. \\n    ?node1 orth:hasHomologousMember* ?protein1.\\n    ?protein1 a orth:Protein;\\n    orth:organism/obo:RO_0002162/up:scientificName 'Oryctolagus cuniculus';\\n    rdfs:label 'APOCI'.\\n    ?protein2 a orth:Protein; \\n    lscr:xrefUniprot ?Prot2_uniprot;\\n    sio:SIO_010079/lscr:xrefEnsemblGene/dc:identifier  ?prot2_ensemblGeneId;\\n    rdfs:label ?geneName2;\\n    orth:organism/obo:RO_0002162/up:scientificName ?species2.\\n    filter(?node1 != ?node2)\\n}\",\n",
    "        \"results\": 56,\n",
    "        \"runtime\": 1\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"Retrieve all Rabbit's proteins encoded by genes that are orthologous to Mouses's hemoglobin Y gene and their cross-reference links to Uniprot.\",\n",
    "        \"endpoint\": \"https://sparql.omabrowser.org/sparql/\",\n",
    "        \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX orth: <http://purl.org/net/orth#>\\nPREFIX obo: <http://purl.obolibrary.org/obo/>\\nPREFIX lscr: <http://purl.org/lscr#>\\nselect distinct ?MOUSE_PROTEIN ?RABIT_PROTEIN ?MOUSE_UNIPROT_XREF ?RABIT_UNIPROT_XREF \\nwhere {\\n    ?cluster a orth:OrthologsCluster.\\n    ?cluster orth:hasHomologousMember ?node1.\\n    ?cluster orth:hasHomologousMember ?node2. \\n    ?node2 orth:hasHomologousMember* ?RABIT_PROTEIN. \\n    ?node1 orth:hasHomologousMember* ?MOUSE_PROTEIN.\\n    ?MOUSE_PROTEIN a orth:Protein.\\n    ?MOUSE_PROTEIN  orth:organism/obo:RO_0002162/up:scientificName 'Mus musculus';\\n    rdfs:label 'HBB-Y';\\n    lscr:xrefUniprot ?MOUSE_UNIPROT_XREF.\\n    ?RABIT_PROTEIN a orth:Protein.\\n    ?RABIT_PROTEIN orth:organism/obo:RO_0002162/up:scientificName 'Oryctolagus cuniculus' .\\n    ?RABIT_PROTEIN lscr:xrefUniprot ?RABIT_UNIPROT_XREF.\\n    filter(?node1 != ?node2)\\n}\",\n",
    "        \"results\": 24,\n",
    "        \"runtime\": 0\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"Select all Rhea reactions that have a given ChEBI ID as reaction participant\",\n",
    "        \"endpoint\": \"https://sparql.rhea-db.org/sparql/\",\n",
    "        \"query\": \"PREFIX rh: <http://rdf.rhea-db.org/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX CHEBI: <http://purl.obolibrary.org/obo/CHEBI_>\\n# Query 14\\n# Select all Rhea reactions that have CHEBI:29985 (L-glutamate) as reaction participant\\n# \\n# This query corresponds to the Rhea website query:\\n# https://www.rhea-db.org/rhea?query=chebi:29985\\nSELECT distinct ?chebi ?rhea ?equation\\nWHERE {\\n  ?rhea rdfs:subClassOf rh:Reaction .\\n  ?rhea rh:equation ?equation .\\n  ?rhea rh:side/rh:contains/rh:compound ?compound .\\n  #\\n  # the ChEBI can be used either as a small molecule, the reactive part of a macromolecule or as a polymer.\\n  #\\n  ?compound (rh:chebi|(rh:reactivePart/rh:chebi)|(rh:underlyingChebi/rh:chebi)) ?chebi .\\n  VALUES (?chebi) { (CHEBI:29985) }\\n}\",\n",
    "        \"results\": 244,\n",
    "        \"runtime\": 0\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"Select all Rhea reactions mapped to KEGG reactions\",\n",
    "        \"endpoint\": \"https://sparql.rhea-db.org/sparql/\",\n",
    "        \"query\": \"PREFIX rh: <http://rdf.rhea-db.org/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\n# Query 5 \\n# Select all Rhea reactions mapped to KEGG reactions\\n# KEGG reactions are mapped to Rhea bidirectional reactions\\n# Rhea web query: https://www.rhea-db.org/rhea?query=kegg:*\\n#\\nSELECT ?rhea ?kegg ?rheaDir \\nWHERE {\\n  ?rhea rdfs:subClassOf rh:Reaction .\\n  ?rhea rh:bidirectionalReaction ?rheaDir .\\n  ?rheaDir rdfs:seeAlso ?kegg .\\n  FILTER (regex(str(?kegg),'kegg'))\\n}\",\n",
    "        \"results\": 6898,\n",
    "        \"runtime\": 0\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"Select all ChEBI compounds used in Rhea as reaction participant\",\n",
    "        \"endpoint\": \"https://sparql.rhea-db.org/sparql/\",\n",
    "        \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rh: <http://rdf.rhea-db.org/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\n# Query 15\\n# Select all ChEBI compounds used in Rhea as reaction participant\\n# \\n# This query can not be expressed in the Rhea website\\nSELECT ?chebi ?name (count(?rhea) as ?countRhea)\\nWHERE {\\n  ?rhea rdfs:subClassOf rh:Reaction .\\n  ?rhea rh:side/rh:contains/rh:compound ?compound .\\n  #\\n  # the ChEBI can be used either as a small molecule, the reactive part of a macromolecule or as a polymer.\\n  #\\n  ?compound (rh:chebi|(rh:reactivePart/rh:chebi)|(rh:underlyingChebi/rh:chebi)) ?chebi .\\n  ?chebi up:name ?name .\\n}\\nGROUP BY ?chebi ?name\\nORDER BY DESC(?countRhea)\",\n",
    "        \"results\": 12594,\n",
    "        \"runtime\": 6\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"Select all Rhea reactions mapped to MetaCyc reactions\",\n",
    "        \"endpoint\": \"https://sparql.rhea-db.org/sparql/\",\n",
    "        \"query\": \"PREFIX rh: <http://rdf.rhea-db.org/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\n# Query 6 \\n# Select all Rhea reactions mapped to MetaCyc reactions\\n# MetaCyc reactions are mapped to either undirected, left-to-right, right-to-left or bidirectional reactions\\n# Rhea web query: https://www.rhea-db.org/rhea?query=metacyc:*\\n#\\nSELECT distinct ?rhea ?rheaDir ?metacyc\\nWHERE {\\n  ?rhea rdfs:subClassOf rh:Reaction .\\n  {\\n    ?rhea rdfs:seeAlso ?metacyc .\\n    FILTER CONTAINS(str(?metacyc), \\\"METACYC\\\") \\n    BIND(?rhea as ?rheaDir)\\n  }\\n  UNION\\n  {\\n    ?rhea rh:directionalReaction ?directionalReaction .\\n    ?directionalReaction rdfs:seeAlso ?metacyc . \\n    FILTER CONTAINS(str(?metacyc), \\\"METACYC\\\") \\n    BIND(?directionalReaction as ?rheaDir  )\\n  }\\n  UNION\\n  {\\n    ?rhea rh:bidirectionalReaction ?bidirectionalReaction .\\n    ?bidirectionalReaction rdfs:seeAlso ?metacyc . \\n    FILTER CONTAINS(str(?metacyc), \\\"METACYC\\\") \\n    BIND(?bidirectionalReaction as ?rheaDir  )\\n  }\\n}\",\n",
    "        \"results\": 9293,\n",
    "        \"runtime\": 3\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"Generate a draft human metabolome\",\n",
    "        \"endpoint\": \"https://sparql.rhea-db.org/sparql/\",\n",
    "        \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX taxon: <http://purl.uniprot.org/taxonomy/>\\nPREFIX rh: <http://rdf.rhea-db.org/>\\nPREFIX chebislash: <http://purl.obolibrary.org/obo/chebi/>\\nSELECT\\n  ?uniprot ?mnemonic ?rhea ?chebi ?smiles ?inchiKey\\nWHERE\\n{\\n  ?rhea rh:side/rh:contains/rh:compound ?compound .\\n  ?compound (rh:chebi|(rh:reactivePart/rh:chebi)|(rh:underlyingChebi/rh:chebi)) ?chebi .\\n  ?chebi chebislash:smiles ?smiles ;\\n          chebislash:inchikey ?inchiKey .\\n  SERVICE <https://sparql.uniprot.org/sparql/> {\\n     ?uniprot up:annotation/up:catalyticActivity/up:catalyzedReaction ?rhea ;\\n                                             up:organism taxon:9606 ;\\n                                             up:mnemonic ?mnemonic .\\n\\n   }\\n}\",\n",
    "        \"results\": 279926,\n",
    "        \"runtime\": 131\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"Select all Rhea reactions linked to an enzyme classification (sub)-class\",\n",
    "        \"endpoint\": \"https://sparql.rhea-db.org/sparql/\",\n",
    "        \"query\": \"PREFIX rh: <http://rdf.rhea-db.org/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX ec: <http://purl.uniprot.org/enzyme/>\\n# Query 10\\n# Select all Rhea reactions mapped to \\n# \\n# This query corresponds to the Rhea website query:\\n# https://www.rhea-db.org/rhea?query=ec:1.*\\n# https://www.rhea-db.org/rhea?query=ec:1.1.*\\n# https://www.rhea-db.org/rhea?query=ec:1.1.1.*\\n#\\nSELECT ?ec ?ecNumber ?rhea ?accession ?equation\\nWHERE {\\n  ?rhea rdfs:subClassOf rh:Reaction .\\n  ?rhea rh:accession ?accession .\\n  ?rhea rh:ec ?ec .\\n  BIND(strafter(str(?ec),str(ec:)) as ?ecNumber)\\n  # class (e.g EC 1.-.-.-)\\n  FILTER (regex(?ecNumber,'^1\\\\\\\\.')) \\n  # sub-class  (e.g EC 1.1.-.-)\\n  #FILTER (regex(?ecNumber,'^1\\\\\\\\.1\\\\\\\\.')) \\n  # sub-sub-class (e.g EC 1.1.1.-)\\n  #FILTER (regex(?ecNumber,'^1\\\\\\\\.1\\\\\\\\.1\\\\\\\\.')) \\n  ?rhea rh:equation ?equation .\\n}\",\n",
    "        \"results\": 2658,\n",
    "        \"runtime\": 0\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"Select all Rhea reactions\",\n",
    "        \"endpoint\": \"https://sparql.rhea-db.org/sparql/\",\n",
    "        \"query\": \"PREFIX rh: <http://rdf.rhea-db.org/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\n# Query 1\\n# Select all Rhea reactions (unspecified direction) \\n# and return identifier (id), accession, boolean attributes (isChemicallyBalanced, isTransport) and chemical equation.\\n#\\nSELECT ?rhea ?id ?accession ?isChemicallyBalanced ?isTransport  ?equation \\nWHERE {\\n  ?rhea rdfs:subClassOf rh:Reaction .\\n  ?rhea rh:id ?id .\\n  ?rhea rh:accession ?accession .\\n  ?rhea rh:equation ?equation .\\n  ?rhea rh:isTransport ?isTransport .\\n  ?rhea rh:isChemicallyBalanced ?isChemicallyBalanced .\\n}\",\n",
    "        \"results\": 16609,\n",
    "        \"runtime\": 2\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"Select all Rhea reactions that have a pair of ChEBI IDs as reaction participant and in opposite side\",\n",
    "        \"endpoint\": \"https://sparql.rhea-db.org/sparql/\",\n",
    "        \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rh: <http://rdf.rhea-db.org/>\\nPREFIX CHEBI: <http://purl.obolibrary.org/obo/CHEBI_>\\n# Query 16\\n# Select all Rhea reactions that have a pair of ChEBI IDs as reaction participant and in opposite side\\n# Return Rhea reactions that have CHEBI:29985 (L-glutamate) as reaction participant in one side\\n# and CHEBI:58359 (L-glutamine) in the other side\\n#\\n# This query cannot be expressed in the Rhea website\\nSELECT ?chebi1 ?name1 ?chebi2 ?name2 ?rhea ?equation\\nWHERE {\\n  VALUES (?chebi1) { (CHEBI:29985) }\\n  ?chebi1 up:name ?name1 .\\n  ?rhea rh:side ?reactionSide1 .\\n  ?reactionSide1  rh:contains / rh:compound / rh:chebi ?chebi1 .\\n\\n  VALUES (?chebi2) { (CHEBI:58359) }\\n  ?chebi2 up:name ?name2 .\\n\\n  ?rhea rh:side ?reactionSide2 .\\n  ?reactionSide2  rh:contains / rh:compound / rh:chebi ?chebi2 .\\n  \\n  ?reactionSide1 rh:transformableTo ?reactionSide2 .\\n  \\n  ?rhea rh:equation ?equation .\\n}\",\n",
    "        \"results\": 32,\n",
    "        \"runtime\": 1\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"Select all Rhea reactions mapped to Reactome reactions\",\n",
    "        \"endpoint\": \"https://sparql.rhea-db.org/sparql/\",\n",
    "        \"query\": \"PREFIX rh: <http://rdf.rhea-db.org/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\n# Query 7 \\n# Select all Rhea reactions mapped to Reactome reactions\\n# MetaCyc reactions are mapped to either undirected, left-to-right or right-to-left reactions\\n# Rhea web query: https://www.rhea-db.org/rhea?query=reactome:*\\n#    \\nSELECT distinct ?rhea ?rheaDir ?reactome  \\nWHERE {\\n  ?rhea rdfs:subClassOf rh:Reaction .\\n  {\\n    ?rhea rdfs:seeAlso ?reactome .\\n    FILTER CONTAINS(str(?reactome), \\\"reactome\\\") \\n    BIND(?rhea as ?rheaDir)\\n  }\\n  UNION\\n  {\\n    ?rhea rh:directionalReaction ?directionalReaction .\\n    ?directionalReaction rdfs:seeAlso ?reactome . \\n    FILTER CONTAINS(str(?reactome), \\\"reactome\\\") \\n    BIND(?directionalReaction as ?rheaDir  )\\n  }\\n}\",\n",
    "        \"results\": 1502,\n",
    "        \"runtime\": 0\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"Distribution of reactions according to the first class of the enzyme classification (federated query)\",\n",
    "        \"endpoint\": \"https://sparql.rhea-db.org/sparql/\",\n",
    "        \"query\": \"PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\\nPREFIX rh: <http://rdf.rhea-db.org/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX ec: <http://purl.uniprot.org/enzyme/>\\n# Query 11\\n# Retrieve the count of reactions mapped to each level (main class) of the enzyme classification\\n# \\n# This query mimics the Filter section of the Rhea website (Browse all reactions)\\n# https://www.rhea-db.org/rhea?query=\\nSELECT ?ecClass (STR(?ecName) AS ?ecClassName) (COUNT(?rhea) AS ?rheaCount)\\nWHERE {\\n  SERVICE <https://sparql.uniprot.org/sparql> {\\n    VALUES (?ecClass) { (ec:1.-.-.-)(ec:2.-.-.-)(ec:3.-.-.-)(ec:4.-.-.-)(ec:5.-.-.-) (ec:6.-.-.-) (ec:7.-.-.-)}\\n    ?ecNumber rdfs:subClassOf ?ecClass .\\n    ?ecClass skos:prefLabel ?ecName .\\n  }\\n  ?rhea rdfs:subClassOf rh:Reaction .\\n  ?rhea rh:ec ?ecNumber .\\n} GROUP BY ?ecClass ?ecName\",\n",
    "        \"results\": 7,\n",
    "        \"runtime\": 4\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"Select all Rhea reactions annotated with a given Pubmed ID\",\n",
    "        \"endpoint\": \"https://sparql.rhea-db.org/sparql/\",\n",
    "        \"query\": \"PREFIX rh: <http://rdf.rhea-db.org/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX pubmed: <http://rdf.ncbi.nlm.nih.gov/pubmed/>\\n# Query 2 \\n# Select all Rhea reactions annotated with a given Pubmed identifier (PMID = 29867142)\\n#\\nSELECT ?pubmed ?rhea ?accession ?isTransport  ?equation \\nWHERE {\\n  ?rhea rdfs:subClassOf rh:Reaction .\\n  ?rhea rh:accession ?accession .\\n  ?rhea rh:citation ?pubmed .\\n  VALUES (?pubmed) { (pubmed:29867142) }\\n  ?rhea rh:isTransport ?isTransport .\\n  ?rhea rh:equation ?equation .\\n} ORDER BY ?rhea\",\n",
    "        \"results\": 14,\n",
    "        \"runtime\": 0\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"Select all Rhea reactions that involve a lipid, i.e. children of CHEBI:18059 in the ChEBI hierarchy.\",\n",
    "        \"endpoint\": \"https://sparql.rhea-db.org/sparql/\",\n",
    "        \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rh: <http://rdf.rhea-db.org/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX owl: <http://www.w3.org/2002/07/owl#>\\nPREFIX chebihash: <http://purl.obolibrary.org/obo/chebi#>\\nPREFIX CHEBI: <http://purl.obolibrary.org/obo/CHEBI_>\\n# Query 17\\n# Select all Rhea reactions that involve a lipid, i.e. children of CHEBI:18059 in the ChEBI hierarchy.\\n# \\n# This query corresponds to the Rhea website query:\\n# https://www.rhea-db.org/rhea?query=chebi:18059\\n#\\nSELECT distinct ?chebi ?name ?rhea ?equation\\nWHERE {\\n  ?rhea rdfs:subClassOf rh:Reaction .\\n  ?rhea rh:equation ?equation .\\n  ?rhea rh:side/rh:contains/rh:compound ?compound .\\n  #\\n  # the ChEBI can be used either as a small molecule, the reactive part of a macromolecule or as a polymer.\\n  #\\n  { \\n    ?chebi rdfs:subClassOf* CHEBI:18059 . # lipid\\n    ?compound (rh:chebi|(rh:reactivePart/rh:chebi)|(rh:underlyingChebi/rh:chebi)) ?chebi .\\n  }\\n  UNION \\n  { # add non-pH 7.3 species\\n    ?not7_3 rdfs:subClassOf* CHEBI:18059 . # lipid\\n    ?not7_3 rdfs:subClassOf ?chebiRestriction .\\n    ?chebiRestriction a owl:Restriction .\\n    ?chebiRestriction owl:onProperty chebihash:has_major_microspecies_at_pH_7_3 .\\n    ?chebiRestriction owl:someValuesFrom ?chebi .\\n    ?compound (rh:chebi|(rh:reactivePart/rh:chebi)|(rh:underlyingChebi/rh:chebi)) ?chebi .\\n  }\\n  ?chebi up:name ?name .\\n}\",\n",
    "        \"results\": 12834,\n",
    "        \"runtime\": 6\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"Select all Rhea reactions mapped to GO molecular functions\",\n",
    "        \"endpoint\": \"https://sparql.rhea-db.org/sparql/\",\n",
    "        \"query\": \"PREFIX rh: <http://rdf.rhea-db.org/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\n# Query 8\\n# Select all Rhea reactions mapped to GO molecular function\\n# \\n# This query corresponds to the Rhea website query:\\n# https://www.rhea-db.org/rhea?query=go:*\\n#\\nSELECT ?go ?rhea ?equation \\nWHERE {\\n  ?rhea rdfs:subClassOf rh:Reaction .\\n  ?rhea rdfs:seeAlso ?go .\\n  FILTER CONTAINS(str(?go), \\\"GO_\\\")   \\n  ?rhea rh:equation ?equation .\\n}\",\n",
    "        \"results\": 4432,\n",
    "        \"runtime\": 0\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"Select all Rhea reactions linked to protein sequences (enzymes and transporters) in UniProtKB\",\n",
    "        \"endpoint\": \"https://sparql.rhea-db.org/sparql/\",\n",
    "        \"query\": \"PREFIX up: <http://purl.uniprot.org/core/>\\nPREFIX rh: <http://rdf.rhea-db.org/>\\n# Query 12\\n# Select all Rhea reactions used to annotate enzyme sequences in UniProtKB\\n# return the number of UniProtKB entries\\n# Federated query using a service to UniProt SPARQL endpoint\\n#\\n# This query corresponds to the Rhea website query:\\n# https://www.rhea-db.org/rhea?query=uniprot:*\\n#\\nSELECT ?uniprotCount ?rhea ?accession ?equation \\nWHERE {\\n  SERVICE <https://sparql.uniprot.org/sparql> { \\n  \\tSELECT ?rhea (count(?uniprot) as ?uniprotCount) {\\n      ?uniprot up:annotation/up:catalyticActivity/up:catalyzedReaction ?rhea . \\n  \\t}\\n  \\tGROUP BY ?rhea\\n  }\\n  ?rhea rh:accession ?accession .\\n  ?rhea rh:equation ?equation .\\n}\",\n",
    "        \"results\": 12402,\n",
    "        \"runtime\": 3\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"Select the specific form of RHEA:11628, a reaction that deals with general chemical classes (i.e. select the equivalent reactions involving instances of the chemical classes in RHEA:11628\",\n",
    "        \"endpoint\": \"https://sparql.rhea-db.org/sparql/\",\n",
    "        \"query\": \"PREFIX rh: <http://rdf.rhea-db.org/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\n# Query 3\\n# Select the specific form of RHEA:11628.\\n# This query mimics the Related reactions sections of\\n# https://www.rhea-db.org/rhea/11628\\n#\\nSELECT ?rhea  ?equation ?childrenRhea ?childrenEquation \\nWHERE {\\n  VALUES (?rhea) {(rh:11628)}\\n  ?rhea rh:equation ?equation .\\n  ?childrenRhea rdfs:subClassOf+ ?rhea .\\n  ?childrenRhea rh:equation ?childrenEquation .\\n}\",\n",
    "        \"results\": 12,\n",
    "        \"runtime\": 0\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"Use IDSM Sachem to find ChEBIs with a a Cholestane skeleton (in SMILES). Then match returned ChEBIs to Rhea undirected reactions.\",\n",
    "        \"endpoint\": \"https://sparql.rhea-db.org/sparql/\",\n",
    "        \"query\": \"PREFIX sachem: <http://bioinfo.uochb.cas.cz/rdf/v1.0/sachem#>\\nPREFIX rh: <http://rdf.rhea-db.org/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\n\\nSELECT \\n  ?rhea \\n  ?chebi\\nWHERE {\\n  SERVICE <https://idsm.elixir-czech.cz/sparql/endpoint/chebi> {\\n    ?chebi sachem:substructureSearch [\\n        sachem:query \\\"[C@]12(CCC3CCCC[C@]3(C)[C@@]1([H])CC[C@]1(C)[C@@]([H])([C@@](C)([H])CCCC(C)C)CC[C@@]21[H])[H]\\\" ].\\n  }\\n  ?rhea rh:side/rh:contains/rh:compound/rdfs:subClassOf ?chebi .\\n}\",\n",
    "        \"results\": 135,\n",
    "        \"runtime\": 0\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"Select all Rhea reactions mapped to enzyme classification (EC numbers)\",\n",
    "        \"endpoint\": \"https://sparql.rhea-db.org/sparql/\",\n",
    "        \"query\": \"PREFIX rh: <http://rdf.rhea-db.org/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\nPREFIX ec: <http://purl.uniprot.org/enzyme/>\\n# Query 9\\n# Select all Rhea reactions mapped to EC numbers (enzyme classification)\\n# \\n# This query corresponds to the Rhea website query:\\n# https://www.rhea-db.org/rhea?query=ec:*\\n#\\nSELECT ?ec ?ecNumber ?rhea ?accession ?equation \\nWHERE {\\n  ?rhea rdfs:subClassOf rh:Reaction .\\n  ?rhea rh:accession ?accession .\\n  ?rhea rh:ec ?ec .\\n  BIND(strafter(str(?ec),str(ec:)) as ?ecNumber)\\n  ?rhea rh:isTransport ?isTransport .\\n  ?rhea rh:equation ?equation .\\n}\",\n",
    "        \"results\": 7675,\n",
    "        \"runtime\": 0\n",
    "    },\n",
    "    {\n",
    "        \"question\": \"Select all cross-references for a given reaction\",\n",
    "        \"endpoint\": \"https://sparql.rhea-db.org/sparql/\",\n",
    "        \"query\": \"PREFIX rh: <http://rdf.rhea-db.org/>\\nPREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\\n# Query 4\\n# Select all cross-references mapped to RHEA:21016\\n# \\n# This query mimics the Cross-references section of \\n# https://www.rhea-db.org/rhea/21016\\n#\\nSELECT distinct ?rhea ?rheaDir ?xref \\nWHERE {\\n  ?rhea rdfs:subClassOf rh:Reaction .\\n  VALUES (?rhea) {(rh:21016)}\\n  {\\n    ?rhea rdfs:seeAlso ?xref .\\n    BIND(?rhea as ?rheaDir)\\n  }\\n  UNION\\n  {\\n    ?rhea rh:directionalReaction ?directionalReaction .\\n    ?directionalReaction rdfs:seeAlso ?xref . \\n    BIND(?directionalReaction as ?rheaDir  )\\n  }\\n  UNION\\n  {\\n    ?rhea rh:bidirectionalReaction ?bidirectionalReaction .\\n    ?bidirectionalReaction rdfs:seeAlso ?xref . \\n    BIND(?bidirectionalReaction as ?rheaDir  )\\n  }\\n}\",\n",
    "        \"results\": 6,\n",
    "        \"runtime\": 0\n",
    "    }\n",
    "]\n",
    "\n",
    "success = 0\n",
    "mismatch = 0\n",
    "fail = 0\n",
    "\n",
    "# Extract the SPARQL query from markdown\n",
    "pattern = re.compile(r\"```sparql(.*?)```\", re.DOTALL)\n",
    "for i, query in enumerate(example_queries):\n",
    "    # if i < 1:\n",
    "    #     continue\n",
    "    resp = requests.post(\"http://localhost:8000/chat\",\n",
    "        json={\n",
    "            \"messages\": [{\"role\": \"user\", \"content\": query[\"question\"]}],\n",
    "            \"model\": \"gpt-4o\",\n",
    "            \"max_tokens\": 50,\n",
    "            \"stream\": False,\n",
    "        },\n",
    "        timeout=60,\n",
    "    )\n",
    "    chat_resp_md = resp.json()[\"choices\"][0][\"message\"][\"content\"]\n",
    "    generated_sparql = pattern.findall(chat_resp_md)[-1].strip()\n",
    "    print(generated_sparql)\n",
    "    try:\n",
    "        # print(f\"[{query['question']}]\")\n",
    "        if generated_sparql.strip() == query[\"query\"].strip():\n",
    "            print(f\"✅ {query['question']}. EXACT MATCH\")\n",
    "            success += 1\n",
    "            continue\n",
    "        sparql_endpoint = SPARQLWrapper(query[\"endpoint\"])\n",
    "        sparql_endpoint.setReturnFormat(JSON)\n",
    "        sparql_endpoint.setQuery(generated_sparql)\n",
    "        sparql_endpoint.setTimeout(200)\n",
    "        results = sparql_endpoint.query().convert()\n",
    "        res_from_generated = len(results[\"results\"][\"bindings\"])\n",
    "        if res_from_generated != query[\"results\"]:\n",
    "            if res_from_generated > 0:\n",
    "                mismatch += 1\n",
    "            raise Exception(f\"{query['results']} != {res_from_generated}\")\n",
    "            # print(results)\n",
    "        else:\n",
    "            print(f\"✅ {query['question']} {query['results']} == {res_from_generated}\")\n",
    "            success += 1\n",
    "    except Exception as e:\n",
    "        fail += 1\n",
    "        print(f\"❌ {query['question']} {e}. Correct query:\")\n",
    "        print(query[\"query\"])\n",
    "        print(\"\")\n",
    "    # print(results)\n",
    "    # print(f\"Results: {len(results['results']['bindings'])}\")\n",
    "\n",
    "    print(f\"⚖️ Success: {success}, Mismatch: {mismatch}, Fail: {fail}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "expasy-scripts",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
