{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Testing Expasy query helper\n",
    "\n",
    "These tests requires to start the expasy chat stack locally with `docker compose -f compose.dev.yml up`\n",
    "\n",
    "Testing our Expasy query helper reusing the example queries that succeeded retrieved from `test_example_queries.ipynb`\n",
    "\n",
    "Potential interesting questions to use for evaluation:\n",
    "- Which are the human genes associated with cancer and their orthologs expressed in the rat brain?\n",
    "- Select the number of proteins for each of the subclasses of ec:1.1.1.- EC (Enzyme Commission) class (from biosoda)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/vemonet/dev/expasy/expasy-chat/.venv/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "üß™ Testing 10 queries\n",
      "‚ùå 1/3 What is the accession number in uniprot of the human gene LCT? Return only unique protein URIs QueryBadFormed: A bad request has been sent to the endpoint: probably the SPARQL query is badly formed. \n",
      "\n",
      "Response:\n",
      "b'<!DOCTYPE html SYSTEM \"about:legacy-compat\">\\n<html xmlns=\"http://www.w3.org/1999/xhtml\" lang=\"en\" xml:lang=\"en\"><head><title>UniProt</title><meta content=\"text/html; charset=UTF-8\" http-equiv=\"Content-Type\"/><link href=\"/\" rel=\"home\"/><link href=\"/base.css\" type=\"text/css\" rel=\"stylesheet\"/><link type=\"image/vnd.microsoft.icon\" href=\"/uniprot-favicon.ico\" rel=\"shortcut icon\"/><link href=\"/uniprot-sparql.css\" type=\"text/css\" rel=\"stylesheet\"/><link href=\"http://creativecommons.org/licenses/by/4.0/\" rel=\"license\"/><script src=\"https://cdn.jsdelivr.net/npm/yasgui-yasqe@2.11.22/dist/yasqe.bundled.min.js\"></script><link type=\"text/css\" rel=\"stylesheet\" href=\"https://cdn.jsdelivr.net/yasqe/2.2/yasqe.min.css\"/><meta content=\"width=device-width, initial-scale=1\" name=\"viewport\"/><meta content=\"no-referrer\" name=\"referrer\"/></head><body typeof=\"schema:WebPage\"><nav id=\"nav-header\"><a accesskey=\"1\" href=\"https://www.uniprot.org/\" id=\"www\"><img alt=\"\" src=\"data:image/svg+xml;base64,PHN2ZyB2aWV3Qm94PSIwIDAgNjQwIDI5MyIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj48ZyBmaWxsPSIjZmZmIiBmaWxsLXJ1bGU9Im5vbnplcm8iPjxwYXRoIGQ9Ik0xOTkuNjU5LDEwNi4xMDMgQzE5OS42NTksMTEyLjM5OCAxOTQuNTU2LDExNy41MDEgMTg4LjI2LDExNy41MDEgQzE4MS45NjQsMTE3LjUwMSAxNzYuODYxLDExMi4zOTcgMTc2Ljg2MSwxMDYuMTAzIEMxNzYuODYxLDk5LjgwNyAxODEuOTY1LDk0LjcwNCAxODguMjYsOTQuNzA0IEMxOTQuNTU1LDk0LjcwNCAxOTkuNjU5LDk5LjgwNyAxOTkuNjU5LDEwNi4xMDMgWiI+PC9wYXRoPjxwYXRoIGQ9Ik0zMzQuMzgzLDE0NC40NTkgQzMyNS4yNDIsMTQ0LjQ1OSAzMTcuODI5LDE1MS44NzIgMzE3LjgyOSwxNjEuMDE1IEwzMTcuNzk3LDE2NS4xODUgTDMxNy43OTcsMTk0LjIxIEMzMTcuNzk3LDIwMC4zMTYgMzEyLjg0OCwyMDUuMjY1IDMwNi43NDIsMjA1LjI2NSBDMzAwLjYzOCwyMDUuMjY1IDI5NS42ODksMjAwLjMxNyAyOTUuNjg5LDE5NC4yMSBMMjk1LjY4OSwxNjUuMTg1IEwyOTUuNzIsMTYxLjAxNSBDMjk1LjcyLDEzOS42NjEgMzEzLjAzLDEyMi4zNTEgMzM0LjM4MywxMjIuMzUxIEwzNDEuMjg1LDEyMi4zNTEgQzM0Ny4zOSwxMjIuMzUxIDM1Mi4zNCwxMjcuMyAzNTIuMzQsMTMzLjQwNSBDMzUyLjM0LDEzOS41MDkgMzQ3LjM5MSwxNDQuNDU5IDM0MS4yODUsMTQ0LjQ1OSBMMzM0LjM4MywxNDQuNDU5IFoiPjwvcGF0aD48cGF0aCBkPSJNNDMzLjg2MSwxNjMuODEzIEM0MzMuODYxLDE4Ny40MTcgNDE0LjcyNCwyMDYuNTU1IDM5MS4xMTksMjA2LjU1NSBDMzY3LjUxNCwyMDYuNTU1IDM0OC4zNzUsMTg3LjQxNyAzNDguMzc1LDE2My44MTMgQzM0OC4zNzUsMTQwLjIwNiAzNjcuNTE0LDEyMS4wNyAzOTEuMTE5LDEyMS4wNyBDNDE0LjcyNCwxMjEuMDcgNDMzLjg2MSwxNDAuMjA2IDQzMy44NjEsMTYzLjgxMyBaIE0zOTEuMDI5LDE0MS42OTQgQzM3OC44MTgsMTQxLjY5NCAzNjguOTIsMTUxLjU5MiAzNjguOTIsMTYzLjgwMSBDMzY4LjkyLDE3Ni4wMTEgMzc4LjgxOCwxODUuOTA5IDM5MS4wMjksMTg1LjkwOSBDNDAzLjIzOCwxODUuOTA5IDQxMy4xMzYsMTc2LjAxMSA0MTMuMTM2LDE2My44MDEgQzQxMy4xMzcsMTUxLjU5MyA0MDMuMjM4LDE0MS42OTQgMzkxLjAyOSwxNDEuNjk0IFoiPjwvcGF0aD48cGF0aCBkPSJNNDgzLjcwMywxODMuMTU2IEw0NzkuNDY3LDE4My4xNTUgQzQ3MC4zMjIsMTgzLjE1NSA0NjIuOTEsMTc1Ljc0MyA0NjIuOTEsMTY2LjU5OCBMNDYyLjg3NywxNjIuNDMgTDQ2Mi44NzcsMTQ0LjQ1NyBMNDgyLjIyMSwxNDQuNDU4IEM0ODguMzI2LDE0NC40NTggNDkzLjI3NiwxMzkuNTEgNDkzLjI3NiwxMzMuNDA0IEM0OTMuMjc2LDEyNy4yOTggNDg4LjMyNywxMjIuMzQ5IDQ4Mi4yMjEsMTIyLjM0OSBMNDYyLjg3NywxMjIuMzQ5IEw0NjIuODc3LDExNi44MjQgQzQ2Mi44NzcsMTEwLjcxOSA0NTcuOTMsMTA1Ljc2OSA0NTEuODI0LDEwNS43NjkgQzQ0NS43MTgsMTA1Ljc2OSA0NDAuNzY5LDExMC43MTggNDQwLjc2OSwxMTYuODI0IEw0NDAuNzY5LDE2Mi40MjkgTDQ0MC44LDE2Ni41OTkgQzQ0MC44LDE4Ny45NTIgNDU4LjExMywyMDUuMjYyIDQ3OS40NjYsMjA1LjI2MiBMNDgzLjcwMiwyMDUuMjYzIEM0ODkuODA3LDIwNS4yNjMgNDk0Ljc1NywyMDAuMzE1IDQ5NC43NTcsMTk0LjIwOCBDNDk0Ljc1OCwxODguMTA1IDQ4OS44MDksMTgzLjE1NiA0ODMuNzAzLDE4My4xNTYgWiI+PC9wYXRoPjxwYXRoIGQ9Ik0yOTIuOTgxLDEzNS4wNDIgQzI5Mi45ODEsMTEzLjY4OCAyNzUuNjcsOTYuMzc4IDI1NC4zMTcsOTYuMzc4IEwyNTAuMTQ3LDk2LjM0NiBMMjIyLjQ1OCw5Ni4zNDYgQzIxNi4zNTMsOTYuMzQ2IDIxMS40MDQsMTAxLjI5NSAyMTEuNDA0LDEwNy40IEwyMTEuNDA0LDE5NC4yIEMyMTEuNDA0LDIwMC4zMDYgMjE2LjM1MywyMDUuMjU2IDIyMi40NTgsMjA1LjI1NiBDMjI4LjU2MiwyMDUuMjU2IDIzMy41MTIsMjAwLjMwNyAyMzMuNTEyLDE5NC4yIEwyMzMuNTEyLDExOC40NTUgTDI1MC4xNDgsMTE4LjQ1NSBMMjU0LjMxNywxMTguNDg3IEMyNjMuNDYxLDExOC40ODcgMjcwLjg3NCwxMjUuODk5IDI3MC44NzQsMTM1LjA0MyBDMjcwLjg3NCwxNDQuMTg3IDI2My40NjEsMTUxLjYgMjU0LjMxNywxNTEuNiBMMjU1LjY3NCwxNTEuNjE2IEwyNTEuNTIsMTUxLjYxNiBDMjQ1LjQxNiwxNTEuNjE2IDI0MC40NjYsMTU2LjU2MyAyNDAuNDY2LDE2Mi42NyBDMjQwLjQ2NiwxNjguNzc0IDI0NS40MTUsMTczLjcyNCAyNTEuNTIsMTczLjcyNCBMMjU1LjY3NCwxNzMuNzI0IEwyNTQuMzE4LDE3My43MDggQzI3NS42NzEsMTczLjcwNiAyOTIuOTgxLDE1Ni4zOTYgMjkyLjk4MSwxMzUuMDQyIFoiPjwvcGF0aD48cGF0aCBkPSJNNDYyLjYwNywyNi4zMTkgQzQ2Mi42MDcsMzEuNDk0IDQ1OC40MSwzNS42ODkgNDUzLjIzNiwzNS42ODkgQzQ0OC4wNiwzNS42ODkgNDQzLjg2NSwzMS40OTQgNDQzLjg2NSwyNi4zMTkgQzQ0My44NjUsMjEuMTQ0IDQ0OC4wNiwxNi45NDggNDUzLjIzNiwxNi45NDggQzQ1OC40MSwxNi45NDggNDYyLjYwNywyMS4xNDUgNDYyLjYwNywyNi4zMTkgWiI+PC9wYXRoPjxwYXRoIGQ9Ik00MzAuNTE0LDU0LjY4OCBDNDMwLjUxNCw1OC44MjggNDI3LjE1Nyw2Mi4xODUgNDIzLjAxNiw2Mi4xODUgQzQxOC44NzcsNjIuMTg1IDQxNS41Miw1OC44MjggNDE1LjUyLDU0LjY4OCBDNDE1LjUyLDUwLjU0OCA0MTguODc3LDQ3LjE5MDk5OTggNDIzLjAxNiw0Ny4xOTA5OTk4IEM0MjcuMTU2LDQ3LjE5IDQzMC41MTQsNTAuNTQ4IDQzMC41MTQsNTQuNjg4IFoiPjwvcGF0aD48cGF0aCBkPSJNNTA1LjU1NywxMS43MTMgQzUwNS41NTcsMTguMTgzIDUwMC4zMTEsMjMuNDI2IDQ5My44NDIsMjMuNDI2IEM0ODcuMzczLDIzLjQyNiA0ODIuMTI5LDE4LjE4MyA0ODIuMTI5LDExLjcxMyBDNDgyLjEyOSw1LjI0MyA0ODcuMzczLDAgNDkzLjg0MiwwIEM1MDAuMzExLDAgNTA1LjU1Nyw1LjI0MyA1MDUuNTU3LDExLjcxMyBaIj48L3BhdGg+PHBhdGggZD0iTTU1NS4xMTksMTYuMzE4IEM1NTUuMTE5LDI0LjQwNCA1NDguNTY0LDMwLjk2IDU0MC40NzYsMzAuOTYgQzUzMi4zOTIsMzAuOTYgNTI1LjgzNSwyNC40MDQgNTI1LjgzNSwxNi4zMTggQzUyNS44MzUsOC4yMzEgNTMyLjM5MiwxLjY3NiA1NDAuNDc2LDEuNjc2IEM1NDguNTY0LDEuNjc3IDU1NS4xMTksOC4yMzEgNTU1LjExOSwxNi4zMTggWiI+PC9wYXRoPjxwYXRoIGQ9Ik02MDIuMTAyLDQyLjA0NiBDNjAyLjEwMiw1Mi4xNTMgNTkzLjkwOSw2MC4zNDcgNTgzLjgwMSw2MC4zNDcgQzU3My42OTIsNjAuMzQ3IDU2NS40OTgsNTIuMTU0IDU2NS40OTgsNDIuMDQ2IEM1NjUuNDk4LDMxLjkzNyA1NzMuNjkxLDIzLjc0NCA1ODMuODAxLDIzLjc0NCBDNTkzLjkwOCwyMy43NDQgNjAyLjEwMiwzMS45MzcgNjAyLjEwMiw0Mi4wNDYgWiI+PC9wYXRoPjxwYXRoIGQ9Ik02MzUuNDM2LDkwLjI4MiBDNjM1LjQzNiwxMDEuOTA3IDYyNi4wMTIsMTExLjMyOSA2MTQuMzg5LDExMS4zMjkgQzYwMi43NjYsMTExLjMyOSA1OTMuMzQyLDEwMS45MDcgNTkzLjM0Miw5MC4yODIgQzU5My4zNDIsNzguNjU3IDYwMi43NjYsNjkuMjM1IDYxNC4zODksNjkuMjM1IEM2MjYuMDEyLDY5LjIzNSA2MzUuNDM2LDc4LjY1NyA2MzUuNDM2LDkwLjI4MiBaIj48L3BhdGg+PHBhdGggZD0iTTYzOS4zODcsMTUzLjU5NSBDNjM5LjM4NywxNjYuOTYgNjI4LjU1MSwxNzcuNzk2IDYxNS4xODQsMTc3Ljc5NiBDNjAxLjgxNSwxNzcuNzk2IDU5MC45NzksMTY2Ljk2IDU5MC45NzksMTUzLjU5NSBDNTkwLjk3OSwxNDAuMjMgNjAxLjgxNSwxMjkuMzk1IDYxNS4xODQsMTI5LjM5NSBDNjI4LjU1MSwxMjkuMzk1IDYzOS4zODcsMTQwLjIyOSA2MzkuMzg3LDE1My41OTUgWiI+PC9wYXRoPjxwYXRoIGQ9Ik02MTIuNzY2LDIxNS44ODYgQzYxMi43NjYsMjMxLjMyNiA2MDAuMjUsMjQzLjg0MSA1ODQuODExLDI0My44NDEgQzU2OS4zNzIsMjQzLjg0MSA1NTYuODU2LDIzMS4zMjYgNTU2Ljg1NiwyMTUuODg2IEM1NTYuODU2LDIwMC40NDUgNTY5LjM3MiwxODcuOTMxIDU4NC44MTEsMTg3LjkzMSBDNjAwLjI1LDE4Ny45MzEgNjEyLjc2NiwyMDAuNDQ0IDYxMi43NjYsMjE1Ljg4NiBaIj48L3BhdGg+PHBhdGggZD0iTTU1MS43NzMsMjU5LjE5NCBDNTUxLjc3MywyNzYuNTAyIDUzNy43NDQsMjkwLjUzMSA1MjAuNDM3LDI5MC41MzEgQzUwMy4xMywyOTAuNTMxIDQ4OS4xMDEsMjc2LjUwMiA0ODkuMTAxLDI1OS4xOTQgQzQ4OS4xMDEsMjQxLjg4NiA1MDMuMTMsMjI3Ljg1OCA1MjAuNDM3LDIyNy44NTggQzUzNy43NDQsMjI3Ljg1OCA1NTEuNzczLDI0MS44ODcgNTUxLjc3MywyNTkuMTk0IFoiPjwvcGF0aD48cGF0aCBkPSJNNDcyLjEwNywyNTcuMzggQzQ3Mi4xMDcsMjc2Ljc3NCA0NTYuMzg0LDI5Mi40OTUgNDM2Ljk5MiwyOTIuNDk1IEM0MTcuNiwyOTIuNDk1IDQwMS44NzcsMjc2Ljc3MyA0MDEuODc3LDI1Ny4zOCBDNDAxLjg3NywyMzcuOTg0IDQxNy42LDIyMi4yNjcgNDM2Ljk5MiwyMjIuMjY3IEM0NTYuMzg0LDIyMi4yNjcgNDcyLjEwNywyMzcuOTg0IDQ3Mi4xMDcsMjU3LjM4IFoiPjwvcGF0aD48cGF0aCBkPSJNMTY1LjgwNSwxNTkuNjU4IEwxNjUuODA1LDE5NC4yMSBDMTY1LjgwNSwyMDAuMzE2IDE2MC44NTYsMjA1LjI2NSAxNTQuNzUxLDIwNS4yNjUgQzE0OC42NDcsMjA1LjI2NSAxNDMuNjk3LDIwMC4zMTcgMTQzLjY5NywxOTQuMjEgTDE0My42OTcsMTU5LjY1OCBMMTQzLjY4MSwxNjEuMDE0IEMxNDMuNjgxLDE1MS44NzEgMTM2LjI2OSwxNDQuNDU4IDEyNy4xMjUsMTQ0LjQ1OCBDMTE3Ljk4MSwxNDQuNDU4IDExMC41NjksMTUxLjg3MSAxMTAuNTY5LDE2MS4wMTQgTDExMC41MzYsMTY1LjE4NCBMMTEwLjUzNiwxOTQuMjA5IEMxMTAuNTM2LDIwMC4zMTUgMTA1LjU4NywyMDUuMjY0IDk5LjQ4MiwyMDUuMjY0IEM5My4zNzcsMjA1LjI2NCA4OC40MjgsMjAwLjMxNiA4OC40MjgsMTk0LjIwOSBMODguNDI4LDE2NS4xODQgTDg4LjQ2LDE2MS4wMTQgQzg4LjQ2LDEzOS42NiAxMDUuNzcsMTIyLjM1IDEyNy4xMjQsMTIyLjM1IEMxNDguNDc4LDEyMi4zNSAxNjUuNzg4LDEzOS42NjEgMTY1Ljc4OCwxNjEuMDE0IEwxNjUuODA1LDE1OS42NTggWiI+PC9wYXRoPjxwYXRoIGQ9Ik0wLDE2Ny45NTcgTDAsMTA3LjM5OSBDMCwxMDEuMjk1IDQuOTQ5LDk2LjM0NSAxMS4wNTQsOTYuMzQ1IEMxNy4xNTksOTYuMzQ1IDIyLjEwOCwxMDEuMjk0IDIyLjEwOCwxMDcuMzk5IEwyMi4xMDgsMTY3Ljk1NyBMMjIuMTIzLDE2Ni42MDEgQzIyLjEyMywxNzUuNzQ0IDI5LjUzNiwxODMuMTU3IDM4LjY3OSwxODMuMTU3IEM0Ny44MjIsMTgzLjE1NyA1NS4yMzUsMTc1Ljc0NCA1NS4yMzUsMTY2LjYwMSBMNTUuMjY4LDE2Mi40MzEgTDU1LjI2OCwxMDcuNCBDNTUuMjY4LDEwMS4yOTYgNjAuMjE3LDk2LjM0NSA2Ni4zMjIsOTYuMzQ1IEM3Mi40MjcsOTYuMzQ1IDc3LjM3NiwxMDEuMjk1IDc3LjM3NiwxMDcuNCBMNzcuMzc2LDE2Mi40MzEgTDc3LjM0NCwxNjYuNjAyIEM3Ny4zNDQsMTg3Ljk1NiA2MC4wMzQsMjA1LjI2NSAzOC42OCwyMDUuMjY1IEMxNy4zMjYsMjA1LjI2NSAwLjAxNiwxODcuOTU1IDAuMDE2LDE2Ni42MDIgTDAsMTY3Ljk1NyBaIj48L3BhdGg+PHBhdGggZD0iTTE5OS4zMTQsMTk0LjIxIEMxOTkuMzE0LDIwMC4zMTYgMTk0LjM2NSwyMDUuMjY1IDE4OC4yNiwyMDUuMjY1IEMxODIuMTU2LDIwNS4yNjUgMTc3LjIwNiwyMDAuMzE3IDE3Ny4yMDYsMTk0LjIxIEwxNzcuMjA2LDEzMy40NzggQzE3Ny4yMDYsMTI3LjM3MyAxODIuMTU1LDEyMi40MjMgMTg4LjI2LDEyMi40MjMgQzE5NC4zNjUsMTIyLjQyMyAxOTkuMzE0LDEyNy4zNzIgMTk5LjMxNCwxMzMuNDc4IEwxOTkuMzE0LDE5NC4yMSBaIj48L3BhdGg+PC9nPjwvc3ZnPgo=\" title=\"UniProt home\"/></a><a href=\"/sparql\">SPARQL</a><a href=\"https://www.uniprot.org/downloads\" id=\"download\">Downloads</a><a accesskey=\"9\" href=\"https://www.uniprot.org/contact\" id=\"contact\"><img alt=\"Contact\" src=\"data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZpZXdCb3g9IjAgMCAxMyAxMCIgZmlsbD0iI2ZmZiIgd2lkdGg9IjEuNGVtIj48cGF0aCBmaWxsLXJ1bGU9ImV2ZW5vZGQiIGQ9Ik0xMiAxLjI1QzEyIC41NiAxMS40NiAwIDEwLjggMEgxLjJDLjU0IDAgMCAuNTYgMCAxLjI1djcuNUMwIDkuNDQuNTQgMTAgMS4yIDEwaDkuNmMuNjYgMCAxLjItLjU2IDEuMi0xLjI1di03LjV6bS0xLjIgMEw2IDQuMzggMS4yIDEuMjVoOS42ek02IDUuNjNMMS4yIDIuNXY2LjI1aDkuNlYyLjVMNiA1LjYzeiI+PC9wYXRoPjwvc3ZnPgo=\"/></a><a rel=\"help\" href=\"https://www.uniprot.org/help/\" id=\"help\">Documentation/Help</a></nav><a accesskey=\"2\" href=\"#content\" hidden=\"hidden\">Skip Header</a><main class=\"sparql results\" id=\"content\"><section class=\"error\" id=\"page-header\"><h2 class=\"page-title\">Malformed query.</h2><p>org.eclipse.rdf4j.query.parser.sparql.ast.VisitorException: QName \\'up:Protein\\' uses an undefined prefix</p></section><div class=\"main-aside\"><div id=\"queryform\"><form method=\"get\" action=\"/sparql/\" focus=\"query\" name=\"sparql-form\" id=\"sparql-form\"><button id=\"addPrefix\" type=\"button\">Add common prefixes</button><textarea accesskey=\"1\" rows=\"10\" cols=\"80\" name=\"query\" id=\"query\"># https://sparql.uniprot.org/sparql\\n\\nSELECT DISTINCT ?protein\\nWHERE {\\n  ?protein a up:Protein ;\\n           up:organism &lt;http://purl.uniprot.org/taxonomy/9606&gt; ;\\n           up:encodedBy &lt;http://purl.uniprot.org/uniprot/LCT&gt; .\\n}</textarea><input type=\"hidden\" name=\"format\" value=\"html\"><button id=\"sparqlFormSubmitter\" onclick=\"return false;\">Submit Query</button></form></div></div></main><footer id=\"page-footer\"><section class=\"small\" id=\"copyright\">\\n\\t\\t\\t&copy; 2002&ndash; <span property=\"schema:copyrightYear\">2024</span>\\n\\t\\t\\t<a accesskey=\"9\" href=\"https://www.uniprot.org/help/about\">UniProt Consortium</a> |\\n\\t\\t\\t<a href=\"https://www.uniprot.org/help/license\" property=\"schema:license\">License\\n\\t\\t\\t\\t&amp; Disclaimer</a> | <a href=\"https://www.uniprot.org/help/privacy\">Privacy\\n\\t\\t\\t\\tNotice</a></section><section><div resource=\"/help/about#UniProt\" id=\"logos\" typeof=\"NGO\" property=\"creator\"><span resource=\"/help/about#EBI\" typeof=\"Organization\" class=\"ebi\" property=\"member\"><span hidden=\"hidden\" property=\"name\">European Bioinformatics Institute</span><a href=\"https://www.ebi.ac.uk/\" property=\"url\"></a></span><span resource=\"/help/about#PIR\" typeof=\"Organization\" class=\"pir\" property=\"member\"><span hidden=\"hidden\" property=\"name\">Protein Information Resource</span><a href=\"http://pir.georgetown.edu/\" property=\"url\"></a></span><span resource=\"/help/about#SIB\" typeof=\"Organization\" class=\"sib\" property=\"member\"><span hidden=\"hidden\" property=\"name\">SIB Swiss Institute of Bioinformatics</span><a href=\"https://www.sib.swiss/\" property=\"url\"></a></span></div></section><section id=\"poweredby\"><h3>Powered by</h3><span class=\"vitalit\" resource=\"https://www.sib.swiss/alan-bridge-group/\" typeof=\"schema:NGO\" property=\"schema:sponsor\"><h4>Hardware</h4><a href=\"https://www.sib.swiss/alan-bridge-group\" property=\"schema:url\"><img alt=\"Swiss-Prot\" src=\"https://www.sib.swiss//templates/sib/images/SIB_LogoQ_GBv.svg\"/><h5 property=\"schema:name\">Swiss-Prot</h5></a></span><span class=\"openlinksw\" resource=\"http://www.openlinksw.com/#this\" typeof=\"schema:Corporation\" property=\"schema:sponsor\"><h4>Software</h4><a href=\"http://virtuoso.openlinksw.com/\" property=\"schema:url\"><img alt=\"Virtuoso\" src=\"/images/logos/virtuoso.png\"/><h5 property=\"schema:name\">Virtuoso</h5></a></span></section></footer><script src=\"/sparql.js\" type=\"text/javascript\"> </script></body></html>'. Generated query:\n",
      "# https://sparql.uniprot.org/sparql\n",
      "\n",
      "SELECT DISTINCT ?protein\n",
      "WHERE {\n",
      "  ?protein a up:Protein ;\n",
      "           up:organism <http://purl.uniprot.org/taxonomy/9606> ;\n",
      "           up:encodedBy <http://purl.uniprot.org/uniprot/LCT> .\n",
      "}\n",
      "Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
      "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
      "SELECT DISTINCT ?protein\n",
      "WHERE{\n",
      "    ?protein a up:Protein .\n",
      "    ?protein up:organism taxon:9606 .\n",
      "    ?protein up:encodedBy ?gene .\n",
      "    ?gene skos:prefLabel \"LCT\" .\n",
      "}\n",
      "\n",
      "‚ùå 2/3 What is the accession number in uniprot of the human gene LCT? Return only unique protein URIs QueryBadFormed: A bad request has been sent to the endpoint: probably the SPARQL query is badly formed. \n",
      "\n",
      "Response:\n",
      "b'<!DOCTYPE html SYSTEM \"about:legacy-compat\">\\n<html xmlns=\"http://www.w3.org/1999/xhtml\" lang=\"en\" xml:lang=\"en\"><head><title>UniProt</title><meta content=\"text/html; charset=UTF-8\" http-equiv=\"Content-Type\"/><link href=\"/\" rel=\"home\"/><link href=\"/base.css\" type=\"text/css\" rel=\"stylesheet\"/><link type=\"image/vnd.microsoft.icon\" href=\"/uniprot-favicon.ico\" rel=\"shortcut icon\"/><link href=\"/uniprot-sparql.css\" type=\"text/css\" rel=\"stylesheet\"/><link href=\"http://creativecommons.org/licenses/by/4.0/\" rel=\"license\"/><script src=\"https://cdn.jsdelivr.net/npm/yasgui-yasqe@2.11.22/dist/yasqe.bundled.min.js\"></script><link type=\"text/css\" rel=\"stylesheet\" href=\"https://cdn.jsdelivr.net/yasqe/2.2/yasqe.min.css\"/><meta content=\"width=device-width, initial-scale=1\" name=\"viewport\"/><meta content=\"no-referrer\" name=\"referrer\"/></head><body typeof=\"schema:WebPage\"><nav id=\"nav-header\"><a accesskey=\"1\" href=\"https://www.uniprot.org/\" id=\"www\"><img alt=\"\" src=\"data:image/svg+xml;base64,PHN2ZyB2aWV3Qm94PSIwIDAgNjQwIDI5MyIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj48ZyBmaWxsPSIjZmZmIiBmaWxsLXJ1bGU9Im5vbnplcm8iPjxwYXRoIGQ9Ik0xOTkuNjU5LDEwNi4xMDMgQzE5OS42NTksMTEyLjM5OCAxOTQuNTU2LDExNy41MDEgMTg4LjI2LDExNy41MDEgQzE4MS45NjQsMTE3LjUwMSAxNzYuODYxLDExMi4zOTcgMTc2Ljg2MSwxMDYuMTAzIEMxNzYuODYxLDk5LjgwNyAxODEuOTY1LDk0LjcwNCAxODguMjYsOTQuNzA0IEMxOTQuNTU1LDk0LjcwNCAxOTkuNjU5LDk5LjgwNyAxOTkuNjU5LDEwNi4xMDMgWiI+PC9wYXRoPjxwYXRoIGQ9Ik0zMzQuMzgzLDE0NC40NTkgQzMyNS4yNDIsMTQ0LjQ1OSAzMTcuODI5LDE1MS44NzIgMzE3LjgyOSwxNjEuMDE1IEwzMTcuNzk3LDE2NS4xODUgTDMxNy43OTcsMTk0LjIxIEMzMTcuNzk3LDIwMC4zMTYgMzEyLjg0OCwyMDUuMjY1IDMwNi43NDIsMjA1LjI2NSBDMzAwLjYzOCwyMDUuMjY1IDI5NS42ODksMjAwLjMxNyAyOTUuNjg5LDE5NC4yMSBMMjk1LjY4OSwxNjUuMTg1IEwyOTUuNzIsMTYxLjAxNSBDMjk1LjcyLDEzOS42NjEgMzEzLjAzLDEyMi4zNTEgMzM0LjM4MywxMjIuMzUxIEwzNDEuMjg1LDEyMi4zNTEgQzM0Ny4zOSwxMjIuMzUxIDM1Mi4zNCwxMjcuMyAzNTIuMzQsMTMzLjQwNSBDMzUyLjM0LDEzOS41MDkgMzQ3LjM5MSwxNDQuNDU5IDM0MS4yODUsMTQ0LjQ1OSBMMzM0LjM4MywxNDQuNDU5IFoiPjwvcGF0aD48cGF0aCBkPSJNNDMzLjg2MSwxNjMuODEzIEM0MzMuODYxLDE4Ny40MTcgNDE0LjcyNCwyMDYuNTU1IDM5MS4xMTksMjA2LjU1NSBDMzY3LjUxNCwyMDYuNTU1IDM0OC4zNzUsMTg3LjQxNyAzNDguMzc1LDE2My44MTMgQzM0OC4zNzUsMTQwLjIwNiAzNjcuNTE0LDEyMS4wNyAzOTEuMTE5LDEyMS4wNyBDNDE0LjcyNCwxMjEuMDcgNDMzLjg2MSwxNDAuMjA2IDQzMy44NjEsMTYzLjgxMyBaIE0zOTEuMDI5LDE0MS42OTQgQzM3OC44MTgsMTQxLjY5NCAzNjguOTIsMTUxLjU5MiAzNjguOTIsMTYzLjgwMSBDMzY4LjkyLDE3Ni4wMTEgMzc4LjgxOCwxODUuOTA5IDM5MS4wMjksMTg1LjkwOSBDNDAzLjIzOCwxODUuOTA5IDQxMy4xMzYsMTc2LjAxMSA0MTMuMTM2LDE2My44MDEgQzQxMy4xMzcsMTUxLjU5MyA0MDMuMjM4LDE0MS42OTQgMzkxLjAyOSwxNDEuNjk0IFoiPjwvcGF0aD48cGF0aCBkPSJNNDgzLjcwMywxODMuMTU2IEw0NzkuNDY3LDE4My4xNTUgQzQ3MC4zMjIsMTgzLjE1NSA0NjIuOTEsMTc1Ljc0MyA0NjIuOTEsMTY2LjU5OCBMNDYyLjg3NywxNjIuNDMgTDQ2Mi44NzcsMTQ0LjQ1NyBMNDgyLjIyMSwxNDQuNDU4IEM0ODguMzI2LDE0NC40NTggNDkzLjI3NiwxMzkuNTEgNDkzLjI3NiwxMzMuNDA0IEM0OTMuMjc2LDEyNy4yOTggNDg4LjMyNywxMjIuMzQ5IDQ4Mi4yMjEsMTIyLjM0OSBMNDYyLjg3NywxMjIuMzQ5IEw0NjIuODc3LDExNi44MjQgQzQ2Mi44NzcsMTEwLjcxOSA0NTcuOTMsMTA1Ljc2OSA0NTEuODI0LDEwNS43NjkgQzQ0NS43MTgsMTA1Ljc2OSA0NDAuNzY5LDExMC43MTggNDQwLjc2OSwxMTYuODI0IEw0NDAuNzY5LDE2Mi40MjkgTDQ0MC44LDE2Ni41OTkgQzQ0MC44LDE4Ny45NTIgNDU4LjExMywyMDUuMjYyIDQ3OS40NjYsMjA1LjI2MiBMNDgzLjcwMiwyMDUuMjYzIEM0ODkuODA3LDIwNS4yNjMgNDk0Ljc1NywyMDAuMzE1IDQ5NC43NTcsMTk0LjIwOCBDNDk0Ljc1OCwxODguMTA1IDQ4OS44MDksMTgzLjE1NiA0ODMuNzAzLDE4My4xNTYgWiI+PC9wYXRoPjxwYXRoIGQ9Ik0yOTIuOTgxLDEzNS4wNDIgQzI5Mi45ODEsMTEzLjY4OCAyNzUuNjcsOTYuMzc4IDI1NC4zMTcsOTYuMzc4IEwyNTAuMTQ3LDk2LjM0NiBMMjIyLjQ1OCw5Ni4zNDYgQzIxNi4zNTMsOTYuMzQ2IDIxMS40MDQsMTAxLjI5NSAyMTEuNDA0LDEwNy40IEwyMTEuNDA0LDE5NC4yIEMyMTEuNDA0LDIwMC4zMDYgMjE2LjM1MywyMDUuMjU2IDIyMi40NTgsMjA1LjI1NiBDMjI4LjU2MiwyMDUuMjU2IDIzMy41MTIsMjAwLjMwNyAyMzMuNTEyLDE5NC4yIEwyMzMuNTEyLDExOC40NTUgTDI1MC4xNDgsMTE4LjQ1NSBMMjU0LjMxNywxMTguNDg3IEMyNjMuNDYxLDExOC40ODcgMjcwLjg3NCwxMjUuODk5IDI3MC44NzQsMTM1LjA0MyBDMjcwLjg3NCwxNDQuMTg3IDI2My40NjEsMTUxLjYgMjU0LjMxNywxNTEuNiBMMjU1LjY3NCwxNTEuNjE2IEwyNTEuNTIsMTUxLjYxNiBDMjQ1LjQxNiwxNTEuNjE2IDI0MC40NjYsMTU2LjU2MyAyNDAuNDY2LDE2Mi42NyBDMjQwLjQ2NiwxNjguNzc0IDI0NS40MTUsMTczLjcyNCAyNTEuNTIsMTczLjcyNCBMMjU1LjY3NCwxNzMuNzI0IEwyNTQuMzE4LDE3My43MDggQzI3NS42NzEsMTczLjcwNiAyOTIuOTgxLDE1Ni4zOTYgMjkyLjk4MSwxMzUuMDQyIFoiPjwvcGF0aD48cGF0aCBkPSJNNDYyLjYwNywyNi4zMTkgQzQ2Mi42MDcsMzEuNDk0IDQ1OC40MSwzNS42ODkgNDUzLjIzNiwzNS42ODkgQzQ0OC4wNiwzNS42ODkgNDQzLjg2NSwzMS40OTQgNDQzLjg2NSwyNi4zMTkgQzQ0My44NjUsMjEuMTQ0IDQ0OC4wNiwxNi45NDggNDUzLjIzNiwxNi45NDggQzQ1OC40MSwxNi45NDggNDYyLjYwNywyMS4xNDUgNDYyLjYwNywyNi4zMTkgWiI+PC9wYXRoPjxwYXRoIGQ9Ik00MzAuNTE0LDU0LjY4OCBDNDMwLjUxNCw1OC44MjggNDI3LjE1Nyw2Mi4xODUgNDIzLjAxNiw2Mi4xODUgQzQxOC44NzcsNjIuMTg1IDQxNS41Miw1OC44MjggNDE1LjUyLDU0LjY4OCBDNDE1LjUyLDUwLjU0OCA0MTguODc3LDQ3LjE5MDk5OTggNDIzLjAxNiw0Ny4xOTA5OTk4IEM0MjcuMTU2LDQ3LjE5IDQzMC41MTQsNTAuNTQ4IDQzMC41MTQsNTQuNjg4IFoiPjwvcGF0aD48cGF0aCBkPSJNNTA1LjU1NywxMS43MTMgQzUwNS41NTcsMTguMTgzIDUwMC4zMTEsMjMuNDI2IDQ5My44NDIsMjMuNDI2IEM0ODcuMzczLDIzLjQyNiA0ODIuMTI5LDE4LjE4MyA0ODIuMTI5LDExLjcxMyBDNDgyLjEyOSw1LjI0MyA0ODcuMzczLDAgNDkzLjg0MiwwIEM1MDAuMzExLDAgNTA1LjU1Nyw1LjI0MyA1MDUuNTU3LDExLjcxMyBaIj48L3BhdGg+PHBhdGggZD0iTTU1NS4xMTksMTYuMzE4IEM1NTUuMTE5LDI0LjQwNCA1NDguNTY0LDMwLjk2IDU0MC40NzYsMzAuOTYgQzUzMi4zOTIsMzAuOTYgNTI1LjgzNSwyNC40MDQgNTI1LjgzNSwxNi4zMTggQzUyNS44MzUsOC4yMzEgNTMyLjM5MiwxLjY3NiA1NDAuNDc2LDEuNjc2IEM1NDguNTY0LDEuNjc3IDU1NS4xMTksOC4yMzEgNTU1LjExOSwxNi4zMTggWiI+PC9wYXRoPjxwYXRoIGQ9Ik02MDIuMTAyLDQyLjA0NiBDNjAyLjEwMiw1Mi4xNTMgNTkzLjkwOSw2MC4zNDcgNTgzLjgwMSw2MC4zNDcgQzU3My42OTIsNjAuMzQ3IDU2NS40OTgsNTIuMTU0IDU2NS40OTgsNDIuMDQ2IEM1NjUuNDk4LDMxLjkzNyA1NzMuNjkxLDIzLjc0NCA1ODMuODAxLDIzLjc0NCBDNTkzLjkwOCwyMy43NDQgNjAyLjEwMiwzMS45MzcgNjAyLjEwMiw0Mi4wNDYgWiI+PC9wYXRoPjxwYXRoIGQ9Ik02MzUuNDM2LDkwLjI4MiBDNjM1LjQzNiwxMDEuOTA3IDYyNi4wMTIsMTExLjMyOSA2MTQuMzg5LDExMS4zMjkgQzYwMi43NjYsMTExLjMyOSA1OTMuMzQyLDEwMS45MDcgNTkzLjM0Miw5MC4yODIgQzU5My4zNDIsNzguNjU3IDYwMi43NjYsNjkuMjM1IDYxNC4zODksNjkuMjM1IEM2MjYuMDEyLDY5LjIzNSA2MzUuNDM2LDc4LjY1NyA2MzUuNDM2LDkwLjI4MiBaIj48L3BhdGg+PHBhdGggZD0iTTYzOS4zODcsMTUzLjU5NSBDNjM5LjM4NywxNjYuOTYgNjI4LjU1MSwxNzcuNzk2IDYxNS4xODQsMTc3Ljc5NiBDNjAxLjgxNSwxNzcuNzk2IDU5MC45NzksMTY2Ljk2IDU5MC45NzksMTUzLjU5NSBDNTkwLjk3OSwxNDAuMjMgNjAxLjgxNSwxMjkuMzk1IDYxNS4xODQsMTI5LjM5NSBDNjI4LjU1MSwxMjkuMzk1IDYzOS4zODcsMTQwLjIyOSA2MzkuMzg3LDE1My41OTUgWiI+PC9wYXRoPjxwYXRoIGQ9Ik02MTIuNzY2LDIxNS44ODYgQzYxMi43NjYsMjMxLjMyNiA2MDAuMjUsMjQzLjg0MSA1ODQuODExLDI0My44NDEgQzU2OS4zNzIsMjQzLjg0MSA1NTYuODU2LDIzMS4zMjYgNTU2Ljg1NiwyMTUuODg2IEM1NTYuODU2LDIwMC40NDUgNTY5LjM3MiwxODcuOTMxIDU4NC44MTEsMTg3LjkzMSBDNjAwLjI1LDE4Ny45MzEgNjEyLjc2NiwyMDAuNDQ0IDYxMi43NjYsMjE1Ljg4NiBaIj48L3BhdGg+PHBhdGggZD0iTTU1MS43NzMsMjU5LjE5NCBDNTUxLjc3MywyNzYuNTAyIDUzNy43NDQsMjkwLjUzMSA1MjAuNDM3LDI5MC41MzEgQzUwMy4xMywyOTAuNTMxIDQ4OS4xMDEsMjc2LjUwMiA0ODkuMTAxLDI1OS4xOTQgQzQ4OS4xMDEsMjQxLjg4NiA1MDMuMTMsMjI3Ljg1OCA1MjAuNDM3LDIyNy44NTggQzUzNy43NDQsMjI3Ljg1OCA1NTEuNzczLDI0MS44ODcgNTUxLjc3MywyNTkuMTk0IFoiPjwvcGF0aD48cGF0aCBkPSJNNDcyLjEwNywyNTcuMzggQzQ3Mi4xMDcsMjc2Ljc3NCA0NTYuMzg0LDI5Mi40OTUgNDM2Ljk5MiwyOTIuNDk1IEM0MTcuNiwyOTIuNDk1IDQwMS44NzcsMjc2Ljc3MyA0MDEuODc3LDI1Ny4zOCBDNDAxLjg3NywyMzcuOTg0IDQxNy42LDIyMi4yNjcgNDM2Ljk5MiwyMjIuMjY3IEM0NTYuMzg0LDIyMi4yNjcgNDcyLjEwNywyMzcuOTg0IDQ3Mi4xMDcsMjU3LjM4IFoiPjwvcGF0aD48cGF0aCBkPSJNMTY1LjgwNSwxNTkuNjU4IEwxNjUuODA1LDE5NC4yMSBDMTY1LjgwNSwyMDAuMzE2IDE2MC44NTYsMjA1LjI2NSAxNTQuNzUxLDIwNS4yNjUgQzE0OC42NDcsMjA1LjI2NSAxNDMuNjk3LDIwMC4zMTcgMTQzLjY5NywxOTQuMjEgTDE0My42OTcsMTU5LjY1OCBMMTQzLjY4MSwxNjEuMDE0IEMxNDMuNjgxLDE1MS44NzEgMTM2LjI2OSwxNDQuNDU4IDEyNy4xMjUsMTQ0LjQ1OCBDMTE3Ljk4MSwxNDQuNDU4IDExMC41NjksMTUxLjg3MSAxMTAuNTY5LDE2MS4wMTQgTDExMC41MzYsMTY1LjE4NCBMMTEwLjUzNiwxOTQuMjA5IEMxMTAuNTM2LDIwMC4zMTUgMTA1LjU4NywyMDUuMjY0IDk5LjQ4MiwyMDUuMjY0IEM5My4zNzcsMjA1LjI2NCA4OC40MjgsMjAwLjMxNiA4OC40MjgsMTk0LjIwOSBMODguNDI4LDE2NS4xODQgTDg4LjQ2LDE2MS4wMTQgQzg4LjQ2LDEzOS42NiAxMDUuNzcsMTIyLjM1IDEyNy4xMjQsMTIyLjM1IEMxNDguNDc4LDEyMi4zNSAxNjUuNzg4LDEzOS42NjEgMTY1Ljc4OCwxNjEuMDE0IEwxNjUuODA1LDE1OS42NTggWiI+PC9wYXRoPjxwYXRoIGQ9Ik0wLDE2Ny45NTcgTDAsMTA3LjM5OSBDMCwxMDEuMjk1IDQuOTQ5LDk2LjM0NSAxMS4wNTQsOTYuMzQ1IEMxNy4xNTksOTYuMzQ1IDIyLjEwOCwxMDEuMjk0IDIyLjEwOCwxMDcuMzk5IEwyMi4xMDgsMTY3Ljk1NyBMMjIuMTIzLDE2Ni42MDEgQzIyLjEyMywxNzUuNzQ0IDI5LjUzNiwxODMuMTU3IDM4LjY3OSwxODMuMTU3IEM0Ny44MjIsMTgzLjE1NyA1NS4yMzUsMTc1Ljc0NCA1NS4yMzUsMTY2LjYwMSBMNTUuMjY4LDE2Mi40MzEgTDU1LjI2OCwxMDcuNCBDNTUuMjY4LDEwMS4yOTYgNjAuMjE3LDk2LjM0NSA2Ni4zMjIsOTYuMzQ1IEM3Mi40MjcsOTYuMzQ1IDc3LjM3NiwxMDEuMjk1IDc3LjM3NiwxMDcuNCBMNzcuMzc2LDE2Mi40MzEgTDc3LjM0NCwxNjYuNjAyIEM3Ny4zNDQsMTg3Ljk1NiA2MC4wMzQsMjA1LjI2NSAzOC42OCwyMDUuMjY1IEMxNy4zMjYsMjA1LjI2NSAwLjAxNiwxODcuOTU1IDAuMDE2LDE2Ni42MDIgTDAsMTY3Ljk1NyBaIj48L3BhdGg+PHBhdGggZD0iTTE5OS4zMTQsMTk0LjIxIEMxOTkuMzE0LDIwMC4zMTYgMTk0LjM2NSwyMDUuMjY1IDE4OC4yNiwyMDUuMjY1IEMxODIuMTU2LDIwNS4yNjUgMTc3LjIwNiwyMDAuMzE3IDE3Ny4yMDYsMTk0LjIxIEwxNzcuMjA2LDEzMy40NzggQzE3Ny4yMDYsMTI3LjM3MyAxODIuMTU1LDEyMi40MjMgMTg4LjI2LDEyMi40MjMgQzE5NC4zNjUsMTIyLjQyMyAxOTkuMzE0LDEyNy4zNzIgMTk5LjMxNCwxMzMuNDc4IEwxOTkuMzE0LDE5NC4yMSBaIj48L3BhdGg+PC9nPjwvc3ZnPgo=\" title=\"UniProt home\"/></a><a href=\"/sparql\">SPARQL</a><a href=\"https://www.uniprot.org/downloads\" id=\"download\">Downloads</a><a accesskey=\"9\" href=\"https://www.uniprot.org/contact\" id=\"contact\"><img alt=\"Contact\" src=\"data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZpZXdCb3g9IjAgMCAxMyAxMCIgZmlsbD0iI2ZmZiIgd2lkdGg9IjEuNGVtIj48cGF0aCBmaWxsLXJ1bGU9ImV2ZW5vZGQiIGQ9Ik0xMiAxLjI1QzEyIC41NiAxMS40NiAwIDEwLjggMEgxLjJDLjU0IDAgMCAuNTYgMCAxLjI1djcuNUMwIDkuNDQuNTQgMTAgMS4yIDEwaDkuNmMuNjYgMCAxLjItLjU2IDEuMi0xLjI1di03LjV6bS0xLjIgMEw2IDQuMzggMS4yIDEuMjVoOS42ek02IDUuNjNMMS4yIDIuNXY2LjI1aDkuNlYyLjVMNiA1LjYzeiI+PC9wYXRoPjwvc3ZnPgo=\"/></a><a rel=\"help\" href=\"https://www.uniprot.org/help/\" id=\"help\">Documentation/Help</a></nav><a accesskey=\"2\" href=\"#content\" hidden=\"hidden\">Skip Header</a><main class=\"sparql results\" id=\"content\"><section class=\"error\" id=\"page-header\"><h2 class=\"page-title\">Malformed query.</h2><p>org.eclipse.rdf4j.query.parser.sparql.ast.VisitorException: QName \\'up:Protein\\' uses an undefined prefix</p></section><div class=\"main-aside\"><div id=\"queryform\"><form method=\"get\" action=\"/sparql/\" focus=\"query\" name=\"sparql-form\" id=\"sparql-form\"><button id=\"addPrefix\" type=\"button\">Add common prefixes</button><textarea accesskey=\"1\" rows=\"10\" cols=\"80\" name=\"query\" id=\"query\"># https://sparql.uniprot.org/sparql\\n\\nSELECT DISTINCT ?protein\\nWHERE {\\n  ?protein a up:Protein ;\\n           up:organism &lt;http://purl.uniprot.org/taxonomy/9606&gt; ;\\n           up:encodedBy &lt;http://purl.uniprot.org/uniprot/LCT&gt; .\\n}</textarea><input type=\"hidden\" name=\"format\" value=\"html\"><button id=\"sparqlFormSubmitter\" onclick=\"return false;\">Submit Query</button></form></div></div></main><footer id=\"page-footer\"><section class=\"small\" id=\"copyright\">\\n\\t\\t\\t&copy; 2002&ndash; <span property=\"schema:copyrightYear\">2024</span>\\n\\t\\t\\t<a accesskey=\"9\" href=\"https://www.uniprot.org/help/about\">UniProt Consortium</a> |\\n\\t\\t\\t<a href=\"https://www.uniprot.org/help/license\" property=\"schema:license\">License\\n\\t\\t\\t\\t&amp; Disclaimer</a> | <a href=\"https://www.uniprot.org/help/privacy\">Privacy\\n\\t\\t\\t\\tNotice</a></section><section><div resource=\"/help/about#UniProt\" id=\"logos\" typeof=\"NGO\" property=\"creator\"><span resource=\"/help/about#EBI\" typeof=\"Organization\" class=\"ebi\" property=\"member\"><span hidden=\"hidden\" property=\"name\">European Bioinformatics Institute</span><a href=\"https://www.ebi.ac.uk/\" property=\"url\"></a></span><span resource=\"/help/about#PIR\" typeof=\"Organization\" class=\"pir\" property=\"member\"><span hidden=\"hidden\" property=\"name\">Protein Information Resource</span><a href=\"http://pir.georgetown.edu/\" property=\"url\"></a></span><span resource=\"/help/about#SIB\" typeof=\"Organization\" class=\"sib\" property=\"member\"><span hidden=\"hidden\" property=\"name\">SIB Swiss Institute of Bioinformatics</span><a href=\"https://www.sib.swiss/\" property=\"url\"></a></span></div></section><section id=\"poweredby\"><h3>Powered by</h3><span class=\"vitalit\" resource=\"https://www.sib.swiss/alan-bridge-group/\" typeof=\"schema:NGO\" property=\"schema:sponsor\"><h4>Hardware</h4><a href=\"https://www.sib.swiss/alan-bridge-group\" property=\"schema:url\"><img alt=\"Swiss-Prot\" src=\"https://www.sib.swiss//templates/sib/images/SIB_LogoQ_GBv.svg\"/><h5 property=\"schema:name\">Swiss-Prot</h5></a></span><span class=\"openlinksw\" resource=\"http://www.openlinksw.com/#this\" typeof=\"schema:Corporation\" property=\"schema:sponsor\"><h4>Software</h4><a href=\"http://virtuoso.openlinksw.com/\" property=\"schema:url\"><img alt=\"Virtuoso\" src=\"/images/logos/virtuoso.png\"/><h5 property=\"schema:name\">Virtuoso</h5></a></span></section></footer><script src=\"/sparql.js\" type=\"text/javascript\"> </script></body></html>'. Generated query:\n",
      "# https://sparql.uniprot.org/sparql\n",
      "\n",
      "SELECT DISTINCT ?protein\n",
      "WHERE {\n",
      "  ?protein a up:Protein ;\n",
      "           up:organism <http://purl.uniprot.org/taxonomy/9606> ;\n",
      "           up:encodedBy <http://purl.uniprot.org/uniprot/LCT> .\n",
      "}\n",
      "Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
      "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
      "SELECT DISTINCT ?protein\n",
      "WHERE{\n",
      "    ?protein a up:Protein .\n",
      "    ?protein up:organism taxon:9606 .\n",
      "    ?protein up:encodedBy ?gene .\n",
      "    ?gene skos:prefLabel \"LCT\" .\n",
      "}\n",
      "\n",
      "‚ùå 3/3 What is the accession number in uniprot of the human gene LCT? Return only unique protein URIs QueryBadFormed: A bad request has been sent to the endpoint: probably the SPARQL query is badly formed. \n",
      "\n",
      "Response:\n",
      "b'<!DOCTYPE html SYSTEM \"about:legacy-compat\">\\n<html xmlns=\"http://www.w3.org/1999/xhtml\" lang=\"en\" xml:lang=\"en\"><head><title>UniProt</title><meta content=\"text/html; charset=UTF-8\" http-equiv=\"Content-Type\"/><link href=\"/\" rel=\"home\"/><link href=\"/base.css\" type=\"text/css\" rel=\"stylesheet\"/><link type=\"image/vnd.microsoft.icon\" href=\"/uniprot-favicon.ico\" rel=\"shortcut icon\"/><link href=\"/uniprot-sparql.css\" type=\"text/css\" rel=\"stylesheet\"/><link href=\"http://creativecommons.org/licenses/by/4.0/\" rel=\"license\"/><script src=\"https://cdn.jsdelivr.net/npm/yasgui-yasqe@2.11.22/dist/yasqe.bundled.min.js\"></script><link type=\"text/css\" rel=\"stylesheet\" href=\"https://cdn.jsdelivr.net/yasqe/2.2/yasqe.min.css\"/><meta content=\"width=device-width, initial-scale=1\" name=\"viewport\"/><meta content=\"no-referrer\" name=\"referrer\"/></head><body typeof=\"schema:WebPage\"><nav id=\"nav-header\"><a accesskey=\"1\" href=\"https://www.uniprot.org/\" id=\"www\"><img alt=\"\" src=\"data:image/svg+xml;base64,PHN2ZyB2aWV3Qm94PSIwIDAgNjQwIDI5MyIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj48ZyBmaWxsPSIjZmZmIiBmaWxsLXJ1bGU9Im5vbnplcm8iPjxwYXRoIGQ9Ik0xOTkuNjU5LDEwNi4xMDMgQzE5OS42NTksMTEyLjM5OCAxOTQuNTU2LDExNy41MDEgMTg4LjI2LDExNy41MDEgQzE4MS45NjQsMTE3LjUwMSAxNzYuODYxLDExMi4zOTcgMTc2Ljg2MSwxMDYuMTAzIEMxNzYuODYxLDk5LjgwNyAxODEuOTY1LDk0LjcwNCAxODguMjYsOTQuNzA0IEMxOTQuNTU1LDk0LjcwNCAxOTkuNjU5LDk5LjgwNyAxOTkuNjU5LDEwNi4xMDMgWiI+PC9wYXRoPjxwYXRoIGQ9Ik0zMzQuMzgzLDE0NC40NTkgQzMyNS4yNDIsMTQ0LjQ1OSAzMTcuODI5LDE1MS44NzIgMzE3LjgyOSwxNjEuMDE1IEwzMTcuNzk3LDE2NS4xODUgTDMxNy43OTcsMTk0LjIxIEMzMTcuNzk3LDIwMC4zMTYgMzEyLjg0OCwyMDUuMjY1IDMwNi43NDIsMjA1LjI2NSBDMzAwLjYzOCwyMDUuMjY1IDI5NS42ODksMjAwLjMxNyAyOTUuNjg5LDE5NC4yMSBMMjk1LjY4OSwxNjUuMTg1IEwyOTUuNzIsMTYxLjAxNSBDMjk1LjcyLDEzOS42NjEgMzEzLjAzLDEyMi4zNTEgMzM0LjM4MywxMjIuMzUxIEwzNDEuMjg1LDEyMi4zNTEgQzM0Ny4zOSwxMjIuMzUxIDM1Mi4zNCwxMjcuMyAzNTIuMzQsMTMzLjQwNSBDMzUyLjM0LDEzOS41MDkgMzQ3LjM5MSwxNDQuNDU5IDM0MS4yODUsMTQ0LjQ1OSBMMzM0LjM4MywxNDQuNDU5IFoiPjwvcGF0aD48cGF0aCBkPSJNNDMzLjg2MSwxNjMuODEzIEM0MzMuODYxLDE4Ny40MTcgNDE0LjcyNCwyMDYuNTU1IDM5MS4xMTksMjA2LjU1NSBDMzY3LjUxNCwyMDYuNTU1IDM0OC4zNzUsMTg3LjQxNyAzNDguMzc1LDE2My44MTMgQzM0OC4zNzUsMTQwLjIwNiAzNjcuNTE0LDEyMS4wNyAzOTEuMTE5LDEyMS4wNyBDNDE0LjcyNCwxMjEuMDcgNDMzLjg2MSwxNDAuMjA2IDQzMy44NjEsMTYzLjgxMyBaIE0zOTEuMDI5LDE0MS42OTQgQzM3OC44MTgsMTQxLjY5NCAzNjguOTIsMTUxLjU5MiAzNjguOTIsMTYzLjgwMSBDMzY4LjkyLDE3Ni4wMTEgMzc4LjgxOCwxODUuOTA5IDM5MS4wMjksMTg1LjkwOSBDNDAzLjIzOCwxODUuOTA5IDQxMy4xMzYsMTc2LjAxMSA0MTMuMTM2LDE2My44MDEgQzQxMy4xMzcsMTUxLjU5MyA0MDMuMjM4LDE0MS42OTQgMzkxLjAyOSwxNDEuNjk0IFoiPjwvcGF0aD48cGF0aCBkPSJNNDgzLjcwMywxODMuMTU2IEw0NzkuNDY3LDE4My4xNTUgQzQ3MC4zMjIsMTgzLjE1NSA0NjIuOTEsMTc1Ljc0MyA0NjIuOTEsMTY2LjU5OCBMNDYyLjg3NywxNjIuNDMgTDQ2Mi44NzcsMTQ0LjQ1NyBMNDgyLjIyMSwxNDQuNDU4IEM0ODguMzI2LDE0NC40NTggNDkzLjI3NiwxMzkuNTEgNDkzLjI3NiwxMzMuNDA0IEM0OTMuMjc2LDEyNy4yOTggNDg4LjMyNywxMjIuMzQ5IDQ4Mi4yMjEsMTIyLjM0OSBMNDYyLjg3NywxMjIuMzQ5IEw0NjIuODc3LDExNi44MjQgQzQ2Mi44NzcsMTEwLjcxOSA0NTcuOTMsMTA1Ljc2OSA0NTEuODI0LDEwNS43NjkgQzQ0NS43MTgsMTA1Ljc2OSA0NDAuNzY5LDExMC43MTggNDQwLjc2OSwxMTYuODI0IEw0NDAuNzY5LDE2Mi40MjkgTDQ0MC44LDE2Ni41OTkgQzQ0MC44LDE4Ny45NTIgNDU4LjExMywyMDUuMjYyIDQ3OS40NjYsMjA1LjI2MiBMNDgzLjcwMiwyMDUuMjYzIEM0ODkuODA3LDIwNS4yNjMgNDk0Ljc1NywyMDAuMzE1IDQ5NC43NTcsMTk0LjIwOCBDNDk0Ljc1OCwxODguMTA1IDQ4OS44MDksMTgzLjE1NiA0ODMuNzAzLDE4My4xNTYgWiI+PC9wYXRoPjxwYXRoIGQ9Ik0yOTIuOTgxLDEzNS4wNDIgQzI5Mi45ODEsMTEzLjY4OCAyNzUuNjcsOTYuMzc4IDI1NC4zMTcsOTYuMzc4IEwyNTAuMTQ3LDk2LjM0NiBMMjIyLjQ1OCw5Ni4zNDYgQzIxNi4zNTMsOTYuMzQ2IDIxMS40MDQsMTAxLjI5NSAyMTEuNDA0LDEwNy40IEwyMTEuNDA0LDE5NC4yIEMyMTEuNDA0LDIwMC4zMDYgMjE2LjM1MywyMDUuMjU2IDIyMi40NTgsMjA1LjI1NiBDMjI4LjU2MiwyMDUuMjU2IDIzMy41MTIsMjAwLjMwNyAyMzMuNTEyLDE5NC4yIEwyMzMuNTEyLDExOC40NTUgTDI1MC4xNDgsMTE4LjQ1NSBMMjU0LjMxNywxMTguNDg3IEMyNjMuNDYxLDExOC40ODcgMjcwLjg3NCwxMjUuODk5IDI3MC44NzQsMTM1LjA0MyBDMjcwLjg3NCwxNDQuMTg3IDI2My40NjEsMTUxLjYgMjU0LjMxNywxNTEuNiBMMjU1LjY3NCwxNTEuNjE2IEwyNTEuNTIsMTUxLjYxNiBDMjQ1LjQxNiwxNTEuNjE2IDI0MC40NjYsMTU2LjU2MyAyNDAuNDY2LDE2Mi42NyBDMjQwLjQ2NiwxNjguNzc0IDI0NS40MTUsMTczLjcyNCAyNTEuNTIsMTczLjcyNCBMMjU1LjY3NCwxNzMuNzI0IEwyNTQuMzE4LDE3My43MDggQzI3NS42NzEsMTczLjcwNiAyOTIuOTgxLDE1Ni4zOTYgMjkyLjk4MSwxMzUuMDQyIFoiPjwvcGF0aD48cGF0aCBkPSJNNDYyLjYwNywyNi4zMTkgQzQ2Mi42MDcsMzEuNDk0IDQ1OC40MSwzNS42ODkgNDUzLjIzNiwzNS42ODkgQzQ0OC4wNiwzNS42ODkgNDQzLjg2NSwzMS40OTQgNDQzLjg2NSwyNi4zMTkgQzQ0My44NjUsMjEuMTQ0IDQ0OC4wNiwxNi45NDggNDUzLjIzNiwxNi45NDggQzQ1OC40MSwxNi45NDggNDYyLjYwNywyMS4xNDUgNDYyLjYwNywyNi4zMTkgWiI+PC9wYXRoPjxwYXRoIGQ9Ik00MzAuNTE0LDU0LjY4OCBDNDMwLjUxNCw1OC44MjggNDI3LjE1Nyw2Mi4xODUgNDIzLjAxNiw2Mi4xODUgQzQxOC44NzcsNjIuMTg1IDQxNS41Miw1OC44MjggNDE1LjUyLDU0LjY4OCBDNDE1LjUyLDUwLjU0OCA0MTguODc3LDQ3LjE5MDk5OTggNDIzLjAxNiw0Ny4xOTA5OTk4IEM0MjcuMTU2LDQ3LjE5IDQzMC41MTQsNTAuNTQ4IDQzMC41MTQsNTQuNjg4IFoiPjwvcGF0aD48cGF0aCBkPSJNNTA1LjU1NywxMS43MTMgQzUwNS41NTcsMTguMTgzIDUwMC4zMTEsMjMuNDI2IDQ5My44NDIsMjMuNDI2IEM0ODcuMzczLDIzLjQyNiA0ODIuMTI5LDE4LjE4MyA0ODIuMTI5LDExLjcxMyBDNDgyLjEyOSw1LjI0MyA0ODcuMzczLDAgNDkzLjg0MiwwIEM1MDAuMzExLDAgNTA1LjU1Nyw1LjI0MyA1MDUuNTU3LDExLjcxMyBaIj48L3BhdGg+PHBhdGggZD0iTTU1NS4xMTksMTYuMzE4IEM1NTUuMTE5LDI0LjQwNCA1NDguNTY0LDMwLjk2IDU0MC40NzYsMzAuOTYgQzUzMi4zOTIsMzAuOTYgNTI1LjgzNSwyNC40MDQgNTI1LjgzNSwxNi4zMTggQzUyNS44MzUsOC4yMzEgNTMyLjM5MiwxLjY3NiA1NDAuNDc2LDEuNjc2IEM1NDguNTY0LDEuNjc3IDU1NS4xMTksOC4yMzEgNTU1LjExOSwxNi4zMTggWiI+PC9wYXRoPjxwYXRoIGQ9Ik02MDIuMTAyLDQyLjA0NiBDNjAyLjEwMiw1Mi4xNTMgNTkzLjkwOSw2MC4zNDcgNTgzLjgwMSw2MC4zNDcgQzU3My42OTIsNjAuMzQ3IDU2NS40OTgsNTIuMTU0IDU2NS40OTgsNDIuMDQ2IEM1NjUuNDk4LDMxLjkzNyA1NzMuNjkxLDIzLjc0NCA1ODMuODAxLDIzLjc0NCBDNTkzLjkwOCwyMy43NDQgNjAyLjEwMiwzMS45MzcgNjAyLjEwMiw0Mi4wNDYgWiI+PC9wYXRoPjxwYXRoIGQ9Ik02MzUuNDM2LDkwLjI4MiBDNjM1LjQzNiwxMDEuOTA3IDYyNi4wMTIsMTExLjMyOSA2MTQuMzg5LDExMS4zMjkgQzYwMi43NjYsMTExLjMyOSA1OTMuMzQyLDEwMS45MDcgNTkzLjM0Miw5MC4yODIgQzU5My4zNDIsNzguNjU3IDYwMi43NjYsNjkuMjM1IDYxNC4zODksNjkuMjM1IEM2MjYuMDEyLDY5LjIzNSA2MzUuNDM2LDc4LjY1NyA2MzUuNDM2LDkwLjI4MiBaIj48L3BhdGg+PHBhdGggZD0iTTYzOS4zODcsMTUzLjU5NSBDNjM5LjM4NywxNjYuOTYgNjI4LjU1MSwxNzcuNzk2IDYxNS4xODQsMTc3Ljc5NiBDNjAxLjgxNSwxNzcuNzk2IDU5MC45NzksMTY2Ljk2IDU5MC45NzksMTUzLjU5NSBDNTkwLjk3OSwxNDAuMjMgNjAxLjgxNSwxMjkuMzk1IDYxNS4xODQsMTI5LjM5NSBDNjI4LjU1MSwxMjkuMzk1IDYzOS4zODcsMTQwLjIyOSA2MzkuMzg3LDE1My41OTUgWiI+PC9wYXRoPjxwYXRoIGQ9Ik02MTIuNzY2LDIxNS44ODYgQzYxMi43NjYsMjMxLjMyNiA2MDAuMjUsMjQzLjg0MSA1ODQuODExLDI0My44NDEgQzU2OS4zNzIsMjQzLjg0MSA1NTYuODU2LDIzMS4zMjYgNTU2Ljg1NiwyMTUuODg2IEM1NTYuODU2LDIwMC40NDUgNTY5LjM3MiwxODcuOTMxIDU4NC44MTEsMTg3LjkzMSBDNjAwLjI1LDE4Ny45MzEgNjEyLjc2NiwyMDAuNDQ0IDYxMi43NjYsMjE1Ljg4NiBaIj48L3BhdGg+PHBhdGggZD0iTTU1MS43NzMsMjU5LjE5NCBDNTUxLjc3MywyNzYuNTAyIDUzNy43NDQsMjkwLjUzMSA1MjAuNDM3LDI5MC41MzEgQzUwMy4xMywyOTAuNTMxIDQ4OS4xMDEsMjc2LjUwMiA0ODkuMTAxLDI1OS4xOTQgQzQ4OS4xMDEsMjQxLjg4NiA1MDMuMTMsMjI3Ljg1OCA1MjAuNDM3LDIyNy44NTggQzUzNy43NDQsMjI3Ljg1OCA1NTEuNzczLDI0MS44ODcgNTUxLjc3MywyNTkuMTk0IFoiPjwvcGF0aD48cGF0aCBkPSJNNDcyLjEwNywyNTcuMzggQzQ3Mi4xMDcsMjc2Ljc3NCA0NTYuMzg0LDI5Mi40OTUgNDM2Ljk5MiwyOTIuNDk1IEM0MTcuNiwyOTIuNDk1IDQwMS44NzcsMjc2Ljc3MyA0MDEuODc3LDI1Ny4zOCBDNDAxLjg3NywyMzcuOTg0IDQxNy42LDIyMi4yNjcgNDM2Ljk5MiwyMjIuMjY3IEM0NTYuMzg0LDIyMi4yNjcgNDcyLjEwNywyMzcuOTg0IDQ3Mi4xMDcsMjU3LjM4IFoiPjwvcGF0aD48cGF0aCBkPSJNMTY1LjgwNSwxNTkuNjU4IEwxNjUuODA1LDE5NC4yMSBDMTY1LjgwNSwyMDAuMzE2IDE2MC44NTYsMjA1LjI2NSAxNTQuNzUxLDIwNS4yNjUgQzE0OC42NDcsMjA1LjI2NSAxNDMuNjk3LDIwMC4zMTcgMTQzLjY5NywxOTQuMjEgTDE0My42OTcsMTU5LjY1OCBMMTQzLjY4MSwxNjEuMDE0IEMxNDMuNjgxLDE1MS44NzEgMTM2LjI2OSwxNDQuNDU4IDEyNy4xMjUsMTQ0LjQ1OCBDMTE3Ljk4MSwxNDQuNDU4IDExMC41NjksMTUxLjg3MSAxMTAuNTY5LDE2MS4wMTQgTDExMC41MzYsMTY1LjE4NCBMMTEwLjUzNiwxOTQuMjA5IEMxMTAuNTM2LDIwMC4zMTUgMTA1LjU4NywyMDUuMjY0IDk5LjQ4MiwyMDUuMjY0IEM5My4zNzcsMjA1LjI2NCA4OC40MjgsMjAwLjMxNiA4OC40MjgsMTk0LjIwOSBMODguNDI4LDE2NS4xODQgTDg4LjQ2LDE2MS4wMTQgQzg4LjQ2LDEzOS42NiAxMDUuNzcsMTIyLjM1IDEyNy4xMjQsMTIyLjM1IEMxNDguNDc4LDEyMi4zNSAxNjUuNzg4LDEzOS42NjEgMTY1Ljc4OCwxNjEuMDE0IEwxNjUuODA1LDE1OS42NTggWiI+PC9wYXRoPjxwYXRoIGQ9Ik0wLDE2Ny45NTcgTDAsMTA3LjM5OSBDMCwxMDEuMjk1IDQuOTQ5LDk2LjM0NSAxMS4wNTQsOTYuMzQ1IEMxNy4xNTksOTYuMzQ1IDIyLjEwOCwxMDEuMjk0IDIyLjEwOCwxMDcuMzk5IEwyMi4xMDgsMTY3Ljk1NyBMMjIuMTIzLDE2Ni42MDEgQzIyLjEyMywxNzUuNzQ0IDI5LjUzNiwxODMuMTU3IDM4LjY3OSwxODMuMTU3IEM0Ny44MjIsMTgzLjE1NyA1NS4yMzUsMTc1Ljc0NCA1NS4yMzUsMTY2LjYwMSBMNTUuMjY4LDE2Mi40MzEgTDU1LjI2OCwxMDcuNCBDNTUuMjY4LDEwMS4yOTYgNjAuMjE3LDk2LjM0NSA2Ni4zMjIsOTYuMzQ1IEM3Mi40MjcsOTYuMzQ1IDc3LjM3NiwxMDEuMjk1IDc3LjM3NiwxMDcuNCBMNzcuMzc2LDE2Mi40MzEgTDc3LjM0NCwxNjYuNjAyIEM3Ny4zNDQsMTg3Ljk1NiA2MC4wMzQsMjA1LjI2NSAzOC42OCwyMDUuMjY1IEMxNy4zMjYsMjA1LjI2NSAwLjAxNiwxODcuOTU1IDAuMDE2LDE2Ni42MDIgTDAsMTY3Ljk1NyBaIj48L3BhdGg+PHBhdGggZD0iTTE5OS4zMTQsMTk0LjIxIEMxOTkuMzE0LDIwMC4zMTYgMTk0LjM2NSwyMDUuMjY1IDE4OC4yNiwyMDUuMjY1IEMxODIuMTU2LDIwNS4yNjUgMTc3LjIwNiwyMDAuMzE3IDE3Ny4yMDYsMTk0LjIxIEwxNzcuMjA2LDEzMy40NzggQzE3Ny4yMDYsMTI3LjM3MyAxODIuMTU1LDEyMi40MjMgMTg4LjI2LDEyMi40MjMgQzE5NC4zNjUsMTIyLjQyMyAxOTkuMzE0LDEyNy4zNzIgMTk5LjMxNCwxMzMuNDc4IEwxOTkuMzE0LDE5NC4yMSBaIj48L3BhdGg+PC9nPjwvc3ZnPgo=\" title=\"UniProt home\"/></a><a href=\"/sparql\">SPARQL</a><a href=\"https://www.uniprot.org/downloads\" id=\"download\">Downloads</a><a accesskey=\"9\" href=\"https://www.uniprot.org/contact\" id=\"contact\"><img alt=\"Contact\" src=\"data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZpZXdCb3g9IjAgMCAxMyAxMCIgZmlsbD0iI2ZmZiIgd2lkdGg9IjEuNGVtIj48cGF0aCBmaWxsLXJ1bGU9ImV2ZW5vZGQiIGQ9Ik0xMiAxLjI1QzEyIC41NiAxMS40NiAwIDEwLjggMEgxLjJDLjU0IDAgMCAuNTYgMCAxLjI1djcuNUMwIDkuNDQuNTQgMTAgMS4yIDEwaDkuNmMuNjYgMCAxLjItLjU2IDEuMi0xLjI1di03LjV6bS0xLjIgMEw2IDQuMzggMS4yIDEuMjVoOS42ek02IDUuNjNMMS4yIDIuNXY2LjI1aDkuNlYyLjVMNiA1LjYzeiI+PC9wYXRoPjwvc3ZnPgo=\"/></a><a rel=\"help\" href=\"https://www.uniprot.org/help/\" id=\"help\">Documentation/Help</a></nav><a accesskey=\"2\" href=\"#content\" hidden=\"hidden\">Skip Header</a><main class=\"sparql results\" id=\"content\"><section class=\"error\" id=\"page-header\"><h2 class=\"page-title\">Malformed query.</h2><p>org.eclipse.rdf4j.query.parser.sparql.ast.VisitorException: QName \\'up:Protein\\' uses an undefined prefix</p></section><div class=\"main-aside\"><div id=\"queryform\"><form method=\"get\" action=\"/sparql/\" focus=\"query\" name=\"sparql-form\" id=\"sparql-form\"><button id=\"addPrefix\" type=\"button\">Add common prefixes</button><textarea accesskey=\"1\" rows=\"10\" cols=\"80\" name=\"query\" id=\"query\"># https://sparql.uniprot.org/sparql\\n\\nSELECT DISTINCT ?protein\\nWHERE {\\n  ?protein a up:Protein ;\\n           up:organism &lt;http://purl.uniprot.org/taxonomy/9606&gt; ;\\n           up:encodedBy &lt;http://purl.uniprot.org/uniprot/LCT&gt; .\\n}</textarea><input type=\"hidden\" name=\"format\" value=\"html\"><button id=\"sparqlFormSubmitter\" onclick=\"return false;\">Submit Query</button></form></div></div></main><footer id=\"page-footer\"><section class=\"small\" id=\"copyright\">\\n\\t\\t\\t&copy; 2002&ndash; <span property=\"schema:copyrightYear\">2024</span>\\n\\t\\t\\t<a accesskey=\"9\" href=\"https://www.uniprot.org/help/about\">UniProt Consortium</a> |\\n\\t\\t\\t<a href=\"https://www.uniprot.org/help/license\" property=\"schema:license\">License\\n\\t\\t\\t\\t&amp; Disclaimer</a> | <a href=\"https://www.uniprot.org/help/privacy\">Privacy\\n\\t\\t\\t\\tNotice</a></section><section><div resource=\"/help/about#UniProt\" id=\"logos\" typeof=\"NGO\" property=\"creator\"><span resource=\"/help/about#EBI\" typeof=\"Organization\" class=\"ebi\" property=\"member\"><span hidden=\"hidden\" property=\"name\">European Bioinformatics Institute</span><a href=\"https://www.ebi.ac.uk/\" property=\"url\"></a></span><span resource=\"/help/about#PIR\" typeof=\"Organization\" class=\"pir\" property=\"member\"><span hidden=\"hidden\" property=\"name\">Protein Information Resource</span><a href=\"http://pir.georgetown.edu/\" property=\"url\"></a></span><span resource=\"/help/about#SIB\" typeof=\"Organization\" class=\"sib\" property=\"member\"><span hidden=\"hidden\" property=\"name\">SIB Swiss Institute of Bioinformatics</span><a href=\"https://www.sib.swiss/\" property=\"url\"></a></span></div></section><section id=\"poweredby\"><h3>Powered by</h3><span class=\"vitalit\" resource=\"https://www.sib.swiss/alan-bridge-group/\" typeof=\"schema:NGO\" property=\"schema:sponsor\"><h4>Hardware</h4><a href=\"https://www.sib.swiss/alan-bridge-group\" property=\"schema:url\"><img alt=\"Swiss-Prot\" src=\"https://www.sib.swiss//templates/sib/images/SIB_LogoQ_GBv.svg\"/><h5 property=\"schema:name\">Swiss-Prot</h5></a></span><span class=\"openlinksw\" resource=\"http://www.openlinksw.com/#this\" typeof=\"schema:Corporation\" property=\"schema:sponsor\"><h4>Software</h4><a href=\"http://virtuoso.openlinksw.com/\" property=\"schema:url\"><img alt=\"Virtuoso\" src=\"/images/logos/virtuoso.png\"/><h5 property=\"schema:name\">Virtuoso</h5></a></span></section></footer><script src=\"/sparql.js\" type=\"text/javascript\"> </script></body></html>'. Generated query:\n",
      "# https://sparql.uniprot.org/sparql\n",
      "\n",
      "SELECT DISTINCT ?protein\n",
      "WHERE {\n",
      "  ?protein a up:Protein ;\n",
      "           up:organism <http://purl.uniprot.org/taxonomy/9606> ;\n",
      "           up:encodedBy <http://purl.uniprot.org/uniprot/LCT> .\n",
      "}\n",
      "Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
      "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
      "SELECT DISTINCT ?protein\n",
      "WHERE{\n",
      "    ?protein a up:Protein .\n",
      "    ?protein up:organism taxon:9606 .\n",
      "    ?protein up:encodedBy ?gene .\n",
      "    ?gene skos:prefLabel \"LCT\" .\n",
      "}\n",
      "\n",
      "‚úÖ 1/3 What is the accession number in uniprot of the human gene LCT? Return only unique protein URIs = 4\n",
      "‚úÖ 2/3 What is the accession number in uniprot of the human gene LCT? Return only unique protein URIs = 4\n",
      "‚úÖ 3/3 What is the accession number in uniprot of the human gene LCT? Return only unique protein URIs = 4\n",
      "‚úÖ 1/3 What is the accession number in uniprot of the human gene LCT? Return only unique protein URIs = 4\n",
      "‚úÖ 2/3 What is the accession number in uniprot of the human gene LCT? Return only unique protein URIs = 4\n",
      "‚úÖ 3/3 What is the accession number in uniprot of the human gene LCT? Return only unique protein URIs = 4\n",
      "üéØ No RAG - Success: 0, Mismatch: 0, Error: 3\n",
      "üéØ RAG without validation - Success: 3, Mismatch: 0, Error: 0\n",
      "üéØ RAG with validation - Success: 3, Mismatch: 0, Error: 0\n",
      "‚úÖ 1/3 How do I filter for reviewed mouse proteins which carry an N-terminal glycine? Return protein URI and AA sequence = 20\n",
      "‚úÖ 2/3 How do I filter for reviewed mouse proteins which carry an N-terminal glycine? Return protein URI and AA sequence = 20\n",
      "‚úÖ 3/3 How do I filter for reviewed mouse proteins which carry an N-terminal glycine? Return protein URI and AA sequence = 20\n",
      "‚úÖ 1/3 How do I filter for reviewed mouse proteins which carry an N-terminal glycine? Return protein URI and AA sequence = 20\n",
      "‚úÖ 2/3 How do I filter for reviewed mouse proteins which carry an N-terminal glycine? Return protein URI and AA sequence = 20\n",
      "‚úÖ 3/3 How do I filter for reviewed mouse proteins which carry an N-terminal glycine? Return protein URI and AA sequence = 20\n",
      "‚úÖ 1/3 How do I filter for reviewed mouse proteins which carry an N-terminal glycine? Return protein URI and AA sequence = 20\n",
      "‚úÖ 2/3 How do I filter for reviewed mouse proteins which carry an N-terminal glycine? Return protein URI and AA sequence = 20\n",
      "‚úÖ 3/3 How do I filter for reviewed mouse proteins which carry an N-terminal glycine? Return protein URI and AA sequence = 20\n",
      "üéØ No RAG - Success: 3, Mismatch: 0, Error: 3\n",
      "üéØ RAG without validation - Success: 6, Mismatch: 0, Error: 0\n",
      "üéØ RAG with validation - Success: 6, Mismatch: 0, Error: 0\n",
      "Missing from generated: {'rhea': {'type': 'uri', 'value': 'http://rdf.rhea-db.org/21744'}}\n",
      "‚ùå 1/3 How could I download a table that only includes the Rhea reactions for which there is experimental evidence? Return only the rhea URI \n",
      "Results mismatch. Ref: 11650 != gen: 0\n",
      ". Generated query:\n",
      "# https://sparql.rhea-db.org/sparql\n",
      "SELECT DISTINCT ?rhea_uri\n",
      "WHERE {\n",
      "  ?rhea_uri a <http://rdf.rhea-db.org/Reactions> ;\n",
      "            <http://rdf.rhea-db.org/experimentalEvidence> ?evidence .\n",
      "}\n",
      "Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\n",
      "SELECT DISTINCT ?rhea\n",
      "WHERE {\n",
      "  # ECO 269 is experimental evidence\n",
      "  BIND (<http://purl.obolibrary.org/obo/ECO_0000269> as ?evidence)\n",
      "  #GRAPH <http://sparql.uniprot.org/uniprot> {\n",
      "    ?protein up:reviewed true ;\n",
      "      up:annotation ?a ;\n",
      "      up:attribution ?attribution  .\n",
      "\n",
      "    ?a a up:Catalytic_Activity_Annotation ;\n",
      "      up:catalyticActivity ?ca .\n",
      "    ?ca up:catalyzedReaction ?rhea .\n",
      "\n",
      "    [] rdf:subject ?a ;\n",
      "      rdf:predicate up:catalyticActivity ;\n",
      "      rdf:object ?ca ;\n",
      "      up:attribution ?attribution .\n",
      "\n",
      "    ?attribution up:evidence ?evidence .\n",
      "  #}\n",
      "}\n",
      "\n",
      "Missing from generated: {'rhea': {'type': 'uri', 'value': 'http://rdf.rhea-db.org/21744'}}\n",
      "‚ùå 2/3 How could I download a table that only includes the Rhea reactions for which there is experimental evidence? Return only the rhea URI \n",
      "Results mismatch. Ref: 11650 != gen: 0\n",
      ". Generated query:\n",
      "# https://sparql.rhea-db.org/sparql\n",
      "SELECT DISTINCT ?rhea_uri\n",
      "WHERE {\n",
      "  ?rhea_uri a <http://rdf.rhea-db.org/Reactions> ;\n",
      "            <http://rdf.rhea-db.org/evidence> <http://rdf.rhea-db.org/ExperimentalEvidence> .\n",
      "}\n",
      "Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\n",
      "SELECT DISTINCT ?rhea\n",
      "WHERE {\n",
      "  # ECO 269 is experimental evidence\n",
      "  BIND (<http://purl.obolibrary.org/obo/ECO_0000269> as ?evidence)\n",
      "  #GRAPH <http://sparql.uniprot.org/uniprot> {\n",
      "    ?protein up:reviewed true ;\n",
      "      up:annotation ?a ;\n",
      "      up:attribution ?attribution  .\n",
      "\n",
      "    ?a a up:Catalytic_Activity_Annotation ;\n",
      "      up:catalyticActivity ?ca .\n",
      "    ?ca up:catalyzedReaction ?rhea .\n",
      "\n",
      "    [] rdf:subject ?a ;\n",
      "      rdf:predicate up:catalyticActivity ;\n",
      "      rdf:object ?ca ;\n",
      "      up:attribution ?attribution .\n",
      "\n",
      "    ?attribution up:evidence ?evidence .\n",
      "  #}\n",
      "}\n",
      "\n",
      "Missing from generated: {'rhea': {'type': 'uri', 'value': 'http://rdf.rhea-db.org/21744'}}\n",
      "‚ùå 3/3 How could I download a table that only includes the Rhea reactions for which there is experimental evidence? Return only the rhea URI \n",
      "Results mismatch. Ref: 11650 != gen: 0\n",
      ". Generated query:\n",
      "# https://sparql.rhea-db.org/sparql\n",
      "SELECT DISTINCT ?rhea_uri\n",
      "WHERE {\n",
      "  ?rhea_uri a <http://rdf.rhea-db.org/Reactions> ;\n",
      "            <http://rdf.rhea-db.org/experimentalEvidence> ?evidence .\n",
      "}\n",
      "Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\n",
      "SELECT DISTINCT ?rhea\n",
      "WHERE {\n",
      "  # ECO 269 is experimental evidence\n",
      "  BIND (<http://purl.obolibrary.org/obo/ECO_0000269> as ?evidence)\n",
      "  #GRAPH <http://sparql.uniprot.org/uniprot> {\n",
      "    ?protein up:reviewed true ;\n",
      "      up:annotation ?a ;\n",
      "      up:attribution ?attribution  .\n",
      "\n",
      "    ?a a up:Catalytic_Activity_Annotation ;\n",
      "      up:catalyticActivity ?ca .\n",
      "    ?ca up:catalyzedReaction ?rhea .\n",
      "\n",
      "    [] rdf:subject ?a ;\n",
      "      rdf:predicate up:catalyticActivity ;\n",
      "      rdf:object ?ca ;\n",
      "      up:attribution ?attribution .\n",
      "\n",
      "    ?attribution up:evidence ?evidence .\n",
      "  #}\n",
      "}\n",
      "\n",
      "‚úÖ 1/3 How could I download a table that only includes the Rhea reactions for which there is experimental evidence? Return only the rhea URI = 11650\n",
      "‚úÖ 2/3 How could I download a table that only includes the Rhea reactions for which there is experimental evidence? Return only the rhea URI = 11650\n",
      "‚úÖ 3/3 How could I download a table that only includes the Rhea reactions for which there is experimental evidence? Return only the rhea URI = 11650\n",
      "‚úÖ 1/3 How could I download a table that only includes the Rhea reactions for which there is experimental evidence? Return only the rhea URI = 11650\n",
      "‚úÖ 2/3 How could I download a table that only includes the Rhea reactions for which there is experimental evidence? Return only the rhea URI = 11650\n",
      "‚úÖ 3/3 How could I download a table that only includes the Rhea reactions for which there is experimental evidence? Return only the rhea URI = 11650\n",
      "üéØ No RAG - Success: 3, Mismatch: 3, Error: 3\n",
      "üéØ RAG without validation - Success: 9, Mismatch: 0, Error: 0\n",
      "üéØ RAG with validation - Success: 9, Mismatch: 0, Error: 0\n",
      "Missing from generated: {'reaction': {'type': 'uri', 'value': 'http://rdf.rhea-db.org/10100'}, 'sterol': {'type': 'uri', 'value': 'http://purl.obolibrary.org/obo/CHEBI_15889'}, 'protein': {'type': 'uri', 'value': 'http://purl.uniprot.org/uniprot/B3KRG8'}}\n",
      "‚ùå 1/3 Which human proteins are enzymes catalyzing a reaction involving sterols? Return the protein, sterol and reaction URI \n",
      "Results mismatch. Ref: 710 != gen: 0\n",
      ". Generated query:\n",
      "# https://sparql.rhea-db.org/sparql\n",
      "PREFIX rh:<http://rdf.rhea-db.org/>\n",
      "PREFIX up:<http://purl.uniprot.org/core/>\n",
      "PREFIX rdfs:<http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX obo:<http://purl.obolibrary.org/obo/>\n",
      "\n",
      "SELECT DISTINCT ?protein ?sterol ?reaction\n",
      "WHERE {\n",
      "  ?reaction rh:ec ?enzyme .\n",
      "  ?enzyme up:organism <http://purl.uniprot.org/taxonomy/9606> .\n",
      "  ?reaction rh:substrate ?sterol .\n",
      "  ?sterol rdfs:label ?sterolLabel .\n",
      "  FILTER(CONTAINS(LCASE(?sterolLabel), \"sterol\"))\n",
      "  ?enzyme up:recommendedName ?protein .\n",
      "}\n",
      "Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX rh: <http://rdf.rhea-db.org/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX CHEBI: <http://purl.obolibrary.org/obo/CHEBI_>\n",
      "SELECT DISTINCT ?protein ?sterol ?reaction\n",
      "WHERE {\n",
      "  SERVICE <https://sparql.rhea-db.org/sparql> {\n",
      "    ?reaction rdfs:subClassOf rh:Reaction .\n",
      "    ?reaction rh:side/rh:contains/rh:compound ?compound .\n",
      "    ?compound rh:chebi ?sterol .\n",
      "    ?sterol rdfs:subClassOf* CHEBI:15889 .\n",
      "  }\n",
      "  ?protein a up:Protein ;\n",
      "    up:organism taxon:9606 ;\n",
      "    up:annotation/up:catalyticActivity/up:catalyzedReaction ?reaction .\n",
      "}\n",
      "\n",
      "Missing from generated: {'reaction': {'type': 'uri', 'value': 'http://rdf.rhea-db.org/10100'}, 'sterol': {'type': 'uri', 'value': 'http://purl.obolibrary.org/obo/CHEBI_15889'}, 'protein': {'type': 'uri', 'value': 'http://purl.uniprot.org/uniprot/B3KRG8'}}\n",
      "‚ùå 2/3 Which human proteins are enzymes catalyzing a reaction involving sterols? Return the protein, sterol and reaction URI \n",
      "Results mismatch. Ref: 710 != gen: 0\n",
      ". Generated query:\n",
      "# https://sparql.uniprot.org/sparql\n",
      "\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX ec: <http://purl.uniprot.org/enzyme/>\n",
      "\n",
      "SELECT ?protein ?sterol ?reaction\n",
      "WHERE {\n",
      "  ?protein a up:Protein ;\n",
      "           up:organism <http://purl.uniprot.org/taxonomy/9606> ;\n",
      "           up:enzyme ?enzyme ;\n",
      "           up:reaction ?reaction .\n",
      "  ?reaction rdfs:comment ?reactionComment .\n",
      "  FILTER(CONTAINS(LCASE(?reactionComment), \"sterol\"))\n",
      "  ?reaction up:substrate ?sterol .\n",
      "}\n",
      "Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX rh: <http://rdf.rhea-db.org/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX CHEBI: <http://purl.obolibrary.org/obo/CHEBI_>\n",
      "SELECT DISTINCT ?protein ?sterol ?reaction\n",
      "WHERE {\n",
      "  SERVICE <https://sparql.rhea-db.org/sparql> {\n",
      "    ?reaction rdfs:subClassOf rh:Reaction .\n",
      "    ?reaction rh:side/rh:contains/rh:compound ?compound .\n",
      "    ?compound rh:chebi ?sterol .\n",
      "    ?sterol rdfs:subClassOf* CHEBI:15889 .\n",
      "  }\n",
      "  ?protein a up:Protein ;\n",
      "    up:organism taxon:9606 ;\n",
      "    up:annotation/up:catalyticActivity/up:catalyzedReaction ?reaction .\n",
      "}\n",
      "\n",
      "Missing from generated: {'reaction': {'type': 'uri', 'value': 'http://rdf.rhea-db.org/10100'}, 'sterol': {'type': 'uri', 'value': 'http://purl.obolibrary.org/obo/CHEBI_15889'}, 'protein': {'type': 'uri', 'value': 'http://purl.uniprot.org/uniprot/B3KRG8'}}\n",
      "‚ùå 3/3 Which human proteins are enzymes catalyzing a reaction involving sterols? Return the protein, sterol and reaction URI \n",
      "Results mismatch. Ref: 710 != gen: 0\n",
      ". Generated query:\n",
      "# https://sparql.rhea-db.org/sparql\n",
      "PREFIX rh:<http://rdf.rhea-db.org/>\n",
      "PREFIX rhv:<http://rdf.rhea-db.org/terms/>\n",
      "PREFIX uniprot:<http://purl.uniprot.org/core/>\n",
      "PREFIX up:<http://purl.uniprot.org/uniprot/>\n",
      "\n",
      "SELECT DISTINCT ?protein ?sterol ?reaction\n",
      "WHERE {\n",
      "  ?reaction rhv:ec ?enzyme .\n",
      "  ?enzyme rhv:enzymeType \"human\" .\n",
      "  ?reaction rhv:substrate ?substrate .\n",
      "  ?substrate rhv:compound ?sterol .\n",
      "  ?sterol rhv:compoundType \"sterol\" .\n",
      "  ?reaction rhv:participant ?participant .\n",
      "  ?participant rhv:protein ?protein .\n",
      "  ?protein a up:Protein .\n",
      "}\n",
      "LIMIT 100\n",
      "Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX rh: <http://rdf.rhea-db.org/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX CHEBI: <http://purl.obolibrary.org/obo/CHEBI_>\n",
      "SELECT DISTINCT ?protein ?sterol ?reaction\n",
      "WHERE {\n",
      "  SERVICE <https://sparql.rhea-db.org/sparql> {\n",
      "    ?reaction rdfs:subClassOf rh:Reaction .\n",
      "    ?reaction rh:side/rh:contains/rh:compound ?compound .\n",
      "    ?compound rh:chebi ?sterol .\n",
      "    ?sterol rdfs:subClassOf* CHEBI:15889 .\n",
      "  }\n",
      "  ?protein a up:Protein ;\n",
      "    up:organism taxon:9606 ;\n",
      "    up:annotation/up:catalyticActivity/up:catalyzedReaction ?reaction .\n",
      "}\n",
      "\n",
      "‚úÖ 1/3 Which human proteins are enzymes catalyzing a reaction involving sterols? Return the protein, sterol and reaction URI = 710\n",
      "‚úÖ 2/3 Which human proteins are enzymes catalyzing a reaction involving sterols? Return the protein, sterol and reaction URI = 710\n",
      "‚úÖ 3/3 Which human proteins are enzymes catalyzing a reaction involving sterols? Return the protein, sterol and reaction URI = 710\n",
      "‚úÖ 1/3 Which human proteins are enzymes catalyzing a reaction involving sterols? Return the protein, sterol and reaction URI = 710\n",
      "‚úÖ 2/3 Which human proteins are enzymes catalyzing a reaction involving sterols? Return the protein, sterol and reaction URI = 710\n",
      "‚úÖ 3/3 Which human proteins are enzymes catalyzing a reaction involving sterols? Return the protein, sterol and reaction URI = 710\n",
      "üéØ No RAG - Success: 3, Mismatch: 6, Error: 3\n",
      "üéØ RAG without validation - Success: 12, Mismatch: 0, Error: 0\n",
      "üéØ RAG with validation - Success: 12, Mismatch: 0, Error: 0\n",
      "Missing from generated: {'hgncSymbol': {'type': 'literal', 'value': 'BRAF'}, 'diseaseLabel': {'type': 'literal', 'value': 'Colorectal cancer'}}\n",
      "‚ùå 1/3 Which are the human proteins associated with cancer (which have cancer in their disease label)? Return the unique disease label (?diseaseLabel), and HGNC symbol (?hgncSymbol) \n",
      "Results mismatch. Ref: 121 != gen: 0\n",
      ". Generated query:\n",
      "# https://sparql.uniprot.org/sparql\n",
      "\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX dc: <http://purl.org/dc/terms/>\n",
      "PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
      "PREFIX dcterms: <http://purl.org/dc/terms/>\n",
      "\n",
      "SELECT DISTINCT ?diseaseLabel ?hgncSymbol\n",
      "WHERE {\n",
      "  ?protein a up:Protein ;\n",
      "           up:organism <http://purl.uniprot.org/taxonomy/9606> ;\n",
      "           up:annotation ?annotation .\n",
      "  ?annotation a up:Disease_Annotation ;\n",
      "              up:disease ?disease .\n",
      "  ?disease skos:prefLabel ?diseaseLabel .\n",
      "  FILTER(CONTAINS(LCASE(?diseaseLabel), \"cancer\"))\n",
      "  ?protein up:recommendedName ?recName .\n",
      "  ?recName up:fullName ?fullName .\n",
      "  ?protein up:gene ?gene .\n",
      "  ?gene skos:prefLabel ?hgncSymbol .\n",
      "}\n",
      "ORDER BY ?diseaseLabel ?hgncSymbol\n",
      "Correct query:\n",
      "PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
      "PREFIX up:<http://purl.uniprot.org/core/>\n",
      "PREFIX taxon:<http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX rdfs:<http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX orth:<http://purl.org/net/orth#>\n",
      "PREFIX dcterms:<http://purl.org/dc/terms/>\n",
      "PREFIX obo:<http://purl.obolibrary.org/obo/>\n",
      "PREFIX lscr:<http://purl.org/lscr#>\n",
      "PREFIX genex:<http://purl.org/genex#>\n",
      "PREFIX sio: <http://semanticscience.org/resource/>\n",
      "SELECT DISTINCT ?diseaseLabel ?hgncSymbol\n",
      "WHERE {\n",
      "    ?humanProtein a up:Protein ;\n",
      "        # up:organism/up:scientificName 'Homo sapiens' ;\n",
      "        up:organism taxon:9606 ;\n",
      "        up:annotation ?annotation ;\n",
      "        rdfs:seeAlso ?hgnc .\n",
      "    ?hgnc up:database <http://purl.uniprot.org/database/HGNC> ;\n",
      "        rdfs:comment ?hgncSymbol .\n",
      "    ?annotation a up:Disease_Annotation ;\n",
      "        up:disease ?disease .\n",
      "    ?disease skos:prefLabel ?diseaseLabel.\n",
      "    FILTER CONTAINS (LCASE(?diseaseLabel), \"cancer\")\n",
      "}\n",
      "\n",
      "Missing from generated: {'hgncSymbol': {'type': 'literal', 'value': 'BRAF'}, 'diseaseLabel': {'type': 'literal', 'value': 'Colorectal cancer'}}\n",
      "‚ùå 2/3 Which are the human proteins associated with cancer (which have cancer in their disease label)? Return the unique disease label (?diseaseLabel), and HGNC symbol (?hgncSymbol) \n",
      "Results mismatch. Ref: 121 != gen: 0\n",
      ". Generated query:\n",
      "# https://sparql.uniprot.org/sparql\n",
      "\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX dc: <http://purl.org/dc/terms/>\n",
      "PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
      "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
      "\n",
      "SELECT DISTINCT ?diseaseLabel ?hgncSymbol\n",
      "WHERE {\n",
      "  ?protein a up:Protein ;\n",
      "           up:organism <http://purl.uniprot.org/taxonomy/9606> ;\n",
      "           up:annotation ?annotation .\n",
      "  ?annotation a up:Disease_Annotation ;\n",
      "              up:disease ?disease .\n",
      "  ?disease skos:prefLabel ?diseaseLabel .\n",
      "  FILTER(CONTAINS(LCASE(?diseaseLabel), \"cancer\"))\n",
      "  ?protein up:encodedBy ?gene .\n",
      "  ?gene a up:Gene ;\n",
      "        skos:exactMatch/obo:HGNC_symbol ?hgncSymbol .\n",
      "}\n",
      "Correct query:\n",
      "PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
      "PREFIX up:<http://purl.uniprot.org/core/>\n",
      "PREFIX taxon:<http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX rdfs:<http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX orth:<http://purl.org/net/orth#>\n",
      "PREFIX dcterms:<http://purl.org/dc/terms/>\n",
      "PREFIX obo:<http://purl.obolibrary.org/obo/>\n",
      "PREFIX lscr:<http://purl.org/lscr#>\n",
      "PREFIX genex:<http://purl.org/genex#>\n",
      "PREFIX sio: <http://semanticscience.org/resource/>\n",
      "SELECT DISTINCT ?diseaseLabel ?hgncSymbol\n",
      "WHERE {\n",
      "    ?humanProtein a up:Protein ;\n",
      "        # up:organism/up:scientificName 'Homo sapiens' ;\n",
      "        up:organism taxon:9606 ;\n",
      "        up:annotation ?annotation ;\n",
      "        rdfs:seeAlso ?hgnc .\n",
      "    ?hgnc up:database <http://purl.uniprot.org/database/HGNC> ;\n",
      "        rdfs:comment ?hgncSymbol .\n",
      "    ?annotation a up:Disease_Annotation ;\n",
      "        up:disease ?disease .\n",
      "    ?disease skos:prefLabel ?diseaseLabel.\n",
      "    FILTER CONTAINS (LCASE(?diseaseLabel), \"cancer\")\n",
      "}\n",
      "\n",
      "Missing from generated: {'hgncSymbol': {'type': 'literal', 'value': 'BRAF'}, 'diseaseLabel': {'type': 'literal', 'value': 'Colorectal cancer'}}\n",
      "‚ùå 3/3 Which are the human proteins associated with cancer (which have cancer in their disease label)? Return the unique disease label (?diseaseLabel), and HGNC symbol (?hgncSymbol) \n",
      "Results mismatch. Ref: 121 != gen: 0\n",
      ". Generated query:\n",
      "# https://sparql.uniprot.org/sparql\n",
      "\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX dc: <http://purl.org/dc/terms/>\n",
      "PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
      "PREFIX dcterms: <http://purl.org/dc/terms/>\n",
      "\n",
      "SELECT DISTINCT ?diseaseLabel ?hgncSymbol\n",
      "WHERE {\n",
      "  ?protein a up:Protein ;\n",
      "           up:organism <http://purl.uniprot.org/taxonomy/9606> ;\n",
      "           up:annotation ?annotation .\n",
      "  ?annotation a up:Disease_Annotation ;\n",
      "              up:disease ?disease .\n",
      "  ?disease skos:prefLabel ?diseaseLabel .\n",
      "  FILTER(CONTAINS(LCASE(?diseaseLabel), \"cancer\"))\n",
      "  ?protein up:encodedBy ?gene .\n",
      "  ?gene skos:exactMatch ?hgnc .\n",
      "  ?hgnc a <http://identifiers.org/hgnc/> ;\n",
      "        skos:prefLabel ?hgncSymbol .\n",
      "}\n",
      "Correct query:\n",
      "PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
      "PREFIX up:<http://purl.uniprot.org/core/>\n",
      "PREFIX taxon:<http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX rdfs:<http://www.w3.org/2000/01/rdf-schema#>\n",
      "PREFIX orth:<http://purl.org/net/orth#>\n",
      "PREFIX dcterms:<http://purl.org/dc/terms/>\n",
      "PREFIX obo:<http://purl.obolibrary.org/obo/>\n",
      "PREFIX lscr:<http://purl.org/lscr#>\n",
      "PREFIX genex:<http://purl.org/genex#>\n",
      "PREFIX sio: <http://semanticscience.org/resource/>\n",
      "SELECT DISTINCT ?diseaseLabel ?hgncSymbol\n",
      "WHERE {\n",
      "    ?humanProtein a up:Protein ;\n",
      "        # up:organism/up:scientificName 'Homo sapiens' ;\n",
      "        up:organism taxon:9606 ;\n",
      "        up:annotation ?annotation ;\n",
      "        rdfs:seeAlso ?hgnc .\n",
      "    ?hgnc up:database <http://purl.uniprot.org/database/HGNC> ;\n",
      "        rdfs:comment ?hgncSymbol .\n",
      "    ?annotation a up:Disease_Annotation ;\n",
      "        up:disease ?disease .\n",
      "    ?disease skos:prefLabel ?diseaseLabel.\n",
      "    FILTER CONTAINS (LCASE(?diseaseLabel), \"cancer\")\n",
      "}\n",
      "\n",
      "‚úÖ 1/3 Which are the human proteins associated with cancer (which have cancer in their disease label)? Return the unique disease label (?diseaseLabel), and HGNC symbol (?hgncSymbol) = 121\n",
      "‚úÖ 2/3 Which are the human proteins associated with cancer (which have cancer in their disease label)? Return the unique disease label (?diseaseLabel), and HGNC symbol (?hgncSymbol) = 121\n",
      "‚úÖ 3/3 Which are the human proteins associated with cancer (which have cancer in their disease label)? Return the unique disease label (?diseaseLabel), and HGNC symbol (?hgncSymbol) = 121\n",
      "‚úÖ 1/3 Which are the human proteins associated with cancer (which have cancer in their disease label)? Return the unique disease label (?diseaseLabel), and HGNC symbol (?hgncSymbol) = 121\n",
      "‚úÖ 2/3 Which are the human proteins associated with cancer (which have cancer in their disease label)? Return the unique disease label (?diseaseLabel), and HGNC symbol (?hgncSymbol) = 121\n",
      "‚úÖ 3/3 Which are the human proteins associated with cancer (which have cancer in their disease label)? Return the unique disease label (?diseaseLabel), and HGNC symbol (?hgncSymbol) = 121\n",
      "üéØ No RAG - Success: 3, Mismatch: 9, Error: 3\n",
      "üéØ RAG without validation - Success: 15, Mismatch: 0, Error: 0\n",
      "üéØ RAG with validation - Success: 15, Mismatch: 0, Error: 0\n",
      "Missing from generated: {'gene': {'type': 'uri', 'value': 'http://omabrowser.org/ontology/oma#GENE_WBGene00001030'}, 'confidence': {'type': 'uri', 'value': 'http://purl.obolibrary.org/obo/CIO_0000029'}, 'fdr': {'type': 'typed-literal', 'datatype': 'http://www.w3.org/2001/XMLSchema#double', 'value': '0.0016728772206653400815'}}\n",
      "‚ùå 1/3 In bgee how can I retrieve the confidence level and false discovery rate of a gene expression? Return distinct ?gene, ?confidence and ?fdr, limit to 10 \n",
      "Results mismatch. Ref: 10 != gen: 0\n",
      ". Generated query:\n",
      "# https://bgee.org/sparql\n",
      "PREFIX bgee: <http://bgee.org/ontology#>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "\n",
      "SELECT DISTINCT ?gene ?confidence ?fdr\n",
      "WHERE {\n",
      "  ?expression a bgee:GeneExpression ;\n",
      "              bgee:hasGene ?gene ;\n",
      "              bgee:hasConfidenceLevel ?confidence ;\n",
      "              bgee:hasFalseDiscoveryRate ?fdr .\n",
      "}\n",
      "LIMIT 10\n",
      "Correct query:\n",
      "PREFIX genex: <http://purl.org/genex#>\n",
      "PREFIX bgee: <http://bgee.org/#>\n",
      "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
      "SELECT DISTINCT ?gene ?confidence ?fdr\n",
      "WHERE {\n",
      "    ?expression a genex:Expression ;\n",
      "        genex:hasConfidenceLevel ?confidence ;\n",
      "        genex:hasFDRpvalue ?fdr ;\n",
      "        genex:hasSequenceUnit ?gene .\n",
      "} LIMIT 10\n",
      "\n",
      "Missing from generated: {'gene': {'type': 'uri', 'value': 'http://omabrowser.org/ontology/oma#GENE_WBGene00001030'}, 'confidence': {'type': 'uri', 'value': 'http://purl.obolibrary.org/obo/CIO_0000029'}, 'fdr': {'type': 'typed-literal', 'datatype': 'http://www.w3.org/2001/XMLSchema#double', 'value': '0.0016728772206653400815'}}\n",
      "‚ùå 2/3 In bgee how can I retrieve the confidence level and false discovery rate of a gene expression? Return distinct ?gene, ?confidence and ?fdr, limit to 10 \n",
      "Results mismatch. Ref: 10 != gen: 0\n",
      ". Generated query:\n",
      "# https://bgee.org/sparql\n",
      "PREFIX bgee: <http://bgee.org/ontology#>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "\n",
      "SELECT DISTINCT ?gene ?confidence ?fdr\n",
      "WHERE {\n",
      "  ?expression a bgee:GeneExpression ;\n",
      "              bgee:hasGene ?gene ;\n",
      "              bgee:hasConfidenceLevel ?confidence ;\n",
      "              bgee:hasFalseDiscoveryRate ?fdr .\n",
      "}\n",
      "LIMIT 10\n",
      "Correct query:\n",
      "PREFIX genex: <http://purl.org/genex#>\n",
      "PREFIX bgee: <http://bgee.org/#>\n",
      "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
      "SELECT DISTINCT ?gene ?confidence ?fdr\n",
      "WHERE {\n",
      "    ?expression a genex:Expression ;\n",
      "        genex:hasConfidenceLevel ?confidence ;\n",
      "        genex:hasFDRpvalue ?fdr ;\n",
      "        genex:hasSequenceUnit ?gene .\n",
      "} LIMIT 10\n",
      "\n",
      "Missing from generated: {'gene': {'type': 'uri', 'value': 'http://omabrowser.org/ontology/oma#GENE_WBGene00001030'}, 'confidence': {'type': 'uri', 'value': 'http://purl.obolibrary.org/obo/CIO_0000029'}, 'fdr': {'type': 'typed-literal', 'datatype': 'http://www.w3.org/2001/XMLSchema#double', 'value': '0.0016728772206653400815'}}\n",
      "‚ùå 3/3 In bgee how can I retrieve the confidence level and false discovery rate of a gene expression? Return distinct ?gene, ?confidence and ?fdr, limit to 10 \n",
      "Results mismatch. Ref: 10 != gen: 0\n",
      ". Generated query:\n",
      "# https://bgee.org/sparql\n",
      "PREFIX bgee: <http://bgee.org/ontology#>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "\n",
      "SELECT DISTINCT ?gene ?confidence ?fdr\n",
      "WHERE {\n",
      "  ?expression a bgee:GeneExpression ;\n",
      "              bgee:hasGene ?gene ;\n",
      "              bgee:hasConfidenceLevel ?confidence ;\n",
      "              bgee:hasFalseDiscoveryRate ?fdr .\n",
      "}\n",
      "LIMIT 10\n",
      "Correct query:\n",
      "PREFIX genex: <http://purl.org/genex#>\n",
      "PREFIX bgee: <http://bgee.org/#>\n",
      "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
      "SELECT DISTINCT ?gene ?confidence ?fdr\n",
      "WHERE {\n",
      "    ?expression a genex:Expression ;\n",
      "        genex:hasConfidenceLevel ?confidence ;\n",
      "        genex:hasFDRpvalue ?fdr ;\n",
      "        genex:hasSequenceUnit ?gene .\n",
      "} LIMIT 10\n",
      "\n",
      "‚úÖ 1/3 In bgee how can I retrieve the confidence level and false discovery rate of a gene expression? Return distinct ?gene, ?confidence and ?fdr, limit to 10 = 10\n",
      "‚úÖ 2/3 In bgee how can I retrieve the confidence level and false discovery rate of a gene expression? Return distinct ?gene, ?confidence and ?fdr, limit to 10 = 10\n",
      "‚úÖ 3/3 In bgee how can I retrieve the confidence level and false discovery rate of a gene expression? Return distinct ?gene, ?confidence and ?fdr, limit to 10 = 10\n",
      "‚úÖ 1/3 In bgee how can I retrieve the confidence level and false discovery rate of a gene expression? Return distinct ?gene, ?confidence and ?fdr, limit to 10 = 10\n",
      "‚úÖ 2/3 In bgee how can I retrieve the confidence level and false discovery rate of a gene expression? Return distinct ?gene, ?confidence and ?fdr, limit to 10 = 10\n",
      "‚úÖ 3/3 In bgee how can I retrieve the confidence level and false discovery rate of a gene expression? Return distinct ?gene, ?confidence and ?fdr, limit to 10 = 10\n",
      "üéØ No RAG - Success: 3, Mismatch: 12, Error: 3\n",
      "üéØ RAG without validation - Success: 18, Mismatch: 0, Error: 0\n",
      "üéØ RAG with validation - Success: 18, Mismatch: 0, Error: 0\n",
      "Missing from generated: {'protein': {'type': 'uri', 'value': 'http://purl.uniprot.org/uniprot/A0AA38YEC8'}, 'enzymeClassName': {'type': 'literal', 'value': 'methylglyoxal reductase (NADPH)'}, 'enzymeClass': {'type': 'uri', 'value': 'http://purl.uniprot.org/enzyme/1.1.1.283'}}\n",
      "‚ùå 1/3 How can I get a list of proteins that are enzymes? Return the proteins URI, and their enzyme class URI and name, limit to 20 \n",
      "Results mismatch. Ref: 20 != gen: 0\n",
      ". Generated query:\n",
      "# https://sparql.uniprot.org/sparql\n",
      "\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "\n",
      "SELECT ?protein ?enzymeClass ?enzymeClassName\n",
      "WHERE {\n",
      "  ?protein a up:Protein ;\n",
      "           up:enzyme ?enzymeClass .\n",
      "  ?enzymeClass rdfs:label ?enzymeClassName .\n",
      "}\n",
      "LIMIT 20\n",
      "Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
      "SELECT ?protein ?enzymeClass ?enzymeClassName\n",
      "WHERE {\n",
      "    ?protein a up:Protein ;\n",
      "        up:enzyme ?enzymeClass .\n",
      "    ?enzymeClass skos:prefLabel ?enzymeClassName .\n",
      "} LIMIT 20\n",
      "\n",
      "‚ùå 2/3 How can I get a list of proteins that are enzymes? Return the proteins URI, and their enzyme class URI and name, limit to 20 QueryBadFormed: A bad request has been sent to the endpoint: probably the SPARQL query is badly formed. \n",
      "\n",
      "Response:\n",
      "b'<!DOCTYPE html SYSTEM \"about:legacy-compat\">\\n<html xmlns=\"http://www.w3.org/1999/xhtml\" lang=\"en\" xml:lang=\"en\"><head><title>UniProt</title><meta content=\"text/html; charset=UTF-8\" http-equiv=\"Content-Type\"/><link href=\"/\" rel=\"home\"/><link href=\"/base.css\" type=\"text/css\" rel=\"stylesheet\"/><link type=\"image/vnd.microsoft.icon\" href=\"/uniprot-favicon.ico\" rel=\"shortcut icon\"/><link href=\"/uniprot-sparql.css\" type=\"text/css\" rel=\"stylesheet\"/><link href=\"http://creativecommons.org/licenses/by/4.0/\" rel=\"license\"/><script src=\"https://cdn.jsdelivr.net/npm/yasgui-yasqe@2.11.22/dist/yasqe.bundled.min.js\"></script><link type=\"text/css\" rel=\"stylesheet\" href=\"https://cdn.jsdelivr.net/yasqe/2.2/yasqe.min.css\"/><meta content=\"width=device-width, initial-scale=1\" name=\"viewport\"/><meta content=\"no-referrer\" name=\"referrer\"/></head><body typeof=\"schema:WebPage\"><nav id=\"nav-header\"><a accesskey=\"1\" href=\"https://www.uniprot.org/\" id=\"www\"><img alt=\"\" src=\"data:image/svg+xml;base64,PHN2ZyB2aWV3Qm94PSIwIDAgNjQwIDI5MyIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj48ZyBmaWxsPSIjZmZmIiBmaWxsLXJ1bGU9Im5vbnplcm8iPjxwYXRoIGQ9Ik0xOTkuNjU5LDEwNi4xMDMgQzE5OS42NTksMTEyLjM5OCAxOTQuNTU2LDExNy41MDEgMTg4LjI2LDExNy41MDEgQzE4MS45NjQsMTE3LjUwMSAxNzYuODYxLDExMi4zOTcgMTc2Ljg2MSwxMDYuMTAzIEMxNzYuODYxLDk5LjgwNyAxODEuOTY1LDk0LjcwNCAxODguMjYsOTQuNzA0IEMxOTQuNTU1LDk0LjcwNCAxOTkuNjU5LDk5LjgwNyAxOTkuNjU5LDEwNi4xMDMgWiI+PC9wYXRoPjxwYXRoIGQ9Ik0zMzQuMzgzLDE0NC40NTkgQzMyNS4yNDIsMTQ0LjQ1OSAzMTcuODI5LDE1MS44NzIgMzE3LjgyOSwxNjEuMDE1IEwzMTcuNzk3LDE2NS4xODUgTDMxNy43OTcsMTk0LjIxIEMzMTcuNzk3LDIwMC4zMTYgMzEyLjg0OCwyMDUuMjY1IDMwNi43NDIsMjA1LjI2NSBDMzAwLjYzOCwyMDUuMjY1IDI5NS42ODksMjAwLjMxNyAyOTUuNjg5LDE5NC4yMSBMMjk1LjY4OSwxNjUuMTg1IEwyOTUuNzIsMTYxLjAxNSBDMjk1LjcyLDEzOS42NjEgMzEzLjAzLDEyMi4zNTEgMzM0LjM4MywxMjIuMzUxIEwzNDEuMjg1LDEyMi4zNTEgQzM0Ny4zOSwxMjIuMzUxIDM1Mi4zNCwxMjcuMyAzNTIuMzQsMTMzLjQwNSBDMzUyLjM0LDEzOS41MDkgMzQ3LjM5MSwxNDQuNDU5IDM0MS4yODUsMTQ0LjQ1OSBMMzM0LjM4MywxNDQuNDU5IFoiPjwvcGF0aD48cGF0aCBkPSJNNDMzLjg2MSwxNjMuODEzIEM0MzMuODYxLDE4Ny40MTcgNDE0LjcyNCwyMDYuNTU1IDM5MS4xMTksMjA2LjU1NSBDMzY3LjUxNCwyMDYuNTU1IDM0OC4zNzUsMTg3LjQxNyAzNDguMzc1LDE2My44MTMgQzM0OC4zNzUsMTQwLjIwNiAzNjcuNTE0LDEyMS4wNyAzOTEuMTE5LDEyMS4wNyBDNDE0LjcyNCwxMjEuMDcgNDMzLjg2MSwxNDAuMjA2IDQzMy44NjEsMTYzLjgxMyBaIE0zOTEuMDI5LDE0MS42OTQgQzM3OC44MTgsMTQxLjY5NCAzNjguOTIsMTUxLjU5MiAzNjguOTIsMTYzLjgwMSBDMzY4LjkyLDE3Ni4wMTEgMzc4LjgxOCwxODUuOTA5IDM5MS4wMjksMTg1LjkwOSBDNDAzLjIzOCwxODUuOTA5IDQxMy4xMzYsMTc2LjAxMSA0MTMuMTM2LDE2My44MDEgQzQxMy4xMzcsMTUxLjU5MyA0MDMuMjM4LDE0MS42OTQgMzkxLjAyOSwxNDEuNjk0IFoiPjwvcGF0aD48cGF0aCBkPSJNNDgzLjcwMywxODMuMTU2IEw0NzkuNDY3LDE4My4xNTUgQzQ3MC4zMjIsMTgzLjE1NSA0NjIuOTEsMTc1Ljc0MyA0NjIuOTEsMTY2LjU5OCBMNDYyLjg3NywxNjIuNDMgTDQ2Mi44NzcsMTQ0LjQ1NyBMNDgyLjIyMSwxNDQuNDU4IEM0ODguMzI2LDE0NC40NTggNDkzLjI3NiwxMzkuNTEgNDkzLjI3NiwxMzMuNDA0IEM0OTMuMjc2LDEyNy4yOTggNDg4LjMyNywxMjIuMzQ5IDQ4Mi4yMjEsMTIyLjM0OSBMNDYyLjg3NywxMjIuMzQ5IEw0NjIuODc3LDExNi44MjQgQzQ2Mi44NzcsMTEwLjcxOSA0NTcuOTMsMTA1Ljc2OSA0NTEuODI0LDEwNS43NjkgQzQ0NS43MTgsMTA1Ljc2OSA0NDAuNzY5LDExMC43MTggNDQwLjc2OSwxMTYuODI0IEw0NDAuNzY5LDE2Mi40MjkgTDQ0MC44LDE2Ni41OTkgQzQ0MC44LDE4Ny45NTIgNDU4LjExMywyMDUuMjYyIDQ3OS40NjYsMjA1LjI2MiBMNDgzLjcwMiwyMDUuMjYzIEM0ODkuODA3LDIwNS4yNjMgNDk0Ljc1NywyMDAuMzE1IDQ5NC43NTcsMTk0LjIwOCBDNDk0Ljc1OCwxODguMTA1IDQ4OS44MDksMTgzLjE1NiA0ODMuNzAzLDE4My4xNTYgWiI+PC9wYXRoPjxwYXRoIGQ9Ik0yOTIuOTgxLDEzNS4wNDIgQzI5Mi45ODEsMTEzLjY4OCAyNzUuNjcsOTYuMzc4IDI1NC4zMTcsOTYuMzc4IEwyNTAuMTQ3LDk2LjM0NiBMMjIyLjQ1OCw5Ni4zNDYgQzIxNi4zNTMsOTYuMzQ2IDIxMS40MDQsMTAxLjI5NSAyMTEuNDA0LDEwNy40IEwyMTEuNDA0LDE5NC4yIEMyMTEuNDA0LDIwMC4zMDYgMjE2LjM1MywyMDUuMjU2IDIyMi40NTgsMjA1LjI1NiBDMjI4LjU2MiwyMDUuMjU2IDIzMy41MTIsMjAwLjMwNyAyMzMuNTEyLDE5NC4yIEwyMzMuNTEyLDExOC40NTUgTDI1MC4xNDgsMTE4LjQ1NSBMMjU0LjMxNywxMTguNDg3IEMyNjMuNDYxLDExOC40ODcgMjcwLjg3NCwxMjUuODk5IDI3MC44NzQsMTM1LjA0MyBDMjcwLjg3NCwxNDQuMTg3IDI2My40NjEsMTUxLjYgMjU0LjMxNywxNTEuNiBMMjU1LjY3NCwxNTEuNjE2IEwyNTEuNTIsMTUxLjYxNiBDMjQ1LjQxNiwxNTEuNjE2IDI0MC40NjYsMTU2LjU2MyAyNDAuNDY2LDE2Mi42NyBDMjQwLjQ2NiwxNjguNzc0IDI0NS40MTUsMTczLjcyNCAyNTEuNTIsMTczLjcyNCBMMjU1LjY3NCwxNzMuNzI0IEwyNTQuMzE4LDE3My43MDggQzI3NS42NzEsMTczLjcwNiAyOTIuOTgxLDE1Ni4zOTYgMjkyLjk4MSwxMzUuMDQyIFoiPjwvcGF0aD48cGF0aCBkPSJNNDYyLjYwNywyNi4zMTkgQzQ2Mi42MDcsMzEuNDk0IDQ1OC40MSwzNS42ODkgNDUzLjIzNiwzNS42ODkgQzQ0OC4wNiwzNS42ODkgNDQzLjg2NSwzMS40OTQgNDQzLjg2NSwyNi4zMTkgQzQ0My44NjUsMjEuMTQ0IDQ0OC4wNiwxNi45NDggNDUzLjIzNiwxNi45NDggQzQ1OC40MSwxNi45NDggNDYyLjYwNywyMS4xNDUgNDYyLjYwNywyNi4zMTkgWiI+PC9wYXRoPjxwYXRoIGQ9Ik00MzAuNTE0LDU0LjY4OCBDNDMwLjUxNCw1OC44MjggNDI3LjE1Nyw2Mi4xODUgNDIzLjAxNiw2Mi4xODUgQzQxOC44NzcsNjIuMTg1IDQxNS41Miw1OC44MjggNDE1LjUyLDU0LjY4OCBDNDE1LjUyLDUwLjU0OCA0MTguODc3LDQ3LjE5MDk5OTggNDIzLjAxNiw0Ny4xOTA5OTk4IEM0MjcuMTU2LDQ3LjE5IDQzMC41MTQsNTAuNTQ4IDQzMC41MTQsNTQuNjg4IFoiPjwvcGF0aD48cGF0aCBkPSJNNTA1LjU1NywxMS43MTMgQzUwNS41NTcsMTguMTgzIDUwMC4zMTEsMjMuNDI2IDQ5My44NDIsMjMuNDI2IEM0ODcuMzczLDIzLjQyNiA0ODIuMTI5LDE4LjE4MyA0ODIuMTI5LDExLjcxMyBDNDgyLjEyOSw1LjI0MyA0ODcuMzczLDAgNDkzLjg0MiwwIEM1MDAuMzExLDAgNTA1LjU1Nyw1LjI0MyA1MDUuNTU3LDExLjcxMyBaIj48L3BhdGg+PHBhdGggZD0iTTU1NS4xMTksMTYuMzE4IEM1NTUuMTE5LDI0LjQwNCA1NDguNTY0LDMwLjk2IDU0MC40NzYsMzAuOTYgQzUzMi4zOTIsMzAuOTYgNTI1LjgzNSwyNC40MDQgNTI1LjgzNSwxNi4zMTggQzUyNS44MzUsOC4yMzEgNTMyLjM5MiwxLjY3NiA1NDAuNDc2LDEuNjc2IEM1NDguNTY0LDEuNjc3IDU1NS4xMTksOC4yMzEgNTU1LjExOSwxNi4zMTggWiI+PC9wYXRoPjxwYXRoIGQ9Ik02MDIuMTAyLDQyLjA0NiBDNjAyLjEwMiw1Mi4xNTMgNTkzLjkwOSw2MC4zNDcgNTgzLjgwMSw2MC4zNDcgQzU3My42OTIsNjAuMzQ3IDU2NS40OTgsNTIuMTU0IDU2NS40OTgsNDIuMDQ2IEM1NjUuNDk4LDMxLjkzNyA1NzMuNjkxLDIzLjc0NCA1ODMuODAxLDIzLjc0NCBDNTkzLjkwOCwyMy43NDQgNjAyLjEwMiwzMS45MzcgNjAyLjEwMiw0Mi4wNDYgWiI+PC9wYXRoPjxwYXRoIGQ9Ik02MzUuNDM2LDkwLjI4MiBDNjM1LjQzNiwxMDEuOTA3IDYyNi4wMTIsMTExLjMyOSA2MTQuMzg5LDExMS4zMjkgQzYwMi43NjYsMTExLjMyOSA1OTMuMzQyLDEwMS45MDcgNTkzLjM0Miw5MC4yODIgQzU5My4zNDIsNzguNjU3IDYwMi43NjYsNjkuMjM1IDYxNC4zODksNjkuMjM1IEM2MjYuMDEyLDY5LjIzNSA2MzUuNDM2LDc4LjY1NyA2MzUuNDM2LDkwLjI4MiBaIj48L3BhdGg+PHBhdGggZD0iTTYzOS4zODcsMTUzLjU5NSBDNjM5LjM4NywxNjYuOTYgNjI4LjU1MSwxNzcuNzk2IDYxNS4xODQsMTc3Ljc5NiBDNjAxLjgxNSwxNzcuNzk2IDU5MC45NzksMTY2Ljk2IDU5MC45NzksMTUzLjU5NSBDNTkwLjk3OSwxNDAuMjMgNjAxLjgxNSwxMjkuMzk1IDYxNS4xODQsMTI5LjM5NSBDNjI4LjU1MSwxMjkuMzk1IDYzOS4zODcsMTQwLjIyOSA2MzkuMzg3LDE1My41OTUgWiI+PC9wYXRoPjxwYXRoIGQ9Ik02MTIuNzY2LDIxNS44ODYgQzYxMi43NjYsMjMxLjMyNiA2MDAuMjUsMjQzLjg0MSA1ODQuODExLDI0My44NDEgQzU2OS4zNzIsMjQzLjg0MSA1NTYuODU2LDIzMS4zMjYgNTU2Ljg1NiwyMTUuODg2IEM1NTYuODU2LDIwMC40NDUgNTY5LjM3MiwxODcuOTMxIDU4NC44MTEsMTg3LjkzMSBDNjAwLjI1LDE4Ny45MzEgNjEyLjc2NiwyMDAuNDQ0IDYxMi43NjYsMjE1Ljg4NiBaIj48L3BhdGg+PHBhdGggZD0iTTU1MS43NzMsMjU5LjE5NCBDNTUxLjc3MywyNzYuNTAyIDUzNy43NDQsMjkwLjUzMSA1MjAuNDM3LDI5MC41MzEgQzUwMy4xMywyOTAuNTMxIDQ4OS4xMDEsMjc2LjUwMiA0ODkuMTAxLDI1OS4xOTQgQzQ4OS4xMDEsMjQxLjg4NiA1MDMuMTMsMjI3Ljg1OCA1MjAuNDM3LDIyNy44NTggQzUzNy43NDQsMjI3Ljg1OCA1NTEuNzczLDI0MS44ODcgNTUxLjc3MywyNTkuMTk0IFoiPjwvcGF0aD48cGF0aCBkPSJNNDcyLjEwNywyNTcuMzggQzQ3Mi4xMDcsMjc2Ljc3NCA0NTYuMzg0LDI5Mi40OTUgNDM2Ljk5MiwyOTIuNDk1IEM0MTcuNiwyOTIuNDk1IDQwMS44NzcsMjc2Ljc3MyA0MDEuODc3LDI1Ny4zOCBDNDAxLjg3NywyMzcuOTg0IDQxNy42LDIyMi4yNjcgNDM2Ljk5MiwyMjIuMjY3IEM0NTYuMzg0LDIyMi4yNjcgNDcyLjEwNywyMzcuOTg0IDQ3Mi4xMDcsMjU3LjM4IFoiPjwvcGF0aD48cGF0aCBkPSJNMTY1LjgwNSwxNTkuNjU4IEwxNjUuODA1LDE5NC4yMSBDMTY1LjgwNSwyMDAuMzE2IDE2MC44NTYsMjA1LjI2NSAxNTQuNzUxLDIwNS4yNjUgQzE0OC42NDcsMjA1LjI2NSAxNDMuNjk3LDIwMC4zMTcgMTQzLjY5NywxOTQuMjEgTDE0My42OTcsMTU5LjY1OCBMMTQzLjY4MSwxNjEuMDE0IEMxNDMuNjgxLDE1MS44NzEgMTM2LjI2OSwxNDQuNDU4IDEyNy4xMjUsMTQ0LjQ1OCBDMTE3Ljk4MSwxNDQuNDU4IDExMC41NjksMTUxLjg3MSAxMTAuNTY5LDE2MS4wMTQgTDExMC41MzYsMTY1LjE4NCBMMTEwLjUzNiwxOTQuMjA5IEMxMTAuNTM2LDIwMC4zMTUgMTA1LjU4NywyMDUuMjY0IDk5LjQ4MiwyMDUuMjY0IEM5My4zNzcsMjA1LjI2NCA4OC40MjgsMjAwLjMxNiA4OC40MjgsMTk0LjIwOSBMODguNDI4LDE2NS4xODQgTDg4LjQ2LDE2MS4wMTQgQzg4LjQ2LDEzOS42NiAxMDUuNzcsMTIyLjM1IDEyNy4xMjQsMTIyLjM1IEMxNDguNDc4LDEyMi4zNSAxNjUuNzg4LDEzOS42NjEgMTY1Ljc4OCwxNjEuMDE0IEwxNjUuODA1LDE1OS42NTggWiI+PC9wYXRoPjxwYXRoIGQ9Ik0wLDE2Ny45NTcgTDAsMTA3LjM5OSBDMCwxMDEuMjk1IDQuOTQ5LDk2LjM0NSAxMS4wNTQsOTYuMzQ1IEMxNy4xNTksOTYuMzQ1IDIyLjEwOCwxMDEuMjk0IDIyLjEwOCwxMDcuMzk5IEwyMi4xMDgsMTY3Ljk1NyBMMjIuMTIzLDE2Ni42MDEgQzIyLjEyMywxNzUuNzQ0IDI5LjUzNiwxODMuMTU3IDM4LjY3OSwxODMuMTU3IEM0Ny44MjIsMTgzLjE1NyA1NS4yMzUsMTc1Ljc0NCA1NS4yMzUsMTY2LjYwMSBMNTUuMjY4LDE2Mi40MzEgTDU1LjI2OCwxMDcuNCBDNTUuMjY4LDEwMS4yOTYgNjAuMjE3LDk2LjM0NSA2Ni4zMjIsOTYuMzQ1IEM3Mi40MjcsOTYuMzQ1IDc3LjM3NiwxMDEuMjk1IDc3LjM3NiwxMDcuNCBMNzcuMzc2LDE2Mi40MzEgTDc3LjM0NCwxNjYuNjAyIEM3Ny4zNDQsMTg3Ljk1NiA2MC4wMzQsMjA1LjI2NSAzOC42OCwyMDUuMjY1IEMxNy4zMjYsMjA1LjI2NSAwLjAxNiwxODcuOTU1IDAuMDE2LDE2Ni42MDIgTDAsMTY3Ljk1NyBaIj48L3BhdGg+PHBhdGggZD0iTTE5OS4zMTQsMTk0LjIxIEMxOTkuMzE0LDIwMC4zMTYgMTk0LjM2NSwyMDUuMjY1IDE4OC4yNiwyMDUuMjY1IEMxODIuMTU2LDIwNS4yNjUgMTc3LjIwNiwyMDAuMzE3IDE3Ny4yMDYsMTk0LjIxIEwxNzcuMjA2LDEzMy40NzggQzE3Ny4yMDYsMTI3LjM3MyAxODIuMTU1LDEyMi40MjMgMTg4LjI2LDEyMi40MjMgQzE5NC4zNjUsMTIyLjQyMyAxOTkuMzE0LDEyNy4zNzIgMTk5LjMxNCwxMzMuNDc4IEwxOTkuMzE0LDE5NC4yMSBaIj48L3BhdGg+PC9nPjwvc3ZnPgo=\" title=\"UniProt home\"/></a><a href=\"/sparql\">SPARQL</a><a href=\"https://www.uniprot.org/downloads\" id=\"download\">Downloads</a><a accesskey=\"9\" href=\"https://www.uniprot.org/contact\" id=\"contact\"><img alt=\"Contact\" src=\"data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZpZXdCb3g9IjAgMCAxMyAxMCIgZmlsbD0iI2ZmZiIgd2lkdGg9IjEuNGVtIj48cGF0aCBmaWxsLXJ1bGU9ImV2ZW5vZGQiIGQ9Ik0xMiAxLjI1QzEyIC41NiAxMS40NiAwIDEwLjggMEgxLjJDLjU0IDAgMCAuNTYgMCAxLjI1djcuNUMwIDkuNDQuNTQgMTAgMS4yIDEwaDkuNmMuNjYgMCAxLjItLjU2IDEuMi0xLjI1di03LjV6bS0xLjIgMEw2IDQuMzggMS4yIDEuMjVoOS42ek02IDUuNjNMMS4yIDIuNXY2LjI1aDkuNlYyLjVMNiA1LjYzeiI+PC9wYXRoPjwvc3ZnPgo=\"/></a><a rel=\"help\" href=\"https://www.uniprot.org/help/\" id=\"help\">Documentation/Help</a></nav><a accesskey=\"2\" href=\"#content\" hidden=\"hidden\">Skip Header</a><main class=\"sparql results\" id=\"content\"><section class=\"error\" id=\"page-header\"><h2 class=\"page-title\">Malformed query.</h2><p>org.eclipse.rdf4j.query.parser.sparql.ast.VisitorException: QName \\'up:Protein\\' uses an undefined prefix</p></section><div class=\"main-aside\"><div id=\"queryform\"><form method=\"get\" action=\"/sparql/\" focus=\"query\" name=\"sparql-form\" id=\"sparql-form\"><button id=\"addPrefix\" type=\"button\">Add common prefixes</button><textarea accesskey=\"1\" rows=\"10\" cols=\"80\" name=\"query\" id=\"query\"># https://sparql.uniprot.org/sparql\\nSELECT ?protein ?enzymeClass ?enzymeClassName\\nWHERE {\\n  ?protein a up:Protein ;\\n           up:recommendedName ?proteinName ;\\n           up:enzyme ?enzymeClass .\\n  ?enzymeClass rdfs:label ?enzymeClassName .\\n}\\nLIMIT 20</textarea><input type=\"hidden\" name=\"format\" value=\"html\"><button id=\"sparqlFormSubmitter\" onclick=\"return false;\">Submit Query</button></form></div></div></main><footer id=\"page-footer\"><section class=\"small\" id=\"copyright\">\\n\\t\\t\\t&copy; 2002&ndash; <span property=\"schema:copyrightYear\">2024</span>\\n\\t\\t\\t<a accesskey=\"9\" href=\"https://www.uniprot.org/help/about\">UniProt Consortium</a> |\\n\\t\\t\\t<a href=\"https://www.uniprot.org/help/license\" property=\"schema:license\">License\\n\\t\\t\\t\\t&amp; Disclaimer</a> | <a href=\"https://www.uniprot.org/help/privacy\">Privacy\\n\\t\\t\\t\\tNotice</a></section><section><div resource=\"/help/about#UniProt\" id=\"logos\" typeof=\"NGO\" property=\"creator\"><span resource=\"/help/about#EBI\" typeof=\"Organization\" class=\"ebi\" property=\"member\"><span hidden=\"hidden\" property=\"name\">European Bioinformatics Institute</span><a href=\"https://www.ebi.ac.uk/\" property=\"url\"></a></span><span resource=\"/help/about#PIR\" typeof=\"Organization\" class=\"pir\" property=\"member\"><span hidden=\"hidden\" property=\"name\">Protein Information Resource</span><a href=\"http://pir.georgetown.edu/\" property=\"url\"></a></span><span resource=\"/help/about#SIB\" typeof=\"Organization\" class=\"sib\" property=\"member\"><span hidden=\"hidden\" property=\"name\">SIB Swiss Institute of Bioinformatics</span><a href=\"https://www.sib.swiss/\" property=\"url\"></a></span></div></section><section id=\"poweredby\"><h3>Powered by</h3><span class=\"vitalit\" resource=\"https://www.sib.swiss/alan-bridge-group/\" typeof=\"schema:NGO\" property=\"schema:sponsor\"><h4>Hardware</h4><a href=\"https://www.sib.swiss/alan-bridge-group\" property=\"schema:url\"><img alt=\"Swiss-Prot\" src=\"https://www.sib.swiss//templates/sib/images/SIB_LogoQ_GBv.svg\"/><h5 property=\"schema:name\">Swiss-Prot</h5></a></span><span class=\"openlinksw\" resource=\"http://www.openlinksw.com/#this\" typeof=\"schema:Corporation\" property=\"schema:sponsor\"><h4>Software</h4><a href=\"http://virtuoso.openlinksw.com/\" property=\"schema:url\"><img alt=\"Virtuoso\" src=\"/images/logos/virtuoso.png\"/><h5 property=\"schema:name\">Virtuoso</h5></a></span></section></footer><script src=\"/sparql.js\" type=\"text/javascript\"> </script></body></html>'. Generated query:\n",
      "# https://sparql.uniprot.org/sparql\n",
      "SELECT ?protein ?enzymeClass ?enzymeClassName\n",
      "WHERE {\n",
      "  ?protein a up:Protein ;\n",
      "           up:recommendedName ?proteinName ;\n",
      "           up:enzyme ?enzymeClass .\n",
      "  ?enzymeClass rdfs:label ?enzymeClassName .\n",
      "}\n",
      "LIMIT 20\n",
      "Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
      "SELECT ?protein ?enzymeClass ?enzymeClassName\n",
      "WHERE {\n",
      "    ?protein a up:Protein ;\n",
      "        up:enzyme ?enzymeClass .\n",
      "    ?enzymeClass skos:prefLabel ?enzymeClassName .\n",
      "} LIMIT 20\n",
      "\n",
      "Missing from generated: {'protein': {'type': 'uri', 'value': 'http://purl.uniprot.org/uniprot/A0AA38YEC8'}, 'enzymeClassName': {'type': 'literal', 'value': 'methylglyoxal reductase (NADPH)'}, 'enzymeClass': {'type': 'uri', 'value': 'http://purl.uniprot.org/enzyme/1.1.1.283'}}\n",
      "‚ùå 3/3 How can I get a list of proteins that are enzymes? Return the proteins URI, and their enzyme class URI and name, limit to 20 \n",
      "Results mismatch. Ref: 20 != gen: 0\n",
      ". Generated query:\n",
      "# https://sparql.uniprot.org/sparql\n",
      "SELECT ?protein ?enzymeClass ?enzymeClassName\n",
      "WHERE {\n",
      "  ?protein a <http://purl.uniprot.org/core/Protein> ;\n",
      "           <http://purl.uniprot.org/core/classifiedWith> ?enzymeClass .\n",
      "  ?enzymeClass a <http://purl.uniprot.org/enzyme/Enzyme> ;\n",
      "               rdfs:label ?enzymeClassName .\n",
      "}\n",
      "LIMIT 20\n",
      "Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
      "SELECT ?protein ?enzymeClass ?enzymeClassName\n",
      "WHERE {\n",
      "    ?protein a up:Protein ;\n",
      "        up:enzyme ?enzymeClass .\n",
      "    ?enzymeClass skos:prefLabel ?enzymeClassName .\n",
      "} LIMIT 20\n",
      "\n",
      "Missing from generated: {'protein': {'type': 'uri', 'value': 'http://purl.uniprot.org/uniprot/A0AA38YEC8'}, 'enzymeClassName': {'type': 'literal', 'value': 'methylglyoxal reductase (NADPH)'}, 'enzymeClass': {'type': 'uri', 'value': 'http://purl.uniprot.org/enzyme/1.1.1.283'}}\n",
      "‚ùå 1/3 How can I get a list of proteins that are enzymes? Return the proteins URI, and their enzyme class URI and name, limit to 20 \n",
      "Results mismatch. Ref: 20 != gen: 0\n",
      ". Generated query:\n",
      "# https://sparql.uniprot.org/sparql/\n",
      "\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
      "\n",
      "SELECT ?protein ?enzymeClass ?enzymeName\n",
      "WHERE {\n",
      "  ?protein a up:Protein ;\n",
      "           up:annotation ?enzymeAnnotation .\n",
      "  ?enzymeAnnotation a up:Enzyme_Annotation ;\n",
      "                    up:enzyme ?enzymeClass .\n",
      "  ?enzymeClass skos:prefLabel ?enzymeName .\n",
      "}\n",
      "LIMIT 20\n",
      "Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
      "SELECT ?protein ?enzymeClass ?enzymeClassName\n",
      "WHERE {\n",
      "    ?protein a up:Protein ;\n",
      "        up:enzyme ?enzymeClass .\n",
      "    ?enzymeClass skos:prefLabel ?enzymeClassName .\n",
      "} LIMIT 20\n",
      "\n",
      "Missing from generated: {'protein': {'type': 'uri', 'value': 'http://purl.uniprot.org/uniprot/A0AA38YEC8'}, 'enzymeClassName': {'type': 'literal', 'value': 'methylglyoxal reductase (NADPH)'}, 'enzymeClass': {'type': 'uri', 'value': 'http://purl.uniprot.org/enzyme/1.1.1.283'}}\n",
      "‚ùå 2/3 How can I get a list of proteins that are enzymes? Return the proteins URI, and their enzyme class URI and name, limit to 20 \n",
      "Results mismatch. Ref: 20 != gen: 0\n",
      ". Generated query:\n",
      "# https://sparql.uniprot.org/sparql/\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "\n",
      "SELECT ?protein ?enzymeClass ?enzymeName\n",
      "WHERE {\n",
      "  ?protein a up:Protein ;\n",
      "           up:annotation ?enzymeAnnotation .\n",
      "  ?enzymeAnnotation a up:Enzyme_Annotation ;\n",
      "                    up:enzyme ?enzymeClass .\n",
      "  ?enzymeClass rdfs:label ?enzymeName .\n",
      "}\n",
      "LIMIT 20\n",
      "Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
      "SELECT ?protein ?enzymeClass ?enzymeClassName\n",
      "WHERE {\n",
      "    ?protein a up:Protein ;\n",
      "        up:enzyme ?enzymeClass .\n",
      "    ?enzymeClass skos:prefLabel ?enzymeClassName .\n",
      "} LIMIT 20\n",
      "\n",
      "Missing from generated: {'protein': {'type': 'uri', 'value': 'http://purl.uniprot.org/uniprot/A0AA38YEC8'}, 'enzymeClassName': {'type': 'literal', 'value': 'methylglyoxal reductase (NADPH)'}, 'enzymeClass': {'type': 'uri', 'value': 'http://purl.uniprot.org/enzyme/1.1.1.283'}}\n",
      "‚ùå 3/3 How can I get a list of proteins that are enzymes? Return the proteins URI, and their enzyme class URI and name, limit to 20 \n",
      "Results mismatch. Ref: 20 != gen: 0\n",
      ". Generated query:\n",
      "# https://sparql.uniprot.org/sparql/\n",
      "\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
      "\n",
      "SELECT ?protein ?enzymeClass ?enzymeName\n",
      "WHERE {\n",
      "  ?protein a up:Protein ;\n",
      "           up:annotation ?enzymeAnnotation .\n",
      "  ?enzymeAnnotation a up:Enzyme_Annotation ;\n",
      "                    up:enzyme ?enzymeClass .\n",
      "  ?enzymeClass skos:prefLabel ?enzymeName .\n",
      "}\n",
      "LIMIT 20\n",
      "Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
      "SELECT ?protein ?enzymeClass ?enzymeClassName\n",
      "WHERE {\n",
      "    ?protein a up:Protein ;\n",
      "        up:enzyme ?enzymeClass .\n",
      "    ?enzymeClass skos:prefLabel ?enzymeClassName .\n",
      "} LIMIT 20\n",
      "\n",
      "Missing from generated: {'protein': {'type': 'uri', 'value': 'http://purl.uniprot.org/uniprot/A0AA38YEC8'}, 'enzymeClassName': {'type': 'literal', 'value': 'methylglyoxal reductase (NADPH)'}, 'enzymeClass': {'type': 'uri', 'value': 'http://purl.uniprot.org/enzyme/1.1.1.283'}}\n",
      "‚ùå 1/3 How can I get a list of proteins that are enzymes? Return the proteins URI, and their enzyme class URI and name, limit to 20 \n",
      "Results mismatch. Ref: 20 != gen: 0\n",
      ". Generated query:\n",
      "# https://sparql.uniprot.org/sparql/\n",
      "\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "\n",
      "SELECT ?protein ?enzymeClass ?enzymeName\n",
      "WHERE {\n",
      "  ?protein a up:Protein ;\n",
      "           up:annotation ?catalyticActivityAnnotation .\n",
      "  ?catalyticActivityAnnotation a up:Catalytic_Activity_Annotation ;\n",
      "                                up:catalyticActivity ?enzymeClass .\n",
      "  ?enzymeClass rdfs:label ?enzymeName .\n",
      "}\n",
      "LIMIT 20\n",
      "Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
      "SELECT ?protein ?enzymeClass ?enzymeClassName\n",
      "WHERE {\n",
      "    ?protein a up:Protein ;\n",
      "        up:enzyme ?enzymeClass .\n",
      "    ?enzymeClass skos:prefLabel ?enzymeClassName .\n",
      "} LIMIT 20\n",
      "\n",
      "Missing from generated: {'protein': {'type': 'uri', 'value': 'http://purl.uniprot.org/uniprot/A0AA38YEC8'}, 'enzymeClassName': {'type': 'literal', 'value': 'methylglyoxal reductase (NADPH)'}, 'enzymeClass': {'type': 'uri', 'value': 'http://purl.uniprot.org/enzyme/1.1.1.283'}}\n",
      "‚ùå 2/3 How can I get a list of proteins that are enzymes? Return the proteins URI, and their enzyme class URI and name, limit to 20 \n",
      "Results mismatch. Ref: 20 != gen: 0\n",
      ". Generated query:\n",
      "# https://sparql.uniprot.org/sparql/\n",
      "\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "\n",
      "SELECT ?protein ?enzymeClass ?enzymeClassName\n",
      "WHERE {\n",
      "  ?protein a up:Protein ;\n",
      "           up:annotation ?activityAnnotation .\n",
      "  ?activityAnnotation a up:Catalytic_Activity_Annotation ;\n",
      "                      up:catalyticActivity ?enzymeClass .\n",
      "  ?enzymeClass rdfs:label ?enzymeClassName .\n",
      "}\n",
      "LIMIT 20\n",
      "Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
      "SELECT ?protein ?enzymeClass ?enzymeClassName\n",
      "WHERE {\n",
      "    ?protein a up:Protein ;\n",
      "        up:enzyme ?enzymeClass .\n",
      "    ?enzymeClass skos:prefLabel ?enzymeClassName .\n",
      "} LIMIT 20\n",
      "\n",
      "Missing from generated: {'protein': {'type': 'uri', 'value': 'http://purl.uniprot.org/uniprot/A0AA38YEC8'}, 'enzymeClassName': {'type': 'literal', 'value': 'methylglyoxal reductase (NADPH)'}, 'enzymeClass': {'type': 'uri', 'value': 'http://purl.uniprot.org/enzyme/1.1.1.283'}}\n",
      "‚ùå 3/3 How can I get a list of proteins that are enzymes? Return the proteins URI, and their enzyme class URI and name, limit to 20 \n",
      "Results mismatch. Ref: 20 != gen: 0\n",
      ". Generated query:\n",
      "# https://sparql.uniprot.org/sparql/\n",
      "\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "\n",
      "SELECT ?protein ?enzymeClass ?enzymeName\n",
      "WHERE {\n",
      "  ?protein a up:Protein ;\n",
      "           up:annotation ?enzymeAnnotation .\n",
      "  ?enzymeAnnotation a up:Catalytic_Activity_Annotation ;\n",
      "                    up:catalyticActivity ?catalyticActivity .\n",
      "  ?catalyticActivity up:enzymeClass ?enzymeClass .\n",
      "  ?enzymeClass rdfs:label ?enzymeName .\n",
      "}\n",
      "LIMIT 20\n",
      "Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
      "SELECT ?protein ?enzymeClass ?enzymeClassName\n",
      "WHERE {\n",
      "    ?protein a up:Protein ;\n",
      "        up:enzyme ?enzymeClass .\n",
      "    ?enzymeClass skos:prefLabel ?enzymeClassName .\n",
      "} LIMIT 20\n",
      "\n",
      "üéØ No RAG - Success: 3, Mismatch: 14, Error: 4\n",
      "üéØ RAG without validation - Success: 18, Mismatch: 3, Error: 0\n",
      "üéØ RAG with validation - Success: 18, Mismatch: 3, Error: 0\n",
      "Missing from generated: {'ensemblURI': {'type': 'uri', 'value': 'http://rdf.ebi.ac.uk/resource/ensembl.protein/ENSBIXP00000024889'}}\n",
      "‚ùå 1/3 How can I get the cross ref to the ensembl protein for the LCT protein in OMA? Return only the distinct ?ensemblURI \n",
      "Results mismatch. Ref: 96 != gen: 0\n",
      ". Generated query:\n",
      "# https://sparql.omabrowser.org/sparql\n",
      "SELECT DISTINCT ?ensemblURI WHERE {\n",
      "  ?protein a up:Protein ;\n",
      "           up:recommendedName ?name ;\n",
      "           rdfs:seeAlso ?ensemblURI .\n",
      "  ?name rdfs:label \"Lactase\"@en .\n",
      "  FILTER(CONTAINS(STR(?ensemblURI), \"ensembl\"))\n",
      "}\n",
      "Correct query:\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "SELECT DISTINCT ?ensemblURI\n",
      "WHERE {\n",
      "    ?protein a orth:Protein ;\n",
      "        rdfs:label 'LCT' ;\n",
      "        lscr:xrefEnsemblProtein ?ensemblURI .\n",
      "}\n",
      "\n",
      "Missing from generated: {'ensemblURI': {'type': 'uri', 'value': 'http://rdf.ebi.ac.uk/resource/ensembl.protein/ENSBIXP00000024889'}}\n",
      "‚ùå 2/3 How can I get the cross ref to the ensembl protein for the LCT protein in OMA? Return only the distinct ?ensemblURI \n",
      "Results mismatch. Ref: 96 != gen: 0\n",
      ". Generated query:\n",
      "# https://sparql.omabrowser.org/sparql\n",
      "\n",
      "SELECT DISTINCT ?ensemblURI\n",
      "WHERE {\n",
      "  ?protein a up:Protein ;\n",
      "           up:recommendedName ?name ;\n",
      "           rdfs:seeAlso ?ensemblURI .\n",
      "  ?name rdfs:label \"Lactase-phlorizin hydrolase\"@en .\n",
      "  FILTER(CONTAINS(STR(?ensemblURI), \"ensembl\"))\n",
      "}\n",
      "Correct query:\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "SELECT DISTINCT ?ensemblURI\n",
      "WHERE {\n",
      "    ?protein a orth:Protein ;\n",
      "        rdfs:label 'LCT' ;\n",
      "        lscr:xrefEnsemblProtein ?ensemblURI .\n",
      "}\n",
      "\n",
      "Missing from generated: {'ensemblURI': {'type': 'uri', 'value': 'http://rdf.ebi.ac.uk/resource/ensembl.protein/ENSBIXP00000024889'}}\n",
      "‚ùå 3/3 How can I get the cross ref to the ensembl protein for the LCT protein in OMA? Return only the distinct ?ensemblURI \n",
      "Results mismatch. Ref: 96 != gen: 0\n",
      ". Generated query:\n",
      "# https://sparql.omabrowser.org/sparql\n",
      "SELECT DISTINCT ?ensemblURI WHERE {\n",
      "  ?protein a up:Protein ;\n",
      "           up:recommendedName ?name ;\n",
      "           rdfs:seeAlso ?ensemblURI .\n",
      "  ?name rdfs:label \"Lactase-phlorizin hydrolase\"@en .\n",
      "  FILTER(CONTAINS(STR(?ensemblURI), \"ensembl\"))\n",
      "}\n",
      "Correct query:\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "PREFIX lscr: <http://purl.org/lscr#>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "SELECT DISTINCT ?ensemblURI\n",
      "WHERE {\n",
      "    ?protein a orth:Protein ;\n",
      "        rdfs:label 'LCT' ;\n",
      "        lscr:xrefEnsemblProtein ?ensemblURI .\n",
      "}\n",
      "\n",
      "‚úÖ 1/3 How can I get the cross ref to the ensembl protein for the LCT protein in OMA? Return only the distinct ?ensemblURI = 96\n",
      "‚úÖ 2/3 How can I get the cross ref to the ensembl protein for the LCT protein in OMA? Return only the distinct ?ensemblURI = 96\n",
      "‚úÖ 3/3 How can I get the cross ref to the ensembl protein for the LCT protein in OMA? Return only the distinct ?ensemblURI = 96\n",
      "‚úÖ 1/3 How can I get the cross ref to the ensembl protein for the LCT protein in OMA? Return only the distinct ?ensemblURI = 96\n",
      "‚úÖ 2/3 How can I get the cross ref to the ensembl protein for the LCT protein in OMA? Return only the distinct ?ensemblURI = 96\n",
      "‚úÖ 3/3 How can I get the cross ref to the ensembl protein for the LCT protein in OMA? Return only the distinct ?ensemblURI = 96\n",
      "üéØ No RAG - Success: 3, Mismatch: 17, Error: 4\n",
      "üéØ RAG without validation - Success: 21, Mismatch: 3, Error: 0\n",
      "üéØ RAG with validation - Success: 21, Mismatch: 3, Error: 0\n",
      "Missing from generated: {'orthologCluster': {'type': 'uri', 'value': 'https://omabrowser.org/oma/hog/resolve/HOG:D0000193_-2035759834'}, 'datasetURI': {'type': 'uri', 'value': 'https://omabrowser.org/oma/current/#DATASET_OMA'}}\n",
      "‚ùå 1/3 How can I get the URI of a dataset to which an ortholog cluster belongs in OMA? Return orthologCluster, datasetURI and limit to 20 \n",
      "Results mismatch. Ref: 20 != gen: 0\n",
      ". Generated query:\n",
      "# https://sparql.omabrowser.org/sparql\n",
      "SELECT ?orthologCluster ?datasetURI\n",
      "WHERE {\n",
      "  ?orthologCluster a oma:OrthologCluster .\n",
      "  ?orthologCluster oma:inDataset ?datasetURI .\n",
      "}\n",
      "LIMIT 20\n",
      "Correct query:\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "SELECT DISTINCT ?orthologCluster ?datasetURI\n",
      "WHERE {\n",
      "    ?orthologCluster a orth:OrthologsCluster ;\n",
      "        orth:inDataset ?datasetURI .\n",
      "} LIMIT 20\n",
      "\n",
      "Missing from generated: {'orthologCluster': {'type': 'uri', 'value': 'https://omabrowser.org/oma/hog/resolve/HOG:D0000193_-2035759834'}, 'datasetURI': {'type': 'uri', 'value': 'https://omabrowser.org/oma/current/#DATASET_OMA'}}\n",
      "‚ùå 2/3 How can I get the URI of a dataset to which an ortholog cluster belongs in OMA? Return orthologCluster, datasetURI and limit to 20 \n",
      "Results mismatch. Ref: 20 != gen: 0\n",
      ". Generated query:\n",
      "# https://sparql.omabrowser.org/sparql\n",
      "SELECT ?orthologCluster ?datasetURI\n",
      "WHERE {\n",
      "  ?orthologCluster a oma:OrthologCluster .\n",
      "  ?orthologCluster oma:inDataset ?datasetURI .\n",
      "}\n",
      "LIMIT 20\n",
      "Correct query:\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "SELECT DISTINCT ?orthologCluster ?datasetURI\n",
      "WHERE {\n",
      "    ?orthologCluster a orth:OrthologsCluster ;\n",
      "        orth:inDataset ?datasetURI .\n",
      "} LIMIT 20\n",
      "\n",
      "Missing from generated: {'orthologCluster': {'type': 'uri', 'value': 'https://omabrowser.org/oma/hog/resolve/HOG:D0000193_-2035759834'}, 'datasetURI': {'type': 'uri', 'value': 'https://omabrowser.org/oma/current/#DATASET_OMA'}}\n",
      "‚ùå 3/3 How can I get the URI of a dataset to which an ortholog cluster belongs in OMA? Return orthologCluster, datasetURI and limit to 20 \n",
      "Results mismatch. Ref: 20 != gen: 0\n",
      ". Generated query:\n",
      "# https://sparql.omabrowser.org/sparql\n",
      "SELECT ?orthologCluster ?datasetURI\n",
      "WHERE {\n",
      "  ?orthologCluster a oma:OrthologCluster .\n",
      "  ?orthologCluster oma:inDataset ?datasetURI .\n",
      "}\n",
      "LIMIT 20\n",
      "Correct query:\n",
      "PREFIX orth: <http://purl.org/net/orth#>\n",
      "SELECT DISTINCT ?orthologCluster ?datasetURI\n",
      "WHERE {\n",
      "    ?orthologCluster a orth:OrthologsCluster ;\n",
      "        orth:inDataset ?datasetURI .\n",
      "} LIMIT 20\n",
      "\n",
      "‚úÖ 1/3 How can I get the URI of a dataset to which an ortholog cluster belongs in OMA? Return orthologCluster, datasetURI and limit to 20 = 20\n",
      "‚úÖ 2/3 How can I get the URI of a dataset to which an ortholog cluster belongs in OMA? Return orthologCluster, datasetURI and limit to 20 = 20\n",
      "‚úÖ 3/3 How can I get the URI of a dataset to which an ortholog cluster belongs in OMA? Return orthologCluster, datasetURI and limit to 20 = 20\n",
      "‚úÖ 1/3 How can I get the URI of a dataset to which an ortholog cluster belongs in OMA? Return orthologCluster, datasetURI and limit to 20 = 20\n",
      "‚úÖ 2/3 How can I get the URI of a dataset to which an ortholog cluster belongs in OMA? Return orthologCluster, datasetURI and limit to 20 = 20\n",
      "‚úÖ 3/3 How can I get the URI of a dataset to which an ortholog cluster belongs in OMA? Return orthologCluster, datasetURI and limit to 20 = 20\n",
      "üéØ No RAG - Success: 3, Mismatch: 20, Error: 4\n",
      "üéØ RAG without validation - Success: 24, Mismatch: 3, Error: 0\n",
      "üéØ RAG with validation - Success: 24, Mismatch: 3, Error: 0\n",
      "Missing from generated: {'strain': {'type': 'uri', 'value': 'http://purl.uniprot.org/taxonomy/101028#strain-D19B933A18EA91CE5E571899A51CBD16'}, 'name': {'type': 'literal', 'value': 'CBS 109949 / FRC R-5291 / MAFF 237835 / NRRL 28334'}}\n",
      "‚ùå 1/3 Give me the list of strains associated to the Escherichia coli taxon and their scientific name. Return ?strain, ?name, limit to 20 \n",
      "Results mismatch. Ref: 20 != gen: 0\n",
      ". Generated query:\n",
      "# https://sparql.uniprot.org/sparql\n",
      "\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
      "\n",
      "SELECT ?strain ?name\n",
      "WHERE {\n",
      "  ?strain a up:Strain ;\n",
      "          up:scientificName ?name ;\n",
      "          up:organism taxon:562 .\n",
      "}\n",
      "LIMIT 20\n",
      "Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "SELECT ?strain ?name\n",
      "WHERE {\n",
      "    ?strain a up:Strain ;\n",
      "        up:name ?name .\n",
      "    ?organism rdfs:subClassOf taxon:562 . # Escherichia coli taxon ID\n",
      "} LIMIT 20\n",
      "\n",
      "Missing from generated: {'strain': {'type': 'uri', 'value': 'http://purl.uniprot.org/taxonomy/101028#strain-D19B933A18EA91CE5E571899A51CBD16'}, 'name': {'type': 'literal', 'value': 'CBS 109949 / FRC R-5291 / MAFF 237835 / NRRL 28334'}}\n",
      "‚ùå 2/3 Give me the list of strains associated to the Escherichia coli taxon and their scientific name. Return ?strain, ?name, limit to 20 \n",
      "Results mismatch. Ref: 20 != gen: 0\n",
      ". Generated query:\n",
      "# https://sparql.uniprot.org/sparql\n",
      "\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
      "\n",
      "SELECT ?strain ?name\n",
      "WHERE {\n",
      "  ?strain a up:Strain ;\n",
      "          up:scientificName ?name ;\n",
      "          up:organism taxon:562 .\n",
      "}\n",
      "LIMIT 20\n",
      "Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "SELECT ?strain ?name\n",
      "WHERE {\n",
      "    ?strain a up:Strain ;\n",
      "        up:name ?name .\n",
      "    ?organism rdfs:subClassOf taxon:562 . # Escherichia coli taxon ID\n",
      "} LIMIT 20\n",
      "\n",
      "Missing from generated: {'strain': {'type': 'uri', 'value': 'http://purl.uniprot.org/taxonomy/101028#strain-D19B933A18EA91CE5E571899A51CBD16'}, 'name': {'type': 'literal', 'value': 'CBS 109949 / FRC R-5291 / MAFF 237835 / NRRL 28334'}}\n",
      "‚ùå 3/3 Give me the list of strains associated to the Escherichia coli taxon and their scientific name. Return ?strain, ?name, limit to 20 \n",
      "Results mismatch. Ref: 20 != gen: 0\n",
      ". Generated query:\n",
      "# https://sparql.uniprot.org/sparql\n",
      "\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "\n",
      "SELECT ?strain ?name\n",
      "WHERE {\n",
      "  ?strain a up:Strain ;\n",
      "          up:scientificName ?name ;\n",
      "          up:organism <http://purl.uniprot.org/taxonomy/562> .\n",
      "}\n",
      "LIMIT 20\n",
      "Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "SELECT ?strain ?name\n",
      "WHERE {\n",
      "    ?strain a up:Strain ;\n",
      "        up:name ?name .\n",
      "    ?organism rdfs:subClassOf taxon:562 . # Escherichia coli taxon ID\n",
      "} LIMIT 20\n",
      "\n",
      "Missing from generated: {'strain': {'type': 'uri', 'value': 'http://purl.uniprot.org/taxonomy/101028#strain-D19B933A18EA91CE5E571899A51CBD16'}, 'name': {'type': 'literal', 'value': 'CBS 109949 / FRC R-5291 / MAFF 237835 / NRRL 28334'}}\n",
      "‚ùå 1/3 Give me the list of strains associated to the Escherichia coli taxon and their scientific name. Return ?strain, ?name, limit to 20 \n",
      "Results mismatch. Ref: 20 != gen: 0\n",
      ". Generated query:\n",
      "# https://sparql.uniprot.org/sparql/\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "\n",
      "SELECT ?strain ?name\n",
      "WHERE\n",
      "{\n",
      "    ?strain a up:Strain .\n",
      "    ?strain up:scientificName ?name .\n",
      "    ?strain rdfs:subClassOf taxon:562 .\n",
      "}\n",
      "LIMIT 20\n",
      "Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "SELECT ?strain ?name\n",
      "WHERE {\n",
      "    ?strain a up:Strain ;\n",
      "        up:name ?name .\n",
      "    ?organism rdfs:subClassOf taxon:562 . # Escherichia coli taxon ID\n",
      "} LIMIT 20\n",
      "\n",
      "Missing from generated: {'strain': {'type': 'uri', 'value': 'http://purl.uniprot.org/taxonomy/101028#strain-D19B933A18EA91CE5E571899A51CBD16'}, 'name': {'type': 'literal', 'value': 'CBS 109949 / FRC R-5291 / MAFF 237835 / NRRL 28334'}}\n",
      "‚ùå 2/3 Give me the list of strains associated to the Escherichia coli taxon and their scientific name. Return ?strain, ?name, limit to 20 \n",
      "Results mismatch. Ref: 20 != gen: 0\n",
      ". Generated query:\n",
      "# https://sparql.uniprot.org/sparql/\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "\n",
      "SELECT ?strain ?name\n",
      "WHERE\n",
      "{\n",
      "    ?strain a up:Strain .\n",
      "    ?strain up:scientificName ?name .\n",
      "    ?strain rdfs:subClassOf taxon:562 .\n",
      "}\n",
      "LIMIT 20\n",
      "Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "SELECT ?strain ?name\n",
      "WHERE {\n",
      "    ?strain a up:Strain ;\n",
      "        up:name ?name .\n",
      "    ?organism rdfs:subClassOf taxon:562 . # Escherichia coli taxon ID\n",
      "} LIMIT 20\n",
      "\n",
      "Missing from generated: {'strain': {'type': 'uri', 'value': 'http://purl.uniprot.org/taxonomy/101028#strain-D19B933A18EA91CE5E571899A51CBD16'}, 'name': {'type': 'literal', 'value': 'CBS 109949 / FRC R-5291 / MAFF 237835 / NRRL 28334'}}\n",
      "‚ùå 3/3 Give me the list of strains associated to the Escherichia coli taxon and their scientific name. Return ?strain, ?name, limit to 20 \n",
      "Results mismatch. Ref: 20 != gen: 0\n",
      ". Generated query:\n",
      "# https://sparql.uniprot.org/sparql/\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "\n",
      "SELECT ?strain ?name\n",
      "WHERE\n",
      "{\n",
      "    ?strain a up:Strain .\n",
      "    ?strain up:scientificName ?name .\n",
      "    ?strain rdfs:subClassOf taxon:562 . # Escherichia coli taxon ID\n",
      "}\n",
      "LIMIT 20\n",
      "Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "SELECT ?strain ?name\n",
      "WHERE {\n",
      "    ?strain a up:Strain ;\n",
      "        up:name ?name .\n",
      "    ?organism rdfs:subClassOf taxon:562 . # Escherichia coli taxon ID\n",
      "} LIMIT 20\n",
      "\n",
      "Missing from generated: {'strain': {'type': 'uri', 'value': 'http://purl.uniprot.org/taxonomy/101028#strain-D19B933A18EA91CE5E571899A51CBD16'}, 'name': {'type': 'literal', 'value': 'CBS 109949 / FRC R-5291 / MAFF 237835 / NRRL 28334'}}\n",
      "‚ùå 1/3 Give me the list of strains associated to the Escherichia coli taxon and their scientific name. Return ?strain, ?name, limit to 20 \n",
      "Results mismatch. Ref: 20 != gen: 0\n",
      ". Generated query:\n",
      "# https://sparql.uniprot.org/sparql/\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "\n",
      "SELECT ?strain ?name\n",
      "WHERE\n",
      "{\n",
      "    ?strain a up:Strain ;\n",
      "            rdfs:label ?name .\n",
      "    ?taxon rdfs:subClassOf taxon:562 .\n",
      "    ?taxon up:hasStrain ?strain .\n",
      "}\n",
      "LIMIT 20\n",
      "Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "SELECT ?strain ?name\n",
      "WHERE {\n",
      "    ?strain a up:Strain ;\n",
      "        up:name ?name .\n",
      "    ?organism rdfs:subClassOf taxon:562 . # Escherichia coli taxon ID\n",
      "} LIMIT 20\n",
      "\n",
      "‚ùå 2/3 Give me the list of strains associated to the Escherichia coli taxon and their scientific name. Return ?strain, ?name, limit to 20 The read operation timed out. Generated query:\n",
      "# https://sparql.uniprot.org/sparql/\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "\n",
      "SELECT ?strain ?name\n",
      "WHERE\n",
      "{\n",
      "    ?strain a up:Strain .\n",
      "    ?strain up:name ?name .\n",
      "    ?strain up:organism ?organism .\n",
      "    ?organism rdfs:subClassOf* taxon:562 . # Escherichia coli taxon and its subclasses\n",
      "}\n",
      "LIMIT 20\n",
      "Correct query:\n",
      "PREFIX up: <http://purl.uniprot.org/core/>\n",
      "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
      "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
      "SELECT ?strain ?name\n",
      "WHERE {\n",
      "    ?strain a up:Strain ;\n",
      "        up:name ?name .\n",
      "    ?organism rdfs:subClassOf taxon:562 . # Escherichia coli taxon ID\n",
      "} LIMIT 20\n",
      "\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[1], line 345\u001b[0m\n\u001b[1;32m    343\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m approach_index, approach \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28menumerate\u001b[39m(list_of_approaches):\n\u001b[1;32m    344\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m t \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mrange\u001b[39m(number_of_tries):\n\u001b[0;32m--> 345\u001b[0m         chat_resp_md \u001b[38;5;241m=\u001b[39m \u001b[43mapproach\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtest_query\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mquestion\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    346\u001b[0m         \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    347\u001b[0m             generated_sparql \u001b[38;5;241m=\u001b[39m extract_sparql_queries(chat_resp_md)[\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m1\u001b[39m]\n",
      "Cell \u001b[0;32mIn[1], line 305\u001b[0m, in \u001b[0;36manswer_rag_with_validation\u001b[0;34m(question)\u001b[0m\n\u001b[1;32m    304\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21manswer_rag_with_validation\u001b[39m(question: \u001b[38;5;28mstr\u001b[39m):\n\u001b[0;32m--> 305\u001b[0m     response \u001b[38;5;241m=\u001b[39m \u001b[43mrequests\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpost\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mhttp://localhost:8000/chat\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    306\u001b[0m \u001b[43m        \u001b[49m\u001b[43mjson\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m{\u001b[49m\n\u001b[1;32m    307\u001b[0m \u001b[43m            \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mmessages\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43m[\u001b[49m\u001b[43m{\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mrole\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43muser\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mcontent\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mquestion\u001b[49m\u001b[43m}\u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    308\u001b[0m \u001b[43m            \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mstream\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    309\u001b[0m \u001b[43m            \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mapi_key\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mexpasy_api_key\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    310\u001b[0m \u001b[43m        \u001b[49m\u001b[43m}\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    311\u001b[0m \u001b[43m        \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m60\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    312\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    313\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m response\u001b[38;5;241m.\u001b[39mjson()[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mchoices\u001b[39m\u001b[38;5;124m\"\u001b[39m][\u001b[38;5;241m0\u001b[39m][\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmessage\u001b[39m\u001b[38;5;124m\"\u001b[39m][\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcontent\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n",
      "File \u001b[0;32m~/dev/expasy/expasy-chat/.venv/lib/python3.10/site-packages/requests/api.py:115\u001b[0m, in \u001b[0;36mpost\u001b[0;34m(url, data, json, **kwargs)\u001b[0m\n\u001b[1;32m    103\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mpost\u001b[39m(url, data\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, json\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[1;32m    104\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124mr\u001b[39m\u001b[38;5;124;03m\"\"\"Sends a POST request.\u001b[39;00m\n\u001b[1;32m    105\u001b[0m \n\u001b[1;32m    106\u001b[0m \u001b[38;5;124;03m    :param url: URL for the new :class:`Request` object.\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    112\u001b[0m \u001b[38;5;124;03m    :rtype: requests.Response\u001b[39;00m\n\u001b[1;32m    113\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m--> 115\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mrequest\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mpost\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdata\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdata\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mjson\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mjson\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/dev/expasy/expasy-chat/.venv/lib/python3.10/site-packages/requests/api.py:59\u001b[0m, in \u001b[0;36mrequest\u001b[0;34m(method, url, **kwargs)\u001b[0m\n\u001b[1;32m     55\u001b[0m \u001b[38;5;66;03m# By using the 'with' statement we are sure the session is closed, thus we\u001b[39;00m\n\u001b[1;32m     56\u001b[0m \u001b[38;5;66;03m# avoid leaving sockets open which can trigger a ResourceWarning in some\u001b[39;00m\n\u001b[1;32m     57\u001b[0m \u001b[38;5;66;03m# cases, and look like a memory leak in others.\u001b[39;00m\n\u001b[1;32m     58\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m sessions\u001b[38;5;241m.\u001b[39mSession() \u001b[38;5;28;01mas\u001b[39;00m session:\n\u001b[0;32m---> 59\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43msession\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmethod\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43murl\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/dev/expasy/expasy-chat/.venv/lib/python3.10/site-packages/requests/sessions.py:589\u001b[0m, in \u001b[0;36mSession.request\u001b[0;34m(self, method, url, params, data, headers, cookies, files, auth, timeout, allow_redirects, proxies, hooks, stream, verify, cert, json)\u001b[0m\n\u001b[1;32m    584\u001b[0m send_kwargs \u001b[38;5;241m=\u001b[39m {\n\u001b[1;32m    585\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtimeout\u001b[39m\u001b[38;5;124m\"\u001b[39m: timeout,\n\u001b[1;32m    586\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mallow_redirects\u001b[39m\u001b[38;5;124m\"\u001b[39m: allow_redirects,\n\u001b[1;32m    587\u001b[0m }\n\u001b[1;32m    588\u001b[0m send_kwargs\u001b[38;5;241m.\u001b[39mupdate(settings)\n\u001b[0;32m--> 589\u001b[0m resp \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msend\u001b[49m\u001b[43m(\u001b[49m\u001b[43mprep\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43msend_kwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    591\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m resp\n",
      "File \u001b[0;32m~/dev/expasy/expasy-chat/.venv/lib/python3.10/site-packages/requests/sessions.py:703\u001b[0m, in \u001b[0;36mSession.send\u001b[0;34m(self, request, **kwargs)\u001b[0m\n\u001b[1;32m    700\u001b[0m start \u001b[38;5;241m=\u001b[39m preferred_clock()\n\u001b[1;32m    702\u001b[0m \u001b[38;5;66;03m# Send the request\u001b[39;00m\n\u001b[0;32m--> 703\u001b[0m r \u001b[38;5;241m=\u001b[39m \u001b[43madapter\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msend\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrequest\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    705\u001b[0m \u001b[38;5;66;03m# Total elapsed time of the request (approximately)\u001b[39;00m\n\u001b[1;32m    706\u001b[0m elapsed \u001b[38;5;241m=\u001b[39m preferred_clock() \u001b[38;5;241m-\u001b[39m start\n",
      "File \u001b[0;32m~/dev/expasy/expasy-chat/.venv/lib/python3.10/site-packages/requests/adapters.py:667\u001b[0m, in \u001b[0;36mHTTPAdapter.send\u001b[0;34m(self, request, stream, timeout, verify, cert, proxies)\u001b[0m\n\u001b[1;32m    664\u001b[0m     timeout \u001b[38;5;241m=\u001b[39m TimeoutSauce(connect\u001b[38;5;241m=\u001b[39mtimeout, read\u001b[38;5;241m=\u001b[39mtimeout)\n\u001b[1;32m    666\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 667\u001b[0m     resp \u001b[38;5;241m=\u001b[39m \u001b[43mconn\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43murlopen\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    668\u001b[0m \u001b[43m        \u001b[49m\u001b[43mmethod\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    669\u001b[0m \u001b[43m        \u001b[49m\u001b[43murl\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    670\u001b[0m \u001b[43m        \u001b[49m\u001b[43mbody\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mbody\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    671\u001b[0m \u001b[43m        \u001b[49m\u001b[43mheaders\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mheaders\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    672\u001b[0m \u001b[43m        \u001b[49m\u001b[43mredirect\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    673\u001b[0m \u001b[43m        \u001b[49m\u001b[43massert_same_host\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    674\u001b[0m \u001b[43m        \u001b[49m\u001b[43mpreload_content\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    675\u001b[0m \u001b[43m        \u001b[49m\u001b[43mdecode_content\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    676\u001b[0m \u001b[43m        \u001b[49m\u001b[43mretries\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmax_retries\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    677\u001b[0m \u001b[43m        \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    678\u001b[0m \u001b[43m        \u001b[49m\u001b[43mchunked\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mchunked\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    679\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    681\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m (ProtocolError, \u001b[38;5;167;01mOSError\u001b[39;00m) \u001b[38;5;28;01mas\u001b[39;00m err:\n\u001b[1;32m    682\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mConnectionError\u001b[39;00m(err, request\u001b[38;5;241m=\u001b[39mrequest)\n",
      "File \u001b[0;32m~/dev/expasy/expasy-chat/.venv/lib/python3.10/site-packages/urllib3/connectionpool.py:789\u001b[0m, in \u001b[0;36mHTTPConnectionPool.urlopen\u001b[0;34m(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, preload_content, decode_content, **response_kw)\u001b[0m\n\u001b[1;32m    786\u001b[0m response_conn \u001b[38;5;241m=\u001b[39m conn \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m release_conn \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m    788\u001b[0m \u001b[38;5;66;03m# Make the request on the HTTPConnection object\u001b[39;00m\n\u001b[0;32m--> 789\u001b[0m response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_make_request\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    790\u001b[0m \u001b[43m    \u001b[49m\u001b[43mconn\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    791\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    792\u001b[0m \u001b[43m    \u001b[49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    793\u001b[0m \u001b[43m    \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout_obj\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    794\u001b[0m \u001b[43m    \u001b[49m\u001b[43mbody\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mbody\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    795\u001b[0m \u001b[43m    \u001b[49m\u001b[43mheaders\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mheaders\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    796\u001b[0m \u001b[43m    \u001b[49m\u001b[43mchunked\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mchunked\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    797\u001b[0m \u001b[43m    \u001b[49m\u001b[43mretries\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mretries\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    798\u001b[0m \u001b[43m    \u001b[49m\u001b[43mresponse_conn\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mresponse_conn\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    799\u001b[0m \u001b[43m    \u001b[49m\u001b[43mpreload_content\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mpreload_content\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    800\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdecode_content\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdecode_content\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    801\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mresponse_kw\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    802\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    804\u001b[0m \u001b[38;5;66;03m# Everything went great!\u001b[39;00m\n\u001b[1;32m    805\u001b[0m clean_exit \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n",
      "File \u001b[0;32m~/dev/expasy/expasy-chat/.venv/lib/python3.10/site-packages/urllib3/connectionpool.py:536\u001b[0m, in \u001b[0;36mHTTPConnectionPool._make_request\u001b[0;34m(self, conn, method, url, body, headers, retries, timeout, chunked, response_conn, preload_content, decode_content, enforce_content_length)\u001b[0m\n\u001b[1;32m    534\u001b[0m \u001b[38;5;66;03m# Receive the response from the server\u001b[39;00m\n\u001b[1;32m    535\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 536\u001b[0m     response \u001b[38;5;241m=\u001b[39m \u001b[43mconn\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgetresponse\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    537\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m (BaseSSLError, \u001b[38;5;167;01mOSError\u001b[39;00m) \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m    538\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_raise_timeout(err\u001b[38;5;241m=\u001b[39me, url\u001b[38;5;241m=\u001b[39murl, timeout_value\u001b[38;5;241m=\u001b[39mread_timeout)\n",
      "File \u001b[0;32m~/dev/expasy/expasy-chat/.venv/lib/python3.10/site-packages/urllib3/connection.py:464\u001b[0m, in \u001b[0;36mHTTPConnection.getresponse\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    461\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mresponse\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m HTTPResponse\n\u001b[1;32m    463\u001b[0m \u001b[38;5;66;03m# Get the response from http.client.HTTPConnection\u001b[39;00m\n\u001b[0;32m--> 464\u001b[0m httplib_response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgetresponse\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    466\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    467\u001b[0m     assert_header_parsing(httplib_response\u001b[38;5;241m.\u001b[39mmsg)\n",
      "File \u001b[0;32m/usr/lib/python3.10/http/client.py:1375\u001b[0m, in \u001b[0;36mHTTPConnection.getresponse\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1373\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m   1374\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 1375\u001b[0m         \u001b[43mresponse\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mbegin\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1376\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mConnectionError\u001b[39;00m:\n\u001b[1;32m   1377\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mclose()\n",
      "File \u001b[0;32m/usr/lib/python3.10/http/client.py:318\u001b[0m, in \u001b[0;36mHTTPResponse.begin\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    316\u001b[0m \u001b[38;5;66;03m# read until we get a non-100 response\u001b[39;00m\n\u001b[1;32m    317\u001b[0m \u001b[38;5;28;01mwhile\u001b[39;00m \u001b[38;5;28;01mTrue\u001b[39;00m:\n\u001b[0;32m--> 318\u001b[0m     version, status, reason \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_read_status\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    319\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m status \u001b[38;5;241m!=\u001b[39m CONTINUE:\n\u001b[1;32m    320\u001b[0m         \u001b[38;5;28;01mbreak\u001b[39;00m\n",
      "File \u001b[0;32m/usr/lib/python3.10/http/client.py:279\u001b[0m, in \u001b[0;36mHTTPResponse._read_status\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    278\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_read_status\u001b[39m(\u001b[38;5;28mself\u001b[39m):\n\u001b[0;32m--> 279\u001b[0m     line \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mstr\u001b[39m(\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfp\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mreadline\u001b[49m\u001b[43m(\u001b[49m\u001b[43m_MAXLINE\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m)\u001b[49m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124miso-8859-1\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    280\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(line) \u001b[38;5;241m>\u001b[39m _MAXLINE:\n\u001b[1;32m    281\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m LineTooLong(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mstatus line\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[0;32m/usr/lib/python3.10/socket.py:705\u001b[0m, in \u001b[0;36mSocketIO.readinto\u001b[0;34m(self, b)\u001b[0m\n\u001b[1;32m    703\u001b[0m \u001b[38;5;28;01mwhile\u001b[39;00m \u001b[38;5;28;01mTrue\u001b[39;00m:\n\u001b[1;32m    704\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 705\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_sock\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrecv_into\u001b[49m\u001b[43m(\u001b[49m\u001b[43mb\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    706\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m timeout:\n\u001b[1;32m    707\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_timeout_occurred \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "import requests\n",
    "from dotenv import load_dotenv\n",
    "from openai import OpenAI\n",
    "from qdrant_client.models import FieldCondition, Filter, MatchValue, ScoredPoint\n",
    "from rdflib.plugins.sparql import prepareQuery, prepareUpdate\n",
    "from rdflib.plugins.sparql.algebra import translateQuery\n",
    "from SPARQLWrapper import JSON, TURTLE, XML, SPARQLWrapper\n",
    "\n",
    "from expasy_chat.config import settings\n",
    "from expasy_chat.validate_sparql import extract_sparql_queries\n",
    "\n",
    "load_dotenv()\n",
    "expasy_api_key = os.getenv('EXPASY_API_KEY')\n",
    "\n",
    "example_queries = [\n",
    "  {\n",
    "    \"question\": \"What is the accession number in uniprot of the human gene LCT? Return only unique protein URIs\",\n",
    "    \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    \"query\": \"\"\"PREFIX up: <http://purl.uniprot.org/core/>\n",
    "PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
    "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
    "SELECT DISTINCT ?protein\n",
    "WHERE{\n",
    "    ?protein a up:Protein .\n",
    "    ?protein up:organism taxon:9606 .\n",
    "    ?protein up:encodedBy ?gene .\n",
    "    ?gene skos:prefLabel \"LCT\" .\n",
    "}\"\"\",\n",
    "  },\n",
    "  {\n",
    "    # NOTE: The \"mature\" part in the question makes it harder to answer\n",
    "    # \"question\": \"How do I filter for reviewed (mouse) proteins whose mature form carries an N-terminal glycine? Return protein URI and AA sequence\",\n",
    "    \"question\": \"How do I filter for reviewed mouse proteins which carry an N-terminal glycine? Return protein URI and AA sequence\",\n",
    "    \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    \"query\": \"\"\"PREFIX up: <http://purl.uniprot.org/core/>\n",
    "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
    "PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\n",
    "SELECT ?protein ?sequence\n",
    "WHERE\n",
    "{\n",
    "    ?protein a up:Protein ;\n",
    "        up:organism taxon:10090 ;  # Taxonomy ID for Mus musculus (Mouse)\n",
    "        up:reviewed true ;\n",
    "        up:sequence ?isoform .\n",
    "    ?isoform rdf:value ?sequence .\n",
    "    # Ensure the N-terminal amino acid is Glycine (G)\n",
    "    FILTER (STRSTARTS(?sequence, \"G\"))\n",
    "}\"\"\",\n",
    "  },\n",
    "\n",
    "  {\n",
    "    \"question\": \"How could I download a table that only includes the Rhea reactions for which there is experimental evidence? Return only the rhea URI\",\n",
    "    \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    \"query\": \"\"\"PREFIX up: <http://purl.uniprot.org/core/>\n",
    "PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\n",
    "SELECT DISTINCT ?rhea\n",
    "WHERE {\n",
    "  # ECO 269 is experimental evidence\n",
    "  BIND (<http://purl.obolibrary.org/obo/ECO_0000269> as ?evidence)\n",
    "  #GRAPH <http://sparql.uniprot.org/uniprot> {\n",
    "    ?protein up:reviewed true ;\n",
    "      up:annotation ?a ;\n",
    "      up:attribution ?attribution  .\n",
    "\n",
    "    ?a a up:Catalytic_Activity_Annotation ;\n",
    "      up:catalyticActivity ?ca .\n",
    "    ?ca up:catalyzedReaction ?rhea .\n",
    "\n",
    "    [] rdf:subject ?a ;\n",
    "      rdf:predicate up:catalyticActivity ;\n",
    "      rdf:object ?ca ;\n",
    "      up:attribution ?attribution .\n",
    "\n",
    "    ?attribution up:evidence ?evidence .\n",
    "  #}\n",
    "}\"\"\",\n",
    "  },\n",
    "\n",
    "  {\n",
    "    \"question\": \"Which human proteins are enzymes catalyzing a reaction involving sterols? Return the protein, sterol and reaction URI\",\n",
    "    \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    \"query\": \"\"\"PREFIX up: <http://purl.uniprot.org/core/>\n",
    "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
    "PREFIX rh: <http://rdf.rhea-db.org/>\n",
    "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
    "PREFIX CHEBI: <http://purl.obolibrary.org/obo/CHEBI_>\n",
    "SELECT DISTINCT ?protein ?sterol ?reaction\n",
    "WHERE {\n",
    "  SERVICE <https://sparql.rhea-db.org/sparql> {\n",
    "    ?reaction rdfs:subClassOf rh:Reaction .\n",
    "    ?reaction rh:side/rh:contains/rh:compound ?compound .\n",
    "    ?compound rh:chebi ?sterol .\n",
    "    ?sterol rdfs:subClassOf* CHEBI:15889 .\n",
    "  }\n",
    "  ?protein a up:Protein ;\n",
    "    up:organism taxon:9606 ;\n",
    "    up:annotation/up:catalyticActivity/up:catalyzedReaction ?reaction .\n",
    "}\"\"\",\n",
    "  },\n",
    "\n",
    "  {\n",
    "    \"question\": \"Which are the human proteins associated with cancer (which have cancer in their disease label)? Return the unique disease label (?diseaseLabel), and HGNC symbol (?hgncSymbol)\",\n",
    "    \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    \"query\": \"\"\"PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
    "PREFIX up:<http://purl.uniprot.org/core/>\n",
    "PREFIX taxon:<http://purl.uniprot.org/taxonomy/>\n",
    "PREFIX rdfs:<http://www.w3.org/2000/01/rdf-schema#>\n",
    "PREFIX orth:<http://purl.org/net/orth#>\n",
    "PREFIX dcterms:<http://purl.org/dc/terms/>\n",
    "PREFIX obo:<http://purl.obolibrary.org/obo/>\n",
    "PREFIX lscr:<http://purl.org/lscr#>\n",
    "PREFIX genex:<http://purl.org/genex#>\n",
    "PREFIX sio: <http://semanticscience.org/resource/>\n",
    "SELECT DISTINCT ?diseaseLabel ?hgncSymbol\n",
    "WHERE {\n",
    "    ?humanProtein a up:Protein ;\n",
    "        # up:organism/up:scientificName 'Homo sapiens' ;\n",
    "        up:organism taxon:9606 ;\n",
    "        up:annotation ?annotation ;\n",
    "        rdfs:seeAlso ?hgnc .\n",
    "    ?hgnc up:database <http://purl.uniprot.org/database/HGNC> ;\n",
    "        rdfs:comment ?hgncSymbol .\n",
    "    ?annotation a up:Disease_Annotation ;\n",
    "        up:disease ?disease .\n",
    "    ?disease skos:prefLabel ?diseaseLabel.\n",
    "    FILTER CONTAINS (LCASE(?diseaseLabel), \"cancer\")\n",
    "}\"\"\",\n",
    "  },\n",
    "\n",
    "  {\n",
    "    \"question\": \"In bgee how can I retrieve the confidence level and false discovery rate of a gene expression? Return distinct ?gene, ?confidence and ?fdr, limit to 10\",\n",
    "    \"endpoint\": \"https://www.bgee.org/sparql/\",\n",
    "    \"query\": \"\"\"PREFIX genex: <http://purl.org/genex#>\n",
    "PREFIX bgee: <http://bgee.org/#>\n",
    "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
    "SELECT DISTINCT ?gene ?confidence ?fdr\n",
    "WHERE {\n",
    "    ?expression a genex:Expression ;\n",
    "        genex:hasConfidenceLevel ?confidence ;\n",
    "        genex:hasFDRpvalue ?fdr ;\n",
    "        genex:hasSequenceUnit ?gene .\n",
    "} LIMIT 10\"\"\",\n",
    "  },\n",
    "\n",
    "  {\n",
    "    # Asking for \"full name\" will cause the RAG to fail, but validation don't catch it yet because VoID don't register the class when we have\n",
    "    # \"up:Protein up:enzyme ??\"...\n",
    "    \"question\": \"How can I get a list of proteins that are enzymes? Return the proteins URI, and their enzyme class URI and name, limit to 20\",\n",
    "    \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    \"query\": \"\"\"PREFIX up: <http://purl.uniprot.org/core/>\n",
    "PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
    "SELECT ?protein ?enzymeClass ?enzymeClassName\n",
    "WHERE {\n",
    "    ?protein a up:Protein ;\n",
    "        up:enzyme ?enzymeClass .\n",
    "    ?enzymeClass skos:prefLabel ?enzymeClassName .\n",
    "} LIMIT 20\"\"\",\n",
    "  },\n",
    "\n",
    "  {\n",
    "    # There are no example with xrefEnsembl, so the RAG without validation usually fails\n",
    "    \"question\": \"How can I get the cross ref to the ensembl protein for the LCT protein in OMA? Return only the distinct ?ensemblURI\",\n",
    "    \"endpoint\": \"https://sparql.omabrowser.org/sparql/\",\n",
    "    \"query\": \"\"\"PREFIX orth: <http://purl.org/net/orth#>\n",
    "PREFIX lscr: <http://purl.org/lscr#>\n",
    "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
    "SELECT DISTINCT ?ensemblURI\n",
    "WHERE {\n",
    "    ?protein a orth:Protein ;\n",
    "        rdfs:label 'LCT' ;\n",
    "        lscr:xrefEnsemblProtein ?ensemblURI .\n",
    "}\"\"\",\n",
    "  },\n",
    "\n",
    "  {\n",
    "    # There are no example with inDataset\n",
    "    \"question\": \"How can I get the URI of a dataset to which an ortholog cluster belongs in OMA? Return orthologCluster, datasetURI and limit to 20\",\n",
    "    \"endpoint\": \"https://sparql.omabrowser.org/sparql/\",\n",
    "    \"query\": \"\"\"PREFIX orth: <http://purl.org/net/orth#>\n",
    "SELECT DISTINCT ?orthologCluster ?datasetURI\n",
    "WHERE {\n",
    "    ?orthologCluster a orth:OrthologsCluster ;\n",
    "        orth:inDataset ?datasetURI .\n",
    "} LIMIT 20\"\"\",\n",
    "  },\n",
    "\n",
    "  {\n",
    "    \"question\": \"Give me the list of strains associated to the Escherichia coli taxon and their name. Return ?taxon, ?strain, ?name, limit to 20\",\n",
    "    \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "    \"query\": \"\"\"PREFIX up: <http://purl.uniprot.org/core/>\n",
    "PREFIX taxon: <http://purl.uniprot.org/taxonomy/>\n",
    "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
    "SELECT ?taxon ?strain ?name\n",
    "WHERE {\n",
    "    ?taxon a up:Taxon .\n",
    "    ?taxon rdfs:subClassOf taxon:562 . # Escherichia coli taxon\n",
    "    ?taxon up:strain ?strain .\n",
    "    ?strain up:name ?name .\n",
    "} LIMIT 20\"\"\",\n",
    "  },\n",
    "\n",
    "    # Give me the list of strains associated to the Escherichia coli taxon and their scientific name\n",
    "\n",
    "    # Give me the query to retrieve the common name, scientific name of taxons, with the scientific name of their strains limit to 100 results\n",
    "    # Retrieve strains of taxons. Return scientific and common name of taxons, and scientific name of the strains, limit to 100 results\n",
    "\n",
    "#   {\n",
    "#     \"question\": \"Which are the human genes associated with cancer (which have cancer in their disease label) and their orthologs expressed in the rat brain? Return the disease label, human gene URI, human gene HGNC symbol, ortholog rat gene URI\",\n",
    "#     \"endpoint\": \"https://sparql.uniprot.org/sparql/\",\n",
    "#     \"query\": \"\"\"PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
    "# PREFIX up:<http://purl.uniprot.org/core/>\n",
    "# PREFIX taxon:<http://purl.uniprot.org/taxonomy/>\n",
    "# PREFIX rdfs:<http://www.w3.org/2000/01/rdf-schema#>\n",
    "# PREFIX orth:<http://purl.org/net/orth#>\n",
    "# PREFIX dcterms:<http://purl.org/dc/terms/>\n",
    "# PREFIX obo:<http://purl.obolibrary.org/obo/>\n",
    "# PREFIX lscr:<http://purl.org/lscr#>\n",
    "# PREFIX genex:<http://purl.org/genex#>\n",
    "# PREFIX sio: <http://semanticscience.org/resource/>\n",
    "# SELECT DISTINCT ?diseaseLabel ?humanProtein ?hgncSymbol ?orthologRatProtein ?orthologRatGene\n",
    "# WHERE {\n",
    "#   SERVICE <https://sparql.uniprot.org/sparql> {\n",
    "# \tSELECT DISTINCT * WHERE {\n",
    "#   \t?humanProtein a up:Protein ;\n",
    "#       \tup:organism/up:scientificName 'Homo sapiens' ;\n",
    "#       \tup:annotation ?annotation ;\n",
    "#       \trdfs:seeAlso ?hgnc .\n",
    "#   \t?hgnc up:database <http://purl.uniprot.org/database/HGNC> ;\n",
    "#       \trdfs:comment ?hgncSymbol .\n",
    "#   \t?annotation a up:Disease_Annotation ;\n",
    "#       \t\tup:disease ?disease .\n",
    "#   \t?disease skos:prefLabel ?diseaseLabel.\n",
    "#   \tFILTER CONTAINS (?diseaseLabel, \"cancer\")\n",
    "# \t}\n",
    "#   }\n",
    "#   SERVICE <https://sparql.omabrowser.org/sparql/> {\n",
    "# \tSELECT ?humanProtein ?orthologRatProtein ?orthologRatGene WHERE {\n",
    "#   \t?humanProteinOma a orth:Protein ;\n",
    "#       \t\tlscr:xrefUniprot ?humanProtein .\n",
    "#   \t?orthologRatProtein a orth:Protein ;\n",
    "#       \tsio:SIO_010079 ?orthologRatGene ;\n",
    "#       \torth:organism/obo:RO_0002162/up:scientificName 'Rattus norvegicus' .\n",
    "#   \t?cluster a orth:OrthologsCluster .\n",
    "#   \t?cluster orth:hasHomologousMember ?node1 .\n",
    "#   \t?cluster orth:hasHomologousMember ?node2 .\n",
    "#   \t?node1 orth:hasHomologousMember* ?humanProteinOma .\n",
    "#   \t?node2 orth:hasHomologousMember* ?orthologRatProtein .\n",
    "#   \tFILTER(?node1 != ?node2)\n",
    "# \t}\n",
    "#   }\n",
    "#   SERVICE <https://www.bgee.org/sparql/> {\n",
    "# \t?orthologRatGene genex:isExpressedIn ?anatEntity ;\n",
    "#   \t\torth:organism ?ratOrganism .\n",
    "# \t?anatEntity rdfs:label 'brain' .\n",
    "# \t?ratOrganism obo:RO_0002162 taxon:10116 .\n",
    "#   }\n",
    "# }\"\"\",\n",
    "#   },\n",
    "\n",
    "]\n",
    "\n",
    "def result_sets_are_same(gen_set, ref_set) -> bool:\n",
    "    gen_set, ref_set = list(gen_set), list(ref_set)\n",
    "    for item in gen_set:\n",
    "        if item not in ref_set:\n",
    "            print(f\"Missing from reference: {item}\")\n",
    "            return False\n",
    "    for item in ref_set:\n",
    "        if item not in gen_set:\n",
    "            print(f\"Missing from generated: {item}\")\n",
    "            return False\n",
    "    return True\n",
    "\n",
    "QLEVER_UNIPROT = \"https://qlever.cs.uni-freiburg.de/api/uniprot\"\n",
    "\n",
    "client = OpenAI()\n",
    "def answer_no_rag(question: str):\n",
    "    response = client.chat.completions.create(\n",
    "        model=settings.llm_model,\n",
    "        messages=[\n",
    "            {\"role\": \"system\", \"content\": settings.system_prompt},\n",
    "            {\"role\": \"user\", \"content\": question},\n",
    "        ],\n",
    "        stream=False,\n",
    "        temperature=0,\n",
    "        #   response_format={ \"type\": \"json_object\" },\n",
    "    )\n",
    "    return response.choices[0].message.content\n",
    "\n",
    "def answer_rag_without_validation(question: str):\n",
    "    response = requests.post(\"http://localhost:8000/chat\",\n",
    "        json={\n",
    "            \"messages\": [{\"role\": \"user\", \"content\": question}],\n",
    "            \"stream\": False,\n",
    "            \"api_key\": expasy_api_key,\n",
    "            \"validate\": False,\n",
    "        },\n",
    "        timeout=60,\n",
    "    )\n",
    "    return response.json()[\"choices\"][0][\"message\"][\"content\"]\n",
    "\n",
    "def answer_rag_with_validation(question: str):\n",
    "    response = requests.post(\"http://localhost:8000/chat\",\n",
    "        json={\n",
    "            \"messages\": [{\"role\": \"user\", \"content\": question}],\n",
    "            \"stream\": False,\n",
    "            \"api_key\": expasy_api_key,\n",
    "        },\n",
    "        timeout=60,\n",
    "    )\n",
    "    return response.json()[\"choices\"][0][\"message\"][\"content\"]\n",
    "\n",
    "\n",
    "list_of_approaches = [\n",
    "    answer_no_rag,\n",
    "    answer_rag_without_validation,\n",
    "    answer_rag_with_validation,\n",
    "]\n",
    "\n",
    "number_of_tries = 3\n",
    "\n",
    "success = [0, 0, 0]\n",
    "mismatch = [0, 0, 0]\n",
    "fail = [0, 0, 0]\n",
    "\n",
    "\n",
    "print(f\"üß™ Testing {len(example_queries)} queries\")\n",
    "for i, test_query in enumerate(example_queries):\n",
    "    # if i < 1:\n",
    "    #     continue\n",
    "    # Execute the reference query\n",
    "    ref_endpoint = SPARQLWrapper(test_query[\"endpoint\"])\n",
    "    # ref_endpoint = SPARQLWrapper(QLEVER_UNIPROT)\n",
    "    ref_endpoint.setReturnFormat(JSON)\n",
    "    # ref_endpoint.setTimeout(200)\n",
    "    ref_endpoint.setQuery(test_query[\"query\"])\n",
    "    results = ref_endpoint.query().convert()\n",
    "    res_from_ref = results[\"results\"][\"bindings\"]\n",
    "\n",
    "\n",
    "    for approach_index, approach in enumerate(list_of_approaches):\n",
    "        for t in range(number_of_tries):\n",
    "            chat_resp_md = approach(test_query[\"question\"])\n",
    "            try:\n",
    "                generated_sparql = extract_sparql_queries(chat_resp_md)[-1]\n",
    "                if generated_sparql[\"query\"].strip() == test_query[\"query\"].strip():\n",
    "                    print(f\"‚úÖ {t+1}/{number_of_tries} {test_query['question']}. EXACT MATCH\")\n",
    "                    success[approach_index] += 1\n",
    "                    continue\n",
    "\n",
    "                # Execute the generated query\n",
    "                sparql_endpoint = SPARQLWrapper(generated_sparql[\"endpoint_url\"])\n",
    "                # sparql_endpoint = SPARQLWrapper(QLEVER_UNIPROT)\n",
    "                sparql_endpoint.setReturnFormat(JSON)\n",
    "                sparql_endpoint.setTimeout(200)\n",
    "                sparql_endpoint.setQuery(generated_sparql[\"query\"])\n",
    "                results = sparql_endpoint.query().convert()\n",
    "                res_from_generated = results[\"results\"][\"bindings\"]\n",
    "\n",
    "                if not result_sets_are_same(res_from_generated, res_from_ref):\n",
    "                    mismatch[approach_index] += 1\n",
    "                    raise Exception(f\"\\nResults mismatch. Ref: {len(res_from_ref)} != gen: {len(res_from_generated)}\\n\")\n",
    "                else:\n",
    "                    print(f\"‚úÖ {t+1}/{number_of_tries} {test_query['question']} = {len(res_from_generated)}\")\n",
    "                    success[approach_index] += 1\n",
    "\n",
    "            except Exception as e:\n",
    "                fail[approach_index] += 1\n",
    "                print(f\"‚ùå {t+1}/{number_of_tries} {test_query['question']} {e}. Generated query:\")\n",
    "                print(generated_sparql[\"query\"])\n",
    "                print(\"Correct query:\")\n",
    "                print(test_query[\"query\"])\n",
    "                print(\"\")\n",
    "            # print(f\"Results: {len(results['results']['bindings'])}\")\n",
    "\n",
    "    print(f\"üéØ No RAG - Success: {success[0]}, Mismatch: {mismatch[0]}, Error: {fail[0]-mismatch[0]}\")\n",
    "    print(f\"üéØ RAG without validation - Success: {success[1]}, Mismatch: {mismatch[1]}, Error: {fail[1]-mismatch[1]}\")\n",
    "    print(f\"üéØ RAG with validation - Success: {success[2]}, Mismatch: {mismatch[2]}, Error: {fail[2]-mismatch[2]}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "expasy-scripts",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
